{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate RNA Secondary structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import re\n",
    "import numpy as np\n",
    "import forgi.graph.bulge_graph as fgb\n",
    "\n",
    "v19db_filename = '/projects/ps-yeolab/genomes/hg19/gencode/v19/gencode.v19.annotation.gtf.db'\n",
    "v19db = gffutils.FeatureDB(v19db_filename)\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "csv_folder = '{}/csvs_for_paper/'.format(folder)\n",
    "bed_folder = '{}/bed'.format(folder)\n",
    "\n",
    "alt_exons_bedfile = '{}/exon2.bed'.format(bed_folder)\n",
    "constitutive_bedfile = '{}/constitutive_exons.bed'.format(bed_folder)\n",
    "\n",
    "\n",
    "splicing_feature_folder = '{}/splicing_feature_data'.format(csv_folder)\n",
    "alternative_feature_folder = '{}/alternative'.format(splicing_feature_folder)\n",
    "constitutive_feature_folder = '{}/constitutive'.format(splicing_feature_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "figure_folder = '/home/obotvinnik/Dropbox/figures2/singlecell_pnms/isoform_rna_properties'\n",
    "! mkdir $figure_folder\n",
    "\n",
    "prefix = 'isoform_transcriptions'\n",
    "transcribed_fasta = '{}/{}.fa'.format(folder, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from qtools import Submitter\n",
    "\n",
    "rnafold_results = transcribed_fasta.replace('.fa', '_rnafold.txt')\n",
    "command = 'RNAfold < {} > {}'.format(transcribed_fasta, rnafold_results)\n",
    "print command\n",
    "sub = Submitter([command], 'RNAfold', walltime='168:00:00', write_and_submit=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "! wc -l $transcribed_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCACAGUGCCAGCUCCCUGCGCCCGGCCGACCUGCUUGCCCUCAUCCUCCUGGUUCAGGACCUCUACCCCAGCGAGAGCACAGCAGAGGACGACAUUCAGCCUUCCCCGCGGAGGGCCCGGAGCAGCCAGAACAUCCCCGUGCAGCAGGCCUGGAGCCCUCACUCCACGGGCCCAACUGGGGGGAGCUCUGCAGAGACGGAGACAGACAGCUUCUCCCUCCCUGAGGAGUACUUCACACCAGCUCCUUCCCCUGGCGAUCAGAGCUCAG..........(((((....((((.((((...((((((((.(((....(((((...))))).............))).))).))))).))..........(((((((....))))))).((((((..........(((((...(((..(((((((((......))))..)))))...))).)))))))))))(((.((.(((((.((....))))))).)).)))((((((...........))))))...)).))))....)))))...(-95.90)\n",
      "CCACAGUGCCAGCUCCCUGCGCCCGGCCGACCUGCUUGCCCUCAUCCUCCUGGUUCAGGACCUCUACCCCAGCGAGAGCACAGCAGAGGACGACAUUCAGGAGACAGACAGCUUCUCCCUCCCUGAGGAGUACUUCACACCAGCUCCUUCCCCUGGCGAUCAGAGCUCAG..........(((((....((((.((((...((((((((.(((....(((((...))))).............))).))).))))).))..........(((((.((....)))))))......(((((((...........)))))))..)).))))....)))))...(-47.10)\n",
      "GCUGCGGCCCCCAGACCUGGCGCAGCGUGUCCAGCUGUGGGAGCACUUCCAGAGCCUGCUGUGGACCUACAGCCGCCUGCGGGAGCAGGAGCAGUGCUUCGCCGUGGAGGUGACUACUGGGCGCGGGAGUCCUCAGGACUUGAAGCUGUCAGGCCCUGGAGCGACUGAUUCACCCCCAGCUCUGUGAGCUGUGCAUAGAGGCGCUGGAGCGGCACGUCAUCCAGGCUGUCAACACCAGCCCCGAGCGGGGAGGCGAGGAGGCCCUGCAUGCCUUCCUGCUCGUGCACUCCAAGCUGCUGGCAUUCUACUCUAGCCACAGUGCCAGCUCCCUGCGCCCGGCCGACCUGCUUGCCCUCAUCCUCCUGGUUCAGGACCUCUACCCCAGCGAGAGCACAGCAGAGGACGACAUUCAG((((((.((.........))))))))((((((.((((((((......((((.((....)).))))))))))))..(((((....))))).((.(((((((((.((((((((.....(((((((((((((((((((((.(((((.....))))).))).))).))))..........((((((...))))))........(((..(((((..((((........(((((.......)))))..(((((((((((((.(.((...)).).))))))))))))))))).))))).)))((((((((((((...)))....)))))))))..)))))))))))....((((...(((...........))).))))))))))))....)))).))))).))...)))).))......(-183.10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>event_id</th>\n",
       "      <th>isoform</th>\n",
       "      <th>rnafold_double_stranded_percentage</th>\n",
       "      <th>rnafold_five_prime_percentage</th>\n",
       "      <th>rnafold_hairpin_percentage</th>\n",
       "      <th>rnafold_interior_loop_percentage</th>\n",
       "      <th>rnafold_minimum_free_energy</th>\n",
       "      <th>rnafold_multiloop_percentage</th>\n",
       "      <th>rnafold_stem_percentage</th>\n",
       "      <th>rnafold_three_prime_percentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>isoform2</td>\n",
       "      <td>0.555957</td>\n",
       "      <td>0.037175</td>\n",
       "      <td>0.104089</td>\n",
       "      <td>0.223048</td>\n",
       "      <td>-95.9</td>\n",
       "      <td>0.052045</td>\n",
       "      <td>0.572491</td>\n",
       "      <td>0.011152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>isoform1</td>\n",
       "      <td>0.483146</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>-47.1</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.505882</td>\n",
       "      <td>0.017647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>isoform2</td>\n",
       "      <td>0.644550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>0.096852</td>\n",
       "      <td>-183.1</td>\n",
       "      <td>0.111380</td>\n",
       "      <td>0.658596</td>\n",
       "      <td>0.014528</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            event_id   isoform  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...  isoform2   \n",
       "1  exon:chr10:100190328-100190427:-@exon:chr10:10...  isoform1   \n",
       "2  exon:chr10:100193697-100193848:-@exon:chr10:10...  isoform2   \n",
       "\n",
       "   rnafold_double_stranded_percentage  rnafold_five_prime_percentage  \\\n",
       "0                            0.555957                       0.037175   \n",
       "1                            0.483146                       0.058824   \n",
       "2                            0.644550                            NaN   \n",
       "\n",
       "   rnafold_hairpin_percentage  rnafold_interior_loop_percentage  \\\n",
       "0                    0.104089                          0.223048   \n",
       "1                    0.105882                          0.205882   \n",
       "2                    0.118644                          0.096852   \n",
       "\n",
       "   rnafold_minimum_free_energy  rnafold_multiloop_percentage  \\\n",
       "0                        -95.9                      0.052045   \n",
       "1                        -47.1                      0.105882   \n",
       "2                       -183.1                      0.111380   \n",
       "\n",
       "   rnafold_stem_percentage  rnafold_three_prime_percentage  \n",
       "0                 0.572491                        0.011152  \n",
       "1                 0.505882                        0.017647  \n",
       "2                 0.658596                        0.014528  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "seriess = []\n",
    "\n",
    "for record in SeqIO.parse('rnafold_output', 'fasta'):\n",
    "    event_name, isoform = record.id.split('|')\n",
    "    print record.seq\n",
    "    s = str(record.seq).lstrip('ACGUT')\n",
    "    search = re.search('([\\+.(+)+]+)(\\(-?\\d+\\.\\d+\\))', s)\n",
    "    structure, mfe = search.groups()\n",
    "    mfe = float(mfe.strip('()'))\n",
    "    bg = fgb.BulgeGraph(dotbracket_str=structure)\n",
    "    element_counts = pd.Series(Counter(bg.to_element_string()))\n",
    "    element_counts = element_counts.rename(acronym_to_full)\n",
    "    element_percentage = element_counts/element_counts.sum()\n",
    "    element_percentage.index = element_percentage.index.map(lambda x: x + '_percentage')\n",
    "    element_percentage['double_stranded_percentage'] = sum(1 for _ in structure if _ != '.')/float(len(s))\n",
    "    element_percentage['minimum_free_energy'] = mfe\n",
    "    element_percentage.index = element_percentage.index.map(lambda x: 'rnafold_'+ x)\n",
    "    element_percentage['event_id'] = event_name\n",
    "    element_percentage['isoform'] = isoform\n",
    "\n",
    "    seriess.append(element_percentage)\n",
    "structure_df = pd.DataFrame(seriess)\n",
    "structure_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_double_stranded_percentage</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_five_prime_percentage</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_hairpin_percentage</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_interior_loop_percentage</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_minimum_free_energy</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_multiloop_percentage</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_stem_percentage</th>\n",
       "      <th colspan=\"2\" halign=\"left\">rnafold_three_prime_percentage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>0.483146</td>\n",
       "      <td>0.555957</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.037175</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.104089</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>0.223048</td>\n",
       "      <td>-47.1</td>\n",
       "      <td>-95.9</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.052045</td>\n",
       "      <td>0.505882</td>\n",
       "      <td>0.572491</td>\n",
       "      <td>0.017647</td>\n",
       "      <td>0.011152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.644550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.096852</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-183.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.111380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.658596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.014528</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   rnafold_double_stranded_percentage  \\\n",
       "isoform                                                                      isoform1   \n",
       "event_id                                                                                \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                           0.483146   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                NaN   \n",
       "\n",
       "                                                              \\\n",
       "isoform                                             isoform2   \n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.555957   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  0.644550   \n",
       "\n",
       "                                                   rnafold_five_prime_percentage  \\\n",
       "isoform                                                                 isoform1   \n",
       "event_id                                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                      0.058824   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                           NaN   \n",
       "\n",
       "                                                              \\\n",
       "isoform                                             isoform2   \n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.037175   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...       NaN   \n",
       "\n",
       "                                                   rnafold_hairpin_percentage  \\\n",
       "isoform                                                              isoform1   \n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                   0.105882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                        NaN   \n",
       "\n",
       "                                                              \\\n",
       "isoform                                             isoform2   \n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.104089   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  0.118644   \n",
       "\n",
       "                                                   rnafold_interior_loop_percentage  \\\n",
       "isoform                                                                    isoform1   \n",
       "event_id                                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                         0.205882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                              NaN   \n",
       "\n",
       "                                                              \\\n",
       "isoform                                             isoform2   \n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.223048   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  0.096852   \n",
       "\n",
       "                                                   rnafold_minimum_free_energy  \\\n",
       "isoform                                                               isoform1   \n",
       "event_id                                                                         \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       -47.1   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                         NaN   \n",
       "\n",
       "                                                             \\\n",
       "isoform                                            isoform2   \n",
       "event_id                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    -95.9   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...   -183.1   \n",
       "\n",
       "                                                   rnafold_multiloop_percentage  \\\n",
       "isoform                                                                isoform1   \n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     0.105882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                          NaN   \n",
       "\n",
       "                                                              \\\n",
       "isoform                                             isoform2   \n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.052045   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  0.111380   \n",
       "\n",
       "                                                   rnafold_stem_percentage  \\\n",
       "isoform                                                           isoform1   \n",
       "event_id                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                0.505882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                     NaN   \n",
       "\n",
       "                                                              \\\n",
       "isoform                                             isoform2   \n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.572491   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  0.658596   \n",
       "\n",
       "                                                   rnafold_three_prime_percentage  \\\n",
       "isoform                                                                  isoform1   \n",
       "event_id                                                                            \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       0.017647   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                            NaN   \n",
       "\n",
       "                                                              \n",
       "isoform                                             isoform2  \n",
       "event_id                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  0.011152  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  0.014528  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structure_df_2d = structure_df.pivot(index='event_id', columns='isoform')\n",
    "structure_df_2d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform1_rnafold_double_stranded_percentage</th>\n",
       "      <th>isoform2_rnafold_double_stranded_percentage</th>\n",
       "      <th>isoform1_rnafold_five_prime_percentage</th>\n",
       "      <th>isoform2_rnafold_five_prime_percentage</th>\n",
       "      <th>isoform1_rnafold_hairpin_percentage</th>\n",
       "      <th>isoform2_rnafold_hairpin_percentage</th>\n",
       "      <th>isoform1_rnafold_interior_loop_percentage</th>\n",
       "      <th>isoform2_rnafold_interior_loop_percentage</th>\n",
       "      <th>isoform1_rnafold_minimum_free_energy</th>\n",
       "      <th>isoform2_rnafold_minimum_free_energy</th>\n",
       "      <th>isoform1_rnafold_multiloop_percentage</th>\n",
       "      <th>isoform2_rnafold_multiloop_percentage</th>\n",
       "      <th>isoform1_rnafold_stem_percentage</th>\n",
       "      <th>isoform2_rnafold_stem_percentage</th>\n",
       "      <th>isoform1_rnafold_three_prime_percentage</th>\n",
       "      <th>isoform2_rnafold_three_prime_percentage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>0.483146</td>\n",
       "      <td>0.555957</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.037175</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.104089</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>0.223048</td>\n",
       "      <td>-47.1</td>\n",
       "      <td>-95.9</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.052045</td>\n",
       "      <td>0.505882</td>\n",
       "      <td>0.572491</td>\n",
       "      <td>0.017647</td>\n",
       "      <td>0.011152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.644550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.096852</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-183.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.111380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.658596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.014528</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    isoform1_rnafold_double_stranded_percentage  \\\n",
       "event_id                                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     0.483146   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                          NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_double_stranded_percentage  \\\n",
       "event_id                                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     0.555957   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                     0.644550   \n",
       "\n",
       "                                                    isoform1_rnafold_five_prime_percentage  \\\n",
       "event_id                                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                0.058824   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                     NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_five_prime_percentage  \\\n",
       "event_id                                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                0.037175   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                     NaN   \n",
       "\n",
       "                                                    isoform1_rnafold_hairpin_percentage  \\\n",
       "event_id                                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                             0.105882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                  NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_hairpin_percentage  \\\n",
       "event_id                                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                             0.104089   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                             0.118644   \n",
       "\n",
       "                                                    isoform1_rnafold_interior_loop_percentage  \\\n",
       "event_id                                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                   0.205882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                        NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_interior_loop_percentage  \\\n",
       "event_id                                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                   0.223048   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                   0.096852   \n",
       "\n",
       "                                                    isoform1_rnafold_minimum_free_energy  \\\n",
       "event_id                                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                 -47.1   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                   NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_minimum_free_energy  \\\n",
       "event_id                                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                 -95.9   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                -183.1   \n",
       "\n",
       "                                                    isoform1_rnafold_multiloop_percentage  \\\n",
       "event_id                                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                               0.105882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                    NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_multiloop_percentage  \\\n",
       "event_id                                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                               0.052045   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                               0.111380   \n",
       "\n",
       "                                                    isoform1_rnafold_stem_percentage  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                          0.505882   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                               NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_stem_percentage  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                          0.572491   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                          0.658596   \n",
       "\n",
       "                                                    isoform1_rnafold_three_prime_percentage  \\\n",
       "event_id                                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                 0.017647   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                      NaN   \n",
       "\n",
       "                                                    isoform2_rnafold_three_prime_percentage  \n",
       "event_id                                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                 0.011152  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                 0.014528  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structure_df_2d.columns = ['_'.join(reversed(col)).strip() for col in structure_df_2d.columns.values]\n",
    "structure_df_2d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'..........(((((....((((.((((...((((((((.(((....(((((...))))).............))).))).))))).))..........(((((((....))))))).((((((..........(((((...(((..(((((((((......))))..)))))...))).)))))))))))(((.((.(((((.((....))))))).)).)))((((((...........))))))...)).))))....)))))...'"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = '..........(((((....((((.((((...((((((((.(((....(((((...))))).............))).))).))))).))..........(((((((....))))))).((((((..........(((((...(((..(((((((((......))))..)))))...))).)))))))))))(((.((.(((((.((....))))))).)).)))((((((...........))))))...)).))))....)))))... (-95.90)'\n",
    "structure, mfe = s.split()\n",
    "structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ffffffffffsssssiiiissssissssiiissssssssisssiiiissssshhhsssssiiiiiiiiiiiiisssisssisssssissmmmmmmmmmmssssssshhhhsssssssmssssssiiiiiiiiiisssssiiisssiissssssssshhhhhhssssiisssssiiisssissssssssssssssississsssisshhhhsssssssississssssssshhhhhhhhhhhssssssmmmssissssiiiisssssttt'"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import forgi.graph.bulge_graph as fgb\n",
    "structure = '..........(((((....((((.((((...((((((((.(((....(((((...))))).............))).))).))))).))..........(((((((....))))))).((((((..........(((((...(((..(((((((((......))))..)))))...))).)))))))))))(((.((.(((((.((....))))))).)).)))((((((...........))))))...)).))))....)))))...'\n",
    "bg = fgb.BulgeGraph(dotbracket_str=structure)\n",
    "bg.to_element_string()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bg??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pprint import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "five_prime        10\n",
       "hairpin           28\n",
       "interior_loop     60\n",
       "multiloop         14\n",
       "stem             154\n",
       "three_prime        3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "element_counts = pd.Series(Counter(bg.to_element_string()))\n",
    "element_counts.rename(acronym_to_full)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'s0': 5,\n",
       " 's1': 4,\n",
       " 's10': 5,\n",
       " 's11': 3,\n",
       " 's12': 5,\n",
       " 's13': 4,\n",
       " 's14': 3,\n",
       " 's15': 2,\n",
       " 's16': 5,\n",
       " 's17': 2,\n",
       " 's18': 6,\n",
       " 's2': 2,\n",
       " 's3': 2,\n",
       " 's4': 5,\n",
       " 's5': 3,\n",
       " 's6': 3,\n",
       " 's7': 5,\n",
       " 's8': 7,\n",
       " 's9': 6}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict((s, bg.stem_length(s)) for s in bg.stem_iterator())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 obotvinnik  404 Oct 20 15:22 RNAfold.sh\r\n",
      "-rw------- 1 obotvinnik 4.0K Oct 21 21:37 RNAfold.sh.err\r\n",
      "-rw------- 1 obotvinnik   24 Oct 21 21:41 RNAfold.sh.out\r\n"
     ]
    }
   ],
   "source": [
    "ll -lha RNAfold.sh*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning from traverse_loop. Loop 18 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 18 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 6 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 8 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 147 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 142 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 7 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 7 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 2 has crossed regions\r\n",
      "Warning from traverse_loop. Loop 24 has crossed regions\r\n"
     ]
    }
   ],
   "source": [
    "! tail RNAfold.sh.err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 obotvinnik yeo-group 40M Oct 20 11:28 /projects/ps-yeolab/obotvinnik/singlecell_pnms/isoform_transcriptions.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls -lha $transcribed_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 obotvinnik yeo-group 2.2M Oct 20 15:59 /projects/ps-yeolab/obotvinnik/singlecell_pnms/isoform_transcriptions_rnafold.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls -lha $rnafold_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/home/obotvinnik/Dropbox/figures2/singlecell_pnms/isoform_rna_properties': File exists\r\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCACAGUGCCAGCUCCCUGCGCCCGGCCGACCUGCUUGCCCUCAUCCUCCUGGUUCAGGACCUCUACCCCAGCGAGAGCACAGCAGAGGACGACAUUCAGCCUUCCCCGCGGAGGGCCCGGAGCAGCCAGAACAUCCCCGUGCAGCAGGCCUGGAGCCCUCACUCCACGGGCCCAACUGGGGGGAGCUCUGCAGAGACGGAGACAGACAGCUUCUCCCUCCCUGAGGAGUACUUCACACCAGCUCCUUCCCCUGGCGAUCAGAGCUCAG..........(((((....((((.((((...((((((((.(((....(((((...))))).............))).))).))))).))..........(((((((....))))))).((((((..........(((((...(((..(((((((((......))))..)))))...))).)))))))))))(((.((.(((((.((....))))))).)).)))((((((...........))))))...)).))))....)))))...(-95.90)\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'acronym_to_full' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-8-30e3c24e6647>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     16\u001b[0m     \u001b[0mbg\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfgb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mBulgeGraph\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdotbracket_str\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mstructure\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     17\u001b[0m     \u001b[0melement_counts\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mCounter\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mbg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_element_string\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 18\u001b[1;33m     \u001b[0melement_counts\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0melement_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0macronym_to_full\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     19\u001b[0m     \u001b[0melement_percentage\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0melement_counts\u001b[0m\u001b[1;33m/\u001b[0m\u001b[0melement_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msum\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     20\u001b[0m     \u001b[0melement_percentage\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0melement_percentage\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mx\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'_percentage'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'acronym_to_full' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!head $transcribed_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNAfold 2.1.8\n",
      "\n",
      "Calculate minimum free energy secondary structures and partition function of \n",
      "RNAs\n",
      "\n",
      "Usage: RNAfold [OPTIONS]...\n",
      "\n",
      "The program reads RNA sequences from stdin, calculates their minimum free \n",
      "energy (mfe) structure and prints to stdout the mfe structure in bracket \n",
      "notation and its free energy. If the -p option was given it also computes the \n",
      "partition function (pf) and base pairing probability matrix, and prints the \n",
      "free energy of the thermodynamic ensemble, the frequency of the mfe structure \n",
      "in the ensemble, and the ensemble diversity to stdout.\n",
      "\n",
      "It also produces PostScript files with plots of the resulting secondary \n",
      "structure graph and a \"dot plot\" of the base pairing matrix.\n",
      "The dot plot shows a matrix of squares with area proportional to the pairing \n",
      "probability in the upper right half, and one square for each pair in the \n",
      "minimum free energy structure in the lower left half. For each pair i-j with \n",
      "probability p>10E-6 there is a line of the form\n",
      "\n",
      "i  j  sqrt(p)  ubox\n",
      "\n",
      "in the PostScript file, so that the pair probabilities can be easily extracted.\n",
      "\n",
      "Sequences may be provided in a simple text format where each sequence occupies \n",
      "a single line. Output files are named \"rna.ps\" and \"dot.ps\". Existing files \n",
      "of the same name will be overwritten.\n",
      "It is also possible to provide sequence data in FASTA format. In this case, the \n",
      "first word (max. 42 char) of the FASTA header will be used for output file \n",
      "names. PostScript files \"name_ss.ps\" and \"name_dp.ps\" are produced for the \n",
      "structure and dot plot, respectively.\n",
      "Once FASTA input was provided all following sequences must be in FASTA format \n",
      "too.\n",
      "The program will continue to read new sequences until a line consisting of the \n",
      "single character @ or an end of file condition is encountered.\n",
      "\n",
      "\n",
      "\n",
      "  -h, --help                    Print help and exit\n",
      "      --detailed-help           Print help, including all details and hidden \n",
      "                                  options, and exit\n",
      "      --full-help               Print help, including hidden options, and exit\n",
      "  -V, --version                 Print version and exit\n",
      "\n",
      "General Options:\n",
      "  Below are command line options which alter the general behavior of this \n",
      "  program\n",
      "\n",
      "\n",
      "  -C, --constraint              Calculate structures subject to constraints.\n",
      "                                    (default=off)\n",
      "      --canonicalBPonly         Remove non-canonical base pairs from the \n",
      "                                  structure constraint\n",
      "\n",
      "                                    (default=off)\n",
      "      --noconv                  Do not automatically substitude nucleotide \n",
      "                                  \"T\" with \"U\"\n",
      "\n",
      "                                    (default=off)\n",
      "      --noPS                    Do not produce postscript drawing of the mfe \n",
      "                                  structure.\n",
      "\n",
      "                                    (default=off)\n",
      "\n",
      "Algorithms:\n",
      "  Select additional algorithms which should be included in the calculations.\n",
      "  The Minimum free energy (MFE) and a structure representative are calculated \n",
      "  in any case.\n",
      "\n",
      "\n",
      "  -p, --partfunc[=INT]          Calculate the partition function and base \n",
      "                                  pairing probability matrix.\n",
      "                                    (default=`1')\n",
      "      --MEA[=gamma]             Calculate an MEA (maximum expected accuracy) \n",
      "                                  structure, where the expected accuracy is \n",
      "                                  computed from the pair probabilities: each \n",
      "                                  base pair (i,j) gets a score 2*gamma*p_ij and \n",
      "                                  the score of an unpaired base is given by the \n",
      "                                  probability of not forming a pair.\n",
      "                                    (default=`1.')\n",
      "  -c, --circ                    Assume a circular (instead of linear) RNA \n",
      "                                  molecule.\n",
      "                                    (default=off)\n",
      "  -g, --gquad                   Incoorporate G-Quadruplex formation into the \n",
      "                                  structure prediction algorithm\n",
      "                                    (default=off)\n",
      "\n",
      "Model Details:\n",
      "  -T, --temp=DOUBLE             Rescale energy parameters to a temperature of \n",
      "                                  temp C. Default is 37C.\n",
      "\n",
      "\n",
      "  -4, --noTetra                 Do not include special tabulated stabilizing \n",
      "                                  energies for tri-, tetra- and hexaloop \n",
      "                                  hairpins. Mostly for testing.\n",
      "\n",
      "                                    (default=off)\n",
      "  -d, --dangles=INT             How to treat \"dangling end\" energies for \n",
      "                                  bases adjacent to helices in free ends and \n",
      "                                  multi-loops\n",
      "                                    (default=`2')\n",
      "      --noLP                    Produce structures without lonely pairs \n",
      "                                  (helices of length 1).\n",
      "                                    (default=off)\n",
      "      --noGU                    Do not allow GU pairs\n",
      "\n",
      "                                    (default=off)\n",
      "      --noClosingGU             Do not allow GU pairs at the end of helices\n",
      "\n",
      "                                    (default=off)\n",
      "  -P, --paramFile=paramfile     Read energy parameters from paramfile, instead \n",
      "                                  of using the default parameter set.\n",
      "\n",
      "\n",
      "If in doubt our program is right, nature is at fault.\n",
      "Comments should be sent to rna@tbi.univie.ac.at.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! RNAfold -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/home/obotvinnik/Dropbox/figures2/singlecell_pnms/isoform_rna_properties': File exists\r\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\r\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:tornado.general:Uncaught exception, closing connection.\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py\", line 414, in _run_callback\n",
      "    callback(*args, **kwargs)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/tornado/stack_context.py\", line 275, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/kernelbase.py\", line 260, in dispatcher\n",
      "    return self.dispatch_shell(stream, msg)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/kernelbase.py\", line 212, in dispatch_shell\n",
      "    handler(stream, idents, msg)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/kernelbase.py\", line 391, in execute_request\n",
      "    ident=ident)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/jupyter_client/session.py\", line 645, in send\n",
      "    if not isinstance(stream, zmq.Socket):\n",
      "KeyboardInterrupt\n",
      "ERROR:tornado.general:Uncaught exception, closing connection.\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py\", line 440, in _handle_events\n",
      "    self._handle_recv()\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py\", line 472, in _handle_recv\n",
      "    self._run_callback(callback, msg)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py\", line 414, in _run_callback\n",
      "    callback(*args, **kwargs)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/tornado/stack_context.py\", line 275, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/kernelbase.py\", line 260, in dispatcher\n",
      "    return self.dispatch_shell(stream, msg)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/kernelbase.py\", line 212, in dispatch_shell\n",
      "    handler(stream, idents, msg)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/kernelbase.py\", line 391, in execute_request\n",
      "    ident=ident)\n",
      "  File \"/home/obotvinnik/anaconda/lib/python2.7/site-packages/jupyter_client/session.py\", line 645, in send\n",
      "    if not isinstance(stream, zmq.Socket):\n",
      "KeyboardInterrupt\n"
     ]
    }
   ],
   "source": [
    "! head $transcribed_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.options.display.max_columns = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4</th>\n",
       "      <th>exon4_length</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>junction_exons12</th>\n",
       "      <th>junction_exons13</th>\n",
       "      <th>junction_exons23</th>\n",
       "      <th>junction_exons24</th>\n",
       "      <th>junction_exons34</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "      <th>splice_type</th>\n",
       "      <th>exon2_merkin2012_ancient</th>\n",
       "      <th>isoform1_gc</th>\n",
       "      <th>isoform1_gc_position1</th>\n",
       "      <th>isoform1_gc_position2</th>\n",
       "      <th>isoform1_gc_position3</th>\n",
       "      <th>isoform2_gc</th>\n",
       "      <th>isoform2_gc_position1</th>\n",
       "      <th>isoform2_gc_position2</th>\n",
       "      <th>isoform2_gc_position3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>928</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "      <td>False</td>\n",
       "      <td>75.714286</td>\n",
       "      <td>87.500000</td>\n",
       "      <td>77.551020</td>\n",
       "      <td>60.465116</td>\n",
       "      <td>77.292576</td>\n",
       "      <td>89.610390</td>\n",
       "      <td>73.076923</td>\n",
       "      <td>68.918919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>3269</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "      <td>False</td>\n",
       "      <td>79.523810</td>\n",
       "      <td>88.311688</td>\n",
       "      <td>74.242424</td>\n",
       "      <td>74.626866</td>\n",
       "      <td>79.178886</td>\n",
       "      <td>73.333333</td>\n",
       "      <td>89.256198</td>\n",
       "      <td>73.913043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>1543</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "      <td>False</td>\n",
       "      <td>78.111588</td>\n",
       "      <td>81.927711</td>\n",
       "      <td>73.239437</td>\n",
       "      <td>78.481013</td>\n",
       "      <td>75.882353</td>\n",
       "      <td>75.206612</td>\n",
       "      <td>53.684211</td>\n",
       "      <td>93.548387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>166</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>2121</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>SE</td>\n",
       "      <td>False</td>\n",
       "      <td>68.899522</td>\n",
       "      <td>62.318841</td>\n",
       "      <td>70.588235</td>\n",
       "      <td>73.611111</td>\n",
       "      <td>71.246006</td>\n",
       "      <td>70.642202</td>\n",
       "      <td>58.585859</td>\n",
       "      <td>83.809524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+</th>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198018.6</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>1857</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>SE</td>\n",
       "      <td>False</td>\n",
       "      <td>74.111675</td>\n",
       "      <td>74.242424</td>\n",
       "      <td>76.271186</td>\n",
       "      <td>72.222222</td>\n",
       "      <td>70.833333</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>71.276596</td>\n",
       "      <td>75.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    criteria  \\\n",
       "event_id                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      best   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      best   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  only one   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                 NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                 NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                 NaN   \n",
       "\n",
       "                                                            criteria_full  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...               only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...               only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...               only one   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                                               exon1  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195392-100195529:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101165513-101165617:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           105   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...            83   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189548-100189646:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190888-100191048:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195029-100195171:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163481-101163631:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419619-101419721:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           151   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           103   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189330-100189399:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163226-101163391:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101421203-101421385:+   \n",
       "\n",
       "                                                    exon3_length exon4  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           166   NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           183   NaN   \n",
       "\n",
       "                                                    exon4_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           NaN   \n",
       "\n",
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...   ENSG00000120053.9   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...   ENSG00000198018.6   \n",
       "\n",
       "                                                   gene_name  intron_length  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      HPS1            928   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...      HPS1           3269   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      HPS1           1543   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...      GOT1           2121   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...    ENTPD7           1857   \n",
       "\n",
       "                                                               junction_exons12  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189647-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100191049-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100195172-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163632-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101419618:+   \n",
       "\n",
       "                                                               junction_exons13  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101421202:+   \n",
       "\n",
       "                                                               junction_exons23  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100189547:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100190887:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195028:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101163480:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419722-101421202:+   \n",
       "\n",
       "                                                   junction_exons24  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              NaN   \n",
       "\n",
       "                                                   junction_exons34 strand  \\\n",
       "event_id                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              NaN      -   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              NaN      -   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              NaN      -   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              NaN      -   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              NaN      +   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                False   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                False   \n",
       "\n",
       "                                                     one_ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                   splice_type  \\\n",
       "event_id                                                         \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...          SE   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...          SE   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...          SE   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...          SE   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...          SE   \n",
       "\n",
       "                                                   exon2_merkin2012_ancient  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                    False   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                    False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                    False   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                    False   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                    False   \n",
       "\n",
       "                                                    isoform1_gc  \\\n",
       "event_id                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    75.714286   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...    79.523810   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    78.111588   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...    68.899522   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...    74.111675   \n",
       "\n",
       "                                                    isoform1_gc_position1  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              87.500000   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              88.311688   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              81.927711   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              62.318841   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              74.242424   \n",
       "\n",
       "                                                    isoform1_gc_position2  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              77.551020   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              74.242424   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              73.239437   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              70.588235   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              76.271186   \n",
       "\n",
       "                                                    isoform1_gc_position3  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              60.465116   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              74.626866   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              78.481013   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              73.611111   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              72.222222   \n",
       "\n",
       "                                                    isoform2_gc  \\\n",
       "event_id                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    77.292576   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...    79.178886   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    75.882353   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...    71.246006   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...    70.833333   \n",
       "\n",
       "                                                    isoform2_gc_position1  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              89.610390   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              73.333333   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              75.206612   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              70.642202   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              66.666667   \n",
       "\n",
       "                                                    isoform2_gc_position2  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              73.076923   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              89.256198   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              53.684211   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              58.585859   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              71.276596   \n",
       "\n",
       "                                                    isoform2_gc_position3  \n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              68.918919  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...              73.913043  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...              93.548387  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              83.809524  \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...              75.000000  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "join_cols = gc_content_2d.columns.difference(splicing_feature_data.columns)\n",
    "\n",
    "if len(join_cols) > 0:\n",
    "    splicing_feature_data = splicing_feature_data.join(gc_content_2d[join_cols])\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "splicing_feature_data.to_csv('{}/splicing_feature_data.csv'.format(folder))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Calculate miRNA hybridization\n",
    "\n",
    "For RNA targets, James Broughton from the Pasquinelli lab recommends `RNAhybrid`, and to use just the first 17 nt of the mature miRNA. Wanted to use `fastx-trimmer` but it only takes DNA sequences  - `U`'s are illegal :("
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import sys\n",
    "import os\n",
    "\n",
    "result_seq = []\n",
    "filename = '/projects/ps-yeolab/genomes/mirbase/release_21/human_mature.fa'\n",
    "with open(filename) as infile:\n",
    "    for seq in SeqIO.parse(infile, 'fasta'):\n",
    "        result_seq.append(seq[:17])\n",
    "\n",
    "trimmed_filename = '/projects/ps-yeolab/genomes/mirbase/release_21/human_mature_17bp.fa'\n",
    "with open(trimmed_filename, 'w') as outfile:\n",
    "    SeqIO.write(result_seq, outfile, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit a compute job to calculate microRNA hybridization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "job ID: 3598068\n"
     ]
    }
   ],
   "source": [
    "from gscripts.qtools import Submitter\n",
    "\n",
    "mirna_seqs = '/projects/ps-yeolab/genomes/mirbase/release_21/human_mature_17bp.fa'\n",
    "rnahybrid_results = '/projects/ps-yeolab/obotvinnik/miso_helpers/hg19/se_exon2_RNAhybrid_mirbase_human_mature_17bp.txt'\n",
    "command = 'RNAhybrid -e -27 -c -s 3utr_human -q {} -t {} > {}'.format(mirna_seqs, transcribed_fasta, rnahybrid_results)\n",
    "sub = Submitter([command], 'RNAhybrid', walltime='168:00:00', write_and_submit=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reading the output takes a LONG time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.53 s, sys: 110 ms, total: 1.64 s\n",
      "Wall time: 2.3 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "rnahybrid = pd.read_csv('/projects/ps-yeolab/obotvinnik/miso_helpers/hg19/se_exon2_RNAhybrid_mirbase_human_mature_17bp.txt', \n",
    "                        sep=':', \n",
    "#                         index_col=[0, 1, 3], \n",
    "                        header=None,\n",
    "                        names=['chrom', 'start-stop', 'exon_length', 'mirna', 'mirna_length',\n",
    "                                                                 'minimum_free_energy', 'p_value', 'target_bound_start', 'mirna_unbound',\n",
    "                                                                 'mirna_bound', 'exon_bound', 'exon_bound'])\n",
    "rnahybrid = rnahybrid.ix[rnahybrid.minimum_free_energy < -28]\n",
    "grouped = rnahybrid.groupby(['chrom', 'start-stop'])\n",
    "chrom_startstop_mirna = grouped.mirna.apply(lambda x: ','.join('{}[{}]'.format(k, v) for k,v in Counter(x).items()))\n",
    "chrom_startstop_mirna.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
