{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40534, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:810: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction23</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    junction13                   junction12  \\\n",
       "0  chr10:100189400-100190327:-  chr10:100189647-100190327:-   \n",
       "1  chr10:100190428-100193696:-  chr10:100191049-100193696:-   \n",
       "2  chr10:100193849-100195391:-  chr10:100195172-100195391:-   \n",
       "3  chr10:101163392-101165512:-  chr10:101163632-101165512:-   \n",
       "4  chr10:101419346-101421202:+  chr10:101419346-101419618:+   \n",
       "\n",
       "                    junction23          criteria_full  \\\n",
       "0  chr10:100189400-100189547:-  best,appris_principal   \n",
       "1  chr10:100190428-100190887:-               only one   \n",
       "2  chr10:100193849-100195028:-  best,appris_principal   \n",
       "3  chr10:101163392-101163480:-               only one   \n",
       "4  chr10:101419722-101421202:+               only one   \n",
       "\n",
       "                                            event_id  criteria  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...      best   \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...  only one   \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...      best   \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...  only one   \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...  only one   \n",
       "\n",
       "  criteria_additional  \n",
       "0    appris_principal  \n",
       "1                 NaN  \n",
       "2    appris_principal  \n",
       "3                 NaN  \n",
       "4                 NaN  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import itertools\n",
    "import re\n",
    "\n",
    "import gffutils\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import six\n",
    "\n",
    "v19db_filename = '/projects/ps-yeolab/genomes/hg19/gencode/v19/gencode.v19.annotation.gtf.db'\n",
    "v19db = gffutils.FeatureDB(v19db_filename)\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "outrigger_folder = '{}/outrigger_index'.format(folder)\n",
    "csv_folder = '{}/csvs_for_paper/'.format(folder)\n",
    "\n",
    "# se_df = pd.read_csv('{}/exons_to_junctions_se.csv'.format(folder), index_col=0)\n",
    "se_df = pd.read_csv('{}/events/se.csv'.format(outrigger_folder), index_col=0)\n",
    "\n",
    "# renamer = dict((x, x.replace('junction_exon', 'junction_exons')) for x in se_df)\n",
    "# print renamer\n",
    "# se_df = se_df.rename(columns=renamer)\n",
    "print se_df.shape\n",
    "se_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Double check that there are no splicing events which share junctions but not flanking exons. I.e. if you group by the junctions, you should only get one event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped = se_df.groupby(['junction13', 'junction12', 'junction23'])\n",
    "size = grouped.size()\n",
    "(size != 1).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(242, 26379)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing = pd.read_csv('{}/splicing.csv'.format(csv_folder), index_col=0)\n",
    "splicing.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40534, 10)\n",
      "(23314, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction23</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    junction13                   junction12  \\\n",
       "0  chr10:100189400-100190327:-  chr10:100189647-100190327:-   \n",
       "1  chr10:100190428-100193696:-  chr10:100191049-100193696:-   \n",
       "2  chr10:100193849-100195391:-  chr10:100195172-100195391:-   \n",
       "3  chr10:101163392-101165512:-  chr10:101163632-101165512:-   \n",
       "4  chr10:101419346-101421202:+  chr10:101419346-101419618:+   \n",
       "\n",
       "                    junction23          criteria_full  \\\n",
       "0  chr10:100189400-100189547:-  best,appris_principal   \n",
       "1  chr10:100190428-100190887:-               only one   \n",
       "2  chr10:100193849-100195028:-  best,appris_principal   \n",
       "3  chr10:101163392-101163480:-               only one   \n",
       "4  chr10:101419722-101421202:+               only one   \n",
       "\n",
       "                                            event_id  criteria  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...      best   \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...  only one   \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...      best   \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...  only one   \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...  only one   \n",
       "\n",
       "  criteria_additional                             exon1  \\\n",
       "0    appris_principal  exon:chr10:100190328-100190427:-   \n",
       "1                 NaN  exon:chr10:100193697-100193848:-   \n",
       "2    appris_principal  exon:chr10:100195392-100195529:-   \n",
       "3                 NaN  exon:chr10:101165513-101165617:-   \n",
       "4                 NaN  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                              exon2                             exon3  \n",
       "0  exon:chr10:100189548-100189646:-  exon:chr10:100189330-100189399:-  \n",
       "1  exon:chr10:100190888-100191048:-  exon:chr10:100190328-100190427:-  \n",
       "2  exon:chr10:100195029-100195171:-  exon:chr10:100193697-100193848:-  \n",
       "3  exon:chr10:101163481-101163631:-  exon:chr10:101163226-101163391:-  \n",
       "4  exon:chr10:101419619-101419721:+  exon:chr10:101421203-101421385:+  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_exons = se_df.event_id.str.split('@').apply(pd.Series).rename(columns=dict((i, 'exon{}'.format(i+1)) for i in range(3)))\n",
    "se_df = pd.concat([se_df, se_exons], axis=1)\n",
    "six.print_(se_df.shape)\n",
    "se_df = se_df.loc[se_df.event_id.isin(splicing.columns)]\n",
    "six.print_(se_df.shape)\n",
    "se_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MXE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13217, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr10:101996727-102003454:-</td>\n",
       "      <td>chr10:101995522-101996606:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:101995522-101997778:-</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr10:102003455-102003534:-@exon:chr10:10...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr10:102003535-102013177:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:102010090-102013177:-</td>\n",
       "      <td>chr10:101997989-102010004:-</td>\n",
       "      <td>random,no gencode transcripts</td>\n",
       "      <td>exon:chr10:102013178-102013296:-@exon:chr10:10...</td>\n",
       "      <td>random</td>\n",
       "      <td>no gencode transcripts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr10:102016234-102020722:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102020722:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>random,no good tags</td>\n",
       "      <td>exon:chr10:102020723-102020801:-@exon:chr10:10...</td>\n",
       "      <td>random</td>\n",
       "      <td>no good tags</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr10:102016234-102021735:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102021735:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:102021736-102021820:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr10:102016234-102021735:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020786-102021735:-</td>\n",
       "      <td>chr10:102013297-102020722:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:102021736-102021820:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                   junction13                   junction34  \\\n",
       "0           0  chr10:101996727-102003454:-  chr10:101995522-101996606:-   \n",
       "1           1  chr10:102003535-102013177:-  chr10:101997989-102003454:-   \n",
       "2           2  chr10:102016234-102020722:-  chr10:102013297-102016018:-   \n",
       "3           3  chr10:102016234-102021735:-  chr10:102013297-102016018:-   \n",
       "4           4  chr10:102016234-102021735:-  chr10:102013297-102016018:-   \n",
       "\n",
       "                    junction12                   junction24  \\\n",
       "0  chr10:101997989-102003454:-  chr10:101995522-101997778:-   \n",
       "1  chr10:102010090-102013177:-  chr10:101997989-102010004:-   \n",
       "2  chr10:102020077-102020722:-  chr10:102013297-102019974:-   \n",
       "3  chr10:102020077-102021735:-  chr10:102013297-102019974:-   \n",
       "4  chr10:102020786-102021735:-  chr10:102013297-102020722:-   \n",
       "\n",
       "                        criteria_full  \\\n",
       "0  one event with gencode transcripts   \n",
       "1       random,no gencode transcripts   \n",
       "2                 random,no good tags   \n",
       "3                            only one   \n",
       "4                            only one   \n",
       "\n",
       "                                            event_id  \\\n",
       "0  exon:chr10:102003455-102003534:-@exon:chr10:10...   \n",
       "1  exon:chr10:102013178-102013296:-@exon:chr10:10...   \n",
       "2  exon:chr10:102020723-102020801:-@exon:chr10:10...   \n",
       "3  exon:chr10:102021736-102021820:-@exon:chr10:10...   \n",
       "4  exon:chr10:102021736-102021820:-@exon:chr10:10...   \n",
       "\n",
       "                             criteria     criteria_additional  \n",
       "0  one event with gencode transcripts                     NaN  \n",
       "1                              random  no gencode transcripts  \n",
       "2                              random            no good tags  \n",
       "3                            only one                     NaN  \n",
       "4                            only one                     NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxe_df = pd.read_csv('{}/events/mxe.csv'.format(outrigger_folder))\n",
    "print mxe_df.shape\n",
    "mxe_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr10:101996727-102003454:-</td>\n",
       "      <td>chr10:101995522-101996606:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:101995522-101997778:-</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr10:102003455-102003534:-@exon:chr10:10...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr10:102003535-102013177:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:102010090-102013177:-</td>\n",
       "      <td>chr10:101997989-102010004:-</td>\n",
       "      <td>random,no gencode transcripts</td>\n",
       "      <td>exon:chr10:102013178-102013296:-@exon:chr10:10...</td>\n",
       "      <td>random</td>\n",
       "      <td>no gencode transcripts</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>exon:chr10:102010005-102010089:-</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997940-101997988:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr10:102016234-102020722:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102020722:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>random,no good tags</td>\n",
       "      <td>exon:chr10:102020723-102020801:-@exon:chr10:10...</td>\n",
       "      <td>random</td>\n",
       "      <td>no good tags</td>\n",
       "      <td>exon:chr10:102020723-102020801:-</td>\n",
       "      <td>exon:chr10:102019975-102020076:-</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr10:102016234-102021735:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102021735:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:102021736-102021820:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:102021736-102021820:-</td>\n",
       "      <td>exon:chr10:102019975-102020076:-</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr10:102016234-102021735:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020786-102021735:-</td>\n",
       "      <td>chr10:102013297-102020722:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:102021736-102021820:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:102021736-102021820:-</td>\n",
       "      <td>exon:chr10:102020723-102020785:-</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                   junction13                   junction34  \\\n",
       "0           0  chr10:101996727-102003454:-  chr10:101995522-101996606:-   \n",
       "1           1  chr10:102003535-102013177:-  chr10:101997989-102003454:-   \n",
       "2           2  chr10:102016234-102020722:-  chr10:102013297-102016018:-   \n",
       "3           3  chr10:102016234-102021735:-  chr10:102013297-102016018:-   \n",
       "4           4  chr10:102016234-102021735:-  chr10:102013297-102016018:-   \n",
       "\n",
       "                    junction12                   junction24  \\\n",
       "0  chr10:101997989-102003454:-  chr10:101995522-101997778:-   \n",
       "1  chr10:102010090-102013177:-  chr10:101997989-102010004:-   \n",
       "2  chr10:102020077-102020722:-  chr10:102013297-102019974:-   \n",
       "3  chr10:102020077-102021735:-  chr10:102013297-102019974:-   \n",
       "4  chr10:102020786-102021735:-  chr10:102013297-102020722:-   \n",
       "\n",
       "                        criteria_full  \\\n",
       "0  one event with gencode transcripts   \n",
       "1       random,no gencode transcripts   \n",
       "2                 random,no good tags   \n",
       "3                            only one   \n",
       "4                            only one   \n",
       "\n",
       "                                            event_id  \\\n",
       "0  exon:chr10:102003455-102003534:-@exon:chr10:10...   \n",
       "1  exon:chr10:102013178-102013296:-@exon:chr10:10...   \n",
       "2  exon:chr10:102020723-102020801:-@exon:chr10:10...   \n",
       "3  exon:chr10:102021736-102021820:-@exon:chr10:10...   \n",
       "4  exon:chr10:102021736-102021820:-@exon:chr10:10...   \n",
       "\n",
       "                             criteria     criteria_additional  \\\n",
       "0  one event with gencode transcripts                     NaN   \n",
       "1                              random  no gencode transcripts   \n",
       "2                              random            no good tags   \n",
       "3                            only one                     NaN   \n",
       "4                            only one                     NaN   \n",
       "\n",
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:102003455-102003534:-  exon:chr10:101997779-101997988:-   \n",
       "1  exon:chr10:102013178-102013296:-  exon:chr10:102010005-102010089:-   \n",
       "2  exon:chr10:102020723-102020801:-  exon:chr10:102019975-102020076:-   \n",
       "3  exon:chr10:102021736-102021820:-  exon:chr10:102019975-102020076:-   \n",
       "4  exon:chr10:102021736-102021820:-  exon:chr10:102020723-102020785:-   \n",
       "\n",
       "                              exon3                             exon4  \n",
       "0  exon:chr10:101996607-101996726:-  exon:chr10:101995424-101995521:-  \n",
       "1  exon:chr10:102003455-102003534:-  exon:chr10:101997940-101997988:-  \n",
       "2  exon:chr10:102016019-102016233:-  exon:chr10:102013178-102013296:-  \n",
       "3  exon:chr10:102016019-102016233:-  exon:chr10:102013178-102013296:-  \n",
       "4  exon:chr10:102016019-102016233:-  exon:chr10:102013178-102013296:-  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxe_exons = mxe_df.event_id.str.split('@').apply(pd.Series).rename(columns=dict((i, 'exon{}'.format(i+1)) for i in range(4)))\n",
    "mxe_df = pd.concat([mxe_df, mxe_exons], axis=1)\n",
    "mxe_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from outrigger.region import Region\n",
    "\n",
    "def get_gene_attributes(df, exon_cols, db):\n",
    "    df['gencode_id'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[db[i].attributes['gene_id'] for i in x]))), axis=1)\n",
    "    df['transcript_id'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[db[i].attributes['transcript_id'] for i in x]))), axis=1)\n",
    "    df['gene_name'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[db[i].attributes['gene_name'] for i in x]))), axis=1)\n",
    "    df['ensembl_id'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[map(lambda y: y.split('.')[0], \n",
    "                                                     db[i].attributes['gene_id']) for i in x]))), \n",
    "        axis=1)\n",
    "    \n",
    "    for exon_col in exon_cols:\n",
    "        df['{}_region'.format(exon_col)] = df[exon_col].map(Region)\n",
    "        df['{}_length'.format(exon_col)] = df['{}_region'.format(exon_col)].map(len)\n",
    "    df['strand'] = df[exon_cols[0]].str[-1]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate events with gene ids and exon lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['exon1', 'exon2', 'exon3', 'exon4']\n"
     ]
    }
   ],
   "source": [
    "mxe_exon_cols = ['exon{}'.format(i) for i in range(1, 5)]\n",
    "print(mxe_exon_cols)\n",
    "mxe_df_annotated = get_gene_attributes(mxe_df, mxe_exon_cols, v19db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['exon1', 'exon2', 'exon3']\n"
     ]
    }
   ],
   "source": [
    "se_exon_cols = ['exon{}'.format(i) for i in range(1, 4)]\n",
    "print(se_exon_cols)\n",
    "se_df_annotated = get_gene_attributes(se_df, se_exon_cols, v19db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spot-check some genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>...</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4_region</th>\n",
       "      <th>exon4_length</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2294</th>\n",
       "      <td>2294</td>\n",
       "      <td>chr15:72494962-72499068:-</td>\n",
       "      <td>chr15:72492997-72494794:-</td>\n",
       "      <td>chr15:72495530-72499068:-</td>\n",
       "      <td>chr15:72492997-72495362:-</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr15:72499069-72499221:-@exon:chr15:7249...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr15:72499069-72499221:-</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067225</td>\n",
       "      <td>exon:chr15:72499069-72499221:-</td>\n",
       "      <td>153</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>182</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0                 junction13                 junction34  \\\n",
       "2294        2294  chr15:72494962-72499068:-  chr15:72492997-72494794:-   \n",
       "\n",
       "                     junction12                 junction24  \\\n",
       "2294  chr15:72495530-72499068:-  chr15:72492997-72495362:-   \n",
       "\n",
       "                           criteria_full  \\\n",
       "2294  one event with gencode transcripts   \n",
       "\n",
       "                                               event_id  \\\n",
       "2294  exon:chr15:72499069-72499221:-@exon:chr15:7249...   \n",
       "\n",
       "                                criteria criteria_additional  \\\n",
       "2294  one event with gencode transcripts                 NaN   \n",
       "\n",
       "                               exon1  ...         ensembl_id  \\\n",
       "2294  exon:chr15:72499069-72499221:-  ...    ENSG00000067225   \n",
       "\n",
       "                        exon1_region exon1_length  \\\n",
       "2294  exon:chr15:72499069-72499221:-          153   \n",
       "\n",
       "                        exon2_region exon2_length  \\\n",
       "2294  exon:chr15:72495363-72495529:-          167   \n",
       "\n",
       "                        exon3_region exon3_length  \\\n",
       "2294  exon:chr15:72494795-72494961:-          167   \n",
       "\n",
       "                        exon4_region  exon4_length strand  \n",
       "2294  exon:chr15:72492815-72492996:-           182      -  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxe_df_annotated.loc[mxe_df_annotated.gene_name == 'PKM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction23</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1351</th>\n",
       "      <td>chr10:79797063-79799961:+</td>\n",
       "      <td>chr10:79797063-79797722:+</td>\n",
       "      <td>chr10:79797741-79799961:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:79796952-79797062:+@exon:chr10:7979...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:79796952-79797062:+</td>\n",
       "      <td>exon:chr10:79797723-79797740:+</td>\n",
       "      <td>exon:chr10:79799962-79799983:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000485708.2,ENST00000476545.1,ENST000003...</td>\n",
       "      <td>RPS24</td>\n",
       "      <td>ENSG00000138326</td>\n",
       "      <td>exon:chr10:79796952-79797062:+</td>\n",
       "      <td>111</td>\n",
       "      <td>exon:chr10:79797723-79797740:+</td>\n",
       "      <td>18</td>\n",
       "      <td>exon:chr10:79799962-79799983:+</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1352</th>\n",
       "      <td>chr10:79797063-79800372:+</td>\n",
       "      <td>chr10:79797063-79797722:+</td>\n",
       "      <td>chr10:79797741-79800372:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:79796952-79797062:+@exon:chr10:7979...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:79796952-79797062:+</td>\n",
       "      <td>exon:chr10:79797723-79797740:+</td>\n",
       "      <td>exon:chr10:79800373-79800455:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000485708.2,ENST00000476545.1,ENST000004...</td>\n",
       "      <td>RPS24</td>\n",
       "      <td>ENSG00000138326</td>\n",
       "      <td>exon:chr10:79796952-79797062:+</td>\n",
       "      <td>111</td>\n",
       "      <td>exon:chr10:79797723-79797740:+</td>\n",
       "      <td>18</td>\n",
       "      <td>exon:chr10:79800373-79800455:+</td>\n",
       "      <td>83</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1353</th>\n",
       "      <td>chr10:79797063-79800372:+</td>\n",
       "      <td>chr10:79797063-79799961:+</td>\n",
       "      <td>chr10:79799984-79800372:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:79796952-79797062:+@exon:chr10:7979...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:79796952-79797062:+</td>\n",
       "      <td>exon:chr10:79799962-79799983:+</td>\n",
       "      <td>exon:chr10:79800373-79800471:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000485708.2,ENST00000476545.1,ENST000003...</td>\n",
       "      <td>RPS24</td>\n",
       "      <td>ENSG00000138326</td>\n",
       "      <td>exon:chr10:79796952-79797062:+</td>\n",
       "      <td>111</td>\n",
       "      <td>exon:chr10:79799962-79799983:+</td>\n",
       "      <td>22</td>\n",
       "      <td>exon:chr10:79800373-79800471:+</td>\n",
       "      <td>99</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1354</th>\n",
       "      <td>chr10:79797741-79800372:+</td>\n",
       "      <td>chr10:79797741-79799961:+</td>\n",
       "      <td>chr10:79799984-79800372:+</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr10:79797723-79797740:+@exon:chr10:7979...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:79797723-79797740:+</td>\n",
       "      <td>exon:chr10:79799962-79799983:+</td>\n",
       "      <td>exon:chr10:79800373-79800428:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000485708.2,ENST00000360830.4</td>\n",
       "      <td>RPS24</td>\n",
       "      <td>ENSG00000138326</td>\n",
       "      <td>exon:chr10:79797723-79797740:+</td>\n",
       "      <td>18</td>\n",
       "      <td>exon:chr10:79799962-79799983:+</td>\n",
       "      <td>22</td>\n",
       "      <td>exon:chr10:79800373-79800428:+</td>\n",
       "      <td>56</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     junction13                 junction12  \\\n",
       "1351  chr10:79797063-79799961:+  chr10:79797063-79797722:+   \n",
       "1352  chr10:79797063-79800372:+  chr10:79797063-79797722:+   \n",
       "1353  chr10:79797063-79800372:+  chr10:79797063-79799961:+   \n",
       "1354  chr10:79797741-79800372:+  chr10:79797741-79799961:+   \n",
       "\n",
       "                     junction23                       criteria_full  \\\n",
       "1351  chr10:79797741-79799961:+                            only one   \n",
       "1352  chr10:79797741-79800372:+               best,appris_principal   \n",
       "1353  chr10:79799984-79800372:+               best,appris_principal   \n",
       "1354  chr10:79799984-79800372:+  one event with gencode transcripts   \n",
       "\n",
       "                                               event_id  \\\n",
       "1351  exon:chr10:79796952-79797062:+@exon:chr10:7979...   \n",
       "1352  exon:chr10:79796952-79797062:+@exon:chr10:7979...   \n",
       "1353  exon:chr10:79796952-79797062:+@exon:chr10:7979...   \n",
       "1354  exon:chr10:79797723-79797740:+@exon:chr10:7979...   \n",
       "\n",
       "                                criteria criteria_additional  \\\n",
       "1351                            only one                 NaN   \n",
       "1352                                best    appris_principal   \n",
       "1353                                best    appris_principal   \n",
       "1354  one event with gencode transcripts                 NaN   \n",
       "\n",
       "                               exon1                           exon2  \\\n",
       "1351  exon:chr10:79796952-79797062:+  exon:chr10:79797723-79797740:+   \n",
       "1352  exon:chr10:79796952-79797062:+  exon:chr10:79797723-79797740:+   \n",
       "1353  exon:chr10:79796952-79797062:+  exon:chr10:79799962-79799983:+   \n",
       "1354  exon:chr10:79797723-79797740:+  exon:chr10:79799962-79799983:+   \n",
       "\n",
       "                               exon3   ...    \\\n",
       "1351  exon:chr10:79799962-79799983:+   ...     \n",
       "1352  exon:chr10:79800373-79800455:+   ...     \n",
       "1353  exon:chr10:79800373-79800471:+   ...     \n",
       "1354  exon:chr10:79800373-79800428:+   ...     \n",
       "\n",
       "                                          transcript_id gene_name  \\\n",
       "1351  ENST00000485708.2,ENST00000476545.1,ENST000003...     RPS24   \n",
       "1352  ENST00000485708.2,ENST00000476545.1,ENST000004...     RPS24   \n",
       "1353  ENST00000485708.2,ENST00000476545.1,ENST000003...     RPS24   \n",
       "1354                ENST00000485708.2,ENST00000360830.4     RPS24   \n",
       "\n",
       "           ensembl_id                    exon1_region exon1_length  \\\n",
       "1351  ENSG00000138326  exon:chr10:79796952-79797062:+          111   \n",
       "1352  ENSG00000138326  exon:chr10:79796952-79797062:+          111   \n",
       "1353  ENSG00000138326  exon:chr10:79796952-79797062:+          111   \n",
       "1354  ENSG00000138326  exon:chr10:79797723-79797740:+           18   \n",
       "\n",
       "                        exon2_region exon2_length  \\\n",
       "1351  exon:chr10:79797723-79797740:+           18   \n",
       "1352  exon:chr10:79797723-79797740:+           18   \n",
       "1353  exon:chr10:79799962-79799983:+           22   \n",
       "1354  exon:chr10:79799962-79799983:+           22   \n",
       "\n",
       "                        exon3_region exon3_length  strand  \n",
       "1351  exon:chr10:79799962-79799983:+           22       +  \n",
       "1352  exon:chr10:79800373-79800455:+           83       +  \n",
       "1353  exon:chr10:79800373-79800471:+           99       +  \n",
       "1354  exon:chr10:79800373-79800428:+           56       +  \n",
       "\n",
       "[4 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_df_annotated.loc[se_df_annotated.gene_name == 'RPS24']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction23</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28066</th>\n",
       "      <td>chr3:186502486-186503671:+</td>\n",
       "      <td>chr3:186502486-186502750:+</td>\n",
       "      <td>chr3:186502891-186503671:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr3:186502353-186502485:+@exon:chr3:1865...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr3:186502353-186502485:+</td>\n",
       "      <td>exon:chr3:186502751-186502890:+</td>\n",
       "      <td>exon:chr3:186503672-186503840:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000323963.5,ENST00000356531.5,ENST000004...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186502353-186502485:+</td>\n",
       "      <td>133</td>\n",
       "      <td>exon:chr3:186502751-186502890:+</td>\n",
       "      <td>140</td>\n",
       "      <td>exon:chr3:186503672-186503840:+</td>\n",
       "      <td>169</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28067</th>\n",
       "      <td>chr3:186502486-186503952:+</td>\n",
       "      <td>chr3:186502486-186503671:+</td>\n",
       "      <td>chr3:186503841-186503952:+</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr3:186502353-186502485:+@exon:chr3:1865...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr3:186502353-186502485:+</td>\n",
       "      <td>exon:chr3:186503672-186503840:+</td>\n",
       "      <td>exon:chr3:186503953-186504062:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000323963.5,ENST00000356531.5,ENST000004...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186502353-186502485:+</td>\n",
       "      <td>133</td>\n",
       "      <td>exon:chr3:186503672-186503840:+</td>\n",
       "      <td>169</td>\n",
       "      <td>exon:chr3:186503953-186504062:+</td>\n",
       "      <td>110</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28068</th>\n",
       "      <td>chr3:186502486-186504290:+</td>\n",
       "      <td>chr3:186502486-186503952:+</td>\n",
       "      <td>chr3:186504063-186504290:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr3:186502353-186502485:+@exon:chr3:1865...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr3:186502353-186502485:+</td>\n",
       "      <td>exon:chr3:186503953-186504062:+</td>\n",
       "      <td>exon:chr3:186504291-186504434:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000323963.5,ENST00000356531.5,ENST000004...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186502353-186502485:+</td>\n",
       "      <td>133</td>\n",
       "      <td>exon:chr3:186503953-186504062:+</td>\n",
       "      <td>110</td>\n",
       "      <td>exon:chr3:186504291-186504434:+</td>\n",
       "      <td>144</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28069</th>\n",
       "      <td>chr3:186504063-186504915:+</td>\n",
       "      <td>chr3:186504063-186504290:+</td>\n",
       "      <td>chr3:186504435-186504915:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr3:186503953-186504062:+@exon:chr3:1865...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr3:186503953-186504062:+</td>\n",
       "      <td>exon:chr3:186504291-186504434:+</td>\n",
       "      <td>exon:chr3:186504916-186505053:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000492144.1,ENST00000323963.5,ENST000003...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186503953-186504062:+</td>\n",
       "      <td>110</td>\n",
       "      <td>exon:chr3:186504291-186504434:+</td>\n",
       "      <td>144</td>\n",
       "      <td>exon:chr3:186504916-186505053:+</td>\n",
       "      <td>138</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28070</th>\n",
       "      <td>chr3:186505054-186505591:+</td>\n",
       "      <td>chr3:186505054-186505267:+</td>\n",
       "      <td>chr3:186505374-186505591:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr3:186504916-186505053:+@exon:chr3:1865...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr3:186504916-186505053:+</td>\n",
       "      <td>exon:chr3:186505268-186505373:+</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000492144.1,ENST00000356531.5,ENST000003...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186504916-186505053:+</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr3:186505268-186505373:+</td>\n",
       "      <td>106</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>80</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28071</th>\n",
       "      <td>chr3:186505054-186505591:+</td>\n",
       "      <td>chr3:186505054-186505283:+</td>\n",
       "      <td>chr3:186505374-186505591:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr3:186504916-186505053:+@exon:chr3:1865...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr3:186504916-186505053:+</td>\n",
       "      <td>exon:chr3:186505284-186505373:+</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000492144.1,ENST00000356531.5,ENST000003...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186504916-186505053:+</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr3:186505284-186505373:+</td>\n",
       "      <td>90</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>80</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28072</th>\n",
       "      <td>chr3:186505374-186506098:+</td>\n",
       "      <td>chr3:186505374-186505591:+</td>\n",
       "      <td>chr3:186505672-186506098:+</td>\n",
       "      <td>random,no good tags</td>\n",
       "      <td>exon:chr3:186505268-186505373:+@exon:chr3:1865...</td>\n",
       "      <td>random</td>\n",
       "      <td>no good tags</td>\n",
       "      <td>exon:chr3:186505268-186505373:+</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>exon:chr3:186506099-186506209:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000323963.5,ENST00000356531.5,ENST000004...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186505268-186505373:+</td>\n",
       "      <td>106</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>80</td>\n",
       "      <td>exon:chr3:186506099-186506209:+</td>\n",
       "      <td>111</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28073</th>\n",
       "      <td>chr3:186505672-186506913:+</td>\n",
       "      <td>chr3:186505672-186506098:+</td>\n",
       "      <td>chr3:186506206-186506913:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr3:186505592-186505671:+@exon:chr3:1865...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>exon:chr3:186506099-186506205:+</td>\n",
       "      <td>exon:chr3:186506914-186507689:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000323963.5,ENST00000356531.5,ENST000004...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>80</td>\n",
       "      <td>exon:chr3:186506099-186506205:+</td>\n",
       "      <td>107</td>\n",
       "      <td>exon:chr3:186506914-186507689:+</td>\n",
       "      <td>776</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28074</th>\n",
       "      <td>chr3:186505672-186506913:+</td>\n",
       "      <td>chr3:186505672-186506098:+</td>\n",
       "      <td>chr3:186506210-186506913:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr3:186505592-186505671:+@exon:chr3:1865...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>exon:chr3:186506099-186506209:+</td>\n",
       "      <td>exon:chr3:186506914-186507686:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000323963.5,ENST00000356531.5,ENST000004...</td>\n",
       "      <td>EIF4A2</td>\n",
       "      <td>ENSG00000156976</td>\n",
       "      <td>exon:chr3:186505592-186505671:+</td>\n",
       "      <td>80</td>\n",
       "      <td>exon:chr3:186506099-186506209:+</td>\n",
       "      <td>111</td>\n",
       "      <td>exon:chr3:186506914-186507686:+</td>\n",
       "      <td>773</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       junction13                  junction12  \\\n",
       "28066  chr3:186502486-186503671:+  chr3:186502486-186502750:+   \n",
       "28067  chr3:186502486-186503952:+  chr3:186502486-186503671:+   \n",
       "28068  chr3:186502486-186504290:+  chr3:186502486-186503952:+   \n",
       "28069  chr3:186504063-186504915:+  chr3:186504063-186504290:+   \n",
       "28070  chr3:186505054-186505591:+  chr3:186505054-186505267:+   \n",
       "28071  chr3:186505054-186505591:+  chr3:186505054-186505283:+   \n",
       "28072  chr3:186505374-186506098:+  chr3:186505374-186505591:+   \n",
       "28073  chr3:186505672-186506913:+  chr3:186505672-186506098:+   \n",
       "28074  chr3:186505672-186506913:+  chr3:186505672-186506098:+   \n",
       "\n",
       "                       junction23                       criteria_full  \\\n",
       "28066  chr3:186502891-186503671:+               best,appris_principal   \n",
       "28067  chr3:186503841-186503952:+  one event with gencode transcripts   \n",
       "28068  chr3:186504063-186504290:+                            only one   \n",
       "28069  chr3:186504435-186504915:+               best,appris_principal   \n",
       "28070  chr3:186505374-186505591:+               best,appris_principal   \n",
       "28071  chr3:186505374-186505591:+               best,appris_principal   \n",
       "28072  chr3:186505672-186506098:+                 random,no good tags   \n",
       "28073  chr3:186506206-186506913:+               best,appris_principal   \n",
       "28074  chr3:186506210-186506913:+               best,appris_principal   \n",
       "\n",
       "                                                event_id  \\\n",
       "28066  exon:chr3:186502353-186502485:+@exon:chr3:1865...   \n",
       "28067  exon:chr3:186502353-186502485:+@exon:chr3:1865...   \n",
       "28068  exon:chr3:186502353-186502485:+@exon:chr3:1865...   \n",
       "28069  exon:chr3:186503953-186504062:+@exon:chr3:1865...   \n",
       "28070  exon:chr3:186504916-186505053:+@exon:chr3:1865...   \n",
       "28071  exon:chr3:186504916-186505053:+@exon:chr3:1865...   \n",
       "28072  exon:chr3:186505268-186505373:+@exon:chr3:1865...   \n",
       "28073  exon:chr3:186505592-186505671:+@exon:chr3:1865...   \n",
       "28074  exon:chr3:186505592-186505671:+@exon:chr3:1865...   \n",
       "\n",
       "                                 criteria criteria_additional  \\\n",
       "28066                                best    appris_principal   \n",
       "28067  one event with gencode transcripts                 NaN   \n",
       "28068                            only one                 NaN   \n",
       "28069                                best    appris_principal   \n",
       "28070                                best    appris_principal   \n",
       "28071                                best    appris_principal   \n",
       "28072                              random        no good tags   \n",
       "28073                                best    appris_principal   \n",
       "28074                                best    appris_principal   \n",
       "\n",
       "                                 exon1                            exon2  \\\n",
       "28066  exon:chr3:186502353-186502485:+  exon:chr3:186502751-186502890:+   \n",
       "28067  exon:chr3:186502353-186502485:+  exon:chr3:186503672-186503840:+   \n",
       "28068  exon:chr3:186502353-186502485:+  exon:chr3:186503953-186504062:+   \n",
       "28069  exon:chr3:186503953-186504062:+  exon:chr3:186504291-186504434:+   \n",
       "28070  exon:chr3:186504916-186505053:+  exon:chr3:186505268-186505373:+   \n",
       "28071  exon:chr3:186504916-186505053:+  exon:chr3:186505284-186505373:+   \n",
       "28072  exon:chr3:186505268-186505373:+  exon:chr3:186505592-186505671:+   \n",
       "28073  exon:chr3:186505592-186505671:+  exon:chr3:186506099-186506205:+   \n",
       "28074  exon:chr3:186505592-186505671:+  exon:chr3:186506099-186506209:+   \n",
       "\n",
       "                                 exon3   ...    \\\n",
       "28066  exon:chr3:186503672-186503840:+   ...     \n",
       "28067  exon:chr3:186503953-186504062:+   ...     \n",
       "28068  exon:chr3:186504291-186504434:+   ...     \n",
       "28069  exon:chr3:186504916-186505053:+   ...     \n",
       "28070  exon:chr3:186505592-186505671:+   ...     \n",
       "28071  exon:chr3:186505592-186505671:+   ...     \n",
       "28072  exon:chr3:186506099-186506209:+   ...     \n",
       "28073  exon:chr3:186506914-186507689:+   ...     \n",
       "28074  exon:chr3:186506914-186507686:+   ...     \n",
       "\n",
       "                                           transcript_id gene_name  \\\n",
       "28066  ENST00000323963.5,ENST00000356531.5,ENST000004...    EIF4A2   \n",
       "28067  ENST00000323963.5,ENST00000356531.5,ENST000004...    EIF4A2   \n",
       "28068  ENST00000323963.5,ENST00000356531.5,ENST000004...    EIF4A2   \n",
       "28069  ENST00000492144.1,ENST00000323963.5,ENST000003...    EIF4A2   \n",
       "28070  ENST00000492144.1,ENST00000356531.5,ENST000003...    EIF4A2   \n",
       "28071  ENST00000492144.1,ENST00000356531.5,ENST000003...    EIF4A2   \n",
       "28072  ENST00000323963.5,ENST00000356531.5,ENST000004...    EIF4A2   \n",
       "28073  ENST00000323963.5,ENST00000356531.5,ENST000004...    EIF4A2   \n",
       "28074  ENST00000323963.5,ENST00000356531.5,ENST000004...    EIF4A2   \n",
       "\n",
       "            ensembl_id                     exon1_region exon1_length  \\\n",
       "28066  ENSG00000156976  exon:chr3:186502353-186502485:+          133   \n",
       "28067  ENSG00000156976  exon:chr3:186502353-186502485:+          133   \n",
       "28068  ENSG00000156976  exon:chr3:186502353-186502485:+          133   \n",
       "28069  ENSG00000156976  exon:chr3:186503953-186504062:+          110   \n",
       "28070  ENSG00000156976  exon:chr3:186504916-186505053:+          138   \n",
       "28071  ENSG00000156976  exon:chr3:186504916-186505053:+          138   \n",
       "28072  ENSG00000156976  exon:chr3:186505268-186505373:+          106   \n",
       "28073  ENSG00000156976  exon:chr3:186505592-186505671:+           80   \n",
       "28074  ENSG00000156976  exon:chr3:186505592-186505671:+           80   \n",
       "\n",
       "                          exon2_region exon2_length  \\\n",
       "28066  exon:chr3:186502751-186502890:+          140   \n",
       "28067  exon:chr3:186503672-186503840:+          169   \n",
       "28068  exon:chr3:186503953-186504062:+          110   \n",
       "28069  exon:chr3:186504291-186504434:+          144   \n",
       "28070  exon:chr3:186505268-186505373:+          106   \n",
       "28071  exon:chr3:186505284-186505373:+           90   \n",
       "28072  exon:chr3:186505592-186505671:+           80   \n",
       "28073  exon:chr3:186506099-186506205:+          107   \n",
       "28074  exon:chr3:186506099-186506209:+          111   \n",
       "\n",
       "                          exon3_region exon3_length  strand  \n",
       "28066  exon:chr3:186503672-186503840:+          169       +  \n",
       "28067  exon:chr3:186503953-186504062:+          110       +  \n",
       "28068  exon:chr3:186504291-186504434:+          144       +  \n",
       "28069  exon:chr3:186504916-186505053:+          138       +  \n",
       "28070  exon:chr3:186505592-186505671:+           80       +  \n",
       "28071  exon:chr3:186505592-186505671:+           80       +  \n",
       "28072  exon:chr3:186506099-186506209:+          111       +  \n",
       "28073  exon:chr3:186506914-186507689:+          776       +  \n",
       "28074  exon:chr3:186506914-186507686:+          773       +  \n",
       "\n",
       "[9 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_df_annotated.loc[se_df_annotated.gene_name == 'EIF4A2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get intron lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction23</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>...</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>-</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>...</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>3269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>...</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>-</td>\n",
       "      <td>1543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>...</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>166</td>\n",
       "      <td>-</td>\n",
       "      <td>2121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>183</td>\n",
       "      <td>+</td>\n",
       "      <td>1857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    junction13                   junction12  \\\n",
       "0  chr10:100189400-100190327:-  chr10:100189647-100190327:-   \n",
       "1  chr10:100190428-100193696:-  chr10:100191049-100193696:-   \n",
       "2  chr10:100193849-100195391:-  chr10:100195172-100195391:-   \n",
       "3  chr10:101163392-101165512:-  chr10:101163632-101165512:-   \n",
       "4  chr10:101419346-101421202:+  chr10:101419346-101419618:+   \n",
       "\n",
       "                    junction23          criteria_full  \\\n",
       "0  chr10:100189400-100189547:-  best,appris_principal   \n",
       "1  chr10:100190428-100190887:-               only one   \n",
       "2  chr10:100193849-100195028:-  best,appris_principal   \n",
       "3  chr10:101163392-101163480:-               only one   \n",
       "4  chr10:101419722-101421202:+               only one   \n",
       "\n",
       "                                            event_id  criteria  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...      best   \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...  only one   \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...      best   \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...  only one   \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...  only one   \n",
       "\n",
       "  criteria_additional                             exon1  \\\n",
       "0    appris_principal  exon:chr10:100190328-100190427:-   \n",
       "1                 NaN  exon:chr10:100193697-100193848:-   \n",
       "2    appris_principal  exon:chr10:100195392-100195529:-   \n",
       "3                 NaN  exon:chr10:101165513-101165617:-   \n",
       "4                 NaN  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                              exon2                             exon3  \\\n",
       "0  exon:chr10:100189548-100189646:-  exon:chr10:100189330-100189399:-   \n",
       "1  exon:chr10:100190888-100191048:-  exon:chr10:100190328-100190427:-   \n",
       "2  exon:chr10:100195029-100195171:-  exon:chr10:100193697-100193848:-   \n",
       "3  exon:chr10:101163481-101163631:-  exon:chr10:101163226-101163391:-   \n",
       "4  exon:chr10:101419619-101419721:+  exon:chr10:101421203-101421385:+   \n",
       "\n",
       "       ...       gene_name       ensembl_id                      exon1_region  \\\n",
       "0      ...            HPS1  ENSG00000107521  exon:chr10:100190328-100190427:-   \n",
       "1      ...            HPS1  ENSG00000107521  exon:chr10:100193697-100193848:-   \n",
       "2      ...            HPS1  ENSG00000107521  exon:chr10:100195392-100195529:-   \n",
       "3      ...            GOT1  ENSG00000120053  exon:chr10:101165513-101165617:-   \n",
       "4      ...          ENTPD7  ENSG00000198018  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "  exon1_length                      exon2_region  exon2_length  \\\n",
       "0          100  exon:chr10:100189548-100189646:-            99   \n",
       "1          152  exon:chr10:100190888-100191048:-           161   \n",
       "2          138  exon:chr10:100195029-100195171:-           143   \n",
       "3          105  exon:chr10:101163481-101163631:-           151   \n",
       "4           83  exon:chr10:101419619-101419721:+           103   \n",
       "\n",
       "                       exon3_region  exon3_length strand  intron_length  \n",
       "0  exon:chr10:100189330-100189399:-            70      -            928  \n",
       "1  exon:chr10:100190328-100190427:-           100      -           3269  \n",
       "2  exon:chr10:100193697-100193848:-           152      -           1543  \n",
       "3  exon:chr10:101163226-101163391:-           166      -           2121  \n",
       "4  exon:chr10:101421203-101421385:+           183      +           1857  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_strand = se_df_annotated.strand == '+'\n",
    "se_df_annotated['intron_length'] = np.nan\n",
    "\n",
    "\n",
    "se_df_annotated.loc[positive_strand, 'intron_length'] = \\\n",
    "    se_df_annotated.loc[positive_strand].apply(lambda x: x.exon3_region._start - x.exon1_region._stop - 1, axis=1)\n",
    "se_df_annotated.loc[~positive_strand, 'intron_length'] = \\\n",
    "    se_df_annotated.loc[~positive_strand].apply(lambda x: x.exon1_region._start - x.exon3_region._stop - 1, axis=1)\n",
    "se_df_annotated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>...</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4_region</th>\n",
       "      <th>exon4_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr10:101996727-102003454:-</td>\n",
       "      <td>chr10:101995522-101996606:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:101995522-101997778:-</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr10:102003455-102003534:-@exon:chr10:10...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>...</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>80</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>210</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>120</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>98</td>\n",
       "      <td>-</td>\n",
       "      <td>7933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr10:102003535-102013177:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:102010090-102013177:-</td>\n",
       "      <td>chr10:101997989-102010004:-</td>\n",
       "      <td>random,no gencode transcripts</td>\n",
       "      <td>exon:chr10:102013178-102013296:-@exon:chr10:10...</td>\n",
       "      <td>random</td>\n",
       "      <td>no gencode transcripts</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>...</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>119</td>\n",
       "      <td>exon:chr10:102010005-102010089:-</td>\n",
       "      <td>85</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>80</td>\n",
       "      <td>exon:chr10:101997940-101997988:-</td>\n",
       "      <td>49</td>\n",
       "      <td>-</td>\n",
       "      <td>15189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr10:102016234-102020722:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102020722:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>random,no good tags</td>\n",
       "      <td>exon:chr10:102020723-102020801:-@exon:chr10:10...</td>\n",
       "      <td>random</td>\n",
       "      <td>no good tags</td>\n",
       "      <td>exon:chr10:102020723-102020801:-</td>\n",
       "      <td>...</td>\n",
       "      <td>exon:chr10:102020723-102020801:-</td>\n",
       "      <td>79</td>\n",
       "      <td>exon:chr10:102019975-102020076:-</td>\n",
       "      <td>102</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>215</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>119</td>\n",
       "      <td>-</td>\n",
       "      <td>7426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr10:102016234-102021735:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102021735:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:102021736-102021820:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:102021736-102021820:-</td>\n",
       "      <td>...</td>\n",
       "      <td>exon:chr10:102021736-102021820:-</td>\n",
       "      <td>85</td>\n",
       "      <td>exon:chr10:102019975-102020076:-</td>\n",
       "      <td>102</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>215</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>119</td>\n",
       "      <td>-</td>\n",
       "      <td>8439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr10:102016234-102021735:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020786-102021735:-</td>\n",
       "      <td>chr10:102013297-102020722:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:102021736-102021820:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:102021736-102021820:-</td>\n",
       "      <td>...</td>\n",
       "      <td>exon:chr10:102021736-102021820:-</td>\n",
       "      <td>85</td>\n",
       "      <td>exon:chr10:102020723-102020785:-</td>\n",
       "      <td>63</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>215</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>119</td>\n",
       "      <td>-</td>\n",
       "      <td>8439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                   junction13                   junction34  \\\n",
       "0           0  chr10:101996727-102003454:-  chr10:101995522-101996606:-   \n",
       "1           1  chr10:102003535-102013177:-  chr10:101997989-102003454:-   \n",
       "2           2  chr10:102016234-102020722:-  chr10:102013297-102016018:-   \n",
       "3           3  chr10:102016234-102021735:-  chr10:102013297-102016018:-   \n",
       "4           4  chr10:102016234-102021735:-  chr10:102013297-102016018:-   \n",
       "\n",
       "                    junction12                   junction24  \\\n",
       "0  chr10:101997989-102003454:-  chr10:101995522-101997778:-   \n",
       "1  chr10:102010090-102013177:-  chr10:101997989-102010004:-   \n",
       "2  chr10:102020077-102020722:-  chr10:102013297-102019974:-   \n",
       "3  chr10:102020077-102021735:-  chr10:102013297-102019974:-   \n",
       "4  chr10:102020786-102021735:-  chr10:102013297-102020722:-   \n",
       "\n",
       "                        criteria_full  \\\n",
       "0  one event with gencode transcripts   \n",
       "1       random,no gencode transcripts   \n",
       "2                 random,no good tags   \n",
       "3                            only one   \n",
       "4                            only one   \n",
       "\n",
       "                                            event_id  \\\n",
       "0  exon:chr10:102003455-102003534:-@exon:chr10:10...   \n",
       "1  exon:chr10:102013178-102013296:-@exon:chr10:10...   \n",
       "2  exon:chr10:102020723-102020801:-@exon:chr10:10...   \n",
       "3  exon:chr10:102021736-102021820:-@exon:chr10:10...   \n",
       "4  exon:chr10:102021736-102021820:-@exon:chr10:10...   \n",
       "\n",
       "                             criteria     criteria_additional  \\\n",
       "0  one event with gencode transcripts                     NaN   \n",
       "1                              random  no gencode transcripts   \n",
       "2                              random            no good tags   \n",
       "3                            only one                     NaN   \n",
       "4                            only one                     NaN   \n",
       "\n",
       "                              exon1      ...       \\\n",
       "0  exon:chr10:102003455-102003534:-      ...        \n",
       "1  exon:chr10:102013178-102013296:-      ...        \n",
       "2  exon:chr10:102020723-102020801:-      ...        \n",
       "3  exon:chr10:102021736-102021820:-      ...        \n",
       "4  exon:chr10:102021736-102021820:-      ...        \n",
       "\n",
       "                       exon1_region exon1_length  \\\n",
       "0  exon:chr10:102003455-102003534:-           80   \n",
       "1  exon:chr10:102013178-102013296:-          119   \n",
       "2  exon:chr10:102020723-102020801:-           79   \n",
       "3  exon:chr10:102021736-102021820:-           85   \n",
       "4  exon:chr10:102021736-102021820:-           85   \n",
       "\n",
       "                       exon2_region exon2_length  \\\n",
       "0  exon:chr10:101997779-101997988:-          210   \n",
       "1  exon:chr10:102010005-102010089:-           85   \n",
       "2  exon:chr10:102019975-102020076:-          102   \n",
       "3  exon:chr10:102019975-102020076:-          102   \n",
       "4  exon:chr10:102020723-102020785:-           63   \n",
       "\n",
       "                       exon3_region exon3_length  \\\n",
       "0  exon:chr10:101996607-101996726:-          120   \n",
       "1  exon:chr10:102003455-102003534:-           80   \n",
       "2  exon:chr10:102016019-102016233:-          215   \n",
       "3  exon:chr10:102016019-102016233:-          215   \n",
       "4  exon:chr10:102016019-102016233:-          215   \n",
       "\n",
       "                       exon4_region exon4_length  strand intron_length  \n",
       "0  exon:chr10:101995424-101995521:-           98       -          7933  \n",
       "1  exon:chr10:101997940-101997988:-           49       -         15189  \n",
       "2  exon:chr10:102013178-102013296:-          119       -          7426  \n",
       "3  exon:chr10:102013178-102013296:-          119       -          8439  \n",
       "4  exon:chr10:102013178-102013296:-          119       -          8439  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_strand = mxe_df_annotated.strand == '+'\n",
    "mxe_df_annotated['intron_length'] = np.nan\n",
    "\n",
    "\n",
    "mxe_df_annotated.loc[positive_strand, 'intron_length'] = \\\n",
    "    mxe_df_annotated.loc[positive_strand].apply(lambda x: x.exon4_region._start - x.exon1_region._stop - 1, axis=1)\n",
    "mxe_df_annotated.loc[~positive_strand, 'intron_length'] = \\\n",
    "    mxe_df_annotated.loc[~positive_strand].apply(lambda x: x.exon1_region._start - x.exon4_region._stop - 1, axis=1)\n",
    "mxe_df_annotated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36531, 23)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2710206-2710283:+@exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+</th>\n",
       "      <td>13212</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>exon:chrY:2710206-2710283:+</td>\n",
       "      <td>78</td>\n",
       "      <td>exon:chrY:2712118-2712298:+</td>\n",
       "      <td>181</td>\n",
       "      <td>exon:chrY:2713687-2713784:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000129824.11</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>12357</td>\n",
       "      <td>chrY:2710284-2712117:+</td>\n",
       "      <td>chrY:2710284-2713686:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2712299-2722640:+</td>\n",
       "      <td>chrY:2713785-2722640:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000430575.1,ENST00000250784.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2734834-2734903:+</th>\n",
       "      <td>13213</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>exon:chrY:2712118-2712298:+</td>\n",
       "      <td>181</td>\n",
       "      <td>exon:chrY:2713687-2713784:+</td>\n",
       "      <td>98</td>\n",
       "      <td>exon:chrY:2722641-2722812:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000129824.11</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>22535</td>\n",
       "      <td>chrY:2712299-2713686:+</td>\n",
       "      <td>chrY:2712299-2722640:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2713785-2734833:+</td>\n",
       "      <td>chrY:2722813-2734833:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000430575.1,ENST00000250784.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+@exon:chrY:2734834-2734903:+</th>\n",
       "      <td>13214</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>exon:chrY:2713687-2713784:+</td>\n",
       "      <td>98</td>\n",
       "      <td>exon:chrY:2722641-2722812:+</td>\n",
       "      <td>172</td>\n",
       "      <td>exon:chrY:2733129-2733286:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000129824.11</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>21049</td>\n",
       "      <td>chrY:2713785-2722640:+</td>\n",
       "      <td>chrY:2713785-2733128:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2722813-2734833:+</td>\n",
       "      <td>chrY:2733287-2734833:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000430575.1,ENST00000477725.1,ENST000002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2803112-2803487:+@exon:chrY:2821950-2822038:+@exon:chrY:2829115-2829687:+@exon:chrY:2843136-2843285:+</th>\n",
       "      <td>13215</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>ENSG00000067646</td>\n",
       "      <td>exon:chrY:2803112-2803487:+</td>\n",
       "      <td>376</td>\n",
       "      <td>exon:chrY:2821950-2822038:+</td>\n",
       "      <td>89</td>\n",
       "      <td>exon:chrY:2829115-2829687:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067646.7</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>39648</td>\n",
       "      <td>chrY:2803488-2821949:+</td>\n",
       "      <td>chrY:2803488-2829114:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2822039-2843135:+</td>\n",
       "      <td>chrY:2829688-2843135:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000443793.1,ENST00000431102.1,ENST000003...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2803518-2803810:+@exon:chrY:2821950-2822038:+@exon:chrY:2829115-2829687:+@exon:chrY:2843136-2843285:+</th>\n",
       "      <td>13216</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>ENSG00000067646</td>\n",
       "      <td>exon:chrY:2803518-2803810:+</td>\n",
       "      <td>293</td>\n",
       "      <td>exon:chrY:2821950-2822038:+</td>\n",
       "      <td>89</td>\n",
       "      <td>exon:chrY:2829115-2829687:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067646.7</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>39325</td>\n",
       "      <td>chrY:2803811-2821949:+</td>\n",
       "      <td>chrY:2803811-2829114:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2822039-2843135:+</td>\n",
       "      <td>chrY:2829688-2843135:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000443793.1,ENST00000431102.1,ENST000003...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Unnamed: 0  \\\n",
       "event_id                                                         \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...       13212   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...       13213   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...       13214   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...       13215   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...       13216   \n",
       "\n",
       "                                                                              criteria  \\\n",
       "event_id                                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                            only one   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                            only one   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                            only one   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                 NaN   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                 NaN   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                 NaN   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...                 NaN   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...                 NaN   \n",
       "\n",
       "                                                                         criteria_full  \\\n",
       "event_id                                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                            only one   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                            only one   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                            only one   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  ENSG00000129824   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  ENSG00000129824   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENSG00000129824   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  ENSG00000067646   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  ENSG00000067646   \n",
       "\n",
       "                                                                          exon1  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2710206-2710283:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2712118-2712298:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2713687-2713784:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2803112-2803487:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2803518-2803810:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...            78   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...           181   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...            98   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...           376   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...           293   \n",
       "\n",
       "                                                                          exon2  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2712118-2712298:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2713687-2713784:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2722641-2722812:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2821950-2822038:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2821950-2822038:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...           181   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...            98   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...           172   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...            89   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...            89   \n",
       "\n",
       "                                                                          exon3  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2713687-2713784:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2722641-2722812:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2733129-2733286:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2829115-2829687:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2829115-2829687:+   \n",
       "\n",
       "                                                                          ...                          \\\n",
       "event_id                                                                  ...                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                        ...                           \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                        ...                           \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                        ...                           \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...                        ...                           \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...                        ...                           \n",
       "\n",
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  ENSG00000129824.11   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  ENSG00000129824.11   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENSG00000129824.11   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...   ENSG00000067646.7   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...   ENSG00000067646.7   \n",
       "\n",
       "                                                   gene_name  intron_length  \\\n",
       "event_id                                                                      \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...    RPS4Y1          12357   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...    RPS4Y1          22535   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...    RPS4Y1          21049   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...       ZFY          39648   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...       ZFY          39325   \n",
       "\n",
       "                                                                junction12  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2710284-2712117:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2712299-2713686:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2713785-2722640:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2803488-2821949:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2803811-2821949:+   \n",
       "\n",
       "                                                                junction13  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2710284-2713686:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2712299-2722640:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2713785-2733128:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2803488-2829114:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2803811-2829114:+   \n",
       "\n",
       "                                                    junction23  \\\n",
       "event_id                                                         \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...         NaN   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...         NaN   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...         NaN   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...         NaN   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...         NaN   \n",
       "\n",
       "                                                                junction24  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2712299-2722640:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2713785-2734833:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2722813-2734833:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2822039-2843135:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2822039-2843135:+   \n",
       "\n",
       "                                                                junction34  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2713785-2722640:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2722813-2734833:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2733287-2734833:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2829688-2843135:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2829688-2843135:+   \n",
       "\n",
       "                                                   strand  \\\n",
       "event_id                                                    \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...      +   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...      +   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...      +   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...      +   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...      +   \n",
       "\n",
       "                                                                                        transcript_id  \n",
       "event_id                                                                                               \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                ENST00000430575.1,ENST00000250784.8  \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                ENST00000430575.1,ENST00000250784.8  \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENST00000430575.1,ENST00000477725.1,ENST000002...  \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  ENST00000443793.1,ENST00000431102.1,ENST000003...  \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  ENST00000443793.1,ENST00000431102.1,ENST000003...  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "dfs = []\n",
    "annotated_dfs = se_df_annotated, mxe_df_annotated\n",
    "for df in annotated_dfs:\n",
    "    cols = [x  for x in df.columns if 'region' not in x]\n",
    "    dfs.append(df[cols])\n",
    "\n",
    "splicing_feature_data = pd.concat(dfs, ignore_index=True)\n",
    "splicing_feature_data = splicing_feature_data.set_index('event_id')\n",
    "print splicing_feature_data.shape\n",
    "splicing_feature_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>...</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>928</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>...</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>3269</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>...</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>1543</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000325103.6,ENST000004...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>...</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>2121</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000543866.1,ENST00000370508.5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>...</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>1857</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000370489.4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Unnamed: 0  criteria  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...         NaN  only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...         NaN  only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...         NaN  only one   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                 NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                 NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                 NaN   \n",
       "\n",
       "                                                            criteria_full  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...               only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...               only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...               only one   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                                               exon1  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195392-100195529:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101165513-101165617:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           105   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...            83   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189548-100189646:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190888-100191048:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195029-100195171:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163481-101163631:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419619-101419721:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           151   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           103   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189330-100189399:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163226-101163391:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101421203-101421385:+   \n",
       "\n",
       "                                                           ...           \\\n",
       "event_id                                                   ...            \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...         ...            \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...         ...            \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...         ...            \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...         ...            \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...         ...            \n",
       "\n",
       "                                                    gene_name intron_length  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...       HPS1           928   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...       HPS1          3269   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...       HPS1          1543   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...       GOT1          2121   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...     ENTPD7          1857   \n",
       "\n",
       "                                                                     junction12  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189647-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100191049-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100195172-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163632-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101419618:+   \n",
       "\n",
       "                                                                     junction13  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101421202:+   \n",
       "\n",
       "                                                                     junction23  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100189547:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100190887:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195028:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101163480:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419722-101421202:+   \n",
       "\n",
       "                                                    junction24 junction34  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...         NaN        NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...         NaN        NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...         NaN        NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...         NaN        NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...         NaN        NaN   \n",
       "\n",
       "                                                   strand  \\\n",
       "event_id                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      -   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...      -   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      -   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...      -   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...      +   \n",
       "\n",
       "                                                                                        transcript_id  \\\n",
       "event_id                                                                                                \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENST00000361490.4,ENST00000325103.6,ENST000004...   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                ENST00000543866.1,ENST00000370508.5   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                                  ENST00000370489.4   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                False  \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                False  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data['exon2_divisible_by_3'] = splicing_feature_data['exon2_length'].astype(int).mod(3) == 0\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.35164654676849799"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data['exon2_divisible_by_3'].sum()/float(splicing_feature_data.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2299\n",
      "(36531, 24)\n"
     ]
    }
   ],
   "source": [
    "print splicing_feature_data.ensembl_id.str.contains(',').sum()\n",
    "print splicing_feature_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3,629/53,751 splicing events have multiple gene ids associated with them. Since we want to intersect this information with existing gene annotations, we need a *single* id. To do this coarsely and simply, we will take the first gene ID for intersecting purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.options.display.max_columns = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4</th>\n",
       "      <th>exon4_length</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>928</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>3269</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>1543</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000325103.6,ENST000004...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>166</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>2121</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000543866.1,ENST00000370508.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198018.6</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>1857</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000370489.4</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Unnamed: 0  criteria  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...         NaN  only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...         NaN  only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...         NaN  only one   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                 NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                 NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                 NaN   \n",
       "\n",
       "                                                            criteria_full  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...               only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...               only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...               only one   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                                               exon1  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195392-100195529:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101165513-101165617:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           105   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...            83   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189548-100189646:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190888-100191048:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195029-100195171:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163481-101163631:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419619-101419721:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           151   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           103   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189330-100189399:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163226-101163391:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101421203-101421385:+   \n",
       "\n",
       "                                                    exon3_length exon4  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           166   NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           183   NaN   \n",
       "\n",
       "                                                    exon4_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           NaN   \n",
       "\n",
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...   ENSG00000120053.9   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...   ENSG00000198018.6   \n",
       "\n",
       "                                                   gene_name  intron_length  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      HPS1            928   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...      HPS1           3269   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      HPS1           1543   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...      GOT1           2121   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...    ENTPD7           1857   \n",
       "\n",
       "                                                                     junction12  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189647-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100191049-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100195172-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163632-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101419618:+   \n",
       "\n",
       "                                                                     junction13  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101421202:+   \n",
       "\n",
       "                                                                     junction23  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100189547:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100190887:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195028:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101163480:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419722-101421202:+   \n",
       "\n",
       "                                                   junction24 junction34  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...        NaN        NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...        NaN        NaN   \n",
       "\n",
       "                                                   strand  \\\n",
       "event_id                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      -   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...      -   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      -   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...      -   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...      +   \n",
       "\n",
       "                                                                                        transcript_id  \\\n",
       "event_id                                                                                                \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENST00000361490.4,ENST00000325103.6,ENST000004...   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                ENST00000543866.1,ENST00000370508.5   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                                  ENST00000370489.4   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                False   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                False   \n",
       "\n",
       "                                                     one_ensembl_id  \n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053  \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data['one_ensembl_id'] = splicing_feature_data.ensembl_id.str.split(',').str[0]\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add whether this is an SE or MXE event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4</th>\n",
       "      <th>exon4_length</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "      <th>splice_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>928</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>3269</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>1543</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000325103.6,ENST000004...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>166</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>2121</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000543866.1,ENST00000370508.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198018.6</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>1857</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000370489.4</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Unnamed: 0  criteria  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...         NaN  only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...         NaN  only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...         NaN  only one   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                 NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                 NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                 NaN   \n",
       "\n",
       "                                                            criteria_full  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...               only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...               only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...               only one   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                                               exon1  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195392-100195529:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101165513-101165617:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           105   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...            83   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189548-100189646:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190888-100191048:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195029-100195171:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163481-101163631:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419619-101419721:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           151   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           103   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189330-100189399:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163226-101163391:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101421203-101421385:+   \n",
       "\n",
       "                                                    exon3_length exon4  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           166   NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           183   NaN   \n",
       "\n",
       "                                                    exon4_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           NaN   \n",
       "\n",
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...   ENSG00000120053.9   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...   ENSG00000198018.6   \n",
       "\n",
       "                                                   gene_name  intron_length  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      HPS1            928   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...      HPS1           3269   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      HPS1           1543   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...      GOT1           2121   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...    ENTPD7           1857   \n",
       "\n",
       "                                                                     junction12  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189647-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100191049-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100195172-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163632-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101419618:+   \n",
       "\n",
       "                                                                     junction13  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101421202:+   \n",
       "\n",
       "                                                                     junction23  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100189547:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100190887:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195028:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101163480:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419722-101421202:+   \n",
       "\n",
       "                                                   junction24 junction34  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...        NaN        NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...        NaN        NaN   \n",
       "\n",
       "                                                   strand  \\\n",
       "event_id                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...      -   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...      -   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...      -   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...      -   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...      +   \n",
       "\n",
       "                                                                                        transcript_id  \\\n",
       "event_id                                                                                                \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENST00000361490.4,ENST00000325103.6,ENST000004...   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                ENST00000543866.1,ENST00000370508.5   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                                  ENST00000370489.4   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                False   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                False   \n",
       "\n",
       "                                                     one_ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                   splice_type  \n",
       "event_id                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...          SE  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...          SE  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...          SE  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...          SE  \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...          SE  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data['splice_type'] = splicing_feature_data.index.map(lambda x: 'MXE' if len(x.split('@')) == 4 else 'SE')\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4</th>\n",
       "      <th>exon4_length</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "      <th>splice_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2710206-2710283:+@exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+</th>\n",
       "      <td>13212</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>exon:chrY:2710206-2710283:+</td>\n",
       "      <td>78</td>\n",
       "      <td>exon:chrY:2712118-2712298:+</td>\n",
       "      <td>181</td>\n",
       "      <td>exon:chrY:2713687-2713784:+</td>\n",
       "      <td>98</td>\n",
       "      <td>exon:chrY:2722641-2722812:+</td>\n",
       "      <td>172</td>\n",
       "      <td>ENSG00000129824.11</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>12357</td>\n",
       "      <td>chrY:2710284-2712117:+</td>\n",
       "      <td>chrY:2710284-2713686:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2712299-2722640:+</td>\n",
       "      <td>chrY:2713785-2722640:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000430575.1,ENST00000250784.8</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>MXE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2734834-2734903:+</th>\n",
       "      <td>13213</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>exon:chrY:2712118-2712298:+</td>\n",
       "      <td>181</td>\n",
       "      <td>exon:chrY:2713687-2713784:+</td>\n",
       "      <td>98</td>\n",
       "      <td>exon:chrY:2722641-2722812:+</td>\n",
       "      <td>172</td>\n",
       "      <td>exon:chrY:2734834-2734903:+</td>\n",
       "      <td>70</td>\n",
       "      <td>ENSG00000129824.11</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>22535</td>\n",
       "      <td>chrY:2712299-2713686:+</td>\n",
       "      <td>chrY:2712299-2722640:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2713785-2734833:+</td>\n",
       "      <td>chrY:2722813-2734833:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000430575.1,ENST00000250784.8</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>MXE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+@exon:chrY:2734834-2734903:+</th>\n",
       "      <td>13214</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>exon:chrY:2713687-2713784:+</td>\n",
       "      <td>98</td>\n",
       "      <td>exon:chrY:2722641-2722812:+</td>\n",
       "      <td>172</td>\n",
       "      <td>exon:chrY:2733129-2733286:+</td>\n",
       "      <td>158</td>\n",
       "      <td>exon:chrY:2734834-2734903:+</td>\n",
       "      <td>70</td>\n",
       "      <td>ENSG00000129824.11</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>21049</td>\n",
       "      <td>chrY:2713785-2722640:+</td>\n",
       "      <td>chrY:2713785-2733128:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2722813-2734833:+</td>\n",
       "      <td>chrY:2733287-2734833:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000430575.1,ENST00000477725.1,ENST000002...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000129824</td>\n",
       "      <td>MXE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2803112-2803487:+@exon:chrY:2821950-2822038:+@exon:chrY:2829115-2829687:+@exon:chrY:2843136-2843285:+</th>\n",
       "      <td>13215</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>ENSG00000067646</td>\n",
       "      <td>exon:chrY:2803112-2803487:+</td>\n",
       "      <td>376</td>\n",
       "      <td>exon:chrY:2821950-2822038:+</td>\n",
       "      <td>89</td>\n",
       "      <td>exon:chrY:2829115-2829687:+</td>\n",
       "      <td>573</td>\n",
       "      <td>exon:chrY:2843136-2843285:+</td>\n",
       "      <td>150</td>\n",
       "      <td>ENSG00000067646.7</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>39648</td>\n",
       "      <td>chrY:2803488-2821949:+</td>\n",
       "      <td>chrY:2803488-2829114:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2822039-2843135:+</td>\n",
       "      <td>chrY:2829688-2843135:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000443793.1,ENST00000431102.1,ENST000003...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000067646</td>\n",
       "      <td>MXE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chrY:2803518-2803810:+@exon:chrY:2821950-2822038:+@exon:chrY:2829115-2829687:+@exon:chrY:2843136-2843285:+</th>\n",
       "      <td>13216</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>ENSG00000067646</td>\n",
       "      <td>exon:chrY:2803518-2803810:+</td>\n",
       "      <td>293</td>\n",
       "      <td>exon:chrY:2821950-2822038:+</td>\n",
       "      <td>89</td>\n",
       "      <td>exon:chrY:2829115-2829687:+</td>\n",
       "      <td>573</td>\n",
       "      <td>exon:chrY:2843136-2843285:+</td>\n",
       "      <td>150</td>\n",
       "      <td>ENSG00000067646.7</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>39325</td>\n",
       "      <td>chrY:2803811-2821949:+</td>\n",
       "      <td>chrY:2803811-2829114:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:2822039-2843135:+</td>\n",
       "      <td>chrY:2829688-2843135:+</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000443793.1,ENST00000431102.1,ENST000003...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000067646</td>\n",
       "      <td>MXE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Unnamed: 0  \\\n",
       "event_id                                                         \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...       13212   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...       13213   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...       13214   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...       13215   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...       13216   \n",
       "\n",
       "                                                                              criteria  \\\n",
       "event_id                                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                            only one   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                            only one   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                            only one   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                 NaN   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                 NaN   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                 NaN   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...                 NaN   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...                 NaN   \n",
       "\n",
       "                                                                         criteria_full  \\\n",
       "event_id                                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                            only one   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                            only one   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                            only one   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  one event with gencode transcripts   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  ENSG00000129824   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  ENSG00000129824   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENSG00000129824   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  ENSG00000067646   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  ENSG00000067646   \n",
       "\n",
       "                                                                          exon1  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2710206-2710283:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2712118-2712298:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2713687-2713784:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2803112-2803487:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2803518-2803810:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...            78   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...           181   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...            98   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...           376   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...           293   \n",
       "\n",
       "                                                                          exon2  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2712118-2712298:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2713687-2713784:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2722641-2722812:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2821950-2822038:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2821950-2822038:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...           181   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...            98   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...           172   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...            89   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...            89   \n",
       "\n",
       "                                                                          exon3  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2713687-2713784:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2722641-2722812:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2733129-2733286:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2829115-2829687:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2829115-2829687:+   \n",
       "\n",
       "                                                    exon3_length  \\\n",
       "event_id                                                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...            98   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...           172   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...           158   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...           573   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...           573   \n",
       "\n",
       "                                                                          exon4  \\\n",
       "event_id                                                                          \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  exon:chrY:2722641-2722812:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  exon:chrY:2734834-2734903:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  exon:chrY:2734834-2734903:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  exon:chrY:2843136-2843285:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  exon:chrY:2843136-2843285:+   \n",
       "\n",
       "                                                    exon4_length  \\\n",
       "event_id                                                           \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...           172   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...            70   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...            70   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...           150   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...           150   \n",
       "\n",
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  ENSG00000129824.11   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  ENSG00000129824.11   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENSG00000129824.11   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...   ENSG00000067646.7   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...   ENSG00000067646.7   \n",
       "\n",
       "                                                   gene_name  intron_length  \\\n",
       "event_id                                                                      \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...    RPS4Y1          12357   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...    RPS4Y1          22535   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...    RPS4Y1          21049   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...       ZFY          39648   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...       ZFY          39325   \n",
       "\n",
       "                                                                junction12  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2710284-2712117:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2712299-2713686:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2713785-2722640:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2803488-2821949:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2803811-2821949:+   \n",
       "\n",
       "                                                                junction13  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2710284-2713686:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2712299-2722640:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2713785-2733128:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2803488-2829114:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2803811-2829114:+   \n",
       "\n",
       "                                                   junction23  \\\n",
       "event_id                                                        \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...        NaN   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...        NaN   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...        NaN   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...        NaN   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...        NaN   \n",
       "\n",
       "                                                                junction24  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2712299-2722640:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2713785-2734833:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2722813-2734833:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2822039-2843135:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2822039-2843135:+   \n",
       "\n",
       "                                                                junction34  \\\n",
       "event_id                                                                     \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  chrY:2713785-2722640:+   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  chrY:2722813-2734833:+   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  chrY:2733287-2734833:+   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  chrY:2829688-2843135:+   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  chrY:2829688-2843135:+   \n",
       "\n",
       "                                                   strand  \\\n",
       "event_id                                                    \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...      +   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...      +   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...      +   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...      +   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...      +   \n",
       "\n",
       "                                                                                        transcript_id  \\\n",
       "event_id                                                                                                \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                ENST00000430575.1,ENST00000250784.8   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                ENST00000430575.1,ENST00000250784.8   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENST00000430575.1,ENST00000477725.1,ENST000002...   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  ENST00000443793.1,ENST00000431102.1,ENST000003...   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  ENST00000443793.1,ENST00000431102.1,ENST000003...   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...                False   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...                False   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...                False   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...                False   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...                False   \n",
       "\n",
       "                                                     one_ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...  ENSG00000129824   \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...  ENSG00000129824   \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...  ENSG00000129824   \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...  ENSG00000067646   \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...  ENSG00000067646   \n",
       "\n",
       "                                                   splice_type  \n",
       "event_id                                                        \n",
       "exon:chrY:2710206-2710283:+@exon:chrY:2712118-2...         MXE  \n",
       "exon:chrY:2712118-2712298:+@exon:chrY:2713687-2...         MXE  \n",
       "exon:chrY:2713687-2713784:+@exon:chrY:2722641-2...         MXE  \n",
       "exon:chrY:2803112-2803487:+@exon:chrY:2821950-2...         MXE  \n",
       "exon:chrY:2803518-2803810:+@exon:chrY:2821950-2...         MXE  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use only exon ids that were deemed alternative in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write to a safe place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(36531, 26)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/projects/ps-yeolab/obotvinnik/singlecell_pnms'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "splicing_feature_data.to_csv('{}/splicing_feature_data.csv'.format(csv_folder))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
