{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Translate isoforms to get protein products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:784: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "v19db_filename = '/projects/ps-yeolab/genomes/hg19/gencode/v19/gencode.v19.annotation.gtf.db'\n",
    "v19db = gffutils.FeatureDB(v19db_filename)\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/home/obotvinnik/Dropbox/figures2/singlecell_pnms/isoform_protein_properties': File exists\r\n"
     ]
    }
   ],
   "source": [
    "figure_folder = '/home/obotvinnik/Dropbox/figures2/singlecell_pnms/isoform_protein_properties'\n",
    "! mkdir $figure_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "splicing_feature_data = pd.read_csv('{}/splicing_feature_data.csv'.format(folder), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction_exon12</th>\n",
       "      <th>junction_exon23</th>\n",
       "      <th>junction_exon13</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               junction_exon12              junction_exon23  \\\n",
       "0  chr10:100189647-100190327:-  chr10:100189400-100189547:-   \n",
       "1  chr10:100191049-100193696:-  chr10:100190428-100190887:-   \n",
       "2  chr10:100195172-100195391:-  chr10:100193849-100195028:-   \n",
       "3  chr10:101163632-101165512:-  chr10:101163392-101163480:-   \n",
       "4  chr10:101419346-101419618:+  chr10:101419722-101421202:+   \n",
       "\n",
       "               junction_exon13          criteria_full  \\\n",
       "0  chr10:100189400-100190327:-  best,appris_principal   \n",
       "1  chr10:100190428-100193696:-               only one   \n",
       "2  chr10:100193849-100195391:-  best,appris_principal   \n",
       "3  chr10:101163392-101165512:-               only one   \n",
       "4  chr10:101419346-101421202:+               only one   \n",
       "\n",
       "                                            event_id  criteria  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...      best   \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...  only one   \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...      best   \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...  only one   \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...  only one   \n",
       "\n",
       "  criteria_additional  \n",
       "0    appris_principal  \n",
       "1                 NaN  \n",
       "2    appris_principal  \n",
       "3                 NaN  \n",
       "4                 NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_consolidated_events = pd.read_csv('/projects/ps-yeolab/obotvinnik/singlecell_pnms/skipped_exon_consolidated_junctions.csv')\n",
    "se_consolidated_events.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "se_consolidated_events[['exon1', 'exon2', 'exon3']] = se_consolidated_events.event_id.str.split('@').apply(pd.Series)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction_exon12</th>\n",
       "      <th>junction_exon23</th>\n",
       "      <th>junction_exon13</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               junction_exon12              junction_exon23  \\\n",
       "0  chr10:100189647-100190327:-  chr10:100189400-100189547:-   \n",
       "1  chr10:100191049-100193696:-  chr10:100190428-100190887:-   \n",
       "2  chr10:100195172-100195391:-  chr10:100193849-100195028:-   \n",
       "3  chr10:101163632-101165512:-  chr10:101163392-101163480:-   \n",
       "4  chr10:101419346-101419618:+  chr10:101419722-101421202:+   \n",
       "\n",
       "               junction_exon13          criteria_full  \\\n",
       "0  chr10:100189400-100190327:-  best,appris_principal   \n",
       "1  chr10:100190428-100193696:-               only one   \n",
       "2  chr10:100193849-100195391:-  best,appris_principal   \n",
       "3  chr10:101163392-101165512:-               only one   \n",
       "4  chr10:101419346-101421202:+               only one   \n",
       "\n",
       "                                            event_id  criteria  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...      best   \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...  only one   \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...      best   \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...  only one   \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...  only one   \n",
       "\n",
       "  criteria_additional                             exon1  \\\n",
       "0    appris_principal  exon:chr10:100190328-100190427:-   \n",
       "1                 NaN  exon:chr10:100193697-100193848:-   \n",
       "2    appris_principal  exon:chr10:100195392-100195529:-   \n",
       "3                 NaN  exon:chr10:101165513-101165617:-   \n",
       "4                 NaN  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                              exon2                             exon3  \n",
       "0  exon:chr10:100189548-100189646:-  exon:chr10:100189330-100189399:-  \n",
       "1  exon:chr10:100190888-100191048:-  exon:chr10:100190328-100190427:-  \n",
       "2  exon:chr10:100195029-100195171:-  exon:chr10:100193697-100193848:-  \n",
       "3  exon:chr10:101163481-101163631:-  exon:chr10:101163226-101163391:-  \n",
       "4  exon:chr10:101419619-101419721:+  exon:chr10:101421203-101421385:+  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_consolidated_events.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translate isoform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "prefix = 'skipped_exon_isoform_translations'\n",
    "translated_fasta = '{}/{}.fa'.format(folder, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction_exon12</th>\n",
       "      <th>junction_exon23</th>\n",
       "      <th>junction_exon13</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23861</th>\n",
       "      <td>chr7:100771946-100775155:+</td>\n",
       "      <td>chr7:100775351-100776975:+</td>\n",
       "      <td>chr7:100771946-100776975:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr7:100771810-100771945:+@exon:chr7:1007...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr7:100771810-100771945:+</td>\n",
       "      <td>exon:chr7:100775156-100775350:+</td>\n",
       "      <td>exon:chr7:100776976-100777174:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26060</th>\n",
       "      <td>chr8:82587067-82588468:-</td>\n",
       "      <td>chr8:82586185-82586906:-</td>\n",
       "      <td>chr8:82586185-82588468:-</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr8:82588469-82588500:-@exon:chr8:825869...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr8:82588469-82588500:-</td>\n",
       "      <td>exon:chr8:82586907-82587066:-</td>\n",
       "      <td>exon:chr8:82586076-82586184:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7664</th>\n",
       "      <td>chr16:53340310-53341592:+</td>\n",
       "      <td>chr16:53341818-53342597:+</td>\n",
       "      <td>chr16:53340310-53342597:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr16:53340149-53340309:+@exon:chr16:5334...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr16:53340149-53340309:+</td>\n",
       "      <td>exon:chr16:53341593-53341817:+</td>\n",
       "      <td>exon:chr16:53342598-53342766:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22929</th>\n",
       "      <td>chr6:134274564-134289992:+</td>\n",
       "      <td>chr6:134290096-134301219:+</td>\n",
       "      <td>chr6:134274564-134301219:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr6:134274362-134274563:+@exon:chr6:1342...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr6:134274362-134274563:+</td>\n",
       "      <td>exon:chr6:134289993-134290095:+</td>\n",
       "      <td>exon:chr6:134301220-134301398:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8885</th>\n",
       "      <td>chr17:4700091-4700732:+</td>\n",
       "      <td>chr17:4700865-4700976:+</td>\n",
       "      <td>chr17:4700091-4700976:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr17:4700023-4700090:+@exon:chr17:470073...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr17:4700023-4700090:+</td>\n",
       "      <td>exon:chr17:4700733-4700864:+</td>\n",
       "      <td>exon:chr17:4700977-4701106:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8738</th>\n",
       "      <td>chr17:41150849-41151949:+</td>\n",
       "      <td>chr17:41152120-41154689:+</td>\n",
       "      <td>chr17:41150849-41154689:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr17:41150766-41150848:+@exon:chr17:4115...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr17:41150766-41150848:+</td>\n",
       "      <td>exon:chr17:41151950-41152119:+</td>\n",
       "      <td>exon:chr17:41154690-41154800:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26020</th>\n",
       "      <td>chr8:74882870-74884104:-</td>\n",
       "      <td>chr8:74876861-74882813:-</td>\n",
       "      <td>chr8:74876861-74884104:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr8:74884105-74884143:-@exon:chr8:748828...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr8:74884105-74884143:-</td>\n",
       "      <td>exon:chr8:74882814-74882869:-</td>\n",
       "      <td>exon:chr8:74876722-74876860:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16980</th>\n",
       "      <td>chr2:24981026-24985555:+</td>\n",
       "      <td>chr2:24985646-24991089:+</td>\n",
       "      <td>chr2:24981026-24991089:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr2:24980842-24981025:+@exon:chr2:249855...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr2:24980842-24981025:+</td>\n",
       "      <td>exon:chr2:24985556-24985645:+</td>\n",
       "      <td>exon:chr2:24991090-24991186:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20635</th>\n",
       "      <td>chr4:39357618-39367858:-</td>\n",
       "      <td>chr4:39353097-39357496:-</td>\n",
       "      <td>chr4:39353097-39367858:-</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr4:39367859-39367955:-@exon:chr4:393574...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr4:39367859-39367955:-</td>\n",
       "      <td>exon:chr4:39357497-39357617:-</td>\n",
       "      <td>exon:chr4:39352968-39353096:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22487</th>\n",
       "      <td>chr5:78915907-78919074:+</td>\n",
       "      <td>chr5:78919313-78936673:+</td>\n",
       "      <td>chr5:78915907-78936673:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr5:78915817-78915906:+@exon:chr5:789190...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr5:78915817-78915906:+</td>\n",
       "      <td>exon:chr5:78919075-78919312:+</td>\n",
       "      <td>exon:chr5:78936674-78936788:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27813</th>\n",
       "      <td>chrX:64753616-64754359:-</td>\n",
       "      <td>chrX:64752511-64753489:-</td>\n",
       "      <td>chrX:64752511-64754359:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chrX:64754360-64754634:-@exon:chrX:647534...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chrX:64754360-64754634:-</td>\n",
       "      <td>exon:chrX:64753490-64753615:-</td>\n",
       "      <td>exon:chrX:64752441-64752510:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27360</th>\n",
       "      <td>chrX:118797641-118809516:-</td>\n",
       "      <td>chrX:118787004-118797444:-</td>\n",
       "      <td>chrX:118787004-118809516:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chrX:118809517-118809631:-@exon:chrX:1187...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chrX:118809517-118809631:-</td>\n",
       "      <td>exon:chrX:118797445-118797640:-</td>\n",
       "      <td>exon:chrX:118786817-118787003:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26905</th>\n",
       "      <td>chr9:42712788-42713912:+</td>\n",
       "      <td>chr9:42714043-42714429:+</td>\n",
       "      <td>chr9:42712788-42714429:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr9:42712722-42712787:+@exon:chr9:427139...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr9:42712722-42712787:+</td>\n",
       "      <td>exon:chr9:42713913-42714042:+</td>\n",
       "      <td>exon:chr9:42714430-42714591:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25868</th>\n",
       "      <td>chr8:54975936-55013467:-</td>\n",
       "      <td>chr8:54974885-54975887:-</td>\n",
       "      <td>chr8:54974885-55013467:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr8:55013468-55013499:-@exon:chr8:549758...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr8:55013468-55013499:-</td>\n",
       "      <td>exon:chr8:54975888-54975935:-</td>\n",
       "      <td>exon:chr8:54974814-54974884:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13257</th>\n",
       "      <td>chr1:33138503-33139059:+</td>\n",
       "      <td>chr1:33139182-33145240:+</td>\n",
       "      <td>chr1:33138503-33145240:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr1:33138392-33138502:+@exon:chr1:331390...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr1:33138392-33138502:+</td>\n",
       "      <td>exon:chr1:33139060-33139181:+</td>\n",
       "      <td>exon:chr1:33145241-33145641:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26795</th>\n",
       "      <td>chr9:34329598-34336297:+</td>\n",
       "      <td>chr9:34336340-34338710:+</td>\n",
       "      <td>chr9:34329598-34338710:+</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr9:34329504-34329597:+@exon:chr9:343362...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr9:34329504-34329597:+</td>\n",
       "      <td>exon:chr9:34336298-34336339:+</td>\n",
       "      <td>exon:chr9:34338711-34338845:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4327</th>\n",
       "      <td>chr12:9073662-9075228:+</td>\n",
       "      <td>chr12:9075385-9083030:+</td>\n",
       "      <td>chr12:9073662-9083030:+</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>exon:chr12:9073581-9073661:+@exon:chr12:907522...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>exon:chr12:9073581-9073661:+</td>\n",
       "      <td>exon:chr12:9075229-9075384:+</td>\n",
       "      <td>exon:chr12:9083031-9083523:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12249</th>\n",
       "      <td>chr1:17303694-17304731:-</td>\n",
       "      <td>chr1:17303419-17303603:-</td>\n",
       "      <td>chr1:17303419-17304731:-</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>exon:chr1:17304732-17304771:-@exon:chr1:173036...</td>\n",
       "      <td>one event with gencode transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr1:17304732-17304771:-</td>\n",
       "      <td>exon:chr1:17303604-17303693:-</td>\n",
       "      <td>exon:chr1:17303392-17303418:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12460</th>\n",
       "      <td>chr1:202704716-202705340:-</td>\n",
       "      <td>chr1:202703015-202704556:-</td>\n",
       "      <td>chr1:202703015-202705340:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr1:202705341-202705364:-@exon:chr1:2027...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr1:202705341-202705364:-</td>\n",
       "      <td>exon:chr1:202704557-202704715:-</td>\n",
       "      <td>exon:chr1:202702529-202703014:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13371</th>\n",
       "      <td>chr1:38445206-38446257:-</td>\n",
       "      <td>chr1:38435408-38445136:-</td>\n",
       "      <td>chr1:38435408-38446257:-</td>\n",
       "      <td>only one</td>\n",
       "      <td>exon:chr1:38446258-38446396:-@exon:chr1:384451...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon:chr1:38446258-38446396:-</td>\n",
       "      <td>exon:chr1:38445137-38445205:-</td>\n",
       "      <td>exon:chr1:38435243-38435407:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  junction_exon12             junction_exon23  \\\n",
       "23861  chr7:100771946-100775155:+  chr7:100775351-100776975:+   \n",
       "26060    chr8:82587067-82588468:-    chr8:82586185-82586906:-   \n",
       "7664    chr16:53340310-53341592:+   chr16:53341818-53342597:+   \n",
       "22929  chr6:134274564-134289992:+  chr6:134290096-134301219:+   \n",
       "8885      chr17:4700091-4700732:+     chr17:4700865-4700976:+   \n",
       "8738    chr17:41150849-41151949:+   chr17:41152120-41154689:+   \n",
       "26020    chr8:74882870-74884104:-    chr8:74876861-74882813:-   \n",
       "16980    chr2:24981026-24985555:+    chr2:24985646-24991089:+   \n",
       "20635    chr4:39357618-39367858:-    chr4:39353097-39357496:-   \n",
       "22487    chr5:78915907-78919074:+    chr5:78919313-78936673:+   \n",
       "27813    chrX:64753616-64754359:-    chrX:64752511-64753489:-   \n",
       "27360  chrX:118797641-118809516:-  chrX:118787004-118797444:-   \n",
       "26905    chr9:42712788-42713912:+    chr9:42714043-42714429:+   \n",
       "25868    chr8:54975936-55013467:-    chr8:54974885-54975887:-   \n",
       "13257    chr1:33138503-33139059:+    chr1:33139182-33145240:+   \n",
       "26795    chr9:34329598-34336297:+    chr9:34336340-34338710:+   \n",
       "4327      chr12:9073662-9075228:+     chr12:9075385-9083030:+   \n",
       "12249    chr1:17303694-17304731:-    chr1:17303419-17303603:-   \n",
       "12460  chr1:202704716-202705340:-  chr1:202703015-202704556:-   \n",
       "13371    chr1:38445206-38446257:-    chr1:38435408-38445136:-   \n",
       "\n",
       "                  junction_exon13                       criteria_full  \\\n",
       "23861  chr7:100771946-100776975:+                            only one   \n",
       "26060    chr8:82586185-82588468:-               best,appris_principal   \n",
       "7664    chr16:53340310-53342597:+                            only one   \n",
       "22929  chr6:134274564-134301219:+                            only one   \n",
       "8885      chr17:4700091-4700976:+                            only one   \n",
       "8738    chr17:41150849-41154689:+                            only one   \n",
       "26020    chr8:74876861-74884104:-                            only one   \n",
       "16980    chr2:24981026-24991089:+               best,appris_principal   \n",
       "20635    chr4:39353097-39367858:-  one event with gencode transcripts   \n",
       "22487    chr5:78915907-78936673:+                            only one   \n",
       "27813    chrX:64752511-64754359:-                            only one   \n",
       "27360  chrX:118787004-118809516:-                            only one   \n",
       "26905    chr9:42712788-42714429:+               best,appris_principal   \n",
       "25868    chr8:54974885-55013467:-                            only one   \n",
       "13257    chr1:33138503-33145240:+               best,appris_principal   \n",
       "26795    chr9:34329598-34338710:+                            only one   \n",
       "4327      chr12:9073662-9083030:+               best,appris_principal   \n",
       "12249    chr1:17303419-17304731:-  one event with gencode transcripts   \n",
       "12460  chr1:202703015-202705340:-                            only one   \n",
       "13371    chr1:38435408-38446257:-                            only one   \n",
       "\n",
       "                                                event_id  \\\n",
       "23861  exon:chr7:100771810-100771945:+@exon:chr7:1007...   \n",
       "26060  exon:chr8:82588469-82588500:-@exon:chr8:825869...   \n",
       "7664   exon:chr16:53340149-53340309:+@exon:chr16:5334...   \n",
       "22929  exon:chr6:134274362-134274563:+@exon:chr6:1342...   \n",
       "8885   exon:chr17:4700023-4700090:+@exon:chr17:470073...   \n",
       "8738   exon:chr17:41150766-41150848:+@exon:chr17:4115...   \n",
       "26020  exon:chr8:74884105-74884143:-@exon:chr8:748828...   \n",
       "16980  exon:chr2:24980842-24981025:+@exon:chr2:249855...   \n",
       "20635  exon:chr4:39367859-39367955:-@exon:chr4:393574...   \n",
       "22487  exon:chr5:78915817-78915906:+@exon:chr5:789190...   \n",
       "27813  exon:chrX:64754360-64754634:-@exon:chrX:647534...   \n",
       "27360  exon:chrX:118809517-118809631:-@exon:chrX:1187...   \n",
       "26905  exon:chr9:42712722-42712787:+@exon:chr9:427139...   \n",
       "25868  exon:chr8:55013468-55013499:-@exon:chr8:549758...   \n",
       "13257  exon:chr1:33138392-33138502:+@exon:chr1:331390...   \n",
       "26795  exon:chr9:34329504-34329597:+@exon:chr9:343362...   \n",
       "4327   exon:chr12:9073581-9073661:+@exon:chr12:907522...   \n",
       "12249  exon:chr1:17304732-17304771:-@exon:chr1:173036...   \n",
       "12460  exon:chr1:202705341-202705364:-@exon:chr1:2027...   \n",
       "13371  exon:chr1:38446258-38446396:-@exon:chr1:384451...   \n",
       "\n",
       "                                 criteria criteria_additional  \\\n",
       "23861                            only one                 NaN   \n",
       "26060                                best    appris_principal   \n",
       "7664                             only one                 NaN   \n",
       "22929                            only one                 NaN   \n",
       "8885                             only one                 NaN   \n",
       "8738                             only one                 NaN   \n",
       "26020                            only one                 NaN   \n",
       "16980                                best    appris_principal   \n",
       "20635  one event with gencode transcripts                 NaN   \n",
       "22487                            only one                 NaN   \n",
       "27813                            only one                 NaN   \n",
       "27360                            only one                 NaN   \n",
       "26905                                best    appris_principal   \n",
       "25868                            only one                 NaN   \n",
       "13257                                best    appris_principal   \n",
       "26795                            only one                 NaN   \n",
       "4327                                 best    appris_principal   \n",
       "12249  one event with gencode transcripts                 NaN   \n",
       "12460                            only one                 NaN   \n",
       "13371                            only one                 NaN   \n",
       "\n",
       "                                 exon1                            exon2  \\\n",
       "23861  exon:chr7:100771810-100771945:+  exon:chr7:100775156-100775350:+   \n",
       "26060    exon:chr8:82588469-82588500:-    exon:chr8:82586907-82587066:-   \n",
       "7664    exon:chr16:53340149-53340309:+   exon:chr16:53341593-53341817:+   \n",
       "22929  exon:chr6:134274362-134274563:+  exon:chr6:134289993-134290095:+   \n",
       "8885      exon:chr17:4700023-4700090:+     exon:chr17:4700733-4700864:+   \n",
       "8738    exon:chr17:41150766-41150848:+   exon:chr17:41151950-41152119:+   \n",
       "26020    exon:chr8:74884105-74884143:-    exon:chr8:74882814-74882869:-   \n",
       "16980    exon:chr2:24980842-24981025:+    exon:chr2:24985556-24985645:+   \n",
       "20635    exon:chr4:39367859-39367955:-    exon:chr4:39357497-39357617:-   \n",
       "22487    exon:chr5:78915817-78915906:+    exon:chr5:78919075-78919312:+   \n",
       "27813    exon:chrX:64754360-64754634:-    exon:chrX:64753490-64753615:-   \n",
       "27360  exon:chrX:118809517-118809631:-  exon:chrX:118797445-118797640:-   \n",
       "26905    exon:chr9:42712722-42712787:+    exon:chr9:42713913-42714042:+   \n",
       "25868    exon:chr8:55013468-55013499:-    exon:chr8:54975888-54975935:-   \n",
       "13257    exon:chr1:33138392-33138502:+    exon:chr1:33139060-33139181:+   \n",
       "26795    exon:chr9:34329504-34329597:+    exon:chr9:34336298-34336339:+   \n",
       "4327      exon:chr12:9073581-9073661:+     exon:chr12:9075229-9075384:+   \n",
       "12249    exon:chr1:17304732-17304771:-    exon:chr1:17303604-17303693:-   \n",
       "12460  exon:chr1:202705341-202705364:-  exon:chr1:202704557-202704715:-   \n",
       "13371    exon:chr1:38446258-38446396:-    exon:chr1:38445137-38445205:-   \n",
       "\n",
       "                                 exon3  \n",
       "23861  exon:chr7:100776976-100777174:+  \n",
       "26060    exon:chr8:82586076-82586184:-  \n",
       "7664    exon:chr16:53342598-53342766:+  \n",
       "22929  exon:chr6:134301220-134301398:+  \n",
       "8885      exon:chr17:4700977-4701106:+  \n",
       "8738    exon:chr17:41154690-41154800:+  \n",
       "26020    exon:chr8:74876722-74876860:-  \n",
       "16980    exon:chr2:24991090-24991186:+  \n",
       "20635    exon:chr4:39352968-39353096:-  \n",
       "22487    exon:chr5:78936674-78936788:+  \n",
       "27813    exon:chrX:64752441-64752510:-  \n",
       "27360  exon:chrX:118786817-118787003:-  \n",
       "26905    exon:chr9:42714430-42714591:+  \n",
       "25868    exon:chr8:54974814-54974884:-  \n",
       "13257    exon:chr1:33145241-33145641:+  \n",
       "26795    exon:chr9:34338711-34338845:+  \n",
       "4327      exon:chr12:9083031-9083523:+  \n",
       "12249    exon:chr1:17303392-17303418:-  \n",
       "12460  exon:chr1:202702529-202703014:-  \n",
       "13371    exon:chr1:38435243-38435407:-  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "size = 20\n",
    "\n",
    "random_subset = se_consolidated_events.loc[np.random.choice(se_consolidated_events.index, size=size)]\n",
    "random_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "a must be 1-dimensional or an integer",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-65-3f2b53c4cb13>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     61\u001b[0m \u001b[1;31m#     print '\\t', '\\t'.join(map(lambda x: str(list(x.attributes['tag'])) if 'tag' in x.attributes else '-', isoform2_transcripts))\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     62\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 63\u001b[1;33m     \u001b[0misoform1_transcript\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfilter_on_tags\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0misoform1_transcripts\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbest_tags\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     64\u001b[0m     \u001b[0misoform2_transcript\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfilter_on_tags\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0misoform2_transcripts\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbest_tags\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     65\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-65-3f2b53c4cb13>\u001b[0m in \u001b[0;36mfilter_on_tags\u001b[1;34m(features, tags)\u001b[0m\n\u001b[0;32m     31\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mfeature\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     32\u001b[0m     \u001b[1;31m# If none of the tags are there, return a random feature\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 33\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchoice\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfeatures\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     34\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     35\u001b[0m \u001b[0mseqrecords\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mmtrand.pyx\u001b[0m in \u001b[0;36mmtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:10276)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: a must be 1-dimensional or an integer"
     ]
    }
   ],
   "source": [
    "# %%time\n",
    "\n",
    "import itertools\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import generic_dna, generic_protein\n",
    "\n",
    "hg19_fasta = '/projects/ps-yeolab/genomes/hg19/gencode/v19/GRCh37.p13.genome.fa'\n",
    "\n",
    "\n",
    "best_tags = 'appris_principal', 'appris_candidate', 'CCDS', 'exp_conf', 'basic'\n",
    "\n",
    "def overlap(x, y):\n",
    "    return not ((x.start > y.stop) or (x.stop < y.start))\n",
    "\n",
    "\n",
    "def filter_on_tags(features, tags):\n",
    "    if len(features) == 0:\n",
    "        return None\n",
    "    \n",
    "    if len(features) == 1:\n",
    "        return features.pop()\n",
    "    \n",
    "    feature_tags = map(lambda x: (x, tuple(x.attributes['tag'])) \n",
    "                    if 'tag' in x.attributes else (x, ()), features)\n",
    "#     print 'feature_tags', feature_tags\n",
    "    for tag in tags:\n",
    "        for feature, feature_tag in feature_tags:\n",
    "            if tag in feature_tag:\n",
    "                return feature\n",
    "    # If none of the tags are there, return a random feature\n",
    "    return np.random.choice(features)\n",
    "\n",
    "seqrecords = []\n",
    "\n",
    "for i, row in se_consolidated_events.iterrows():\n",
    "    if (i+1) % 1000 == 0:\n",
    "        print i+1\n",
    "    exon1 = v19db[row.exon1]\n",
    "    exon2 = v19db[row.exon2]\n",
    "    exon3 = v19db[row.exon3]\n",
    "#     print row.event_id\n",
    "#     print '\\tjunction:', row.junction_exon13\n",
    "    \n",
    "    exon_trio = exon1, exon2, exon3\n",
    "    event_id = row.event_id\n",
    "    \n",
    "    isoform_to_exons = {'isoform1': (exon1, exon3), 'isoform2': exon_trio}\n",
    "    \n",
    "    transcripts = map(lambda x: set(v19db.parents(x, featuretype='transcript')), exon_trio)\n",
    "    \n",
    "    # Isoform 1: exclusion of exon2, so all transcripts that have exon1 and exon3 but not exon2\n",
    "    isoform1_transcripts = (transcripts[0] & transcripts[2]) - transcripts[1]\n",
    "#     print '\\tisoform1_transcripts', ','.join(map(lambda x: x.id, isoform1_transcripts))\n",
    "#     print '\\t', '\\t'.join(map(lambda x: str(list(x.attributes['tag'])) if 'tag' in x.attributes else '-', isoform1_transcripts))\n",
    "    \n",
    "    # Isoform 2: inclusion of exon2, so all transcripts that have exon1, exon2, and exon3\n",
    "    isoform2_transcripts = set(itertools.chain(*transcripts))\n",
    "#     print '\\tisoform2_transcripts', ','.join(map(lambda x: x.id, isoform2_transcripts))\n",
    "#     print '\\t', '\\t'.join(map(lambda x: str(list(x.attributes['tag'])) if 'tag' in x.attributes else '-', isoform2_transcripts))\n",
    "    \n",
    "    isoform1_transcript = filter_on_tags(isoform1_transcripts, best_tags)\n",
    "    isoform2_transcript = filter_on_tags(isoform2_transcripts, best_tags)\n",
    "    \n",
    "#     if isoform1_transcript is not None:\n",
    "#         print '\\t--isoform1', isoform1_transcript.id,\n",
    "#         if 'tag' in isoform1_transcript.attributes:\n",
    "#             print isoform1_transcript.attributes['tag']\n",
    "#         else:\n",
    "#             print\n",
    "#     if isoform2_transcript is not None:\n",
    "#         print '\\t--isoform2', isoform2_transcript.id,\n",
    "#         if 'tag' in isoform2_transcript.attributes:\n",
    "#             print isoform2_transcript.attributes['tag']\n",
    "#         else:\n",
    "#             print\n",
    "    \n",
    "#     isoforms = {'isoform1': isoform1_transcripts, \n",
    "#                 'isoform2': isoform2_transcripts}\n",
    "    isoforms = {'isoform1': (isoform1_transcript,), \n",
    "                'isoform2': (isoform2_transcript,)}\n",
    "    \n",
    "    \n",
    "#     isoform_to_cds = {'isoform1': [], 'isoform2': []}\n",
    "    for isoform, transcripts in isoforms.items():\n",
    "        exons = isoform_to_exons[isoform]\n",
    "#         print '\\t', '\\t'.join(map(lambda x: str(list(x.attributes['tag'])) if 'tag' in x.attributes else '-', exons))\n",
    "        \n",
    "        for transcript in transcripts:\n",
    "            if transcript is None:\n",
    "#                 print '\\t\\t', isoform, 'is None'\n",
    "                continue\n",
    "                \n",
    "            reverse = transcript.strand == '-'\n",
    "            cdss = v19db.children(transcript, featuretype='CDS', order_by='start', \n",
    "                                  reverse=reverse)\n",
    "            \n",
    "            cdss = filter(lambda cds: any(map(lambda exon: overlap(cds, exon), exons)), cdss)\n",
    "#             print cdss\n",
    "            if len(cdss) == len(exons):\n",
    "                cds_str = '@'.join(map(lambda x: x.id, cdss))\n",
    "#                 if cds_str in isoform_to_cds[isoform]:\n",
    "#                     continue\n",
    "                \n",
    "#                 isoform_to_cds[isoform].append(cds_str)\n",
    "                    \n",
    "#                 print '\\t\\t', '\\t'.join(map(lambda x: x.id, exons))\n",
    "#                 print '\\t\\t\\t', '\\t'.join(map(lambda x: x.id, cdss))\n",
    "#                 if 'tag' in transcript.attributes:\n",
    "#                     print '\\t\\t\\t', transcript.attributes['tag']\n",
    "#                 else:\n",
    "#                     print '\\t\\t\\t-'\n",
    "\n",
    "                if reverse:\n",
    "                    coding_sequence = Seq(''.join(cds.sequence(hg19_fasta)[::-1] for cds in cdss), generic_dna).complement()\n",
    "                else:\n",
    "                    coding_sequence = Seq(''.join(cds.sequence(hg19_fasta) for cds in cdss), generic_dna)\n",
    "#                 print len(coding_sequence)\n",
    "                coding_sequence = coding_sequence[int(cdss[0].frame):]\n",
    "                translated = coding_sequence.translate()\n",
    "                seqrecord = SeqRecord(translated, id='{0}|{1}|{2}'.format(event_id, cds_str, isoform))\n",
    "#                 print '\\t\\t', translated\n",
    "                seqrecords.append(seqrecord)\n",
    "#     pprint(isoform_to_cds)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open(translated_fasta, 'w') as f:\n",
    "    SeqIO.write(seqrecords, f, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoform1_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
