{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:784: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>junction_exon12</th>\n",
       "      <th>junction_exon23</th>\n",
       "      <th>junction_exon13</th>\n",
       "      <th>event_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "1  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "2  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "3  exon:chr10:100193697-100193848:-  exon:chr10:100190888-100191048:-   \n",
       "4  exon:chr10:100195392-100195529:-  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                              exon3              junction_exon12  \\\n",
       "0  exon:chr10:100188908-100189399:-  chr10:100189647-100190327:-   \n",
       "1  exon:chr10:100188913-100189399:-  chr10:100189647-100190327:-   \n",
       "2  exon:chr10:100189330-100189399:-  chr10:100189647-100190327:-   \n",
       "3  exon:chr10:100190328-100190427:-  chr10:100191049-100193696:-   \n",
       "4  exon:chr10:100193697-100193848:-  chr10:100195172-100195391:-   \n",
       "\n",
       "               junction_exon23              junction_exon13  \\\n",
       "0  chr10:100189400-100189547:-  chr10:100189400-100190327:-   \n",
       "1  chr10:100189400-100189547:-  chr10:100189400-100190327:-   \n",
       "2  chr10:100189400-100189547:-  chr10:100189400-100190327:-   \n",
       "3  chr10:100190428-100190887:-  chr10:100190428-100193696:-   \n",
       "4  chr10:100193849-100195028:-  chr10:100193849-100195391:-   \n",
       "\n",
       "                                            event_id  \n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...  \n",
       "1  exon:chr10:100190328-100190427:-@exon:chr10:10...  \n",
       "2  exon:chr10:100190328-100190427:-@exon:chr10:10...  \n",
       "3  exon:chr10:100193697-100193848:-@exon:chr10:10...  \n",
       "4  exon:chr10:100195392-100195529:-@exon:chr10:10...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "import pandas as pd\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "v19db_filename = '/projects/ps-yeolab/genomes/hg19/gencode/v19/gencode.v19.annotation.gtf.db'\n",
    "v19db = gffutils.FeatureDB(v19db_filename)\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "se_df = pd.read_csv('{}/exons_to_junctions_se.csv'.format(folder), index_col=0)\n",
    "se_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exon_cols = ['exon1', 'exon2', 'exon3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon4</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>event_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:102003455-102003529:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>chr10:101996727-102003454:-</td>\n",
       "      <td>chr10:101995522-101996606:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:101995522-101997778:-</td>\n",
       "      <td>exon:chr10:102003455-102003529:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>chr10:101996727-102003454:-</td>\n",
       "      <td>chr10:101995522-101996606:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:101995522-101997778:-</td>\n",
       "      <td>exon:chr10:102003455-102003534:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>exon:chr10:102010005-102010089:-</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>chr10:102003535-102013177:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:102010090-102013177:-</td>\n",
       "      <td>chr10:101997989-102010004:-</td>\n",
       "      <td>exon:chr10:102013178-102013296:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:102013178-102014376:-</td>\n",
       "      <td>exon:chr10:102010005-102010089:-</td>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>chr10:102003535-102013177:-</td>\n",
       "      <td>chr10:101997989-102003454:-</td>\n",
       "      <td>chr10:102010090-102013177:-</td>\n",
       "      <td>chr10:101997989-102010004:-</td>\n",
       "      <td>exon:chr10:102013178-102014376:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:102020723-102020785:-</td>\n",
       "      <td>exon:chr10:102019975-102020076:-</td>\n",
       "      <td>exon:chr10:102016019-102016233:-</td>\n",
       "      <td>exon:chr10:102013178-102013296:-</td>\n",
       "      <td>chr10:102016234-102020722:-</td>\n",
       "      <td>chr10:102013297-102016018:-</td>\n",
       "      <td>chr10:102020077-102020722:-</td>\n",
       "      <td>chr10:102013297-102019974:-</td>\n",
       "      <td>exon:chr10:102020723-102020785:-@exon:chr10:10...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:102003455-102003529:-  exon:chr10:101997779-101997988:-   \n",
       "1  exon:chr10:102003455-102003534:-  exon:chr10:101997779-101997988:-   \n",
       "2  exon:chr10:102013178-102013296:-  exon:chr10:102010005-102010089:-   \n",
       "3  exon:chr10:102013178-102014376:-  exon:chr10:102010005-102010089:-   \n",
       "4  exon:chr10:102020723-102020785:-  exon:chr10:102019975-102020076:-   \n",
       "\n",
       "                              exon3                             exon4  \\\n",
       "0  exon:chr10:101996607-101996726:-  exon:chr10:101995424-101995521:-   \n",
       "1  exon:chr10:101996607-101996726:-  exon:chr10:101995424-101995521:-   \n",
       "2  exon:chr10:102003455-102003534:-  exon:chr10:101997779-101997988:-   \n",
       "3  exon:chr10:102003455-102003534:-  exon:chr10:101997779-101997988:-   \n",
       "4  exon:chr10:102016019-102016233:-  exon:chr10:102013178-102013296:-   \n",
       "\n",
       "                    junction13                   junction34  \\\n",
       "0  chr10:101996727-102003454:-  chr10:101995522-101996606:-   \n",
       "1  chr10:101996727-102003454:-  chr10:101995522-101996606:-   \n",
       "2  chr10:102003535-102013177:-  chr10:101997989-102003454:-   \n",
       "3  chr10:102003535-102013177:-  chr10:101997989-102003454:-   \n",
       "4  chr10:102016234-102020722:-  chr10:102013297-102016018:-   \n",
       "\n",
       "                    junction12                   junction24  \\\n",
       "0  chr10:101997989-102003454:-  chr10:101995522-101997778:-   \n",
       "1  chr10:101997989-102003454:-  chr10:101995522-101997778:-   \n",
       "2  chr10:102010090-102013177:-  chr10:101997989-102010004:-   \n",
       "3  chr10:102010090-102013177:-  chr10:101997989-102010004:-   \n",
       "4  chr10:102020077-102020722:-  chr10:102013297-102019974:-   \n",
       "\n",
       "                                            event_id  \n",
       "0  exon:chr10:102003455-102003529:-@exon:chr10:10...  \n",
       "1  exon:chr10:102003455-102003534:-@exon:chr10:10...  \n",
       "2  exon:chr10:102013178-102013296:-@exon:chr10:10...  \n",
       "3  exon:chr10:102013178-102014376:-@exon:chr10:10...  \n",
       "4  exon:chr10:102020723-102020785:-@exon:chr10:10...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxe_df = pd.read_csv('/projects/ps-yeolab/obotvinnik/singlecell_pnms/exons_to_junctions_mxe.csv', index_col=0)\n",
    "mxe_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from poshsplice.region import Region\n",
    "\n",
    "def get_gene_attributes(df, exon_cols, db):\n",
    "    df['gencode_id'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[db[i].attributes['gene_id'] for i in x]))), axis=1)\n",
    "    df['gene_name'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[db[i].attributes['gene_name'] for i in x]))), axis=1)\n",
    "    df['ensembl_id'] = df[exon_cols].apply(\n",
    "        lambda x: ','.join(set(itertools.chain(*[map(lambda y: y.split('.')[0], db[i].attributes['gene_id']) for i in x]))), \n",
    "        axis=1)\n",
    "    \n",
    "    for exon_col in exon_cols:\n",
    "        df['{}_region'.format(exon_col)] = df[exon_col].map(Region)\n",
    "        df['{}_length'.format(exon_col)] = df['{}_region'.format(exon_col)].map(len)\n",
    "    df['strand'] = df[exon_cols[0]].str[-1]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mxe_exon_cols = ['exon{}'.format(i) for i in range(1, 5)]\n",
    "mxe_df_annotated = get_gene_attributes(mxe_df, mxe_exon_cols, v19db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "se_exon_cols = ['exon{}'.format(i) for i in range(1, 4)]\n",
    "se_df_annotated = get_gene_attributes(se_df, se_exon_cols, v19db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon4</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>event_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4_region</th>\n",
       "      <th>exon4_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5099</th>\n",
       "      <td>exon:chr15:72499069-72499139:-</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>chr15:72494962-72499068:-</td>\n",
       "      <td>chr15:72492997-72494794:-</td>\n",
       "      <td>chr15:72495530-72499068:-</td>\n",
       "      <td>chr15:72492997-72495362:-</td>\n",
       "      <td>exon:chr15:72499069-72499139:-@exon:chr15:7249...</td>\n",
       "      <td>ENSG00000067225.13</td>\n",
       "      <td>PKM</td>\n",
       "      <td>ENSG00000067225</td>\n",
       "      <td>exon:chr15:72499069-72499139:-</td>\n",
       "      <td>71</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5100</th>\n",
       "      <td>exon:chr15:72499069-72499195:-</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>chr15:72494962-72499068:-</td>\n",
       "      <td>chr15:72492997-72494794:-</td>\n",
       "      <td>chr15:72495530-72499068:-</td>\n",
       "      <td>chr15:72492997-72495362:-</td>\n",
       "      <td>exon:chr15:72499069-72499195:-@exon:chr15:7249...</td>\n",
       "      <td>ENSG00000067225.13</td>\n",
       "      <td>PKM</td>\n",
       "      <td>ENSG00000067225</td>\n",
       "      <td>exon:chr15:72499069-72499195:-</td>\n",
       "      <td>127</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5101</th>\n",
       "      <td>exon:chr15:72499069-72499221:-</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>chr15:72494962-72499068:-</td>\n",
       "      <td>chr15:72492997-72494794:-</td>\n",
       "      <td>chr15:72495530-72499068:-</td>\n",
       "      <td>chr15:72492997-72495362:-</td>\n",
       "      <td>exon:chr15:72499069-72499221:-@exon:chr15:7249...</td>\n",
       "      <td>ENSG00000067225.13</td>\n",
       "      <td>PKM</td>\n",
       "      <td>ENSG00000067225</td>\n",
       "      <td>exon:chr15:72499069-72499221:-</td>\n",
       "      <td>153</td>\n",
       "      <td>exon:chr15:72495363-72495529:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72494795-72494961:-</td>\n",
       "      <td>167</td>\n",
       "      <td>exon:chr15:72492815-72492996:-</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               exon1                           exon2  \\\n",
       "5099  exon:chr15:72499069-72499139:-  exon:chr15:72495363-72495529:-   \n",
       "5100  exon:chr15:72499069-72499195:-  exon:chr15:72495363-72495529:-   \n",
       "5101  exon:chr15:72499069-72499221:-  exon:chr15:72495363-72495529:-   \n",
       "\n",
       "                               exon3                           exon4  \\\n",
       "5099  exon:chr15:72494795-72494961:-  exon:chr15:72492815-72492996:-   \n",
       "5100  exon:chr15:72494795-72494961:-  exon:chr15:72492815-72492996:-   \n",
       "5101  exon:chr15:72494795-72494961:-  exon:chr15:72492815-72492996:-   \n",
       "\n",
       "                     junction13                 junction34  \\\n",
       "5099  chr15:72494962-72499068:-  chr15:72492997-72494794:-   \n",
       "5100  chr15:72494962-72499068:-  chr15:72492997-72494794:-   \n",
       "5101  chr15:72494962-72499068:-  chr15:72492997-72494794:-   \n",
       "\n",
       "                     junction12                 junction24  \\\n",
       "5099  chr15:72495530-72499068:-  chr15:72492997-72495362:-   \n",
       "5100  chr15:72495530-72499068:-  chr15:72492997-72495362:-   \n",
       "5101  chr15:72495530-72499068:-  chr15:72492997-72495362:-   \n",
       "\n",
       "                                               event_id          gencode_id  \\\n",
       "5099  exon:chr15:72499069-72499139:-@exon:chr15:7249...  ENSG00000067225.13   \n",
       "5100  exon:chr15:72499069-72499195:-@exon:chr15:7249...  ENSG00000067225.13   \n",
       "5101  exon:chr15:72499069-72499221:-@exon:chr15:7249...  ENSG00000067225.13   \n",
       "\n",
       "     gene_name       ensembl_id                    exon1_region  exon1_length  \\\n",
       "5099       PKM  ENSG00000067225  exon:chr15:72499069-72499139:-            71   \n",
       "5100       PKM  ENSG00000067225  exon:chr15:72499069-72499195:-           127   \n",
       "5101       PKM  ENSG00000067225  exon:chr15:72499069-72499221:-           153   \n",
       "\n",
       "                        exon2_region  exon2_length  \\\n",
       "5099  exon:chr15:72495363-72495529:-           167   \n",
       "5100  exon:chr15:72495363-72495529:-           167   \n",
       "5101  exon:chr15:72495363-72495529:-           167   \n",
       "\n",
       "                        exon3_region  exon3_length  \\\n",
       "5099  exon:chr15:72494795-72494961:-           167   \n",
       "5100  exon:chr15:72494795-72494961:-           167   \n",
       "5101  exon:chr15:72494795-72494961:-           167   \n",
       "\n",
       "                        exon4_region  exon4_length  \n",
       "5099  exon:chr15:72492815-72492996:-           182  \n",
       "5100  exon:chr15:72492815-72492996:-           182  \n",
       "5101  exon:chr15:72492815-72492996:-           182  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxe_df_annotated.loc[mxe_df_annotated.gene_name == 'PKM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>junctions</th>\n",
       "      <th>event_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron1_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>492</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>487</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>(chr10:100191049-100193696:-, chr10:100190428-...</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>2648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>(chr10:100195172-100195391:-, chr10:100193849-...</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>-</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "1  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "2  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "3  exon:chr10:100193697-100193848:-  exon:chr10:100190888-100191048:-   \n",
       "4  exon:chr10:100195392-100195529:-  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                              exon3  \\\n",
       "0  exon:chr10:100188908-100189399:-   \n",
       "1  exon:chr10:100188913-100189399:-   \n",
       "2  exon:chr10:100189330-100189399:-   \n",
       "3  exon:chr10:100190328-100190427:-   \n",
       "4  exon:chr10:100193697-100193848:-   \n",
       "\n",
       "                                           junctions  \\\n",
       "0  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "1  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "2  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "3  (chr10:100191049-100193696:-, chr10:100190428-...   \n",
       "4  (chr10:100195172-100195391:-, chr10:100193849-...   \n",
       "\n",
       "                                            event_id          gencode_id  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "1  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "2  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "3  exon:chr10:100193697-100193848:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "4  exon:chr10:100195392-100195529:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "\n",
       "  gene_name       ensembl_id  exon2_length                      exon1_region  \\\n",
       "0      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "1      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "2      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "3      HPS1  ENSG00000107521           161  exon:chr10:100193697-100193848:-   \n",
       "4      HPS1  ENSG00000107521           143  exon:chr10:100195392-100195529:-   \n",
       "\n",
       "   exon1_length                      exon2_region  \\\n",
       "0           100  exon:chr10:100189548-100189646:-   \n",
       "1           100  exon:chr10:100189548-100189646:-   \n",
       "2           100  exon:chr10:100189548-100189646:-   \n",
       "3           152  exon:chr10:100190888-100191048:-   \n",
       "4           138  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                       exon3_region  exon3_length strand  intron1_length  \n",
       "0  exon:chr10:100188908-100189399:-           492      -             681  \n",
       "1  exon:chr10:100188913-100189399:-           487      -             681  \n",
       "2  exon:chr10:100189330-100189399:-            70      -             681  \n",
       "3  exon:chr10:100190328-100190427:-           100      -            2648  \n",
       "4  exon:chr10:100193697-100193848:-           152      -             220  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_strand = exons_to_junctions.strand == '+'\n",
    "exons_to_junctions['intron1_length'] = np.nan\n",
    "\n",
    "\n",
    "exons_to_junctions.loc[positive_strand, 'intron1_length'] = \\\n",
    "    exons_to_junctions.loc[positive_strand].apply(lambda x: x.exon2_region._start - x.exon1_region._stop - 1, axis=1)\n",
    "exons_to_junctions.loc[~positive_strand, 'intron1_length'] = \\\n",
    "    exons_to_junctions.loc[~positive_strand].apply(lambda x: x.exon1_region._start - x.exon2_region._stop - 1, axis=1)\n",
    "exons_to_junctions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>junctions</th>\n",
       "      <th>event_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>492</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>487</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>(chr10:100191049-100193696:-, chr10:100190428-...</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>(chr10:100195172-100195391:-, chr10:100193849-...</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>-</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "1  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "2  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "3  exon:chr10:100193697-100193848:-  exon:chr10:100190888-100191048:-   \n",
       "4  exon:chr10:100195392-100195529:-  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                              exon3  \\\n",
       "0  exon:chr10:100188908-100189399:-   \n",
       "1  exon:chr10:100188913-100189399:-   \n",
       "2  exon:chr10:100189330-100189399:-   \n",
       "3  exon:chr10:100190328-100190427:-   \n",
       "4  exon:chr10:100193697-100193848:-   \n",
       "\n",
       "                                           junctions  \\\n",
       "0  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "1  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "2  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "3  (chr10:100191049-100193696:-, chr10:100190428-...   \n",
       "4  (chr10:100195172-100195391:-, chr10:100193849-...   \n",
       "\n",
       "                                            event_id          gencode_id  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "1  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "2  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "3  exon:chr10:100193697-100193848:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "4  exon:chr10:100195392-100195529:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "\n",
       "  gene_name       ensembl_id  exon2_length                      exon1_region  \\\n",
       "0      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "1      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "2      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "3      HPS1  ENSG00000107521           161  exon:chr10:100193697-100193848:-   \n",
       "4      HPS1  ENSG00000107521           143  exon:chr10:100195392-100195529:-   \n",
       "\n",
       "   exon1_length                      exon2_region  \\\n",
       "0           100  exon:chr10:100189548-100189646:-   \n",
       "1           100  exon:chr10:100189548-100189646:-   \n",
       "2           100  exon:chr10:100189548-100189646:-   \n",
       "3           152  exon:chr10:100190888-100191048:-   \n",
       "4           138  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                       exon3_region  exon3_length strand  intron1_length  \\\n",
       "0  exon:chr10:100188908-100189399:-           492      -             681   \n",
       "1  exon:chr10:100188913-100189399:-           487      -             681   \n",
       "2  exon:chr10:100189330-100189399:-            70      -             681   \n",
       "3  exon:chr10:100190328-100190427:-           100      -            2648   \n",
       "4  exon:chr10:100193697-100193848:-           152      -             220   \n",
       "\n",
       "   intron2_length  \n",
       "0             148  \n",
       "1             148  \n",
       "2             148  \n",
       "3             460  \n",
       "4            1180  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_strand = exons_to_junctions.strand == '+'\n",
    "exons_to_junctions['intron2_length'] = np.nan\n",
    "\n",
    "\n",
    "exons_to_junctions.loc[positive_strand, 'intron2_length'] = \\\n",
    "    exons_to_junctions.loc[positive_strand].apply(lambda x: x.exon3_region._start - x.exon2_region._stop - 1, axis=1)\n",
    "exons_to_junctions.loc[~positive_strand, 'intron2_length'] = \\\n",
    "    exons_to_junctions.loc[~positive_strand].apply(lambda x: x.exon2_region._start - x.exon3_region._stop - 1, axis=1)\n",
    "exons_to_junctions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>junctions</th>\n",
       "      <th>event_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "      <th>intron12_length_summed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>492</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>487</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>(chr10:100191049-100193696:-, chr10:100190428-...</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "      <td>3108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>(chr10:100195172-100195391:-, chr10:100193849-...</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>-</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "      <td>1400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "1  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "2  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "3  exon:chr10:100193697-100193848:-  exon:chr10:100190888-100191048:-   \n",
       "4  exon:chr10:100195392-100195529:-  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                              exon3  \\\n",
       "0  exon:chr10:100188908-100189399:-   \n",
       "1  exon:chr10:100188913-100189399:-   \n",
       "2  exon:chr10:100189330-100189399:-   \n",
       "3  exon:chr10:100190328-100190427:-   \n",
       "4  exon:chr10:100193697-100193848:-   \n",
       "\n",
       "                                           junctions  \\\n",
       "0  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "1  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "2  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "3  (chr10:100191049-100193696:-, chr10:100190428-...   \n",
       "4  (chr10:100195172-100195391:-, chr10:100193849-...   \n",
       "\n",
       "                                            event_id          gencode_id  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "1  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "2  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "3  exon:chr10:100193697-100193848:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "4  exon:chr10:100195392-100195529:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "\n",
       "  gene_name       ensembl_id  exon2_length                      exon1_region  \\\n",
       "0      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "1      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "2      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "3      HPS1  ENSG00000107521           161  exon:chr10:100193697-100193848:-   \n",
       "4      HPS1  ENSG00000107521           143  exon:chr10:100195392-100195529:-   \n",
       "\n",
       "   exon1_length                      exon2_region  \\\n",
       "0           100  exon:chr10:100189548-100189646:-   \n",
       "1           100  exon:chr10:100189548-100189646:-   \n",
       "2           100  exon:chr10:100189548-100189646:-   \n",
       "3           152  exon:chr10:100190888-100191048:-   \n",
       "4           138  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                       exon3_region  exon3_length strand  intron1_length  \\\n",
       "0  exon:chr10:100188908-100189399:-           492      -             681   \n",
       "1  exon:chr10:100188913-100189399:-           487      -             681   \n",
       "2  exon:chr10:100189330-100189399:-            70      -             681   \n",
       "3  exon:chr10:100190328-100190427:-           100      -            2648   \n",
       "4  exon:chr10:100193697-100193848:-           152      -             220   \n",
       "\n",
       "   intron2_length  intron12_length_summed  \n",
       "0             148                     829  \n",
       "1             148                     829  \n",
       "2             148                     829  \n",
       "3             460                    3108  \n",
       "4            1180                    1400  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_to_junctions['intron12_length_summed'] = exons_to_junctions.intron1_length + exons_to_junctions.intron2_length\n",
    "exons_to_junctions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon1                      object\n",
       "exon2                      object\n",
       "exon3                      object\n",
       "junctions                  object\n",
       "event_id                   object\n",
       "gencode_id                 object\n",
       "gene_name                  object\n",
       "ensembl_id                 object\n",
       "exon2_length                int64\n",
       "exon1_region               object\n",
       "exon1_length                int64\n",
       "exon2_region               object\n",
       "exon3_region               object\n",
       "exon3_length                int64\n",
       "strand                     object\n",
       "intron1_length            float64\n",
       "intron2_length            float64\n",
       "intron12_length_summed    float64\n",
       "exon2_divisible_by_3         bool\n",
       "dtype: object"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_to_junctions.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>junctions</th>\n",
       "      <th>event_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon1_region</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_region</th>\n",
       "      <th>exon3_region</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "      <th>intron12_length_summed</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188908-100189399:-</td>\n",
       "      <td>492</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>829</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100188913-100189399:-</td>\n",
       "      <td>487</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>829</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>(chr10:100189647-100190327:-, chr10:100189400-...</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>70</td>\n",
       "      <td>-</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>829</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>(chr10:100191049-100193696:-, chr10:100190428-...</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "      <td>3108</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>(chr10:100195172-100195391:-, chr10:100193849-...</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>-</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "      <td>1400</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              exon1                             exon2  \\\n",
       "0  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "1  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "2  exon:chr10:100190328-100190427:-  exon:chr10:100189548-100189646:-   \n",
       "3  exon:chr10:100193697-100193848:-  exon:chr10:100190888-100191048:-   \n",
       "4  exon:chr10:100195392-100195529:-  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                              exon3  \\\n",
       "0  exon:chr10:100188908-100189399:-   \n",
       "1  exon:chr10:100188913-100189399:-   \n",
       "2  exon:chr10:100189330-100189399:-   \n",
       "3  exon:chr10:100190328-100190427:-   \n",
       "4  exon:chr10:100193697-100193848:-   \n",
       "\n",
       "                                           junctions  \\\n",
       "0  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "1  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "2  (chr10:100189647-100190327:-, chr10:100189400-...   \n",
       "3  (chr10:100191049-100193696:-, chr10:100190428-...   \n",
       "4  (chr10:100195172-100195391:-, chr10:100193849-...   \n",
       "\n",
       "                                            event_id          gencode_id  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "1  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "2  exon:chr10:100190328-100190427:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "3  exon:chr10:100193697-100193848:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "4  exon:chr10:100195392-100195529:-@exon:chr10:10...  ENSG00000107521.14   \n",
       "\n",
       "  gene_name       ensembl_id  exon2_length                      exon1_region  \\\n",
       "0      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "1      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "2      HPS1  ENSG00000107521            99  exon:chr10:100190328-100190427:-   \n",
       "3      HPS1  ENSG00000107521           161  exon:chr10:100193697-100193848:-   \n",
       "4      HPS1  ENSG00000107521           143  exon:chr10:100195392-100195529:-   \n",
       "\n",
       "   exon1_length                      exon2_region  \\\n",
       "0           100  exon:chr10:100189548-100189646:-   \n",
       "1           100  exon:chr10:100189548-100189646:-   \n",
       "2           100  exon:chr10:100189548-100189646:-   \n",
       "3           152  exon:chr10:100190888-100191048:-   \n",
       "4           138  exon:chr10:100195029-100195171:-   \n",
       "\n",
       "                       exon3_region  exon3_length strand  intron1_length  \\\n",
       "0  exon:chr10:100188908-100189399:-           492      -             681   \n",
       "1  exon:chr10:100188913-100189399:-           487      -             681   \n",
       "2  exon:chr10:100189330-100189399:-            70      -             681   \n",
       "3  exon:chr10:100190328-100190427:-           100      -            2648   \n",
       "4  exon:chr10:100193697-100193848:-           152      -             220   \n",
       "\n",
       "   intron2_length  intron12_length_summed exon2_divisible_by_3  \n",
       "0             148                     829                 True  \n",
       "1             148                     829                 True  \n",
       "2             148                     829                 True  \n",
       "3             460                    3108                False  \n",
       "4            1180                    1400                False  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_to_junctions['exon2_divisible_by_3'] = exons_to_junctions['exon2_length'].astype(int).mod(3) == 0\n",
    "exons_to_junctions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.38737636382175999"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_to_junctions['exon2_divisible_by_3'].sum()/float(exons_to_junctions.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>intron12_length_summed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188908-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>492</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188913-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>487</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>70</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>152</td>\n",
       "      <td>161</td>\n",
       "      <td>100</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "      <td>False</td>\n",
       "      <td>3108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>138</td>\n",
       "      <td>143</td>\n",
       "      <td>152</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "      <td>False</td>\n",
       "      <td>1400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "\n",
       "                                                         ensembl_id gene_name  \\\n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "\n",
       "                                                    exon3_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           492   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           487   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   \n",
       "\n",
       "                                                    intron1_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...            2648   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...             220   \n",
       "\n",
       "                                                    intron2_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...             460   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...            1180   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "\n",
       "                                                    intron12_length_summed  \n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                    3108  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                    1400  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = ['event_id', 'gencode_id', 'ensembl_id', 'gene_name', 'exon1_length', \n",
    "        'exon2_length', 'exon3_length', 'intron1_length', \n",
    "        'intron2_length', 'exon2_divisible_by_3', 'intron12_length_summed']\n",
    "exon_metadata = exons_to_junctions[cols]\n",
    "exon_metadata = exon_metadata.set_index('event_id')\n",
    "exon_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3928\n",
      "(58842, 10)\n"
     ]
    }
   ],
   "source": [
    "print exon_metadata.ensembl_id.str.contains(',').sum()\n",
    "print exon_metadata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3,928/58,842 splicing events have multiple gene ids associated with them. Since we want to intersect this information with existing gene annotations, we need a *single* id. To do this coarsely and simply, we will take the first gene ID for intersecting purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>intron12_length_summed</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188908-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>492</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188913-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>487</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>70</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>152</td>\n",
       "      <td>161</td>\n",
       "      <td>100</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "      <td>False</td>\n",
       "      <td>3108</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>138</td>\n",
       "      <td>143</td>\n",
       "      <td>152</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "      <td>False</td>\n",
       "      <td>1400</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "\n",
       "                                                         ensembl_id gene_name  \\\n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "\n",
       "                                                    exon3_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           492   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           487   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   \n",
       "\n",
       "                                                    intron1_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...            2648   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...             220   \n",
       "\n",
       "                                                    intron2_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...             460   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...            1180   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "\n",
       "                                                    intron12_length_summed  \\\n",
       "event_id                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                    3108   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                    1400   \n",
       "\n",
       "                                                     one_ensembl_id  \n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_metadata['one_ensembl_id'] = exon_metadata.ensembl_id.str.split(',').str[0]\n",
    "exon_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write to a safe place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exon_metadata.to_csv('{}/splicing_feature_data.csv'.format(folder))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
