{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ancient alternatively spliced exons \n",
    "\n",
    "From Merkin 2012\n",
    "\n",
    "\"Ancient\" is relative here - the authors compared tissue-specific splicing across several mammals (cow, rhesus, mouse, rat) with chicken as the outgroup (estimated as ~300 million years apart in evolutionary time by fossil record), and exons that were alternatively spliced in all species were deemed \"ancient\" alternatively spliced exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab/obotvinnik/merkin2012\n"
     ]
    }
   ],
   "source": [
    "cd /projects/ps-yeolab/obotvinnik/merkin2012/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:784: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTPRD:ENSBTAG00000010178:8:38452684:38452989:+</td>\n",
       "      <td>PTPRD:ENSMMUG00000006952:15:68642301:68642606:+</td>\n",
       "      <td>Ptprd:ENSMUSG00000028399:4:75746351:75746657:-</td>\n",
       "      <td>Q62990_RAT:ENSRNOG00000005711:5:94111899:94112...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UCKL1:ENSBTAG00000015798:13:54576868:54576887:+</td>\n",
       "      <td>UCKL1:ENSMMUG00000015551:10:402713:402732:+</td>\n",
       "      <td>Uckl1:ENSMUSG00000089917:2:181305943:181305962:-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KANK1:ENSBTAG00000010976:8:45921894:45922136:-</td>\n",
       "      <td>KANK1:ENSMMUG00000019862:15:76468463:76468702:-</td>\n",
       "      <td>Kank1:ENSMUSG00000032702:19:25498628:25498873:+</td>\n",
       "      <td>Kank1:ENSRNOG00000016023:1:228857621:228858097:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q29RZ7_BOVIN:ENSBTAG00000014786:22:49256956:49...</td>\n",
       "      <td>Pbrm1:ENSMUSG00000042323:14:31858516:31858611:+</td>\n",
       "      <td>Pbrm1:ENSRNOG00000028227:16:6478444:6478539:+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM179B:ENSBTAG00000039117:21:55239625:55239777:+</td>\n",
       "      <td>Fam179b:ENSMUSG00000035614:12:66091928:66092077:+</td>\n",
       "      <td>Fam179b:ENSRNOG00000004415:6:86344755:86344904:+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0  \\\n",
       "0     PTPRD:ENSBTAG00000010178:8:38452684:38452989:+   \n",
       "1    UCKL1:ENSBTAG00000015798:13:54576868:54576887:+   \n",
       "2     KANK1:ENSBTAG00000010976:8:45921894:45922136:-   \n",
       "3  Q29RZ7_BOVIN:ENSBTAG00000014786:22:49256956:49...   \n",
       "4  FAM179B:ENSBTAG00000039117:21:55239625:55239777:+   \n",
       "\n",
       "                                                   1  \\\n",
       "0    PTPRD:ENSMMUG00000006952:15:68642301:68642606:+   \n",
       "1        UCKL1:ENSMMUG00000015551:10:402713:402732:+   \n",
       "2    KANK1:ENSMMUG00000019862:15:76468463:76468702:-   \n",
       "3    Pbrm1:ENSMUSG00000042323:14:31858516:31858611:+   \n",
       "4  Fam179b:ENSMUSG00000035614:12:66091928:66092077:+   \n",
       "\n",
       "                                                  2  \\\n",
       "0    Ptprd:ENSMUSG00000028399:4:75746351:75746657:-   \n",
       "1  Uckl1:ENSMUSG00000089917:2:181305943:181305962:-   \n",
       "2   Kank1:ENSMUSG00000032702:19:25498628:25498873:+   \n",
       "3     Pbrm1:ENSRNOG00000028227:16:6478444:6478539:+   \n",
       "4  Fam179b:ENSRNOG00000004415:6:86344755:86344904:+   \n",
       "\n",
       "                                                   3  \n",
       "0  Q62990_RAT:ENSRNOG00000005711:5:94111899:94112...  \n",
       "1                                                NaN  \n",
       "2   Kank1:ENSRNOG00000016023:1:228857621:228858097:+  \n",
       "3                                                NaN  \n",
       "4                                                NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_excel('1228186stable4.xls', header=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "df = df.replace('-', np.nan)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get macaque to human liftover\n",
    "\n",
    "genomes:\n",
    "\n",
    "> The 2x80 base libraries were mapped to their respective genomes (musmus9, rhemac2, ratnor4, bostau4, galgal3) using Tophat (35) version 1.1.4 and Ensembl Release 61 (Ensembl) annotations\n",
    "\n",
    "From Gene:\n",
    " - liftover mouse gff to human\n",
    "\t- Check \n",
    "\t\t- [ ] AG/GT splice sites\n",
    "\t\t- [ ] Orthologous Gene by ENSEMBL compara"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab/genomes/rheMac\n"
     ]
    }
   ],
   "source": [
    "cd /projects/ps-yeolab/genomes/rheMac/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `rheMac2': File exists\r\n"
     ]
    }
   ],
   "source": [
    "mkdir rheMac2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab/genomes/rheMac/rheMac2\n"
     ]
    }
   ],
   "source": [
    "cd rheMac2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2015-06-02 15:27:19--  http://hgdownload.cse.ucsc.edu/goldenPath/rheMac2/liftOver/rheMac2ToHg19.over.chain.gz\n",
      "Resolving hgdownload.cse.ucsc.edu... 128.114.119.163\n",
      "Connecting to hgdownload.cse.ucsc.edu|128.114.119.163|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 34213302 (33M) [application/x-gzip]\n",
      "Saving to: `rheMac2ToHg19.over.chain.gz'\n",
      "\n",
      "100%[======================================>] 34,213,302  62.4M/s   in 0.5s    \n",
      "\n",
      "2015-06-02 15:27:20 (62.4 MB/s) - `rheMac2ToHg19.over.chain.gz' saved [34213302/34213302]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget http://hgdownload.cse.ucsc.edu/goldenPath/rheMac2/liftOver/rheMac2ToHg19.over.chain.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rheMac_series = df[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rhemac_df = rheMac_series.str.split(':').apply(pd.Series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0  </th>\n",
       "      <td>        PTPRD</td>\n",
       "      <td> ENSMMUG00000006952</td>\n",
       "      <td> 15</td>\n",
       "      <td>  68642301</td>\n",
       "      <td>  68642606</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1  </th>\n",
       "      <td>        UCKL1</td>\n",
       "      <td> ENSMMUG00000015551</td>\n",
       "      <td> 10</td>\n",
       "      <td>    402713</td>\n",
       "      <td>    402732</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2  </th>\n",
       "      <td>        KANK1</td>\n",
       "      <td> ENSMMUG00000019862</td>\n",
       "      <td> 15</td>\n",
       "      <td>  76468463</td>\n",
       "      <td>  76468702</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3  </th>\n",
       "      <td>        Pbrm1</td>\n",
       "      <td> ENSMUSG00000042323</td>\n",
       "      <td> 14</td>\n",
       "      <td>  31858516</td>\n",
       "      <td>  31858611</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4  </th>\n",
       "      <td>      Fam179b</td>\n",
       "      <td> ENSMUSG00000035614</td>\n",
       "      <td> 12</td>\n",
       "      <td>  66091928</td>\n",
       "      <td>  66092077</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5  </th>\n",
       "      <td>         ANK3</td>\n",
       "      <td> ENSMMUG00000003589</td>\n",
       "      <td>  9</td>\n",
       "      <td>  77118005</td>\n",
       "      <td>  77118311</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6  </th>\n",
       "      <td>    LOC723356</td>\n",
       "      <td> ENSMMUG00000003302</td>\n",
       "      <td> 18</td>\n",
       "      <td>  27874327</td>\n",
       "      <td>  27874375</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7  </th>\n",
       "      <td>        EFHA1</td>\n",
       "      <td> ENSMMUG00000011252</td>\n",
       "      <td> 17</td>\n",
       "      <td>   2170139</td>\n",
       "      <td>   2170174</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8  </th>\n",
       "      <td>        SYNJ2</td>\n",
       "      <td> ENSMMUG00000007153</td>\n",
       "      <td>  4</td>\n",
       "      <td> 104843020</td>\n",
       "      <td> 104843154</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9  </th>\n",
       "      <td>         KTN1</td>\n",
       "      <td> ENSMMUG00000006650</td>\n",
       "      <td>  7</td>\n",
       "      <td> 118677507</td>\n",
       "      <td> 118677596</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10 </th>\n",
       "      <td>         Bptf</td>\n",
       "      <td> ENSMUSG00000040481</td>\n",
       "      <td> 11</td>\n",
       "      <td> 106956316</td>\n",
       "      <td> 106956504</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11 </th>\n",
       "      <td>     Epb4.1l3</td>\n",
       "      <td> ENSMUSG00000024044</td>\n",
       "      <td> 17</td>\n",
       "      <td>  69623441</td>\n",
       "      <td>  69623506</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12 </th>\n",
       "      <td>       STXBP5</td>\n",
       "      <td> ENSMMUG00000017082</td>\n",
       "      <td>  4</td>\n",
       "      <td> 116002128</td>\n",
       "      <td> 116002187</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13 </th>\n",
       "      <td>     PPP1R12A</td>\n",
       "      <td> ENSMMUG00000019214</td>\n",
       "      <td> 11</td>\n",
       "      <td>  76756990</td>\n",
       "      <td>  76757020</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14 </th>\n",
       "      <td>        EDEM3</td>\n",
       "      <td> ENSMMUG00000005410</td>\n",
       "      <td>  1</td>\n",
       "      <td> 185977189</td>\n",
       "      <td> 185977236</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15 </th>\n",
       "      <td>         MTDH</td>\n",
       "      <td> ENSMMUG00000014210</td>\n",
       "      <td>  8</td>\n",
       "      <td> 100229414</td>\n",
       "      <td> 100229512</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16 </th>\n",
       "      <td>       SEMA6A</td>\n",
       "      <td> ENSMMUG00000012265</td>\n",
       "      <td>  6</td>\n",
       "      <td> 112974872</td>\n",
       "      <td> 112974922</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17 </th>\n",
       "      <td>        ACSS2</td>\n",
       "      <td> ENSMMUG00000017932</td>\n",
       "      <td> 10</td>\n",
       "      <td>  29644669</td>\n",
       "      <td>  29645429</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18 </th>\n",
       "      <td>     TP53INP1</td>\n",
       "      <td> ENSMMUG00000005360</td>\n",
       "      <td>  8</td>\n",
       "      <td>  97506088</td>\n",
       "      <td>  97506125</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19 </th>\n",
       "      <td>        PTPRM</td>\n",
       "      <td> ENSMMUG00000000768</td>\n",
       "      <td> 18</td>\n",
       "      <td>   6109310</td>\n",
       "      <td>   6109336</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20 </th>\n",
       "      <td>        PTPRM</td>\n",
       "      <td> ENSMMUG00000000768</td>\n",
       "      <td> 18</td>\n",
       "      <td>   6235336</td>\n",
       "      <td>   6235410</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21 </th>\n",
       "      <td>        PI4KB</td>\n",
       "      <td> ENSMMUG00000013170</td>\n",
       "      <td>  1</td>\n",
       "      <td> 129745837</td>\n",
       "      <td> 129745881</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22 </th>\n",
       "      <td>        HERC4</td>\n",
       "      <td> ENSMMUG00000010105</td>\n",
       "      <td>  9</td>\n",
       "      <td>  69290426</td>\n",
       "      <td>  69290925</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23 </th>\n",
       "      <td>    LOC723689</td>\n",
       "      <td> ENSMMUG00000010767</td>\n",
       "      <td>  7</td>\n",
       "      <td> 131988881</td>\n",
       "      <td> 131988965</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24 </th>\n",
       "      <td>    LOC723689</td>\n",
       "      <td> ENSMMUG00000010767</td>\n",
       "      <td>  7</td>\n",
       "      <td> 131988351</td>\n",
       "      <td> 131988416</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25 </th>\n",
       "      <td> Q3YAN5_MACMU</td>\n",
       "      <td> ENSMMUG00000008379</td>\n",
       "      <td>  1</td>\n",
       "      <td>   1036668</td>\n",
       "      <td>   1036760</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26 </th>\n",
       "      <td>      SLC25A3</td>\n",
       "      <td> ENSMMUG00000023036</td>\n",
       "      <td> 11</td>\n",
       "      <td>  99781096</td>\n",
       "      <td>  99781220</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27 </th>\n",
       "      <td>         OGDH</td>\n",
       "      <td> ENSMMUG00000002651</td>\n",
       "      <td>  3</td>\n",
       "      <td>  81696427</td>\n",
       "      <td>  81696529</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28 </th>\n",
       "      <td>    LOC714322</td>\n",
       "      <td> ENSMMUG00000019672</td>\n",
       "      <td> 14</td>\n",
       "      <td>  25457187</td>\n",
       "      <td>  25457331</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29 </th>\n",
       "      <td>       BAT2L1</td>\n",
       "      <td> ENSMMUG00000009973</td>\n",
       "      <td> 15</td>\n",
       "      <td>   7197573</td>\n",
       "      <td>   7197661</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>        USP28</td>\n",
       "      <td> ENSMMUG00000023773</td>\n",
       "      <td> 14</td>\n",
       "      <td> 112282959</td>\n",
       "      <td> 112283144</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>         NEO1</td>\n",
       "      <td> ENSMMUG00000009584</td>\n",
       "      <td>  7</td>\n",
       "      <td>  51965738</td>\n",
       "      <td>  51965770</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>        EPB41</td>\n",
       "      <td> ENSMMUG00000011377</td>\n",
       "      <td>  1</td>\n",
       "      <td>  31796471</td>\n",
       "      <td>  31796527</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511</th>\n",
       "      <td>       Epb4.1</td>\n",
       "      <td> ENSMUSG00000028906</td>\n",
       "      <td>  4</td>\n",
       "      <td> 131519844</td>\n",
       "      <td> 131519885</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>         EML1</td>\n",
       "      <td> ENSMMUG00000011706</td>\n",
       "      <td>  7</td>\n",
       "      <td> 163157051</td>\n",
       "      <td> 163157107</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>     ARHGAP26</td>\n",
       "      <td> ENSMMUG00000000419</td>\n",
       "      <td>  6</td>\n",
       "      <td> 139638260</td>\n",
       "      <td> 139647215</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>       R3HDM2</td>\n",
       "      <td> ENSMMUG00000021789</td>\n",
       "      <td> 11</td>\n",
       "      <td>  54361203</td>\n",
       "      <td>  54361256</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>         CROP</td>\n",
       "      <td> ENSMMUG00000010117</td>\n",
       "      <td> 16</td>\n",
       "      <td>  34993947</td>\n",
       "      <td>  34994012</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>         NEBL</td>\n",
       "      <td> ENSMMUG00000010589</td>\n",
       "      <td>  9</td>\n",
       "      <td>  21041885</td>\n",
       "      <td>  21041977</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>         NEBL</td>\n",
       "      <td> ENSMMUG00000010589</td>\n",
       "      <td>  9</td>\n",
       "      <td>  21040683</td>\n",
       "      <td>  21040745</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>         TJP1</td>\n",
       "      <td> ENSMMUG00000017878</td>\n",
       "      <td>  7</td>\n",
       "      <td>   8806624</td>\n",
       "      <td>   8806863</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>         NEBL</td>\n",
       "      <td> ENSMMUG00000010589</td>\n",
       "      <td>  9</td>\n",
       "      <td>  21086655</td>\n",
       "      <td>  21086759</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>        ATXN2</td>\n",
       "      <td> ENSMMUG00000002226</td>\n",
       "      <td> 11</td>\n",
       "      <td> 112405606</td>\n",
       "      <td> 112405696</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>         RBM5</td>\n",
       "      <td> ENSMMUG00000004921</td>\n",
       "      <td>  2</td>\n",
       "      <td>  86319462</td>\n",
       "      <td>  86319535</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>      RALGAPB</td>\n",
       "      <td> ENSMMUG00000017350</td>\n",
       "      <td> 10</td>\n",
       "      <td>  25926587</td>\n",
       "      <td>  25926622</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>        FNBP1</td>\n",
       "      <td> ENSMMUG00000003799</td>\n",
       "      <td> 15</td>\n",
       "      <td>   8917701</td>\n",
       "      <td>   8922201</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>        RUFY3</td>\n",
       "      <td> ENSMMUG00000004465</td>\n",
       "      <td>  5</td>\n",
       "      <td>  58897337</td>\n",
       "      <td>  58897390</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td> Q9BDP7_MACMU</td>\n",
       "      <td> ENSMMUG00000004577</td>\n",
       "      <td>  4</td>\n",
       "      <td>  43691711</td>\n",
       "      <td>  43693053</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>        Crim1</td>\n",
       "      <td> ENSMUSG00000024074</td>\n",
       "      <td> 17</td>\n",
       "      <td>  78679313</td>\n",
       "      <td>  78679555</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>    LOC721513</td>\n",
       "      <td> ENSMMUG00000009658</td>\n",
       "      <td> 10</td>\n",
       "      <td>  49701799</td>\n",
       "      <td>  49701831</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>         CLTA</td>\n",
       "      <td> ENSMMUG00000006132</td>\n",
       "      <td> 15</td>\n",
       "      <td>  41260865</td>\n",
       "      <td>  41260918</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>         CLTA</td>\n",
       "      <td> ENSMMUG00000006132</td>\n",
       "      <td> 15</td>\n",
       "      <td>  41259546</td>\n",
       "      <td>  41259581</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>          DMD</td>\n",
       "      <td> ENSMMUG00000019345</td>\n",
       "      <td>  X</td>\n",
       "      <td>  29054887</td>\n",
       "      <td>  29054925</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td> A2D6D3_MACMU</td>\n",
       "      <td> ENSMMUG00000016917</td>\n",
       "      <td>  2</td>\n",
       "      <td> 175891056</td>\n",
       "      <td> 175891163</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>        MYO9B</td>\n",
       "      <td> ENSMMUG00000011588</td>\n",
       "      <td> 19</td>\n",
       "      <td>  16724861</td>\n",
       "      <td>  16725489</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>        SSBP3</td>\n",
       "      <td> ENSMMUG00000020976</td>\n",
       "      <td>  1</td>\n",
       "      <td>  57134755</td>\n",
       "      <td>  57134835</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>      SH3KBP1</td>\n",
       "      <td> ENSMMUG00000004744</td>\n",
       "      <td>  X</td>\n",
       "      <td>  17435360</td>\n",
       "      <td>  17435434</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>        MYO9B</td>\n",
       "      <td> ENSMMUG00000011588</td>\n",
       "      <td> 19</td>\n",
       "      <td>  16759109</td>\n",
       "      <td>  16759156</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>     ARHGEF11</td>\n",
       "      <td> ENSMMUG00000012436</td>\n",
       "      <td>  1</td>\n",
       "      <td> 135565221</td>\n",
       "      <td> 135565250</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>         DTX2</td>\n",
       "      <td> ENSMMUG00000017570</td>\n",
       "      <td>  3</td>\n",
       "      <td>  49873216</td>\n",
       "      <td>  49874395</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>538 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                0                   1   2          3          4  5\n",
       "0           PTPRD  ENSMMUG00000006952  15   68642301   68642606  +\n",
       "1           UCKL1  ENSMMUG00000015551  10     402713     402732  +\n",
       "2           KANK1  ENSMMUG00000019862  15   76468463   76468702  -\n",
       "3           Pbrm1  ENSMUSG00000042323  14   31858516   31858611  +\n",
       "4         Fam179b  ENSMUSG00000035614  12   66091928   66092077  +\n",
       "5            ANK3  ENSMMUG00000003589   9   77118005   77118311  +\n",
       "6       LOC723356  ENSMMUG00000003302  18   27874327   27874375  +\n",
       "7           EFHA1  ENSMMUG00000011252  17    2170139    2170174  -\n",
       "8           SYNJ2  ENSMMUG00000007153   4  104843020  104843154  -\n",
       "9            KTN1  ENSMMUG00000006650   7  118677507  118677596  +\n",
       "10           Bptf  ENSMUSG00000040481  11  106956316  106956504  -\n",
       "11       Epb4.1l3  ENSMUSG00000024044  17   69623441   69623506  +\n",
       "12         STXBP5  ENSMMUG00000017082   4  116002128  116002187  -\n",
       "13       PPP1R12A  ENSMMUG00000019214  11   76756990   76757020  -\n",
       "14          EDEM3  ENSMMUG00000005410   1  185977189  185977236  +\n",
       "15           MTDH  ENSMMUG00000014210   8  100229414  100229512  +\n",
       "16         SEMA6A  ENSMMUG00000012265   6  112974872  112974922  -\n",
       "17          ACSS2  ENSMMUG00000017932  10   29644669   29645429  -\n",
       "18       TP53INP1  ENSMMUG00000005360   8   97506088   97506125  -\n",
       "19          PTPRM  ENSMMUG00000000768  18    6109310    6109336  -\n",
       "20          PTPRM  ENSMMUG00000000768  18    6235336    6235410  -\n",
       "21          PI4KB  ENSMMUG00000013170   1  129745837  129745881  -\n",
       "22          HERC4  ENSMMUG00000010105   9   69290426   69290925  +\n",
       "23      LOC723689  ENSMMUG00000010767   7  131988881  131988965  -\n",
       "24      LOC723689  ENSMMUG00000010767   7  131988351  131988416  -\n",
       "25   Q3YAN5_MACMU  ENSMMUG00000008379   1    1036668    1036760  -\n",
       "26        SLC25A3  ENSMMUG00000023036  11   99781096   99781220  +\n",
       "27           OGDH  ENSMMUG00000002651   3   81696427   81696529  -\n",
       "28      LOC714322  ENSMMUG00000019672  14   25457187   25457331  -\n",
       "29         BAT2L1  ENSMMUG00000009973  15    7197573    7197661  -\n",
       "..            ...                 ...  ..        ...        ... ..\n",
       "508         USP28  ENSMMUG00000023773  14  112282959  112283144  -\n",
       "509          NEO1  ENSMMUG00000009584   7   51965738   51965770  +\n",
       "510         EPB41  ENSMMUG00000011377   1   31796471   31796527  +\n",
       "511        Epb4.1  ENSMUSG00000028906   4  131519844  131519885  -\n",
       "512          EML1  ENSMMUG00000011706   7  163157051  163157107  +\n",
       "513      ARHGAP26  ENSMMUG00000000419   6  139638260  139647215  +\n",
       "514        R3HDM2  ENSMMUG00000021789  11   54361203   54361256  -\n",
       "515          CROP  ENSMMUG00000010117  16   34993947   34994012  +\n",
       "516          NEBL  ENSMMUG00000010589   9   21041885   21041977  -\n",
       "517          NEBL  ENSMMUG00000010589   9   21040683   21040745  -\n",
       "518          TJP1  ENSMMUG00000017878   7    8806624    8806863  -\n",
       "519          NEBL  ENSMMUG00000010589   9   21086655   21086759  -\n",
       "520         ATXN2  ENSMMUG00000002226  11  112405606  112405696  -\n",
       "521          RBM5  ENSMMUG00000004921   2   86319462   86319535  -\n",
       "522       RALGAPB  ENSMMUG00000017350  10   25926587   25926622  -\n",
       "523         FNBP1  ENSMMUG00000003799  15    8917701    8922201  +\n",
       "524         RUFY3  ENSMMUG00000004465   5   58897337   58897390  -\n",
       "525  Q9BDP7_MACMU  ENSMMUG00000004577   4   43691711   43693053  +\n",
       "526         Crim1  ENSMUSG00000024074  17   78679313   78679555  +\n",
       "527     LOC721513  ENSMMUG00000009658  10   49701799   49701831  -\n",
       "528          CLTA  ENSMMUG00000006132  15   41260865   41260918  -\n",
       "529          CLTA  ENSMMUG00000006132  15   41259546   41259581  -\n",
       "530           DMD  ENSMMUG00000019345   X   29054887   29054925  -\n",
       "531  A2D6D3_MACMU  ENSMMUG00000016917   2  175891056  175891163  -\n",
       "532         MYO9B  ENSMMUG00000011588  19   16724861   16725489  +\n",
       "533         SSBP3  ENSMMUG00000020976   1   57134755   57134835  -\n",
       "534       SH3KBP1  ENSMMUG00000004744   X   17435360   17435434  -\n",
       "535         MYO9B  ENSMMUG00000011588  19   16759109   16759156  +\n",
       "536      ARHGEF11  ENSMMUG00000012436   1  135565221  135565250  -\n",
       "537          DTX2  ENSMMUG00000017570   3   49873216   49874395  -\n",
       "\n",
       "[538 rows x 6 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rhemac_df = rhemac_df.dropna()\n",
    "rhemac_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>   PTPRD</td>\n",
       "      <td> ENSMMUG00000006952</td>\n",
       "      <td> 15</td>\n",
       "      <td> 68642301</td>\n",
       "      <td> 68642606</td>\n",
       "      <td> +</td>\n",
       "      <td> chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>   UCKL1</td>\n",
       "      <td> ENSMMUG00000015551</td>\n",
       "      <td> 10</td>\n",
       "      <td>   402713</td>\n",
       "      <td>   402732</td>\n",
       "      <td> +</td>\n",
       "      <td> chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>   KANK1</td>\n",
       "      <td> ENSMMUG00000019862</td>\n",
       "      <td> 15</td>\n",
       "      <td> 76468463</td>\n",
       "      <td> 76468702</td>\n",
       "      <td> -</td>\n",
       "      <td> chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>   Pbrm1</td>\n",
       "      <td> ENSMUSG00000042323</td>\n",
       "      <td> 14</td>\n",
       "      <td> 31858516</td>\n",
       "      <td> 31858611</td>\n",
       "      <td> +</td>\n",
       "      <td> chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> Fam179b</td>\n",
       "      <td> ENSMUSG00000035614</td>\n",
       "      <td> 12</td>\n",
       "      <td> 66091928</td>\n",
       "      <td> 66092077</td>\n",
       "      <td> +</td>\n",
       "      <td> chr12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0                   1   2         3         4  5  chrom\n",
       "0    PTPRD  ENSMMUG00000006952  15  68642301  68642606  +  chr15\n",
       "1    UCKL1  ENSMMUG00000015551  10    402713    402732  +  chr10\n",
       "2    KANK1  ENSMMUG00000019862  15  76468463  76468702  -  chr15\n",
       "3    Pbrm1  ENSMUSG00000042323  14  31858516  31858611  +  chr14\n",
       "4  Fam179b  ENSMUSG00000035614  12  66091928  66092077  +  chr12"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rhemac_df['chrom'] = 'chr' + rhemac_df[2]\n",
    "rhemac_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rhemac_df['strand'] = '+'\n",
    "rhemac_df['score'] = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rhemac_df['name'] = rhemac_df[0] + '_' + rhemac_df[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>   PTPRD</td>\n",
       "      <td> ENSMMUG00000006952</td>\n",
       "      <td> 15</td>\n",
       "      <td> 68642301</td>\n",
       "      <td> 68642606</td>\n",
       "      <td> +</td>\n",
       "      <td> chr15</td>\n",
       "      <td> +</td>\n",
       "      <td> 1000</td>\n",
       "      <td>   PTPRD_ENSMMUG00000006952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>   UCKL1</td>\n",
       "      <td> ENSMMUG00000015551</td>\n",
       "      <td> 10</td>\n",
       "      <td>   402713</td>\n",
       "      <td>   402732</td>\n",
       "      <td> +</td>\n",
       "      <td> chr10</td>\n",
       "      <td> +</td>\n",
       "      <td> 1000</td>\n",
       "      <td>   UCKL1_ENSMMUG00000015551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>   KANK1</td>\n",
       "      <td> ENSMMUG00000019862</td>\n",
       "      <td> 15</td>\n",
       "      <td> 76468463</td>\n",
       "      <td> 76468702</td>\n",
       "      <td> -</td>\n",
       "      <td> chr15</td>\n",
       "      <td> +</td>\n",
       "      <td> 1000</td>\n",
       "      <td>   KANK1_ENSMMUG00000019862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>   Pbrm1</td>\n",
       "      <td> ENSMUSG00000042323</td>\n",
       "      <td> 14</td>\n",
       "      <td> 31858516</td>\n",
       "      <td> 31858611</td>\n",
       "      <td> +</td>\n",
       "      <td> chr14</td>\n",
       "      <td> +</td>\n",
       "      <td> 1000</td>\n",
       "      <td>   Pbrm1_ENSMUSG00000042323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> Fam179b</td>\n",
       "      <td> ENSMUSG00000035614</td>\n",
       "      <td> 12</td>\n",
       "      <td> 66091928</td>\n",
       "      <td> 66092077</td>\n",
       "      <td> +</td>\n",
       "      <td> chr12</td>\n",
       "      <td> +</td>\n",
       "      <td> 1000</td>\n",
       "      <td> Fam179b_ENSMUSG00000035614</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0                   1   2         3         4  5  chrom strand  \\\n",
       "0    PTPRD  ENSMMUG00000006952  15  68642301  68642606  +  chr15      +   \n",
       "1    UCKL1  ENSMMUG00000015551  10    402713    402732  +  chr10      +   \n",
       "2    KANK1  ENSMMUG00000019862  15  76468463  76468702  -  chr15      +   \n",
       "3    Pbrm1  ENSMUSG00000042323  14  31858516  31858611  +  chr14      +   \n",
       "4  Fam179b  ENSMUSG00000035614  12  66091928  66092077  +  chr12      +   \n",
       "\n",
       "   score                        name  \n",
       "0   1000    PTPRD_ENSMMUG00000006952  \n",
       "1   1000    UCKL1_ENSMMUG00000015551  \n",
       "2   1000    KANK1_ENSMMUG00000019862  \n",
       "3   1000    Pbrm1_ENSMUSG00000042323  \n",
       "4   1000  Fam179b_ENSMUSG00000035614  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rhemac_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> chr15</td>\n",
       "      <td> 68642301</td>\n",
       "      <td> 68642606</td>\n",
       "      <td>   PTPRD_ENSMMUG00000006952</td>\n",
       "      <td> 1000</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> chr10</td>\n",
       "      <td>   402713</td>\n",
       "      <td>   402732</td>\n",
       "      <td>   UCKL1_ENSMMUG00000015551</td>\n",
       "      <td> 1000</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> chr15</td>\n",
       "      <td> 76468463</td>\n",
       "      <td> 76468702</td>\n",
       "      <td>   KANK1_ENSMMUG00000019862</td>\n",
       "      <td> 1000</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> chr14</td>\n",
       "      <td> 31858516</td>\n",
       "      <td> 31858611</td>\n",
       "      <td>   Pbrm1_ENSMUSG00000042323</td>\n",
       "      <td> 1000</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> chr12</td>\n",
       "      <td> 66091928</td>\n",
       "      <td> 66092077</td>\n",
       "      <td> Fam179b_ENSMUSG00000035614</td>\n",
       "      <td> 1000</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom         3         4                        name  score strand\n",
       "0  chr15  68642301  68642606    PTPRD_ENSMMUG00000006952   1000      +\n",
       "1  chr10    402713    402732    UCKL1_ENSMMUG00000015551   1000      +\n",
       "2  chr15  76468463  76468702    KANK1_ENSMMUG00000019862   1000      +\n",
       "3  chr14  31858516  31858611    Pbrm1_ENSMUSG00000042323   1000      +\n",
       "4  chr12  66091928  66092077  Fam179b_ENSMUSG00000035614   1000      +"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rhemac_bed = rhemac_df.reindex(columns=['chrom', 3, 4, 'name', 'score', 'strand'])\n",
    "rhemac_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rhemac_bed.to_csv('/projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons.bed', \n",
    "                  sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr16\t55027280\t55043304\tKIF18B_ENSMMUG00000013113\t1000\t+\r\n",
      "chr14\t24791110\t24849594\tLOC713613_ENSMMUG00000015602\t1000\t+\r\n",
      "chr7\t126179652\t126283622\tRHOJ_ENSMMUG00000008203\t1000\t+\r\n",
      "chr7\t92490280\t92873651\tPRKD1_ENSMMUG00000022550\t1000\t+\r\n",
      "chr6\t6532824\t6570828\tSRD5A1_ENSMMUG00000013186\t1000\t+\r\n",
      "chr7\t57472232\t57536686\tIREB2_ENSMMUG00000023474\t1000\t+\r\n",
      "chr2\t43274902\t43438002\tADCY5_ENSMMUG00000002230\t1000\t+\r\n",
      "chr5\t74655913\t74744542\tKIT_ENSMMUG00000003057\t1000\t+\r\n",
      "chr5\t53378257\t53434251\tSCARB2_ENSMMUG00000020913\t1000\t+\r\n",
      "chr1\t144408757\t144417306\tC1orf55_ENSMMUG00000017148\t1000\t+\r\n"
     ]
    }
   ],
   "source": [
    "! head merkin2012_rhemac_ancient_alt_exons.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: liftover: command not found\r\n"
     ]
    }
   ],
   "source": [
    "! liftover -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "liftOver - Move annotations from one assembly to another\r\n",
      "usage:\r\n",
      "   liftOver oldFile map.chain newFile unMapped\r\n",
      "oldFile and newFile are in bed format by default, but can be in GFF and\r\n",
      "maybe eventually others with the appropriate flags below.\r\n",
      "The map.chain file has the old genome as the target and the new genome\r\n",
      "as the query.\r\n",
      "\r\n",
      "***********************************************************************\r\n",
      "WARNING: liftOver was only designed to work between different\r\n",
      "         assemblies of the same organism. It may not do what you want\r\n",
      "         if you are lifting between different organisms. If there has\r\n",
      "         been a rearrangement in one of the species, the size of the\r\n",
      "         region being mapped may change dramatically after mapping.\r\n",
      "***********************************************************************\r\n",
      "\r\n",
      "options:\r\n",
      "   -minMatch=0.N Minimum ratio of bases that must remap. Default 0.95\r\n",
      "   -gff  File is in gff/gtf format.  Note that the gff lines are converted\r\n",
      "         separately.  It would be good to have a separate check after this\r\n",
      "         that the lines that make up a gene model still make a plausible gene\r\n",
      "         after liftOver\r\n",
      "   -genePred - File is in genePred format\r\n",
      "   -sample - File is in sample format\r\n",
      "   -bedPlus=N - File is bed N+ format\r\n",
      "   -positions - File is in browser \"position\" format\r\n",
      "   -hasBin - File has bin value (used only with -bedPlus)\r\n",
      "   -tab - Separate by tabs rather than space (used only with -bedPlus)\r\n",
      "   -pslT - File is in psl format, map target side only\r\n",
      "   -ends=N - Lift the first and last N bases of each record and combine the\r\n",
      "             result. This is useful for lifting large regions like BAC end pairs.\r\n",
      "   -minBlocks=0.N Minimum ratio of alignment blocks or exons that must map\r\n",
      "                  (default 1.00)\r\n",
      "   -fudgeThick    (bed 12 or 12+ only) If thickStart/thickEnd is not mapped,\r\n",
      "                  use the closest mapped base.  Recommended if using \r\n",
      "                  -minBlocks.\r\n",
      "   -multiple               Allow multiple output regions\r\n",
      "   -minChainT, -minChainQ  Minimum chain size in target/query, when mapping\r\n",
      "                           to multiple output regions (default 0, 0)\r\n",
      "   -minSizeT               deprecated synonym for -minChainT (ENCODE compat.)\r\n",
      "   -minSizeQ               Min matching region size in query with -multiple.\r\n",
      "   -chainTable             Used with -multiple, format is db.tablename,\r\n",
      "                               to extend chains from net (preserves dups)\r\n",
      "   -errorHelp              Explain error messages\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! /projects/ps-yeolab/software/bin/liftOver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# ! gunzip -v rheMac2ToHg19.over.chain.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/projects/ps-yeolab/genomes/rheMac/rheMac2'"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!touch /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!chmod ug+rwx /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rwxrwxr-- 1 obotvinnik yeo-group 0 Jun  2 16:12 /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.unmapped\r\n"
     ]
    }
   ],
   "source": [
    "! ls -lha /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading liftover chains\n",
      "Mapping coordinates\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "/projects/ps-yeolab/software/bin/liftOver \\\n",
    "    /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons.bed \\\n",
    "    /projects/ps-yeolab/genomes/rheMac/rheMac2/rheMac2ToHg19.over.chain \\\n",
    "    /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.bed \\ \n",
    "    /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 obotvinnik yeo-group 29K Jun  2 16:12 /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.bed\r\n",
      "-rwxrwxr-- 1 obotvinnik yeo-group   0 Jun  2 16:12 /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.unmapped\r\n"
     ]
    }
   ],
   "source": [
    "!ls -lha /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Intersect with alt exons and add to Splicing feature data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>intron12_length_summed</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "      <th>skipped_exon_exon2_splice_site_3p_score</th>\n",
       "      <th>skipped_exon_exon2_splice_site_3p_seq</th>\n",
       "      <th>skipped_exon_exon2_splice_site_5p_score</th>\n",
       "      <th>skipped_exon_exon2_splice_site_5p_seq</th>\n",
       "      <th>exon2_splice_site_3p_score</th>\n",
       "      <th>exon2_splice_site_3p_seq</th>\n",
       "      <th>exon2_splice_site_5p_score</th>\n",
       "      <th>exon2_splice_site_5p_seq</th>\n",
       "      <th>skipped_exon_exon2_phastcons_placental_mammal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188908-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>492</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>0.291112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188913-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>487</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>0.291112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>70</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>0.291112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>152</td>\n",
       "      <td>161</td>\n",
       "      <td>100</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "      <td>False</td>\n",
       "      <td>3108</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>9.08</td>\n",
       "      <td>TCCCTTCCTGCCCTGGCTAGGCC</td>\n",
       "      <td>7.96</td>\n",
       "      <td>TAGGTGAGC</td>\n",
       "      <td>9.08</td>\n",
       "      <td>TCCCTTCCTGCCCTGGCTAGGCC</td>\n",
       "      <td>7.96</td>\n",
       "      <td>TAGGTGAGC</td>\n",
       "      <td>0.803519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>138</td>\n",
       "      <td>143</td>\n",
       "      <td>152</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "      <td>False</td>\n",
       "      <td>1400</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>7.92</td>\n",
       "      <td>TGGCCTGCCTGTGCCTGCAGTTT</td>\n",
       "      <td>6.05</td>\n",
       "      <td>GGAGTGAGT</td>\n",
       "      <td>7.92</td>\n",
       "      <td>TGGCCTGCCTGTGCCTGCAGTTT</td>\n",
       "      <td>6.05</td>\n",
       "      <td>GGAGTGAGT</td>\n",
       "      <td>0.839394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "\n",
       "                                                         ensembl_id gene_name  \\\n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "\n",
       "                                                    exon3_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           492   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           487   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   \n",
       "\n",
       "                                                    intron1_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...            2648   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...             220   \n",
       "\n",
       "                                                    intron2_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...             460   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...            1180   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "\n",
       "                                                    intron12_length_summed  \\\n",
       "event_id                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                    3108   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                    1400   \n",
       "\n",
       "                                                     one_ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "\n",
       "                                                    skipped_exon_exon2_splice_site_3p_score  \\\n",
       "event_id                                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     7.77   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                     9.08   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                                     7.92   \n",
       "\n",
       "                                                   skipped_exon_exon2_splice_site_3p_seq  \\\n",
       "event_id                                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...               CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...               CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...               CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...               TCCCTTCCTGCCCTGGCTAGGCC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...               TGGCCTGCCTGTGCCTGCAGTTT   \n",
       "\n",
       "                                                    skipped_exon_exon2_splice_site_5p_score  \\\n",
       "event_id                                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                     6.96   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                     7.96   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                                     6.05   \n",
       "\n",
       "                                                   skipped_exon_exon2_splice_site_5p_seq  \\\n",
       "event_id                                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                             ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                             ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                             ACGGTGTGT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                             TAGGTGAGC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                             GGAGTGAGT   \n",
       "\n",
       "                                                    exon2_splice_site_3p_score  \\\n",
       "event_id                                                                         \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                        7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                        7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                        7.77   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                        9.08   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                        7.92   \n",
       "\n",
       "                                                   exon2_splice_site_3p_seq  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  TCCCTTCCTGCCCTGGCTAGGCC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  TGGCCTGCCTGTGCCTGCAGTTT   \n",
       "\n",
       "                                                    exon2_splice_site_5p_score  \\\n",
       "event_id                                                                         \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                        6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                        6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                        6.96   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                        7.96   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                        6.05   \n",
       "\n",
       "                                                   exon2_splice_site_5p_seq  \\\n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                ACGGTGTGT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                TAGGTGAGC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                GGAGTGAGT   \n",
       "\n",
       "                                                    skipped_exon_exon2_phastcons_placental_mammal  \n",
       "event_id                                                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                       0.291112  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                       0.291112  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                       0.291112  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                       0.803519  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                                       0.839394  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "alt_exons_bedfile = '{}/skipped_exon_exon2.bed'.format(folder)\n",
    "constitutive_bedfile = '{}/constitutive_exons.bed'.format(folder)\n",
    "\n",
    "bedfiles = alt_exons_bedfile, constitutive_bedfile\n",
    "\n",
    "splicing_feature_data = pd.read_csv('{}/splicing_feature_data.csv'.format(folder), index_col=0)\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab/obotvinnik/singlecell_pnms/skipped_exon_exon2_merkin2012_ancient_alt_exons.bed\n",
      "/projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons_merkin2012_ancient_alt_exons.bed\n"
     ]
    }
   ],
   "source": [
    "alt_exons_bedfile_intersect = alt_exons_bedfile.replace('.bed', '_merkin2012_ancient_alt_exons.bed')\n",
    "constitutive_bedfile_intersect = constitutive_bedfile.replace('.bed', '_merkin2012_ancient_alt_exons.bed')\n",
    "print alt_exons_bedfile_intersect\n",
    "print constitutive_bedfile_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58842 /projects/ps-yeolab/obotvinnik/singlecell_pnms/skipped_exon_exon2.bed\r\n"
     ]
    }
   ],
   "source": [
    "! wc -l $alt_exons_bedfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1449 /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons.bed\r\n"
     ]
    }
   ],
   "source": [
    "! wc -l $constitutive_bedfile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "bedtools intersect \\\n",
    "    -b /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.bed \\\n",
    "    -a /projects/ps-yeolab/obotvinnik/singlecell_pnms/skipped_exon_exon2.bed -s -f 0.9 \\\n",
    "    -wo > /projects/ps-yeolab/obotvinnik/singlecell_pnms/skipped_exon_exon2_merkin2012_ancient_alt_exons.bed\n",
    "\n",
    "bedtools intersect \\\n",
    "    -b /projects/ps-yeolab/obotvinnik/merkin2012/merkin2012_rhemac_ancient_alt_exons_hg19.bed \\\n",
    "    -a /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons.bed -s -f 0.9 \\\n",
    "    -wo > /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons_merkin2012_ancient_alt_exons.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "368 /projects/ps-yeolab/obotvinnik/singlecell_pnms/skipped_exon_exon2_merkin2012_ancient_alt_exons.bed\r\n"
     ]
    }
   ],
   "source": [
    "! wc -l $alt_exons_bedfile_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>123972857</td>\n",
       "      <td>123972892</td>\n",
       "      <td>exon:chr10:123971201-123971223:+@exon:chr10:12...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>chr10</td>\n",
       "      <td>123972857</td>\n",
       "      <td>123972892</td>\n",
       "      <td>TACC2_ENSMMUG00000013033</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28884662</td>\n",
       "      <td>28884970</td>\n",
       "      <td>exon:chr10:28879649-28879761:+@exon:chr10:2888...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28884662</td>\n",
       "      <td>28884970</td>\n",
       "      <td>WAC_ENSMMUG00000008268</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28884662</td>\n",
       "      <td>28884970</td>\n",
       "      <td>exon:chr10:28879649-28879761:+@exon:chr10:2888...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28884662</td>\n",
       "      <td>28884970</td>\n",
       "      <td>WAC_ENSMMUG00000008268</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>61841908</td>\n",
       "      <td>61841934</td>\n",
       "      <td>exon:chr10:61842373-61842444:-@exon:chr10:6184...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>61841907</td>\n",
       "      <td>61841939</td>\n",
       "      <td>ANK3_ENSMMUG00000003589</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>61926349</td>\n",
       "      <td>61926411</td>\n",
       "      <td>exon:chr10:61926582-61926633:-@exon:chr10:6192...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>61926266</td>\n",
       "      <td>61926571</td>\n",
       "      <td>ANK3_ENSMMUG00000003589</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2   \\\n",
       "0  chr10  123972857  123972892   \n",
       "1  chr10   28884662   28884970   \n",
       "2  chr10   28884662   28884970   \n",
       "3  chr10   61841908   61841934   \n",
       "4  chr10   61926349   61926411   \n",
       "\n",
       "                                                  3     4  5      6   \\\n",
       "0  exon:chr10:123971201-123971223:+@exon:chr10:12...  1000  +  chr10   \n",
       "1  exon:chr10:28879649-28879761:+@exon:chr10:2888...  1000  +  chr10   \n",
       "2  exon:chr10:28879649-28879761:+@exon:chr10:2888...  1000  +  chr10   \n",
       "3  exon:chr10:61842373-61842444:-@exon:chr10:6184...  1000  -  chr10   \n",
       "4  exon:chr10:61926582-61926633:-@exon:chr10:6192...  1000  -  chr10   \n",
       "\n",
       "          7          8                         9     10 11   12  \n",
       "0  123972857  123972892  TACC2_ENSMMUG00000013033  1000  +   35  \n",
       "1   28884662   28884970    WAC_ENSMMUG00000008268  1000  +  308  \n",
       "2   28884662   28884970    WAC_ENSMMUG00000008268  1000  +  308  \n",
       "3   61841907   61841939   ANK3_ENSMMUG00000003589  1000  -   26  \n",
       "4   61926266   61926571   ANK3_ENSMMUG00000003589  1000  -   62  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_table(alt_exons_bedfile_intersect, header=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>intron1_length</th>\n",
       "      <th>intron2_length</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>intron12_length_summed</th>\n",
       "      <th>...</th>\n",
       "      <th>skipped_exon_exon2_splice_site_3p_score</th>\n",
       "      <th>skipped_exon_exon2_splice_site_3p_seq</th>\n",
       "      <th>skipped_exon_exon2_splice_site_5p_score</th>\n",
       "      <th>skipped_exon_exon2_splice_site_5p_seq</th>\n",
       "      <th>exon2_splice_site_3p_score</th>\n",
       "      <th>exon2_splice_site_3p_seq</th>\n",
       "      <th>exon2_splice_site_5p_score</th>\n",
       "      <th>exon2_splice_site_5p_seq</th>\n",
       "      <th>skipped_exon_exon2_phastcons_placental_mammal</th>\n",
       "      <th>exon2_merkin2012_ancient</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188908-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>492</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>0.291112</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100188913-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>487</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>0.291112</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>70</td>\n",
       "      <td>681</td>\n",
       "      <td>148</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>7.77</td>\n",
       "      <td>CTCTCCCCTGCCCTGCACAGCCT</td>\n",
       "      <td>6.96</td>\n",
       "      <td>ACGGTGTGT</td>\n",
       "      <td>0.291112</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>152</td>\n",
       "      <td>161</td>\n",
       "      <td>100</td>\n",
       "      <td>2648</td>\n",
       "      <td>460</td>\n",
       "      <td>False</td>\n",
       "      <td>3108</td>\n",
       "      <td>...</td>\n",
       "      <td>9.08</td>\n",
       "      <td>TCCCTTCCTGCCCTGGCTAGGCC</td>\n",
       "      <td>7.96</td>\n",
       "      <td>TAGGTGAGC</td>\n",
       "      <td>9.08</td>\n",
       "      <td>TCCCTTCCTGCCCTGGCTAGGCC</td>\n",
       "      <td>7.96</td>\n",
       "      <td>TAGGTGAGC</td>\n",
       "      <td>0.803519</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>138</td>\n",
       "      <td>143</td>\n",
       "      <td>152</td>\n",
       "      <td>220</td>\n",
       "      <td>1180</td>\n",
       "      <td>False</td>\n",
       "      <td>1400</td>\n",
       "      <td>...</td>\n",
       "      <td>7.92</td>\n",
       "      <td>TGGCCTGCCTGTGCCTGCAGTTT</td>\n",
       "      <td>6.05</td>\n",
       "      <td>GGAGTGAGT</td>\n",
       "      <td>7.92</td>\n",
       "      <td>TGGCCTGCCTGTGCCTGCAGTTT</td>\n",
       "      <td>6.05</td>\n",
       "      <td>GGAGTGAGT</td>\n",
       "      <td>0.839394</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            gencode_id  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521.14   \n",
       "\n",
       "                                                         ensembl_id gene_name  \\\n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521      HPS1   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "\n",
       "                                                    exon3_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           492   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           487   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            70   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           100   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           152   \n",
       "\n",
       "                                                    intron1_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             681   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...            2648   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...             220   \n",
       "\n",
       "                                                    intron2_length  \\\n",
       "event_id                                                             \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             148   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...             460   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...            1180   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "\n",
       "                                                    intron12_length_summed  \\\n",
       "event_id                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     829   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                    3108   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                    1400   \n",
       "\n",
       "                                                              ...             \\\n",
       "event_id                                                      ...              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            ...              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            ...              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            ...              \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...            ...              \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...            ...              \n",
       "\n",
       "                                                   skipped_exon_exon2_splice_site_3p_score  \\\n",
       "event_id                                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    7.77   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                    9.08   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                                    7.92   \n",
       "\n",
       "                                                    skipped_exon_exon2_splice_site_3p_seq  \\\n",
       "event_id                                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                TCCCTTCCTGCCCTGGCTAGGCC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                TGGCCTGCCTGTGCCTGCAGTTT   \n",
       "\n",
       "                                                   skipped_exon_exon2_splice_site_5p_score  \\\n",
       "event_id                                                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    6.96   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                    7.96   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                                    6.05   \n",
       "\n",
       "                                                    skipped_exon_exon2_splice_site_5p_seq  \\\n",
       "event_id                                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                              ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                              ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                              ACGGTGTGT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                              TAGGTGAGC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                              GGAGTGAGT   \n",
       "\n",
       "                                                   exon2_splice_site_3p_score  \\\n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       7.77   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       7.77   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                       9.08   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                       7.92   \n",
       "\n",
       "                                                    exon2_splice_site_3p_seq  \\\n",
       "event_id                                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...   CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...   CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...   CTCTCCCCTGCCCTGCACAGCCT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...   TCCCTTCCTGCCCTGGCTAGGCC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...   TGGCCTGCCTGTGCCTGCAGTTT   \n",
       "\n",
       "                                                   exon2_splice_site_5p_score  \\\n",
       "event_id                                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       6.96   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                       6.96   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                       7.96   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                       6.05   \n",
       "\n",
       "                                                    exon2_splice_site_5p_seq  \\\n",
       "event_id                                                                       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 ACGGTGTGT   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 ACGGTGTGT   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                 TAGGTGAGC   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                 GGAGTGAGT   \n",
       "\n",
       "                                                   skipped_exon_exon2_phastcons_placental_mammal  \\\n",
       "event_id                                                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                      0.291112   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                      0.291112   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                      0.291112   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                                      0.803519   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                                      0.839394   \n",
       "\n",
       "                                                    exon2_merkin2012_ancient  \n",
       "event_id                                                                      \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     False  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     False  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     False  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                     False  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                     False  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data['exon2_merkin2012_ancient'] = splicing_feature_data.index.isin(df[3])\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(58842, 21)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/projects/ps-yeolab/obotvinnik/singlecell_pnms'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "splicing_feature_data.to_csv('{}/splicing_feature_data.csv'.format(folder))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
