{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:784: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n",
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.py:1170: DtypeWarning: Columns (11,18,19,20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = self._reader.read(nrows)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>event_id</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>isoform1_rnafold_interior_loop_percentage</th>\n",
       "      <th>isoform2_rnafold_interior_loop_percentage</th>\n",
       "      <th>isoform1_rnafold_minimum_free_energy</th>\n",
       "      <th>isoform2_rnafold_minimum_free_energy</th>\n",
       "      <th>isoform1_rnafold_multiloop_percentage</th>\n",
       "      <th>isoform2_rnafold_multiloop_percentage</th>\n",
       "      <th>isoform1_rnafold_stem_percentage</th>\n",
       "      <th>isoform2_rnafold_stem_percentage</th>\n",
       "      <th>isoform1_rnafold_three_prime_percentage</th>\n",
       "      <th>isoform2_rnafold_three_prime_percentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>0.223048</td>\n",
       "      <td>-47.1</td>\n",
       "      <td>-95.9</td>\n",
       "      <td>0.105882</td>\n",
       "      <td>0.052045</td>\n",
       "      <td>0.505882</td>\n",
       "      <td>0.572491</td>\n",
       "      <td>0.017647</td>\n",
       "      <td>0.011152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.096852</td>\n",
       "      <td>-108.7</td>\n",
       "      <td>-183.1</td>\n",
       "      <td>0.079365</td>\n",
       "      <td>0.111380</td>\n",
       "      <td>0.650794</td>\n",
       "      <td>0.658596</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.014528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.182759</td>\n",
       "      <td>0.154734</td>\n",
       "      <td>-116.1</td>\n",
       "      <td>-168.8</td>\n",
       "      <td>0.013793</td>\n",
       "      <td>0.078522</td>\n",
       "      <td>0.675862</td>\n",
       "      <td>0.665127</td>\n",
       "      <td>0.006897</td>\n",
       "      <td>0.004619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.092251</td>\n",
       "      <td>0.191943</td>\n",
       "      <td>-79.6</td>\n",
       "      <td>-131.2</td>\n",
       "      <td>0.103321</td>\n",
       "      <td>0.063981</td>\n",
       "      <td>0.605166</td>\n",
       "      <td>0.630332</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>0.002370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>...</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.073171</td>\n",
       "      <td>-86.2</td>\n",
       "      <td>-122.2</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.135501</td>\n",
       "      <td>0.609023</td>\n",
       "      <td>0.563686</td>\n",
       "      <td>0.041353</td>\n",
       "      <td>0.029810</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 84 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            event_id  criteria  \\\n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...      best   \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...  only one   \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...      best   \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...  only one   \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...  only one   \n",
       "\n",
       "  criteria_additional          criteria_full       ensembl_id  \\\n",
       "0    appris_principal  best,appris_principal  ENSG00000107521   \n",
       "1                 NaN               only one  ENSG00000107521   \n",
       "2    appris_principal  best,appris_principal  ENSG00000107521   \n",
       "3                 NaN               only one  ENSG00000120053   \n",
       "4                 NaN               only one  ENSG00000198018   \n",
       "\n",
       "                              exon1  exon1_length  \\\n",
       "0  exon:chr10:100190328-100190427:-           100   \n",
       "1  exon:chr10:100193697-100193848:-           152   \n",
       "2  exon:chr10:100195392-100195529:-           138   \n",
       "3  exon:chr10:101165513-101165617:-           105   \n",
       "4  exon:chr10:101419263-101419345:+            83   \n",
       "\n",
       "                              exon2  exon2_length  \\\n",
       "0  exon:chr10:100189548-100189646:-            99   \n",
       "1  exon:chr10:100190888-100191048:-           161   \n",
       "2  exon:chr10:100195029-100195171:-           143   \n",
       "3  exon:chr10:101163481-101163631:-           151   \n",
       "4  exon:chr10:101419619-101419721:+           103   \n",
       "\n",
       "                              exon3                   ...                    \\\n",
       "0  exon:chr10:100189330-100189399:-                   ...                     \n",
       "1  exon:chr10:100190328-100190427:-                   ...                     \n",
       "2  exon:chr10:100193697-100193848:-                   ...                     \n",
       "3  exon:chr10:101163226-101163391:-                   ...                     \n",
       "4  exon:chr10:101421203-101421385:+                   ...                     \n",
       "\n",
       "   isoform1_rnafold_interior_loop_percentage  \\\n",
       "0                                   0.205882   \n",
       "1                                   0.095238   \n",
       "2                                   0.182759   \n",
       "3                                   0.092251   \n",
       "4                                   0.105263   \n",
       "\n",
       "  isoform2_rnafold_interior_loop_percentage  \\\n",
       "0                                  0.223048   \n",
       "1                                  0.096852   \n",
       "2                                  0.154734   \n",
       "3                                  0.191943   \n",
       "4                                  0.073171   \n",
       "\n",
       "   isoform1_rnafold_minimum_free_energy isoform2_rnafold_minimum_free_energy  \\\n",
       "0                                 -47.1                                -95.9   \n",
       "1                                -108.7                               -183.1   \n",
       "2                                -116.1                               -168.8   \n",
       "3                                 -79.6                               -131.2   \n",
       "4                                 -86.2                               -122.2   \n",
       "\n",
       "  isoform1_rnafold_multiloop_percentage  \\\n",
       "0                              0.105882   \n",
       "1                              0.079365   \n",
       "2                              0.013793   \n",
       "3                              0.103321   \n",
       "4                              0.142857   \n",
       "\n",
       "   isoform2_rnafold_multiloop_percentage isoform1_rnafold_stem_percentage  \\\n",
       "0                               0.052045                         0.505882   \n",
       "1                               0.111380                         0.650794   \n",
       "2                               0.078522                         0.675862   \n",
       "3                               0.063981                         0.605166   \n",
       "4                               0.135501                         0.609023   \n",
       "\n",
       "  isoform2_rnafold_stem_percentage isoform1_rnafold_three_prime_percentage  \\\n",
       "0                         0.572491                                0.017647   \n",
       "1                         0.658596                                0.023810   \n",
       "2                         0.665127                                0.006897   \n",
       "3                         0.630332                                0.011070   \n",
       "4                         0.563686                                0.041353   \n",
       "\n",
       "  isoform2_rnafold_three_prime_percentage  \n",
       "0                                0.011152  \n",
       "1                                0.014528  \n",
       "2                                0.004619  \n",
       "3                                0.002370  \n",
       "4                                0.029810  \n",
       "\n",
       "[5 rows x 84 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "v19db_filename = '/projects/ps-yeolab/genomes/hg19/gencode/v19/gencode.v19.annotation.gtf.db'\n",
    "v19db = gffutils.FeatureDB(v19db_filename)\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "csv_folder = '/home/obotvinnik/projects/singlecell_pnms/analysis/csvs_for_paper'\n",
    "\n",
    "splicing_feature_data = pd.read_csv('{}/splicing_feature_data.csv'.format(folder))\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(844056, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon</th>\n",
       "      <th>direction</th>\n",
       "      <th>junction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr1:29321-29370:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:18367-29320:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr1:29321-29370:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:24892-29320:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr1:29321-29370:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:27186-29320:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr1:29321-29370:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:18380-29320:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr1:29321-29370:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:18370-29320:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      exon   direction            junction\n",
       "0  exon:chr1:29321-29370:-  downstream  chr1:18367-29320:-\n",
       "1  exon:chr1:29321-29370:-  downstream  chr1:24892-29320:-\n",
       "2  exon:chr1:29321-29370:-  downstream  chr1:27186-29320:-\n",
       "3  exon:chr1:29321-29370:-  downstream  chr1:18380-29320:-\n",
       "4  exon:chr1:29321-29370:-  downstream  chr1:18370-29320:-"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "junction_exons = pd.read_csv('{}/junction_exon_triples.csv'.format(folder))\n",
    "print junction_exons.shape\n",
    "junction_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction13</th>\n",
       "      <th>event_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:51732523-51732902:+</td>\n",
       "      <td>exon:chr10:51740949-51741041:+</td>\n",
       "      <td>exon:chr10:51751436-51751504:+</td>\n",
       "      <td>chr10:51732903-51740948:+</td>\n",
       "      <td>chr10:51741042-51751435:+</td>\n",
       "      <td>chr10:51732903-51751435:+</td>\n",
       "      <td>exon:chr10:51732523-51732902:+@exon:chr10:5174...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr1:32669480-32669646:+</td>\n",
       "      <td>exon:chr1:32669787-32669980:+</td>\n",
       "      <td>exon:chr1:32670199-32670988:+</td>\n",
       "      <td>chr1:32669647-32669786:+</td>\n",
       "      <td>chr1:32669981-32670198:+</td>\n",
       "      <td>chr1:32669647-32670198:+</td>\n",
       "      <td>exon:chr1:32669480-32669646:+@exon:chr1:326697...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:67374548-67375870:+</td>\n",
       "      <td>chr11:67375950-67376030:+</td>\n",
       "      <td>chr11:67376194-67376922:+</td>\n",
       "      <td>exon:chr11:67375871-67375949:+</td>\n",
       "      <td>exon:chr11:67376031-67376193:+</td>\n",
       "      <td>exon:chr11:67375871-67376193:+</td>\n",
       "      <td>chr11:67374548-67375870:+@chr11:67375950-67376...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr11:18425241-18425358:+</td>\n",
       "      <td>exon:chr11:18426996-18427119:+</td>\n",
       "      <td>exon:chr11:18428664-18429394:+</td>\n",
       "      <td>chr11:18425359-18426995:+</td>\n",
       "      <td>chr11:18427120-18428663:+</td>\n",
       "      <td>chr11:18425359-18428663:+</td>\n",
       "      <td>exon:chr11:18425241-18425358:+@exon:chr11:1842...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:99213311-99213420:+</td>\n",
       "      <td>exon:chr10:99213556-99213603:+</td>\n",
       "      <td>exon:chr10:99214471-99214556:+</td>\n",
       "      <td>chr10:99213421-99213555:+</td>\n",
       "      <td>chr10:99213604-99214470:+</td>\n",
       "      <td>chr10:99213421-99214470:+</td>\n",
       "      <td>exon:chr10:99213311-99213420:+@exon:chr10:9921...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            exon1                           exon2  \\\n",
       "0  exon:chr10:51732523-51732902:+  exon:chr10:51740949-51741041:+   \n",
       "1   exon:chr1:32669480-32669646:+   exon:chr1:32669787-32669980:+   \n",
       "2       chr11:67374548-67375870:+       chr11:67375950-67376030:+   \n",
       "3  exon:chr11:18425241-18425358:+  exon:chr11:18426996-18427119:+   \n",
       "4  exon:chr10:99213311-99213420:+  exon:chr10:99213556-99213603:+   \n",
       "\n",
       "                            exon3                      junction12  \\\n",
       "0  exon:chr10:51751436-51751504:+       chr10:51732903-51740948:+   \n",
       "1   exon:chr1:32670199-32670988:+        chr1:32669647-32669786:+   \n",
       "2       chr11:67376194-67376922:+  exon:chr11:67375871-67375949:+   \n",
       "3  exon:chr11:18428664-18429394:+       chr11:18425359-18426995:+   \n",
       "4  exon:chr10:99214471-99214556:+       chr10:99213421-99213555:+   \n",
       "\n",
       "                       junction23                      junction13  \\\n",
       "0       chr10:51741042-51751435:+       chr10:51732903-51751435:+   \n",
       "1        chr1:32669981-32670198:+        chr1:32669647-32670198:+   \n",
       "2  exon:chr11:67376031-67376193:+  exon:chr11:67375871-67376193:+   \n",
       "3       chr11:18427120-18428663:+       chr11:18425359-18428663:+   \n",
       "4       chr10:99213604-99214470:+       chr10:99213421-99214470:+   \n",
       "\n",
       "                                            event_id  \n",
       "0  exon:chr10:51732523-51732902:+@exon:chr10:5174...  \n",
       "1  exon:chr1:32669480-32669646:+@exon:chr1:326697...  \n",
       "2  chr11:67374548-67375870:+@chr11:67375950-67376...  \n",
       "3  exon:chr11:18425241-18425358:+@exon:chr11:1842...  \n",
       "4  exon:chr10:99213311-99213420:+@exon:chr10:9921...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder = '/home/obotvinnik/projects/singlecell_pnms/analysis/csvs_for_paper'\n",
    "se_events = pd.read_csv('{}/se_events_with_stressed.csv'.format(folder))\n",
    "se_events.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon4</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction24</th>\n",
       "      <th>event_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr8:144663224-144663324:-</td>\n",
       "      <td>exon:chr8:144668389-144668460:-</td>\n",
       "      <td>exon:chr8:144663456-144663498:-</td>\n",
       "      <td>exon:chr8:144668899-144669019:-</td>\n",
       "      <td>chr8:144663325-144663455:-</td>\n",
       "      <td>chr8:144663499-144668898:-</td>\n",
       "      <td>chr8:144663325-144668388:-</td>\n",
       "      <td>chr8:144668461-144668898:-</td>\n",
       "      <td>exon:chr8:144663224-144663324:-@exon:chr8:1446...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr19:8376168-8376479:-</td>\n",
       "      <td>exon:chr19:8385741-8385790:-</td>\n",
       "      <td>exon:chr19:8381380-8381529:-</td>\n",
       "      <td>exon:chr19:8386192-8386259:-</td>\n",
       "      <td>chr19:8376480-8381379:-</td>\n",
       "      <td>chr19:8381530-8386191:-</td>\n",
       "      <td>chr19:8376480-8385740:-</td>\n",
       "      <td>chr19:8385791-8386191:-</td>\n",
       "      <td>exon:chr19:8376168-8376479:-@exon:chr19:838574...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon:chr17:33307578-33307641:+</td>\n",
       "      <td>exon:chr17:33310021-33310571:+</td>\n",
       "      <td>exon:chr17:33313007-33313150:+</td>\n",
       "      <td>exon:chr17:33316485-33316899:+</td>\n",
       "      <td>chr17:33307642-33313006:+</td>\n",
       "      <td>chr17:33313151-33316484:+</td>\n",
       "      <td>chr17:33307642-33310020:+</td>\n",
       "      <td>chr17:33310572-33316484:+</td>\n",
       "      <td>exon:chr17:33307578-33307641:+@exon:chr17:3331...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr14:31091525-31091605:+</td>\n",
       "      <td>exon:chr14:31097415-31097485:+</td>\n",
       "      <td>exon:chr14:31099683-31099771:+</td>\n",
       "      <td>exon:chr14:31103153-31103243:+</td>\n",
       "      <td>chr14:31091606-31099682:+</td>\n",
       "      <td>chr14:31099772-31103152:+</td>\n",
       "      <td>chr14:31091606-31097414:+</td>\n",
       "      <td>chr14:31097486-31103152:+</td>\n",
       "      <td>exon:chr14:31091525-31091605:+@exon:chr14:3109...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr1:247002400-247004300:-</td>\n",
       "      <td>exon:chr1:247007097-247007230:-</td>\n",
       "      <td>exon:chr1:247005996-247006078:-</td>\n",
       "      <td>exon:chr1:247012917-247013151:-</td>\n",
       "      <td>chr1:247004301-247005995:-</td>\n",
       "      <td>chr1:247006079-247012916:-</td>\n",
       "      <td>chr1:247004301-247007096:-</td>\n",
       "      <td>chr1:247007231-247012916:-</td>\n",
       "      <td>exon:chr1:247002400-247004300:-@exon:chr1:2470...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             exon1                            exon2  \\\n",
       "0  exon:chr8:144663224-144663324:-  exon:chr8:144668389-144668460:-   \n",
       "1     exon:chr19:8376168-8376479:-     exon:chr19:8385741-8385790:-   \n",
       "2   exon:chr17:33307578-33307641:+   exon:chr17:33310021-33310571:+   \n",
       "3   exon:chr14:31091525-31091605:+   exon:chr14:31097415-31097485:+   \n",
       "4  exon:chr1:247002400-247004300:-  exon:chr1:247007097-247007230:-   \n",
       "\n",
       "                             exon3                            exon4  \\\n",
       "0  exon:chr8:144663456-144663498:-  exon:chr8:144668899-144669019:-   \n",
       "1     exon:chr19:8381380-8381529:-     exon:chr19:8386192-8386259:-   \n",
       "2   exon:chr17:33313007-33313150:+   exon:chr17:33316485-33316899:+   \n",
       "3   exon:chr14:31099683-31099771:+   exon:chr14:31103153-31103243:+   \n",
       "4  exon:chr1:247005996-247006078:-  exon:chr1:247012917-247013151:-   \n",
       "\n",
       "                   junction13                  junction34  \\\n",
       "0  chr8:144663325-144663455:-  chr8:144663499-144668898:-   \n",
       "1     chr19:8376480-8381379:-     chr19:8381530-8386191:-   \n",
       "2   chr17:33307642-33313006:+   chr17:33313151-33316484:+   \n",
       "3   chr14:31091606-31099682:+   chr14:31099772-31103152:+   \n",
       "4  chr1:247004301-247005995:-  chr1:247006079-247012916:-   \n",
       "\n",
       "                   junction12                  junction24  \\\n",
       "0  chr8:144663325-144668388:-  chr8:144668461-144668898:-   \n",
       "1     chr19:8376480-8385740:-     chr19:8385791-8386191:-   \n",
       "2   chr17:33307642-33310020:+   chr17:33310572-33316484:+   \n",
       "3   chr14:31091606-31097414:+   chr14:31097486-31103152:+   \n",
       "4  chr1:247004301-247007096:-  chr1:247007231-247012916:-   \n",
       "\n",
       "                                            event_id  \n",
       "0  exon:chr8:144663224-144663324:-@exon:chr8:1446...  \n",
       "1  exon:chr19:8376168-8376479:-@exon:chr19:838574...  \n",
       "2  exon:chr17:33307578-33307641:+@exon:chr17:3331...  \n",
       "3  exon:chr14:31091525-31091605:+@exon:chr14:3109...  \n",
       "4  exon:chr1:247002400-247004300:-@exon:chr1:2470...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxe_events = pd.read_csv('{}/mxe_events_with_stressed.csv'.format(folder))\n",
    "mxe_events.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(136429, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon2</th>\n",
       "      <th>event_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon:chr10:51740949-51741041:+</td>\n",
       "      <td>exon:chr10:51732523-51732902:+@exon:chr10:5174...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon:chr1:32669787-32669980:+</td>\n",
       "      <td>exon:chr1:32669480-32669646:+@exon:chr1:326697...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:67375950-67376030:+</td>\n",
       "      <td>chr11:67374548-67375870:+@chr11:67375950-67376...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon:chr11:18426996-18427119:+</td>\n",
       "      <td>exon:chr11:18425241-18425358:+@exon:chr11:1842...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon:chr10:99213556-99213603:+</td>\n",
       "      <td>exon:chr10:99213311-99213420:+@exon:chr10:9921...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            exon2  \\\n",
       "0  exon:chr10:51740949-51741041:+   \n",
       "1   exon:chr1:32669787-32669980:+   \n",
       "2       chr11:67375950-67376030:+   \n",
       "3  exon:chr11:18426996-18427119:+   \n",
       "4  exon:chr10:99213556-99213603:+   \n",
       "\n",
       "                                            event_id  \n",
       "0  exon:chr10:51732523-51732902:+@exon:chr10:5174...  \n",
       "1  exon:chr1:32669480-32669646:+@exon:chr1:326697...  \n",
       "2  chr11:67374548-67375870:+@chr11:67375950-67376...  \n",
       "3  exon:chr11:18425241-18425358:+@exon:chr11:1842...  \n",
       "4  exon:chr10:99213311-99213420:+@exon:chr10:9921...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon2s = pd.concat([se_events[['exon2', 'event_id']], mxe_events[['exon2', 'event_id']]], ignore_index=True)\n",
    "print exon2s.shape\n",
    "exon2s.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write Exon2 from SE and MXE to bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33693, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100189548</td>\n",
       "      <td>100189646</td>\n",
       "      <td>exon:chr10:100190328-100190427:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100190888</td>\n",
       "      <td>100191048</td>\n",
       "      <td>exon:chr10:100193697-100193848:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100195029</td>\n",
       "      <td>100195171</td>\n",
       "      <td>exon:chr10:100195392-100195529:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101163481</td>\n",
       "      <td>101163631</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101419619</td>\n",
       "      <td>101419721</td>\n",
       "      <td>exon:chr10:101419263-101419345:+@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start       stop  \\\n",
       "0  chr10  100189548  100189646   \n",
       "1  chr10  100190888  100191048   \n",
       "2  chr10  100195029  100195171   \n",
       "3  chr10  101163481  101163631   \n",
       "4  chr10  101419619  101419721   \n",
       "\n",
       "                                                name  score strand  \n",
       "0  exon:chr10:100190328-100190427:-@exon:chr10:10...   1000      -  \n",
       "1  exon:chr10:100193697-100193848:-@exon:chr10:10...   1000      -  \n",
       "2  exon:chr10:100195392-100195529:-@exon:chr10:10...   1000      -  \n",
       "3  exon:chr10:101165513-101165617:-@exon:chr10:10...   1000      -  \n",
       "4  exon:chr10:101419263-101419345:+@exon:chr10:10...   1000      +  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from outrigger.region import Region\n",
    "\n",
    "splicing_feature_data['exon2_region'] = splicing_feature_data.exon2.map(Region)\n",
    "\n",
    "exon2_bed = pd.DataFrame.from_records(splicing_feature_data.exon2_region.map(\n",
    "    lambda x: pd.Series(dict(chrom=x.chrom, start=x._start, stop=x._stop, strand=x.strand, score=1000))))\n",
    "\n",
    "exon2_bed['name'] = splicing_feature_data.event_id\n",
    "exon2_bed = exon2_bed.reindex(columns=['chrom', 'start', 'stop', 'name', 'score', 'strand'])\n",
    "\n",
    "print exon2_bed.shape\n",
    "exon2_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exon2_bed.to_csv('{}/exon2.bed'.format(folder), index=False, header=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get internal, constitutive exons\n",
    "\n",
    "### Internal: have both downstream and upstream junctions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon                              direction \n",
       "exon:chr10:100003848-100004106:+  downstream    1\n",
       "exon:chr10:100003848-100004321:+  downstream    1\n",
       "exon:chr10:100003848-100004651:+  downstream    1\n",
       "exon:chr10:100003848-100004654:+  downstream    1\n",
       "exon:chr10:100007447-100008748:-  upstream      1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_direction = junction_exons.groupby(['exon', 'direction']).size()\n",
    "exon_direction.head()\n",
    "# constitutive = constitutive[constitutive == 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Group on the number of \"direction\" items they have. Must have two to be internal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon\n",
       "exon:chr10:100010822-100010933:-    2\n",
       "exon:chr10:100144704-100144824:-    2\n",
       "exon:chr10:100146958-100147064:-    2\n",
       "exon:chr10:100150355-100150511:-    2\n",
       "exon:chr10:100155148-100155209:-    2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "internal_exons = exon_direction.groupby(level=0).size()\n",
    "internal_exons = internal_exons[internal_exons == 2]\n",
    "internal_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(448078, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon</th>\n",
       "      <th>direction</th>\n",
       "      <th>junction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>exon:chr1:24738-24891:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr1:24892-29320:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>exon:chr1:24738-24891:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr1:24892-29533:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exon:chr1:24738-24891:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:18062-24737:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>exon:chr1:24738-24891:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:18380-24737:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>exon:chr1:24738-24891:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:18555-24737:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      exon   direction            junction\n",
       "5  exon:chr1:24738-24891:-    upstream  chr1:24892-29320:-\n",
       "6  exon:chr1:24738-24891:-    upstream  chr1:24892-29533:-\n",
       "7  exon:chr1:24738-24891:-  downstream  chr1:18062-24737:-\n",
       "8  exon:chr1:24738-24891:-  downstream  chr1:18380-24737:-\n",
       "9  exon:chr1:24738-24891:-  downstream  chr1:18555-24737:-"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_juction_direction_internal = junction_exons.loc[junction_exons.exon.isin(internal_exons.index)]\n",
    "print exon_juction_direction_internal.shape\n",
    "exon_juction_direction_internal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove all exons that appear as exon2 in any alternative annotations, or exon3 for MXE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(166344,)\n",
      "(22847,)\n",
      "(360186, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon</th>\n",
       "      <th>direction</th>\n",
       "      <th>junction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>exon:chr1:17915-18061:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr1:18062-18267:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>exon:chr1:17915-18061:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr1:18062-24737:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>exon:chr1:17915-18061:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:17743-17914:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>exon:chr1:17915-18061:-</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chr1:17056-17914:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>exon:chr1:16854-17055:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr1:17056-17232:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       exon   direction            junction\n",
       "18  exon:chr1:17915-18061:-    upstream  chr1:18062-18267:-\n",
       "19  exon:chr1:17915-18061:-    upstream  chr1:18062-24737:-\n",
       "20  exon:chr1:17915-18061:-  downstream  chr1:17743-17914:-\n",
       "21  exon:chr1:17915-18061:-  downstream  chr1:17056-17914:-\n",
       "24  exon:chr1:16854-17055:-    upstream  chr1:17056-17232:-"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_exons = pd.concat([se_events['exon2'], mxe_events['exon2'], mxe_events['exon3']], ignore_index=True)\n",
    "print alt_exons.shape\n",
    "alt_exons = alt_exons.drop_duplicates()\n",
    "print alt_exons.shape\n",
    "\n",
    "in_alt_exons = exon_juction_direction_internal.exon.isin(alt_exons.values)\n",
    "\n",
    "constitutive = exon_juction_direction_internal.loc[(~in_alt_exons)]\n",
    "print constitutive.shape\n",
    "constitutive.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(128677,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    exon:chr1:17915-18061:-\n",
       "1    exon:chr1:16854-17055:-\n",
       "2    exon:chr1:16607-16765:-\n",
       "3    exon:chr1:14970-15038:-\n",
       "4    exon:chr1:17606-17742:-\n",
       "dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from poshsplice.region import Region\n",
    "constitutive_regions = pd.Series(constitutive.exon.unique()).map(Region)\n",
    "print constitutive_regions.shape\n",
    "constitutive_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "231088"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_junctions = constitutive.junction.unique()\n",
    "len(constitutive_junctions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(128677, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>exon:chr1:17915-18061:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16854</td>\n",
       "      <td>17055</td>\n",
       "      <td>exon:chr1:16854-17055:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>exon:chr1:16607-16765:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>exon:chr1:14970-15038:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>exon:chr1:17606-17742:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start   stop                     name  score strand\n",
       "0  chr1  17915  18061  exon:chr1:17915-18061:-   1000      -\n",
       "1  chr1  16854  17055  exon:chr1:16854-17055:-   1000      -\n",
       "2  chr1  16607  16765  exon:chr1:16607-16765:-   1000      -\n",
       "3  chr1  14970  15038  exon:chr1:14970-15038:-   1000      -\n",
       "4  chr1  17606  17742  exon:chr1:17606-17742:-   1000      -"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = constitutive_regions.map(\n",
    "    lambda x: pd.Series(dict(chrom=x.chrom, start=x._start, stop=x._stop, strand=x.strand, score=1000, name=x.name)))\n",
    "constitutive_bed = pd.DataFrame.from_records(records)\n",
    "constitutive_bed = constitutive_bed.reindex(columns=['chrom', 'start', 'stop', 'name', 'score', 'strand'])\n",
    "print constitutive_bed.shape\n",
    "constitutive_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(128677, 6)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_bed = constitutive_bed.drop_duplicatescates()\n",
    "constitutive_bed.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "constitutive_bed.to_csv('{}/constitutive_exons.bed'.format(folder), sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit a job to calculate conservation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 2 tasks as an array-job.\n",
      "job ID: 3800347\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bigWigAverageOverBed /projects/ps-yeolab/genomes/hg19/hg19_phastcons_placental_mammal.bw /projects/ps-yeolab/obotvinnik/singlecell_pnms/exon2.bed /projects/ps-yeolab/obotvinnik/singlecell_pnms/exon2_phastcons_placental_mammal.txt -bedOut=/projects/ps-yeolab/obotvinnik/singlecell_pnms/exon2_phastcons_placental_mammal.bed\n",
      "bigWigAverageOverBed /projects/ps-yeolab/genomes/hg19/hg19_phastcons_placental_mammal.bw /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons.bed /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons_phastcons_placental_mammal.txt -bedOut=/projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons_phastcons_placental_mammal.bed\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gscripts.qtools._Submitter.Submitter at 0x2ab918344b10>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "\n",
    "from gscripts.qtools import Submitter\n",
    "\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "alt_exons_bedfile = '{}/exon2.bed'.format(folder)\n",
    "constitutive_bedfile = '{}/constitutive_exons.bed'.format(folder)\n",
    "\n",
    "bedfiles = alt_exons_bedfile, constitutive_bedfile\n",
    "\n",
    "commands = []\n",
    "\n",
    "bw = '/projects/ps-yeolab/genomes/hg19/hg19_phastcons_placental_mammal.bw'\n",
    "\n",
    "for bedfile in bedfiles:\n",
    "    basename = os.path.basename(bedfile)\n",
    "    \n",
    "    prefix = basename.split('.bed')[0]\n",
    "    \n",
    "    prefix += '_phastcons_placental_mammal'\n",
    "    bedout = '{}/{}'.format(folder, prefix + '.bed')\n",
    "    outtab = '{}/{}'.format(folder, prefix + '.txt')\n",
    "    command = 'bigWigAverageOverBed {} {} {} -bedOut={}'.format(bw, bedfile, outtab, bedout)\n",
    "    print command\n",
    "    commands.append(command)\n",
    "\n",
    "jobname = 'exonbody_conservation'\n",
    "Submitter(commands, jobname, array=True, walltime='2:00:00', write_and_submit=True,\n",
    "          err_filename='{}/{}.err'.format(folder, jobname),\n",
    "          out_filename='{}/{}.out'.format(folder, jobname))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Don't do below -- too much work and code crashes and not worth it"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter constitutive junctions for usage\n",
    "\n",
    "Check that the junctions for these exons are seen in 20+ single cells per celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(233, 25836)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <th>exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+</th>\n",
       "      <th>exon:chr10:101421203-101421385:+@exon:chr10:101439018-101439223:+@exon:chr10:101439482-101439632:+</th>\n",
       "      <th>exon:chr10:101480744-101480825:-@exon:chr10:101478103-101478257:-@exon:chr10:101476105-101476218:-</th>\n",
       "      <th>exon:chr10:101507014-101507147:+@exon:chr10:101510126-101510153:+@exon:chr10:101514286-101514391:+</th>\n",
       "      <th>exon:chr10:101507014-101507147:+@exon:chr10:101514286-101514391:+@exon:chr10:101515382-101515891:+</th>\n",
       "      <th>exon:chr10:101964263-101964414:-@exon:chr10:101961845-101961906:-@exon:chr10:101960428-101960537:-</th>\n",
       "      <th>...</th>\n",
       "      <th>exon:chrY:15471647-15471866:-@exon:chrY:15470968-15471102:-@exon:chrY:15470344-15470433:-@exon:chrY:15469757-15469849:-</th>\n",
       "      <th>exon:chrY:16634632-16634821:+@exon:chrY:16733889-16734471:+@exon:chrY:16831339-16831398:+@exon:chrY:16834997-16835149:+</th>\n",
       "      <th>exon:chrY:21237827-21237882:-@exon:chrY:21230610-21230684:-@exon:chrY:21207128-21207177:-@exon:chrY:21206428-21206581:-</th>\n",
       "      <th>exon:chrY:21749096-21749393:+@exon:chrY:21750256-21751498:+@exon:chrY:21753666-21753845:+@exon:chrY:21755285-21755479:+</th>\n",
       "      <th>exon:chrY:21749368-21749393:+@exon:chrY:21751407-21751498:+@exon:chrY:21753666-21753845:+@exon:chrY:21755285-21755479:+</th>\n",
       "      <th>exon:chrY:21903621-21903743:-@exon:chrY:21903204-21903374:-@exon:chrY:21901414-21901548:-@exon:chrY:21897507-21897636:-</th>\n",
       "      <th>exon:chrY:2709527-2709668:+@exon:chrY:2710206-2710283:+@exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+</th>\n",
       "      <th>exon:chrY:2710206-2710283:+@exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+</th>\n",
       "      <th>exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+</th>\n",
       "      <th>exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+@exon:chrY:2734834-2735309:+</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CVN_01</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_02</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_03</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_04</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_05</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25836 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-  \\\n",
       "CVN_01                                                  0                                                    \n",
       "CVN_02                                                  0                                                    \n",
       "CVN_03                                                  0                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                  0                                                    \n",
       "\n",
       "        exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:101421203-101421385:+@exon:chr10:101439018-101439223:+@exon:chr10:101439482-101439632:+  \\\n",
       "CVN_01                                                  0                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:101480744-101480825:-@exon:chr10:101478103-101478257:-@exon:chr10:101476105-101476218:-  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                  0                                                    \n",
       "\n",
       "        exon:chr10:101507014-101507147:+@exon:chr10:101510126-101510153:+@exon:chr10:101514286-101514391:+  \\\n",
       "CVN_01                                                  0                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:101507014-101507147:+@exon:chr10:101514286-101514391:+@exon:chr10:101515382-101515891:+  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "        exon:chr10:101964263-101964414:-@exon:chr10:101961845-101961906:-@exon:chr10:101960428-101960537:-  \\\n",
       "CVN_01                                                NaN                                                    \n",
       "CVN_02                                                NaN                                                    \n",
       "CVN_03                                                NaN                                                    \n",
       "CVN_04                                                NaN                                                    \n",
       "CVN_05                                                NaN                                                    \n",
       "\n",
       "                                                             ...                                                         \\\n",
       "CVN_01                                                       ...                                                          \n",
       "CVN_02                                                       ...                                                          \n",
       "CVN_03                                                       ...                                                          \n",
       "CVN_04                                                       ...                                                          \n",
       "CVN_05                                                       ...                                                          \n",
       "\n",
       "        exon:chrY:15471647-15471866:-@exon:chrY:15470968-15471102:-@exon:chrY:15470344-15470433:-@exon:chrY:15469757-15469849:-  \\\n",
       "CVN_01                                                NaN                                                                         \n",
       "CVN_02                                                NaN                                                                         \n",
       "CVN_03                                                NaN                                                                         \n",
       "CVN_04                                                NaN                                                                         \n",
       "CVN_05                                                NaN                                                                         \n",
       "\n",
       "        exon:chrY:16634632-16634821:+@exon:chrY:16733889-16734471:+@exon:chrY:16831339-16831398:+@exon:chrY:16834997-16835149:+  \\\n",
       "CVN_01                                                  1                                                                         \n",
       "CVN_02                                                NaN                                                                         \n",
       "CVN_03                                                NaN                                                                         \n",
       "CVN_04                                                NaN                                                                         \n",
       "CVN_05                                                NaN                                                                         \n",
       "\n",
       "        exon:chrY:21237827-21237882:-@exon:chrY:21230610-21230684:-@exon:chrY:21207128-21207177:-@exon:chrY:21206428-21206581:-  \\\n",
       "CVN_01                                                NaN                                                                         \n",
       "CVN_02                                                NaN                                                                         \n",
       "CVN_03                                                NaN                                                                         \n",
       "CVN_04                                                NaN                                                                         \n",
       "CVN_05                                                NaN                                                                         \n",
       "\n",
       "        exon:chrY:21749096-21749393:+@exon:chrY:21750256-21751498:+@exon:chrY:21753666-21753845:+@exon:chrY:21755285-21755479:+  \\\n",
       "CVN_01                                                NaN                                                                         \n",
       "CVN_02                                                NaN                                                                         \n",
       "CVN_03                                                NaN                                                                         \n",
       "CVN_04                                                NaN                                                                         \n",
       "CVN_05                                                NaN                                                                         \n",
       "\n",
       "        exon:chrY:21749368-21749393:+@exon:chrY:21751407-21751498:+@exon:chrY:21753666-21753845:+@exon:chrY:21755285-21755479:+  \\\n",
       "CVN_01                                                NaN                                                                         \n",
       "CVN_02                                                NaN                                                                         \n",
       "CVN_03                                                NaN                                                                         \n",
       "CVN_04                                                NaN                                                                         \n",
       "CVN_05                                                NaN                                                                         \n",
       "\n",
       "        exon:chrY:21903621-21903743:-@exon:chrY:21903204-21903374:-@exon:chrY:21901414-21901548:-@exon:chrY:21897507-21897636:-  \\\n",
       "CVN_01                                                NaN                                                                         \n",
       "CVN_02                                                NaN                                                                         \n",
       "CVN_03                                                NaN                                                                         \n",
       "CVN_04                                                NaN                                                                         \n",
       "CVN_05                                                NaN                                                                         \n",
       "\n",
       "        exon:chrY:2709527-2709668:+@exon:chrY:2710206-2710283:+@exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+  \\\n",
       "CVN_01                                                NaN                                                                 \n",
       "CVN_02                                                NaN                                                                 \n",
       "CVN_03                                                NaN                                                                 \n",
       "CVN_04                                                NaN                                                                 \n",
       "CVN_05                                                NaN                                                                 \n",
       "\n",
       "        exon:chrY:2710206-2710283:+@exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+  \\\n",
       "CVN_01                                                NaN                                                                 \n",
       "CVN_02                                                NaN                                                                 \n",
       "CVN_03                                                NaN                                                                 \n",
       "CVN_04                                                NaN                                                                 \n",
       "CVN_05                                                NaN                                                                 \n",
       "\n",
       "        exon:chrY:2712118-2712298:+@exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+  \\\n",
       "CVN_01                                                NaN                                                                 \n",
       "CVN_02                                                NaN                                                                 \n",
       "CVN_03                                                NaN                                                                 \n",
       "CVN_04                                                NaN                                                                 \n",
       "CVN_05                                                NaN                                                                 \n",
       "\n",
       "        exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+@exon:chrY:2734834-2735309:+  \n",
       "CVN_01                                                NaN                                                                \n",
       "CVN_02                                                NaN                                                                \n",
       "CVN_03                                                NaN                                                                \n",
       "CVN_04                                                NaN                                                                \n",
       "CVN_05                                                NaN                                                                \n",
       "\n",
       "[5 rows x 25836 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv_folder = '/home/obotvinnik/projects/singlecell_pnms/analysis/csvs_for_paper'\n",
    "splicing =  pd.read_csv('{}/splicing.csv'.format(csv_folder), index_col=0)\n",
    "print splicing.shape\n",
    "splicing.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>unique_junction_reads</th>\n",
       "      <th>multimap_junction_reads</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>intron_location</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1309283</td>\n",
       "      <td>1309379</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1309283-1309379:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1309826</td>\n",
       "      <td>1310084</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1309826-1310084:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1310171</td>\n",
       "      <td>1310533</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>chr1:1310171-1310533:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1337637</td>\n",
       "      <td>1341188</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>409</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1337637-1341188:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1341267</td>\n",
       "      <td>1342288</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>337</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1341267-1342288:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  intron_start  intron_stop strand intron_motif annotated  \\\n",
       "0  chr1       1309283      1309379      -        GT/AG      True   \n",
       "1  chr1       1309826      1310084      -        GT/AG      True   \n",
       "2  chr1       1310171      1310533      -        GT/AG      True   \n",
       "3  chr1       1337637      1341188      -        GT/AG      True   \n",
       "4  chr1       1341267      1342288      -        GT/AG      True   \n",
       "\n",
       "   unique_junction_reads  multimap_junction_reads  max_overhang  \\\n",
       "0                     37                        0            49   \n",
       "1                     43                        0            49   \n",
       "2                     27                        0            47   \n",
       "3                    409                        0            50   \n",
       "4                    337                        0            50   \n",
       "\n",
       "          intron_location sample_id  \n",
       "0  chr1:1309283-1309379:-    CVN_09  \n",
       "1  chr1:1309826-1310084:-    CVN_09  \n",
       "2  chr1:1310171-1310533:-    CVN_09  \n",
       "3  chr1:1337637-1341188:-    CVN_09  \n",
       "4  chr1:1341267-1342288:-    CVN_09  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder = '/home/obotvinnik/projects/singlecell_pnms/analysis/csvs_for_paper'\n",
    "sj = pd.read_csv('{}/sj_raw.csv'.format(folder))\n",
    "    \n",
    "# print sj.shape\n",
    "# sj = sj.drop_duplicates(['sample_id', 'intron_location'])\n",
    "# print sj.shape\n",
    "    \n",
    "# sj = sj.set_index(['intron_location', 'sample_id'])\n",
    "# sj = sj.sort_index()\n",
    "sj.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get just junctions associated with constitutive exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3693174, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>unique_junction_reads</th>\n",
       "      <th>multimap_junction_reads</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>intron_location</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1309283</td>\n",
       "      <td>1309379</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1309283-1309379:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1309826</td>\n",
       "      <td>1310084</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1309826-1310084:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1337637</td>\n",
       "      <td>1341188</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>409</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1337637-1341188:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1341267</td>\n",
       "      <td>1342288</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>337</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1341267-1342288:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1342400</td>\n",
       "      <td>1342510</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>307</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1342400-1342510:-</td>\n",
       "      <td>CVN_09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  intron_start  intron_stop strand intron_motif annotated  \\\n",
       "0  chr1       1309283      1309379      -        GT/AG      True   \n",
       "1  chr1       1309826      1310084      -        GT/AG      True   \n",
       "3  chr1       1337637      1341188      -        GT/AG      True   \n",
       "4  chr1       1341267      1342288      -        GT/AG      True   \n",
       "5  chr1       1342400      1342510      -        GT/AG      True   \n",
       "\n",
       "   unique_junction_reads  multimap_junction_reads  max_overhang  \\\n",
       "0                     37                        0            49   \n",
       "1                     43                        0            49   \n",
       "3                    409                        0            50   \n",
       "4                    337                        0            50   \n",
       "5                    307                        0            50   \n",
       "\n",
       "          intron_location sample_id  \n",
       "0  chr1:1309283-1309379:-    CVN_09  \n",
       "1  chr1:1309826-1310084:-    CVN_09  \n",
       "3  chr1:1337637-1341188:-    CVN_09  \n",
       "4  chr1:1341267-1342288:-    CVN_09  \n",
       "5  chr1:1342400-1342510:-    CVN_09  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_constitutive = sj.loc[sj['intron_location'].isin(constitutive_junctions)]\n",
    "print sj_constitutive.shape\n",
    "sj_constitutive.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>single</th>\n",
       "      <th>pooled</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>neuron</th>\n",
       "      <th>progenitor</th>\n",
       "      <th>stressed</th>\n",
       "      <th>craig_venter</th>\n",
       "      <th>differentiation_batch</th>\n",
       "      <th>split</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CVN_01</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NPC</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_02</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NPC</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_03</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NPC</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_04</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NPC</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_05</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NPC</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       single pooled phenotype neuron progenitor stressed craig_venter  \\\n",
       "CVN_01   True  False       NPC  False       True    False         True   \n",
       "CVN_02   True  False       NPC  False       True    False         True   \n",
       "CVN_03   True  False       NPC  False       True    False         True   \n",
       "CVN_04   True  False       NPC  False       True    False         True   \n",
       "CVN_05   True  False       NPC  False       True    False         True   \n",
       "\n",
       "       differentiation_batch split  \n",
       "CVN_01                   NaN   NaN  \n",
       "CVN_02                   NaN   NaN  \n",
       "CVN_03                   NaN   NaN  \n",
       "CVN_04                   NaN   NaN  \n",
       "CVN_05                   NaN   NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv('{}/metadata.csv'.format(csv_folder), index_col=0)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "phenotype = metadata.phenotype[metadata.single]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call an exon constitutive if its junction start/stop is detected in at least 20 cells per celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>unique_junction_reads</th>\n",
       "      <th>multimap_junction_reads</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>intron_location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CVN_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1309283</td>\n",
       "      <td>1309379</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1309283-1309379:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1309826</td>\n",
       "      <td>1310084</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1309826-1310084:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1337637</td>\n",
       "      <td>1341188</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>409</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1337637-1341188:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1341267</td>\n",
       "      <td>1342288</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>337</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1341267-1342288:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1342400</td>\n",
       "      <td>1342510</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>307</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1342400-1342510:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  intron_start  intron_stop strand intron_motif annotated  \\\n",
       "sample_id                                                                  \n",
       "CVN_09     chr1       1309283      1309379      -        GT/AG      True   \n",
       "CVN_09     chr1       1309826      1310084      -        GT/AG      True   \n",
       "CVN_09     chr1       1337637      1341188      -        GT/AG      True   \n",
       "CVN_09     chr1       1341267      1342288      -        GT/AG      True   \n",
       "CVN_09     chr1       1342400      1342510      -        GT/AG      True   \n",
       "\n",
       "           unique_junction_reads  multimap_junction_reads  max_overhang  \\\n",
       "sample_id                                                                 \n",
       "CVN_09                        37                        0            49   \n",
       "CVN_09                        43                        0            49   \n",
       "CVN_09                       409                        0            50   \n",
       "CVN_09                       337                        0            50   \n",
       "CVN_09                       307                        0            50   \n",
       "\n",
       "                  intron_location  \n",
       "sample_id                          \n",
       "CVN_09     chr1:1309283-1309379:-  \n",
       "CVN_09     chr1:1309826-1310084:-  \n",
       "CVN_09     chr1:1337637-1341188:-  \n",
       "CVN_09     chr1:1341267-1342288:-  \n",
       "CVN_09     chr1:1342400-1342510:-  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_constitutive = sj_constitutive.set_index('sample_id')\n",
    "sj_constitutive.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('chr1', 155766.0, '-') 27 (35, 10)\n",
      "('chr1', 164262.0, '-') 26 (26, 10)\n",
      "('chr1', 880436.0, '-') 22 (22, 10)\n",
      "('chr1', 880897.0, '-') 29 (29, 10)\n",
      "('chr1', 881552.0, '-') 30 (30, 10)\n",
      "('chr1', 881781.0, '-') 30 (31, 10)\n",
      "('chr1', 883510.0, '-') 30 (30, 10)\n",
      "('chr1', 883869.0, '-') 29 (29, 10)\n",
      "('chr1', 887379.0, '-') 21 (21, 10)\n",
      "('chr1', 887791.0, '-') 24 (24, 10)\n",
      "('chr1', 888554.0, '-') 26 (26, 10)\n",
      "('chr1', 889161.0, '-') 26 (26, 10)\n",
      "('chr1', 889383.0, '-') 24 (24, 10)\n",
      "('chr1', 891302.0, '-') 25 (25, 10)\n",
      "('chr1', 891474.0, '-') 25 (25, 10)\n",
      "('chr1', 892273.0, '-') 22 (22, 10)\n",
      "('chr1', 892478.0, '-') 22 (22, 10)\n",
      "('chr1', 894594.0, '-') 27 (27, 10)\n"
     ]
    }
   ],
   "source": [
    "for p, phenotype_df in sj_constitutive.groupby(phenotype):\n",
    "    for i, (intron, intron_df) in zip(range(50), phenotype_df.groupby(['chrom', 'intron_stop', 'strand'])):\n",
    "        n_unique_samples = len(intron_df.index.unique())\n",
    "        if n_unique_samples >= 20:\n",
    "            print intron, n_unique_samples, intron_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>unique_junction_reads</th>\n",
       "      <th>multimap_junction_reads</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>intron_location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P4_12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_09</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_07</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_05</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1164327</td>\n",
       "      <td>1167271</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:1164327-1167271:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  intron_start  intron_stop strand intron_motif annotated  \\\n",
       "sample_id                                                                  \n",
       "P4_12      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P7_10      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P8_10      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P7_09      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P1_09      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P2_10      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P2_07      chr1       1164327      1167271      -        GT/AG      True   \n",
       "P8_05      chr1       1164327      1167271      -        GT/AG      True   \n",
       "\n",
       "           unique_junction_reads  multimap_junction_reads  max_overhang  \\\n",
       "sample_id                                                                 \n",
       "P4_12                         18                        0            49   \n",
       "P7_10                         32                        0            50   \n",
       "P8_10                         10                        0            49   \n",
       "P7_09                         32                        0            50   \n",
       "P1_09                         15                        0            45   \n",
       "P2_10                         13                        0            46   \n",
       "P2_07                         47                        0            43   \n",
       "P8_05                         67                        0            49   \n",
       "\n",
       "                  intron_location  \n",
       "sample_id                          \n",
       "P4_12      chr1:1164327-1167271:-  \n",
       "P7_10      chr1:1164327-1167271:-  \n",
       "P8_10      chr1:1164327-1167271:-  \n",
       "P7_09      chr1:1164327-1167271:-  \n",
       "P1_09      chr1:1164327-1167271:-  \n",
       "P2_10      chr1:1164327-1167271:-  \n",
       "P2_07      chr1:1164327-1167271:-  \n",
       "P8_05      chr1:1164327-1167271:-  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "148708"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = sj_constitutive.groupby(['chrom', 'intron_stop', 'strand']).size()\n",
    "len(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3693174, 10)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_constitutive.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3147848, 10)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom = 'chrom'\n",
    "start = 'intron_start'\n",
    "stop = 'intron_stop'\n",
    "strand = 'strand'\n",
    "\n",
    "sj_constitutive_stop = sj_constitutive.groupby([chrom, stop, strand]).filter(lambda x: len(x) >= 20)\n",
    "sj_constitutive_stop.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "48075"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(s>=20).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phenotype\n",
      "iPSC    35\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for i, (group, df) in zip(range(20), sj_constitutive.groupby(['chrom', 'intron_stop', 'strand'])):\n",
    "    s = df.groupby(phenotype).size()\n",
    "    if (s >= 20).any():\n",
    "        print s[s >= 20]\n",
    "        break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>unique_junction_reads</th>\n",
       "      <th>multimap_junction_reads</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>intron_location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P4_06</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_03</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>34</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_01</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>44</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>60</td>\n",
       "      <td>103</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>25</td>\n",
       "      <td>16</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9_01</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>129</td>\n",
       "      <td>159</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9_01</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153544</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>58</td>\n",
       "      <td>60</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:153544-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>118</td>\n",
       "      <td>165</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153544</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>20</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:153544-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>29</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>88</td>\n",
       "      <td>72</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_01</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153544</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:153544-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149073</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>17</td>\n",
       "      <td>32</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:149073-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153544</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>13</td>\n",
       "      <td>95</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:153544-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_01</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153544</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "      <td>48</td>\n",
       "      <td>chr1:153544-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_05</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>42</td>\n",
       "      <td>33</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_05</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149073</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:149073-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>143012</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>21</td>\n",
       "      <td>33</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:143012-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_02</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_06</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146340</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>16</td>\n",
       "      <td>76</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146340-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_06</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>31</td>\n",
       "      <td>34</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_01</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>17</td>\n",
       "      <td>43</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3_03</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146340</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>56</td>\n",
       "      <td>138</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146340-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3_03</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>24</td>\n",
       "      <td>45</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_06</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>47</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>73</td>\n",
       "      <td>45</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9_03</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>182</td>\n",
       "      <td>237</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_05</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>46</td>\n",
       "      <td>78</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8_05</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146812</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>48</td>\n",
       "      <td>94</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:146812-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_03</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_07</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>48</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_07</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153544</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>50</td>\n",
       "      <td>chr1:153544-155766:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7_08</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146510</td>\n",
       "      <td>155766</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "      <td>65</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1:146510-155766:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  intron_start  intron_stop strand intron_motif annotated  \\\n",
       "sample_id                                                                  \n",
       "P4_06      chr1        146510       155766      -        GT/AG      True   \n",
       "P2_03      chr1        146510       155766      -        GT/AG      True   \n",
       "P1_01      chr1        146510       155766      -        GT/AG      True   \n",
       "P8_10      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_02      chr1        146510       155766      -        GT/AG      True   \n",
       "P9_01      chr1        146510       155766      -        GT/AG      True   \n",
       "P9_01      chr1        153544       155766      -        GT/AG     False   \n",
       "P9_02      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_11      chr1        153544       155766      -        GT/AG     False   \n",
       "P3_02      chr1        146510       155766      -        GT/AG      True   \n",
       "P4_11      chr1        146510       155766      -        GT/AG      True   \n",
       "P2_13      chr1        146510       155766      -        GT/AG      True   \n",
       "P4_01      chr1        153544       155766      -        GT/AG     False   \n",
       "P8_02      chr1        146510       155766      -        GT/AG      True   \n",
       "P8_02      chr1        149073       155766      -        GT/AG     False   \n",
       "P8_02      chr1        153544       155766      -        GT/AG     False   \n",
       "P7_01      chr1        153544       155766      -        GT/AG     False   \n",
       "P1_05      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_05      chr1        149073       155766      -        GT/AG     False   \n",
       "P1_02      chr1        143012       155766      -        GT/AG     False   \n",
       "P1_02      chr1        146510       155766      -        GT/AG      True   \n",
       "P8_06      chr1        146340       155766      -        GT/AG     False   \n",
       "P8_06      chr1        146510       155766      -        GT/AG      True   \n",
       "P2_01      chr1        146510       155766      -        GT/AG      True   \n",
       "P3_03      chr1        146340       155766      -        GT/AG     False   \n",
       "P3_03      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_06      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_12      chr1        146510       155766      -        GT/AG      True   \n",
       "P9_03      chr1        146510       155766      -        GT/AG      True   \n",
       "P8_05      chr1        146510       155766      -        GT/AG      True   \n",
       "P8_05      chr1        146812       155766      -        GT/AG     False   \n",
       "P1_03      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_07      chr1        146510       155766      -        GT/AG      True   \n",
       "P7_07      chr1        153544       155766      -        GT/AG     False   \n",
       "P7_08      chr1        146510       155766      -        GT/AG      True   \n",
       "\n",
       "           unique_junction_reads  multimap_junction_reads  max_overhang  \\\n",
       "sample_id                                                                 \n",
       "P4_06                         12                       23            49   \n",
       "P2_03                         10                        5            34   \n",
       "P1_01                         19                       22            44   \n",
       "P8_10                         60                      103            49   \n",
       "P7_02                         25                       16            49   \n",
       "P9_01                        129                      159            50   \n",
       "P9_01                         58                       60            49   \n",
       "P9_02                        118                      165            49   \n",
       "P7_11                         20                       33            49   \n",
       "P3_02                         10                        6            29   \n",
       "P4_11                         19                       20            49   \n",
       "P2_13                         88                       72            46   \n",
       "P4_01                         15                        6            46   \n",
       "P8_02                         12                       22            46   \n",
       "P8_02                         17                       32            50   \n",
       "P8_02                         13                       95            49   \n",
       "P7_01                         13                       23            48   \n",
       "P1_05                         42                       33            46   \n",
       "P7_05                         14                       21            46   \n",
       "P1_02                         21                       33            46   \n",
       "P1_02                         37                       38            46   \n",
       "P8_06                         16                       76            49   \n",
       "P8_06                         31                       34            49   \n",
       "P2_01                         11                       17            43   \n",
       "P3_03                         56                      138            46   \n",
       "P3_03                         24                       45            46   \n",
       "P7_06                         13                        7            47   \n",
       "P7_12                         73                       45            50   \n",
       "P9_03                        182                      237            49   \n",
       "P8_05                         46                       78            49   \n",
       "P8_05                         48                       94            50   \n",
       "P1_03                         11                        9            46   \n",
       "P7_07                         10                       24            48   \n",
       "P7_07                         10                       16            50   \n",
       "P7_08                         50                       65            49   \n",
       "\n",
       "                intron_location  \n",
       "sample_id                        \n",
       "P4_06      chr1:146510-155766:-  \n",
       "P2_03      chr1:146510-155766:-  \n",
       "P1_01      chr1:146510-155766:-  \n",
       "P8_10      chr1:146510-155766:-  \n",
       "P7_02      chr1:146510-155766:-  \n",
       "P9_01      chr1:146510-155766:-  \n",
       "P9_01      chr1:153544-155766:-  \n",
       "P9_02      chr1:146510-155766:-  \n",
       "P7_11      chr1:153544-155766:-  \n",
       "P3_02      chr1:146510-155766:-  \n",
       "P4_11      chr1:146510-155766:-  \n",
       "P2_13      chr1:146510-155766:-  \n",
       "P4_01      chr1:153544-155766:-  \n",
       "P8_02      chr1:146510-155766:-  \n",
       "P8_02      chr1:149073-155766:-  \n",
       "P8_02      chr1:153544-155766:-  \n",
       "P7_01      chr1:153544-155766:-  \n",
       "P1_05      chr1:146510-155766:-  \n",
       "P7_05      chr1:149073-155766:-  \n",
       "P1_02      chr1:143012-155766:-  \n",
       "P1_02      chr1:146510-155766:-  \n",
       "P8_06      chr1:146340-155766:-  \n",
       "P8_06      chr1:146510-155766:-  \n",
       "P2_01      chr1:146510-155766:-  \n",
       "P3_03      chr1:146340-155766:-  \n",
       "P3_03      chr1:146510-155766:-  \n",
       "P7_06      chr1:146510-155766:-  \n",
       "P7_12      chr1:146510-155766:-  \n",
       "P9_03      chr1:146510-155766:-  \n",
       "P8_05      chr1:146510-155766:-  \n",
       "P8_05      chr1:146812-155766:-  \n",
       "P1_03      chr1:146510-155766:-  \n",
       "P7_07      chr1:146510-155766:-  \n",
       "P7_07      chr1:153544-155766:-  \n",
       "P7_08      chr1:146510-155766:-  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped = df.groupby(phenotype)\n",
    "s = grouped.size()\n",
    "\n",
    "pd.concat([grouped.get_group(g) for g in s[s >= 20].index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phenotype\n",
       "MN      2\n",
       "NPC     1\n",
       "iPSC    1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5min 21s, sys: 304 ms, total: 5min 22s\n",
      "Wall time: 5min 22s\n"
     ]
    }
   ],
   "source": [
    "def filter_group_size(df, groupby, size):\n",
    "    grouped = df.groupby(phenotype)\n",
    "    s = grouped.size()\n",
    "    ind = s >= 20\n",
    "#     if (ind).any():\n",
    "#         return pd.concat([grouped.get_group(g) for g in s[ind].index])\n",
    "#     else:\n",
    "#         return pd.DataFrame()\n",
    "    return s[ind]\n",
    "\n",
    "%time constitutive_stops = sj_constitutive.groupby(['chrom', 'intron_stop', 'strand', 'intron_location']).apply(filter_group_size, groupby=phenotype, size=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrom  intron_stop  strand  intron_location     \n",
       "chr1   155766       -       chr1:146510-155766:-    iPSC\n",
       "       164262       -       chr1:155832-164262:-    iPSC\n",
       "       880436       -       chr1:880181-880436:-    iPSC\n",
       "       880897       -       chr1:880527-880897:-    iPSC\n",
       "       881552       -       chr1:881034-881552:-    iPSC\n",
       "Name: phenotype, dtype: object"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_stops_phenotype = constitutive_stops.reset_index(level=-1)\n",
    "constitutive_stops_phenotype.phenotype.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5min 15s, sys: 788 ms, total: 5min 16s\n",
      "Wall time: 5min 16s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "constitutive_starts = sj_constitutive.groupby(['chrom', 'intron_start', 'strand', 'intron_location']).apply(\n",
    "    filter_group_size, groupby=phenotype, size=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrom  intron_start  strand  intron_location       phenotype\n",
       "chr1   146510        -       chr1:146510-155766:-  iPSC         23\n",
       "       155832        -       chr1:155832-164262:-  iPSC         25\n",
       "       880181        -       chr1:880181-880436:-  iPSC         22\n",
       "       880527        -       chr1:880527-880897:-  iPSC         29\n",
       "       881034        -       chr1:881034-881552:-  iPSC         30\n",
       "dtype: int64"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_starts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrom  intron_stop  strand  intron_location     \n",
       "chr1   155766       -       chr1:146510-155766:-    iPSC\n",
       "       164262       -       chr1:155832-164262:-    iPSC\n",
       "       880436       -       chr1:880181-880436:-    iPSC\n",
       "       880897       -       chr1:880527-880897:-    iPSC\n",
       "       881552       -       chr1:881034-881552:-    iPSC\n",
       "Name: phenotype, dtype: object"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_stops_phenotype = constitutive_stops.reset_index(level=-1)\n",
    "constitutive_stops_phenotype.phenotype.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrom  intron_start  strand\n",
       "chr1   146510        -         iPSC\n",
       "       155832        -         iPSC\n",
       "       880181        -         iPSC\n",
       "       880527        -         iPSC\n",
       "       881034        -         iPSC\n",
       "Name: phenotype, dtype: object"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_starts_phenotype = constitutive_starts.reset_index(level=-1)\n",
    "constitutive_starts_phenotype.phenotype.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if __name__ == '__main__':\n",
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  from ipykernel import kernelapp as app\n",
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  app.launch_new_instance()\n",
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction_location</th>\n",
       "      <th>exon</th>\n",
       "      <th>direction</th>\n",
       "      <th>junction_region</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100008749-100010821:-</td>\n",
       "      <td>exon:chr10:100010822-100010933:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100008749-100010821:-</td>\n",
       "      <td>100008749</td>\n",
       "      <td>100010821</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100143626-100144703:-</td>\n",
       "      <td>exon:chr10:100144704-100144824:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100143626-100144703:-</td>\n",
       "      <td>100143626</td>\n",
       "      <td>100144703</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:100144825-100146957:-</td>\n",
       "      <td>exon:chr10:100146958-100147064:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100144825-100146957:-</td>\n",
       "      <td>100144825</td>\n",
       "      <td>100146957</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:100147065-100155147:-</td>\n",
       "      <td>exon:chr10:100155148-100155209:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100147065-100155147:-</td>\n",
       "      <td>100147065</td>\n",
       "      <td>100155147</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10:100155210-100157101:-</td>\n",
       "      <td>exon:chr10:100157102-100157255:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100155210-100157101:-</td>\n",
       "      <td>100155210</td>\n",
       "      <td>100157101</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             junction_location                              exon direction  \\\n",
       "0  chr10:100008749-100010821:-  exon:chr10:100010822-100010933:-  upstream   \n",
       "2  chr10:100143626-100144703:-  exon:chr10:100144704-100144824:-  upstream   \n",
       "3  chr10:100144825-100146957:-  exon:chr10:100146958-100147064:-  upstream   \n",
       "4  chr10:100147065-100155147:-  exon:chr10:100155148-100155209:-  upstream   \n",
       "6  chr10:100155210-100157101:-  exon:chr10:100157102-100157255:-  upstream   \n",
       "\n",
       "               junction_region  intron_start  intron_stop  chrom strand  \n",
       "0  chr10:100008749-100010821:-     100008749    100010821  chr10      -  \n",
       "2  chr10:100143626-100144703:-     100143626    100144703  chr10      -  \n",
       "3  chr10:100144825-100146957:-     100144825    100146957  chr10      -  \n",
       "4  chr10:100147065-100155147:-     100147065    100155147  chr10      -  \n",
       "6  chr10:100155210-100157101:-     100155210    100157101  chr10      -  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive['junction_region'] = constitutive.junction_location.apply(Region)\n",
    "constitutive['intron_start'] = constitutive.junction_region.apply(lambda x: x._start)\n",
    "constitutive['intron_stop'] = constitutive.junction_region.apply(lambda x: x._stop)\n",
    "constitutive['chrom'] = constitutive.junction_region.apply(lambda x: x.chrom)\n",
    "constitutive['strand'] = constitutive.junction_region.apply(lambda x: x.strand)\n",
    "constitutive.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>junction_location</th>\n",
       "      <th>exon</th>\n",
       "      <th>direction</th>\n",
       "      <th>junction_region</th>\n",
       "      <th>intron_stop</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr10</th>\n",
       "      <th>100008749</th>\n",
       "      <th>-</th>\n",
       "      <td>chr10:100008749-100010821:-</td>\n",
       "      <td>exon:chr10:100010822-100010933:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100008749-100010821:-</td>\n",
       "      <td>100010821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100143626</th>\n",
       "      <th>-</th>\n",
       "      <td>chr10:100143626-100144703:-</td>\n",
       "      <td>exon:chr10:100144704-100144824:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100143626-100144703:-</td>\n",
       "      <td>100144703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100144825</th>\n",
       "      <th>-</th>\n",
       "      <td>chr10:100144825-100146957:-</td>\n",
       "      <td>exon:chr10:100146958-100147064:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100144825-100146957:-</td>\n",
       "      <td>100146957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100147065</th>\n",
       "      <th>-</th>\n",
       "      <td>chr10:100147065-100155147:-</td>\n",
       "      <td>exon:chr10:100155148-100155209:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100147065-100155147:-</td>\n",
       "      <td>100155147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100155210</th>\n",
       "      <th>-</th>\n",
       "      <td>chr10:100155210-100157101:-</td>\n",
       "      <td>exon:chr10:100157102-100157255:-</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chr10:100155210-100157101:-</td>\n",
       "      <td>100157101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     junction_location  \\\n",
       "chrom intron_start strand                                \n",
       "chr10 100008749    -       chr10:100008749-100010821:-   \n",
       "      100143626    -       chr10:100143626-100144703:-   \n",
       "      100144825    -       chr10:100144825-100146957:-   \n",
       "      100147065    -       chr10:100147065-100155147:-   \n",
       "      100155210    -       chr10:100155210-100157101:-   \n",
       "\n",
       "                                                       exon direction  \\\n",
       "chrom intron_start strand                                               \n",
       "chr10 100008749    -       exon:chr10:100010822-100010933:-  upstream   \n",
       "      100143626    -       exon:chr10:100144704-100144824:-  upstream   \n",
       "      100144825    -       exon:chr10:100146958-100147064:-  upstream   \n",
       "      100147065    -       exon:chr10:100155148-100155209:-  upstream   \n",
       "      100155210    -       exon:chr10:100157102-100157255:-  upstream   \n",
       "\n",
       "                                       junction_region  intron_stop  \n",
       "chrom intron_start strand                                            \n",
       "chr10 100008749    -       chr10:100008749-100010821:-    100010821  \n",
       "      100143626    -       chr10:100143626-100144703:-    100144703  \n",
       "      100144825    -       chr10:100144825-100146957:-    100146957  \n",
       "      100147065    -       chr10:100147065-100155147:-    100155147  \n",
       "      100155210    -       chr10:100155210-100157101:-    100157101  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_index = constitutive.set_index(['chrom', 'intron_start', 'strand'])\n",
    "constitutive_index.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             junction_location                              exon   direction  \\\n",
      "0  chr10:100008749-100010821:-  exon:chr10:100010822-100010933:-    upstream   \n",
      "1  chr10:100010934-100011322:-  exon:chr10:100010822-100010933:-  downstream   \n",
      "\n",
      "               junction_region  intron_start  intron_stop  chrom strand  \n",
      "0  chr10:100008749-100010821:-     100008749    100010821  chr10      -  \n",
      "1  chr10:100010934-100011322:-     100010934    100011322  chr10      -  \n",
      "             junction_location                              exon   direction  \\\n",
      "2  chr10:100143626-100144703:-  exon:chr10:100144704-100144824:-    upstream   \n",
      "3  chr10:100144825-100146957:-  exon:chr10:100144704-100144824:-  downstream   \n",
      "\n",
      "               junction_region  intron_start  intron_stop  chrom strand  \n",
      "2  chr10:100143626-100144703:-     100143626    100144703  chr10      -  \n",
      "3  chr10:100144825-100146957:-     100144825    100146957  chr10      -  \n",
      "             junction_location                              exon   direction  \\\n",
      "3  chr10:100144825-100146957:-  exon:chr10:100146958-100147064:-    upstream   \n",
      "4  chr10:100147065-100155147:-  exon:chr10:100146958-100147064:-  downstream   \n",
      "\n",
      "               junction_region  intron_start  intron_stop  chrom strand  \n",
      "3  chr10:100144825-100146957:-     100144825    100146957  chr10      -  \n",
      "4  chr10:100147065-100155147:-     100147065    100155147  chr10      -  \n",
      "             junction_location                              exon   direction  \\\n",
      "4  chr10:100147065-100155147:-  exon:chr10:100155148-100155209:-    upstream   \n",
      "6  chr10:100155210-100157101:-  exon:chr10:100155148-100155209:-  downstream   \n",
      "\n",
      "               junction_region  intron_start  intron_stop  chrom strand  \n",
      "4  chr10:100147065-100155147:-     100147065    100155147  chr10      -  \n",
      "6  chr10:100155210-100157101:-     100155210    100157101  chr10      -  \n",
      "             junction_location                              exon   direction  \\\n",
      "6  chr10:100155210-100157101:-  exon:chr10:100157102-100157255:-    upstream   \n",
      "7  chr10:100157256-100159858:-  exon:chr10:100157102-100157255:-  downstream   \n",
      "\n",
      "               junction_region  intron_start  intron_stop  chrom strand  \n",
      "6  chr10:100155210-100157101:-     100155210    100157101  chr10      -  \n",
      "7  chr10:100157256-100159858:-     100157256    100159858  chr10      -  \n"
     ]
    }
   ],
   "source": [
    "for i, (exon, df) in zip(range(5), constitutive.groupby('exon')):\n",
    "    print df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# sj_constitutive = sj.loc[idx[constitutive_junctions, :], :]\n",
    "# print sj_constitutive.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sj_constitutive' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-a449682a88db>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0msj_constitutive\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'sj_constitutive' is not defined"
     ]
    }
   ],
   "source": [
    "# sj_constitutive.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sj.reset_index().join"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr10:101157444-101162336:-|5p</th>\n",
       "      <th>chr10:101162480-101163225:-|5p</th>\n",
       "      <th>chr10:101163392-101163480:-|5p</th>\n",
       "      <th>chr10:101163392-101165512:-|5p</th>\n",
       "      <th>chr10:101163632-101165512:-|5p</th>\n",
       "      <th>chr10:101166007-101190204:-|5p</th>\n",
       "      <th>chr10:101437763-101476104:-|5p</th>\n",
       "      <th>chr10:101456299-101476104:-|5p</th>\n",
       "      <th>chr10:101473237-101476104:-|5p</th>\n",
       "      <th>chr10:101474476-101476104:-|5p</th>\n",
       "      <th>...</th>\n",
       "      <th>chrY:2713785-2734833:+|3p</th>\n",
       "      <th>chrY:2714508-2722640:+|3p</th>\n",
       "      <th>chrY:2721733-2722640:+|3p</th>\n",
       "      <th>chrY:2722746-2734833:+|3p</th>\n",
       "      <th>chrY:2722813-2733128:+|3p</th>\n",
       "      <th>chrY:2722813-2734833:+|3p</th>\n",
       "      <th>chrY:2722813-2789783:+|3p</th>\n",
       "      <th>chrY:2722813-2796904:+|3p</th>\n",
       "      <th>chrY:2733287-2734833:+|3p</th>\n",
       "      <th>chrY:2733362-2734833:+|3p</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CVN_01</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_02</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_03</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_04</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_05</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 61766 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr10:101157444-101162336:-|5p  chr10:101162480-101163225:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                  1                               1   \n",
       "CVN_02                                  1                               1   \n",
       "CVN_03                                  1                               1   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                               1   \n",
       "\n",
       "           chr10:101163392-101163480:-|5p  chr10:101163392-101165512:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                  1                               0   \n",
       "CVN_02                                  1                               0   \n",
       "CVN_03                                  1                               0   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                               0   \n",
       "\n",
       "           chr10:101163632-101165512:-|5p  chr10:101166007-101190204:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                  1                               0   \n",
       "CVN_02                                  1                               0   \n",
       "CVN_03                                  1                               0   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                               0   \n",
       "\n",
       "           chr10:101437763-101476104:-|5p  chr10:101456299-101476104:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                NaN                             NaN   \n",
       "CVN_02                                NaN                             NaN   \n",
       "CVN_03                                NaN                             NaN   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  0                               1   \n",
       "\n",
       "           chr10:101473237-101476104:-|5p  chr10:101474476-101476104:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                NaN                             NaN   \n",
       "CVN_02                                NaN                             NaN   \n",
       "CVN_03                                NaN                             NaN   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  0                               1   \n",
       "\n",
       "                     ...              chrY:2713785-2734833:+|3p  \\\n",
       "sample_id            ...                                          \n",
       "CVN_01               ...                                      0   \n",
       "CVN_02               ...                                    NaN   \n",
       "CVN_03               ...                                    NaN   \n",
       "CVN_04               ...                                      0   \n",
       "CVN_05               ...                                      0   \n",
       "\n",
       "           chrY:2714508-2722640:+|3p  chrY:2721733-2722640:+|3p  \\\n",
       "sample_id                                                         \n",
       "CVN_01                             0                          0   \n",
       "CVN_02                           NaN                        NaN   \n",
       "CVN_03                           NaN                        NaN   \n",
       "CVN_04                             0                          0   \n",
       "CVN_05                             0                          0   \n",
       "\n",
       "           chrY:2722746-2734833:+|3p  chrY:2722813-2733128:+|3p  \\\n",
       "sample_id                                                         \n",
       "CVN_01                             0                          1   \n",
       "CVN_02                           NaN                        NaN   \n",
       "CVN_03                           NaN                        NaN   \n",
       "CVN_04                             0                          1   \n",
       "CVN_05                             0                          1   \n",
       "\n",
       "           chrY:2722813-2734833:+|3p  chrY:2722813-2789783:+|3p  \\\n",
       "sample_id                                                         \n",
       "CVN_01                             0                          0   \n",
       "CVN_02                           NaN                        NaN   \n",
       "CVN_03                           NaN                        NaN   \n",
       "CVN_04                             0                          0   \n",
       "CVN_05                             0                          0   \n",
       "\n",
       "           chrY:2722813-2796904:+|3p  chrY:2733287-2734833:+|3p  \\\n",
       "sample_id                                                         \n",
       "CVN_01                             0                          1   \n",
       "CVN_02                           NaN                        NaN   \n",
       "CVN_03                           NaN                        NaN   \n",
       "CVN_04                             0                          1   \n",
       "CVN_05                             0                          1   \n",
       "\n",
       "           chrY:2733362-2734833:+|3p  \n",
       "sample_id                             \n",
       "CVN_01                             0  \n",
       "CVN_02                           NaN  \n",
       "CVN_03                           NaN  \n",
       "CVN_04                             0  \n",
       "CVN_05                             0  \n",
       "\n",
       "[5 rows x 61766 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psi20_all = psi3_psi5.loc[:, (psi3_psi5.groupby(phenotype).count() >= 20).all()]\n",
    "psi20_all.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get Junctions whose psi score is equal to 1 in all celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr10:101157444-101162336:-|5p</th>\n",
       "      <th>chr10:101162480-101163225:-|5p</th>\n",
       "      <th>chr10:101163392-101163480:-|5p</th>\n",
       "      <th>chr10:101163392-101165512:-|5p</th>\n",
       "      <th>chr10:101163632-101165512:-|5p</th>\n",
       "      <th>chr10:101166007-101190204:-|5p</th>\n",
       "      <th>chr10:101437763-101476104:-|5p</th>\n",
       "      <th>chr10:101456299-101476104:-|5p</th>\n",
       "      <th>chr10:101473237-101476104:-|5p</th>\n",
       "      <th>chr10:101474476-101476104:-|5p</th>\n",
       "      <th>...</th>\n",
       "      <th>chrY:2713785-2734833:+|3p</th>\n",
       "      <th>chrY:2714508-2722640:+|3p</th>\n",
       "      <th>chrY:2721733-2722640:+|3p</th>\n",
       "      <th>chrY:2722746-2734833:+|3p</th>\n",
       "      <th>chrY:2722813-2733128:+|3p</th>\n",
       "      <th>chrY:2722813-2734833:+|3p</th>\n",
       "      <th>chrY:2722813-2789783:+|3p</th>\n",
       "      <th>chrY:2722813-2796904:+|3p</th>\n",
       "      <th>chrY:2733287-2734833:+|3p</th>\n",
       "      <th>chrY:2733362-2734833:+|3p</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenotype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MN</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NPC</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iPSC</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 61766 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr10:101157444-101162336:-|5p chr10:101162480-101163225:-|5p  \\\n",
       "phenotype                                                                 \n",
       "MN                                  True                           True   \n",
       "NPC                                 True                           True   \n",
       "iPSC                                True                           True   \n",
       "\n",
       "          chr10:101163392-101163480:-|5p chr10:101163392-101165512:-|5p  \\\n",
       "phenotype                                                                 \n",
       "MN                                 False                          False   \n",
       "NPC                                 True                          False   \n",
       "iPSC                               False                          False   \n",
       "\n",
       "          chr10:101163632-101165512:-|5p chr10:101166007-101190204:-|5p  \\\n",
       "phenotype                                                                 \n",
       "MN                                  True                          False   \n",
       "NPC                                 True                          False   \n",
       "iPSC                                True                          False   \n",
       "\n",
       "          chr10:101437763-101476104:-|5p chr10:101456299-101476104:-|5p  \\\n",
       "phenotype                                                                 \n",
       "MN                                 False                          False   \n",
       "NPC                                False                          False   \n",
       "iPSC                               False                          False   \n",
       "\n",
       "          chr10:101473237-101476104:-|5p chr10:101474476-101476104:-|5p  \\\n",
       "phenotype                                                                 \n",
       "MN                                 False                          False   \n",
       "NPC                                False                          False   \n",
       "iPSC                               False                          False   \n",
       "\n",
       "                     ...            chrY:2713785-2734833:+|3p  \\\n",
       "phenotype            ...                                        \n",
       "MN                   ...                                False   \n",
       "NPC                  ...                                False   \n",
       "iPSC                 ...                                False   \n",
       "\n",
       "          chrY:2714508-2722640:+|3p chrY:2721733-2722640:+|3p  \\\n",
       "phenotype                                                       \n",
       "MN                            False                     False   \n",
       "NPC                           False                     False   \n",
       "iPSC                          False                     False   \n",
       "\n",
       "          chrY:2722746-2734833:+|3p chrY:2722813-2733128:+|3p  \\\n",
       "phenotype                                                       \n",
       "MN                            False                      True   \n",
       "NPC                           False                      True   \n",
       "iPSC                          False                     False   \n",
       "\n",
       "          chrY:2722813-2734833:+|3p chrY:2722813-2789783:+|3p  \\\n",
       "phenotype                                                       \n",
       "MN                            False                     False   \n",
       "NPC                           False                     False   \n",
       "iPSC                          False                     False   \n",
       "\n",
       "          chrY:2722813-2796904:+|3p chrY:2733287-2734833:+|3p  \\\n",
       "phenotype                                                       \n",
       "MN                            False                     False   \n",
       "NPC                           False                      True   \n",
       "iPSC                          False                     False   \n",
       "\n",
       "          chrY:2733362-2734833:+|3p  \n",
       "phenotype                            \n",
       "MN                            False  \n",
       "NPC                           False  \n",
       "iPSC                          False  \n",
       "\n",
       "[3 rows x 61766 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psi_is_1 = psi20_all.groupby(phenotype).apply(lambda x: (x == 1)[x.notnull()].all(axis=0))\n",
    "psi_is_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr10:101157444-101162336:-|5p</th>\n",
       "      <th>chr10:101162480-101163225:-|5p</th>\n",
       "      <th>chr10:101163632-101165512:-|5p</th>\n",
       "      <th>chr10:101480826-101483712:-|5p</th>\n",
       "      <th>chr10:101483881-101486724:-|5p</th>\n",
       "      <th>chr10:101487321-101489309:-|5p</th>\n",
       "      <th>chr10:101489492-101491716:-|5p</th>\n",
       "      <th>chr10:102035251-102039880:-|5p</th>\n",
       "      <th>chr10:102107289-102107820:+|5p</th>\n",
       "      <th>chr10:102286312-102286731:-|5p</th>\n",
       "      <th>...</th>\n",
       "      <th>chrX:77378872-77380370:+|3p</th>\n",
       "      <th>chrX:77380923-77381286:+|3p</th>\n",
       "      <th>chrX:80532669-80533829:+|3p</th>\n",
       "      <th>chrX:80533911-80552693:+|3p</th>\n",
       "      <th>chrX:99887566-99888401:-|3p</th>\n",
       "      <th>chrX:99890250-99890554:-|3p</th>\n",
       "      <th>chrY:1455653-1456171:-|3p</th>\n",
       "      <th>chrY:1456313-1458133:-|3p</th>\n",
       "      <th>chrY:14774638-14776570:+|3p</th>\n",
       "      <th>chrY:2710284-2712117:+|3p</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CVN_01</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_02</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_03</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_04</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVN_05</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 9313 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr10:101157444-101162336:-|5p  chr10:101162480-101163225:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                  1                               1   \n",
       "CVN_02                                  1                               1   \n",
       "CVN_03                                  1                               1   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                               1   \n",
       "\n",
       "           chr10:101163632-101165512:-|5p  chr10:101480826-101483712:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                  1                             NaN   \n",
       "CVN_02                                  1                             NaN   \n",
       "CVN_03                                  1                             NaN   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                               1   \n",
       "\n",
       "           chr10:101483881-101486724:-|5p  chr10:101487321-101489309:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                NaN                             NaN   \n",
       "CVN_02                                NaN                             NaN   \n",
       "CVN_03                                NaN                             NaN   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                               1   \n",
       "\n",
       "           chr10:101489492-101491716:-|5p  chr10:102035251-102039880:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                NaN                               1   \n",
       "CVN_02                                NaN                             NaN   \n",
       "CVN_03                                NaN                               1   \n",
       "CVN_04                                NaN                             NaN   \n",
       "CVN_05                                  1                             NaN   \n",
       "\n",
       "           chr10:102107289-102107820:+|5p  chr10:102286312-102286731:-|5p  \\\n",
       "sample_id                                                                   \n",
       "CVN_01                                NaN                               1   \n",
       "CVN_02                                  1                               1   \n",
       "CVN_03                                  1                               1   \n",
       "CVN_04                                NaN                               1   \n",
       "CVN_05                                NaN                               1   \n",
       "\n",
       "                     ...              chrX:77378872-77380370:+|3p  \\\n",
       "sample_id            ...                                            \n",
       "CVN_01               ...                                        1   \n",
       "CVN_02               ...                                        1   \n",
       "CVN_03               ...                                      NaN   \n",
       "CVN_04               ...                                        1   \n",
       "CVN_05               ...                                        1   \n",
       "\n",
       "           chrX:77380923-77381286:+|3p  chrX:80532669-80533829:+|3p  \\\n",
       "sample_id                                                             \n",
       "CVN_01                               1                            1   \n",
       "CVN_02                               1                            1   \n",
       "CVN_03                             NaN                            1   \n",
       "CVN_04                               1                            1   \n",
       "CVN_05                               1                            1   \n",
       "\n",
       "           chrX:80533911-80552693:+|3p  chrX:99887566-99888401:-|3p  \\\n",
       "sample_id                                                             \n",
       "CVN_01                               1                            1   \n",
       "CVN_02                               1                            1   \n",
       "CVN_03                               1                          NaN   \n",
       "CVN_04                               1                            1   \n",
       "CVN_05                               1                            1   \n",
       "\n",
       "           chrX:99890250-99890554:-|3p  chrY:1455653-1456171:-|3p  \\\n",
       "sample_id                                                           \n",
       "CVN_01                               1                          1   \n",
       "CVN_02                               1                          1   \n",
       "CVN_03                             NaN                          1   \n",
       "CVN_04                               1                          1   \n",
       "CVN_05                               1                          1   \n",
       "\n",
       "           chrY:1456313-1458133:-|3p  chrY:14774638-14776570:+|3p  \\\n",
       "sample_id                                                           \n",
       "CVN_01                             1                            1   \n",
       "CVN_02                             1                          NaN   \n",
       "CVN_03                             1                            1   \n",
       "CVN_04                             1                            1   \n",
       "CVN_05                             1                            1   \n",
       "\n",
       "           chrY:2710284-2712117:+|3p  \n",
       "sample_id                             \n",
       "CVN_01                             1  \n",
       "CVN_02                           NaN  \n",
       "CVN_03                           NaN  \n",
       "CVN_04                             1  \n",
       "CVN_05                             1  \n",
       "\n",
       "[5 rows x 9313 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psi20_is_1_all = psi20_all.loc[:, psi_is_1.all()]\n",
    "psi20_is_1_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9313, 43)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>gerstberger2014_rbp_target_trna</th>\n",
       "      <th>gerstberger2014_tf</th>\n",
       "      <th>rbp</th>\n",
       "      <th>ribosomal</th>\n",
       "      <th>ribosomal_subunit</th>\n",
       "      <th>synapse</th>\n",
       "      <th>transcription_factor</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>exon_stop</th>\n",
       "      <th>exon_start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10:101157444-101162336:-|5p</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4892</td>\n",
       "      <td>101157443</td>\n",
       "      <td>101162337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:101162480-101163225:-|5p</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162480</td>\n",
       "      <td>101163225</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>745</td>\n",
       "      <td>101162479</td>\n",
       "      <td>101163226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:101163632-101165512:-|5p</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101163632</td>\n",
       "      <td>101165512</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>ENSG00000120053.9</td>\n",
       "      <td>GOT1</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1880</td>\n",
       "      <td>101163631</td>\n",
       "      <td>101165513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:101480826-101483712:-|5p</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101480826</td>\n",
       "      <td>101483712</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000014919.8</td>\n",
       "      <td>ENSG00000014919</td>\n",
       "      <td>ENSG00000014919.8</td>\n",
       "      <td>COX15</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2886</td>\n",
       "      <td>101480825</td>\n",
       "      <td>101483713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:101483881-101486724:-|5p</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101483881</td>\n",
       "      <td>101486724</td>\n",
       "      <td>-</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000014919.8</td>\n",
       "      <td>ENSG00000014919</td>\n",
       "      <td>ENSG00000014919.8</td>\n",
       "      <td>COX15</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2843</td>\n",
       "      <td>101483880</td>\n",
       "      <td>101486725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                chrom  intron_start  intron_stop strand  \\\n",
       "chr10:101157444-101162336:-|5p  chr10     101157444    101162336      -   \n",
       "chr10:101162480-101163225:-|5p  chr10     101162480    101163225      -   \n",
       "chr10:101163632-101165512:-|5p  chr10     101163632    101165512      -   \n",
       "chr10:101480826-101483712:-|5p  chr10     101480826    101483712      -   \n",
       "chr10:101483881-101486724:-|5p  chr10     101483881    101486724      -   \n",
       "\n",
       "                               intron_motif annotated         gencode_id  \\\n",
       "chr10:101157444-101162336:-|5p        GT/AG      True  ENSG00000120053.9   \n",
       "chr10:101162480-101163225:-|5p        GT/AG      True  ENSG00000120053.9   \n",
       "chr10:101163632-101165512:-|5p        GT/AG      True  ENSG00000120053.9   \n",
       "chr10:101480826-101483712:-|5p        GT/AG      True  ENSG00000014919.8   \n",
       "chr10:101483881-101486724:-|5p        GT/AG      True  ENSG00000014919.8   \n",
       "\n",
       "                                     ensembl_id            gene_id gene_name  \\\n",
       "chr10:101157444-101162336:-|5p  ENSG00000120053  ENSG00000120053.9      GOT1   \n",
       "chr10:101162480-101163225:-|5p  ENSG00000120053  ENSG00000120053.9      GOT1   \n",
       "chr10:101163632-101165512:-|5p  ENSG00000120053  ENSG00000120053.9      GOT1   \n",
       "chr10:101480826-101483712:-|5p  ENSG00000014919  ENSG00000014919.8     COX15   \n",
       "chr10:101483881-101486724:-|5p  ENSG00000014919  ENSG00000014919.8     COX15   \n",
       "\n",
       "                                  ...     gerstberger2014_rbp_target_trna  \\\n",
       "chr10:101157444-101162336:-|5p    ...                               False   \n",
       "chr10:101162480-101163225:-|5p    ...                               False   \n",
       "chr10:101163632-101165512:-|5p    ...                               False   \n",
       "chr10:101480826-101483712:-|5p    ...                               False   \n",
       "chr10:101483881-101486724:-|5p    ...                               False   \n",
       "\n",
       "                               gerstberger2014_tf    rbp ribosomal  \\\n",
       "chr10:101157444-101162336:-|5p              False  False     False   \n",
       "chr10:101162480-101163225:-|5p              False  False     False   \n",
       "chr10:101163632-101165512:-|5p              False  False     False   \n",
       "chr10:101480826-101483712:-|5p              False  False     False   \n",
       "chr10:101483881-101486724:-|5p              False  False     False   \n",
       "\n",
       "                               ribosomal_subunit synapse transcription_factor  \\\n",
       "chr10:101157444-101162336:-|5p             False   False                False   \n",
       "chr10:101162480-101163225:-|5p             False   False                False   \n",
       "chr10:101163632-101165512:-|5p             False   False                False   \n",
       "chr10:101480826-101483712:-|5p             False   False                False   \n",
       "chr10:101483881-101486724:-|5p             False   False                False   \n",
       "\n",
       "                               intron_length  exon_stop exon_start  \n",
       "chr10:101157444-101162336:-|5p          4892  101157443  101162337  \n",
       "chr10:101162480-101163225:-|5p           745  101162479  101163226  \n",
       "chr10:101163632-101165512:-|5p          1880  101163631  101165513  \n",
       "chr10:101480826-101483712:-|5p          2886  101480825  101483713  \n",
       "chr10:101483881-101486724:-|5p          2843  101483880  101486725  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj20_is_1_all = sj_metadata.loc[psi20_is_1_all.columns, :]\n",
    "print sj20_is_1_all.shape\n",
    "sj20_is_1_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100010822</td>\n",
       "      <td>100010933</td>\n",
       "      <td>exon:chr10:100010822-100010933:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100144704</td>\n",
       "      <td>100144824</td>\n",
       "      <td>exon:chr10:100144704-100144824:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100146958</td>\n",
       "      <td>100147064</td>\n",
       "      <td>exon:chr10:100146958-100147064:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100155148</td>\n",
       "      <td>100155209</td>\n",
       "      <td>exon:chr10:100155148-100155209:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100157102</td>\n",
       "      <td>100157255</td>\n",
       "      <td>exon:chr10:100157102-100157255:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start       stop                              name  score strand\n",
       "0  chr10  100010822  100010933  exon:chr10:100010822-100010933:-   1000      -\n",
       "1  chr10  100144704  100144824  exon:chr10:100144704-100144824:-   1000      -\n",
       "2  chr10  100146958  100147064  exon:chr10:100146958-100147064:-   1000      -\n",
       "3  chr10  100155148  100155209  exon:chr10:100155148-100155209:-   1000      -\n",
       "4  chr10  100157102  100157255  exon:chr10:100157102-100157255:-   1000      -"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = constitutive_regions.map(\n",
    "    lambda x: pd.Series(dict(chrom=x.chrom, start=x._start, stop=x._stop, strand=x.strand, score=1000, name=x.name)))\n",
    "constitutive_bed = pd.DataFrame.from_records(records)\n",
    "constitutive_bed = constitutive_bed.reindex(columns=['chrom', 'start', 'stop', 'name', 'score', 'strand'])\n",
    "constitutive_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "constitutive_exon_start.merge?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "constitutive_exon_stop (3501, 47)\n",
      "constitutive_exon_start (3420, 47)\n"
     ]
    }
   ],
   "source": [
    "constitutive_exon_stop = constitutive_bed.merge(sj20_is_1_all, \n",
    "                                                right_on=['chrom', 'exon_stop', 'strand'], \n",
    "                                                left_on=['chrom', 'stop', 'strand'])\n",
    "print 'constitutive_exon_stop', constitutive_exon_stop.shape\n",
    "constitutive_exon_start = constitutive_bed.merge(sj20_is_1_all, \n",
    "                                                right_on=['chrom', 'exon_start', 'strand'], \n",
    "                                                left_on=['chrom', 'start', 'strand'])\n",
    "print 'constitutive_exon_start', constitutive_exon_start.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_stop</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>...</th>\n",
       "      <th>gerstberger2014_rbp_target_trna</th>\n",
       "      <th>gerstberger2014_tf</th>\n",
       "      <th>rbp</th>\n",
       "      <th>ribosomal</th>\n",
       "      <th>ribosomal_subunit</th>\n",
       "      <th>synapse</th>\n",
       "      <th>transcription_factor</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>exon_stop</th>\n",
       "      <th>exon_start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4892</td>\n",
       "      <td>101157443</td>\n",
       "      <td>101162337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4892</td>\n",
       "      <td>101157443</td>\n",
       "      <td>101162337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101483713</td>\n",
       "      <td>101483880</td>\n",
       "      <td>exon:chr10:101483713-101483880:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101480826</td>\n",
       "      <td>101483712</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2886</td>\n",
       "      <td>101480825</td>\n",
       "      <td>101483713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101483713</td>\n",
       "      <td>101483880</td>\n",
       "      <td>exon:chr10:101483713-101483880:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101480826</td>\n",
       "      <td>101483712</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2886</td>\n",
       "      <td>101480825</td>\n",
       "      <td>101483713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101486725</td>\n",
       "      <td>101486911</td>\n",
       "      <td>exon:chr10:101486725-101486911:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101483881</td>\n",
       "      <td>101486724</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2843</td>\n",
       "      <td>101483880</td>\n",
       "      <td>101486725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start       stop                              name  score  \\\n",
       "0  chr10  101162337  101162479  exon:chr10:101162337-101162479:-   1000   \n",
       "1  chr10  101162337  101162479  exon:chr10:101162337-101162479:-   1000   \n",
       "2  chr10  101483713  101483880  exon:chr10:101483713-101483880:-   1000   \n",
       "3  chr10  101483713  101483880  exon:chr10:101483713-101483880:-   1000   \n",
       "4  chr10  101486725  101486911  exon:chr10:101486725-101486911:-   1000   \n",
       "\n",
       "  strand  intron_start  intron_stop intron_motif annotated    ...      \\\n",
       "0      -     101157444    101162336        GT/AG      True    ...       \n",
       "1      -     101157444    101162336        GT/AG      True    ...       \n",
       "2      -     101480826    101483712        GT/AG      True    ...       \n",
       "3      -     101480826    101483712        GT/AG      True    ...       \n",
       "4      -     101483881    101486724        GT/AG      True    ...       \n",
       "\n",
       "  gerstberger2014_rbp_target_trna gerstberger2014_tf    rbp ribosomal  \\\n",
       "0                           False              False  False     False   \n",
       "1                           False              False  False     False   \n",
       "2                           False              False  False     False   \n",
       "3                           False              False  False     False   \n",
       "4                           False              False  False     False   \n",
       "\n",
       "  ribosomal_subunit synapse transcription_factor intron_length  exon_stop  \\\n",
       "0             False   False                False          4892  101157443   \n",
       "1             False   False                False          4892  101157443   \n",
       "2             False   False                False          2886  101480825   \n",
       "3             False   False                False          2886  101480825   \n",
       "4             False   False                False          2843  101483880   \n",
       "\n",
       "  exon_start  \n",
       "0  101162337  \n",
       "1  101162337  \n",
       "2  101483713  \n",
       "3  101483713  \n",
       "4  101486725  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_exon_start.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5229, 93)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>stop_x</th>\n",
       "      <th>name</th>\n",
       "      <th>score_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>intron_start_x</th>\n",
       "      <th>intron_stop_x</th>\n",
       "      <th>intron_motif_x</th>\n",
       "      <th>annotated_x</th>\n",
       "      <th>...</th>\n",
       "      <th>gerstberger2014_rbp_target_trna_y</th>\n",
       "      <th>gerstberger2014_tf_y</th>\n",
       "      <th>rbp_y</th>\n",
       "      <th>ribosomal_y</th>\n",
       "      <th>ribosomal_subunit_y</th>\n",
       "      <th>synapse_y</th>\n",
       "      <th>transcription_factor_y</th>\n",
       "      <th>intron_length_y</th>\n",
       "      <th>exon_stop_y</th>\n",
       "      <th>exon_start_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>745</td>\n",
       "      <td>101162479</td>\n",
       "      <td>101163226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>745</td>\n",
       "      <td>101162479</td>\n",
       "      <td>101163226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>745</td>\n",
       "      <td>101162479</td>\n",
       "      <td>101163226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101157444</td>\n",
       "      <td>101162336</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>745</td>\n",
       "      <td>101162479</td>\n",
       "      <td>101163226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101483713</td>\n",
       "      <td>101483880</td>\n",
       "      <td>exon:chr10:101483713-101483880:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>101480826</td>\n",
       "      <td>101483712</td>\n",
       "      <td>GT/AG</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2843</td>\n",
       "      <td>101483880</td>\n",
       "      <td>101486725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_x    start_x     stop_x                              name  score_x  \\\n",
       "0   chr10  101162337  101162479  exon:chr10:101162337-101162479:-     1000   \n",
       "1   chr10  101162337  101162479  exon:chr10:101162337-101162479:-     1000   \n",
       "2   chr10  101162337  101162479  exon:chr10:101162337-101162479:-     1000   \n",
       "3   chr10  101162337  101162479  exon:chr10:101162337-101162479:-     1000   \n",
       "4   chr10  101483713  101483880  exon:chr10:101483713-101483880:-     1000   \n",
       "\n",
       "  strand_x  intron_start_x  intron_stop_x intron_motif_x annotated_x  \\\n",
       "0        -       101157444      101162336          GT/AG        True   \n",
       "1        -       101157444      101162336          GT/AG        True   \n",
       "2        -       101157444      101162336          GT/AG        True   \n",
       "3        -       101157444      101162336          GT/AG        True   \n",
       "4        -       101480826      101483712          GT/AG        True   \n",
       "\n",
       "      ...      gerstberger2014_rbp_target_trna_y gerstberger2014_tf_y  rbp_y  \\\n",
       "0     ...                                  False                False  False   \n",
       "1     ...                                  False                False  False   \n",
       "2     ...                                  False                False  False   \n",
       "3     ...                                  False                False  False   \n",
       "4     ...                                  False                False  False   \n",
       "\n",
       "  ribosomal_y ribosomal_subunit_y synapse_y transcription_factor_y  \\\n",
       "0       False               False     False                  False   \n",
       "1       False               False     False                  False   \n",
       "2       False               False     False                  False   \n",
       "3       False               False     False                  False   \n",
       "4       False               False     False                  False   \n",
       "\n",
       "  intron_length_y exon_stop_y exon_start_y  \n",
       "0             745   101162479    101163226  \n",
       "1             745   101162479    101163226  \n",
       "2             745   101162479    101163226  \n",
       "3             745   101162479    101163226  \n",
       "4            2843   101483880    101486725  \n",
       "\n",
       "[5 rows x 93 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "constitutive_exons = constitutive_exon_start.merge(constitutive_exon_stop, left_on='name', right_on='name')\n",
    "print constitutive_exons.shape\n",
    "constitutive_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1449, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101162337</td>\n",
       "      <td>101162479</td>\n",
       "      <td>exon:chr10:101162337-101162479:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101483713</td>\n",
       "      <td>101483880</td>\n",
       "      <td>exon:chr10:101483713-101483880:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101489310</td>\n",
       "      <td>101489491</td>\n",
       "      <td>exon:chr10:101489310-101489491:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr10</td>\n",
       "      <td>104183414</td>\n",
       "      <td>104183474</td>\n",
       "      <td>exon:chr10:104183414-104183474:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr10</td>\n",
       "      <td>104245366</td>\n",
       "      <td>104245490</td>\n",
       "      <td>exon:chr10:104245366-104245490:-</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start       stop                              name  score  \\\n",
       "0   chr10  101162337  101162479  exon:chr10:101162337-101162479:-   1000   \n",
       "4   chr10  101483713  101483880  exon:chr10:101483713-101483880:-   1000   \n",
       "8   chr10  101489310  101489491  exon:chr10:101489310-101489491:-   1000   \n",
       "12  chr10  104183414  104183474  exon:chr10:104183414-104183474:-   1000   \n",
       "16  chr10  104245366  104245490  exon:chr10:104245366-104245490:-   1000   \n",
       "\n",
       "   strand  \n",
       "0       -  \n",
       "4       -  \n",
       "8       -  \n",
       "12      -  \n",
       "16      -  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed = constitutive_exons[['chrom_x', 'start_x', 'stop_x', 'name', 'score_x', 'strand_x']]\n",
    "bed = bed.drop_duplicates()\n",
    "print bed.shape\n",
    "bed.columns = bed.columns.map(lambda x: x.split('_')[0])\n",
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bed.to_csv('{}/constitutive_exons.bed'.format(folder), index=False, header=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit compute job to calculate exon conservation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 2 tasks as an array-job.\n",
      "job ID: 3614584\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bigWigAverageOverBed /projects/ps-yeolab/genomes/hg19/hg19_phastcons_placental_mammal.bw /projects/ps-yeolab/obotvinnik/singlecell_pnms/exon2.bed /projects/ps-yeolab/obotvinnik/singlecell_pnms/exon2_phastcons_placental_mammal.txt -bedOut=/projects/ps-yeolab/obotvinnik/singlecell_pnms/exon2_phastcons_placental_mammal.bed\n",
      "bigWigAverageOverBed /projects/ps-yeolab/genomes/hg19/hg19_phastcons_placental_mammal.bw /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons.bed /projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons_phastcons_placental_mammal.txt -bedOut=/projects/ps-yeolab/obotvinnik/singlecell_pnms/constitutive_exons_phastcons_placental_mammal.bed\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gscripts.qtools._Submitter.Submitter at 0x2ba2a9ceaed0>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "\n",
    "from gscripts.qtools import Submitter\n",
    "\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "alt_exons_bedfile = '{}/exon2.bed'.format(folder)\n",
    "constitutive_bedfile = '{}/constitutive_exons.bed'.format(folder)\n",
    "\n",
    "bedfiles = alt_exons_bedfile, constitutive_bedfile\n",
    "\n",
    "commands = []\n",
    "\n",
    "bw = '/projects/ps-yeolab/genomes/hg19/hg19_phastcons_placental_mammal.bw'\n",
    "\n",
    "for bedfile in bedfiles:\n",
    "    basename = os.path.basename(bedfile)\n",
    "    \n",
    "    prefix = basename.split('.bed')[0]\n",
    "    \n",
    "    prefix += '_phastcons_placental_mammal'\n",
    "    bedout = '{}/{}'.format(folder, prefix + '.bed')\n",
    "    outtab = '{}/{}'.format(folder, prefix + '.txt')\n",
    "    command = 'bigWigAverageOverBed {} {} {} -bedOut={}'.format(bw, bedfile, outtab, bedout)\n",
    "    print command\n",
    "    commands.append(command)\n",
    "\n",
    "jobname = 'exonbody_conservation'\n",
    "Submitter(commands, jobname, array=True, walltime='2:00:00', write_and_submit=True,\n",
    "          err_filename='{}/{}.err'.format(folder, jobname),\n",
    "          out_filename='{}/{}.out'.format(folder, jobname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
