{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:784: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n",
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.py:1170: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = self._reader.read(nrows)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2015-12-02 07:51:26\tReading datapackage from /home/obotvinnik/flotilla_projects/singlecell_pnm_figure2_modalities/datapackage.json\n",
      "2015-12-02 07:51:26\tParsing datapackage to create a Study object\n",
      "2015-12-02 07:52:14\tInitializing Study\n",
      "2015-12-02 07:52:14\tInitializing Predictor configuration manager for Study\n",
      "2015-12-02 07:52:14\tPredictor ExtraTreesClassifier is of type <class 'sklearn.ensemble.forest.ExtraTreesClassifier'>\n",
      "2015-12-02 07:52:14\tAdded ExtraTreesClassifier to default predictors\n",
      "2015-12-02 07:52:14\tPredictor ExtraTreesRegressor is of type <class 'sklearn.ensemble.forest.ExtraTreesRegressor'>\n",
      "2015-12-02 07:52:14\tAdded ExtraTreesRegressor to default predictors\n",
      "2015-12-02 07:52:14\tPredictor GradientBoostingClassifier is of type <class 'sklearn.ensemble.gradient_boosting.GradientBoostingClassifier'>\n",
      "2015-12-02 07:52:14\tAdded GradientBoostingClassifier to default predictors\n",
      "2015-12-02 07:52:14\tPredictor GradientBoostingRegressor is of type <class 'sklearn.ensemble.gradient_boosting.GradientBoostingRegressor'>\n",
      "2015-12-02 07:52:14\tAdded GradientBoostingRegressor to default predictors\n",
      "2015-12-02 07:52:19\tBuilding Gene Ontology database...\n",
      "2015-12-02 07:52:29\t\tDone.\n",
      "2015-12-02 07:52:30\tLoading metadata\n",
      "2015-12-02 07:52:30\tLoading expression data\n",
      "2015-12-02 07:52:30\tInitializing expression\n",
      "2015-12-02 07:52:30\tDone initializing expression\n",
      "2015-12-02 07:52:32\tLoading splicing data\n",
      "2015-12-02 07:52:32\tInitializing splicing\n",
      "2015-12-02 07:52:32\tDone initializing splicing\n",
      "2015-12-02 07:52:32\tSuccessfully initialized a Study object!\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "\n",
    "import modish\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "sns.set(style='ticks', context='talk', rc={'font.sans-serif':'Arial', 'pdf.fonttype': 42})\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "folder = '/home/obotvinnik/Dropbox/figures2/singlecell_pnm/figure2_modalities_motifs'\n",
    "\n",
    "import flotilla\n",
    "flotilla_dir = '/projects/ps-yeolab/obotvinnik/flotilla_projects'\n",
    "\n",
    "study = flotilla.embark('singlecell_pnm_figure2_modalities', flotilla_dir=flotilla_dir)\n",
    "not_outliers = study.splicing.singles.index.difference(study.splicing.outliers.index)\n",
    "\n",
    "psi = study.splicing.singles.ix[not_outliers]\n",
    "grouped = psi.groupby(study.sample_id_to_phenotype)\n",
    "psi_filtered = grouped.apply(lambda x: x.dropna(axis=1, thresh=20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make BED files of each modality within each celltype\n",
    "\n",
    "Use the other modalities, within that celltype, as background.\n",
    "\n",
    "  - What properties are unique to events that are bimodal in iPSC, rather than ~0 or ~1?\n",
    "    - Foreground: Bimodal events in iPSC\n",
    "    - Background: All other events in iPSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# %%time \n",
    "\n",
    "# bed_folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities'\n",
    "# ! mkdir $bed_folder\n",
    "# import pybedtools\n",
    "\n",
    "# DIR = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "# exon_bedfile = '{}/exon2.bed'.format(DIR)\n",
    "# exon_bed = pybedtools.BedTool(exon_bedfile)\n",
    "\n",
    "\n",
    "\n",
    "# for phenotype, phenotype_df in study.supplemental.modalities_tidy.groupby('phenotype'):\n",
    "#     background_events = set(phenotype_df.event_name)\n",
    "#     for modality, modality_df in phenotype_df.groupby('modality'):\n",
    "#         event_names = set(modality_df.event_name)\n",
    "#         format_args = bed_folder, phenotype, modality\n",
    "#         filtered_bed = exon_bed.filter(lambda x: x.name in event_names)\n",
    "#         filtered_filename = '{}/exon2_{}_{}.bed'.format(*format_args)\n",
    "#         filtered_bed.saveas(filtered_filename)\n",
    "\n",
    "#         background_bed = exon_bed.filter(lambda x: x.name not in event_names and x.name in background_events)\n",
    "#         background_filename = '{}/exon2_{}_{}_background.bed'.format(*format_args)\n",
    "#         background_bed.saveas(background_filename)\n",
    "\n",
    "#         ! wc -l $filtered_filename\n",
    "#         ! wc -l $background_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities': File exists\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_foreground.bed (390) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background.bed (2206 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_placental_foreground.bed (599) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_placental_background.bed (3763 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_primate_foreground.bed (297) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_primate_background.bed (1650 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_foreground.bed (390) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_background.bed (2206 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_placental_foreground.bed (768) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_placental_background.bed (3914 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_primate_foreground.bed (343) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_primate_background.bed (1963 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_foreground.bed (390) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_background.bed (2206 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_placental_foreground.bed (768) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_placental_background.bed (3914 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_primate_foreground.bed (343) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400_primate_background.bed (1963 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background.bed (2595 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_placental_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_placental_background.bed (4361 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_primate_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_primate_background.bed (1946 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_background.bed (2595 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_placental_foreground.bed (9) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_placental_background.bed (4673 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_primate_foreground.bed (2) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_primate_background.bed (2304 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_background.bed (2595 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_placental_foreground.bed (9) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_placental_background.bed (4673 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_primate_foreground.bed (2) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400_primate_background.bed (2304 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_foreground.bed (566) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background.bed (2030 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_placental_foreground.bed (332) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_placental_background.bed (4030 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_primate_foreground.bed (149) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_primate_background.bed (1798 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_foreground.bed (566) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_background.bed (2030 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_placental_foreground.bed (728) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_placental_background.bed (3954 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_primate_foreground.bed (261) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_primate_background.bed (2045 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_foreground.bed (566) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_background.bed (2030 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_placental_foreground.bed (728) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_placental_background.bed (3954 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_primate_foreground.bed (261) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400_primate_background.bed (2045 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_foreground.bed (1639) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background.bed (957 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_placental_foreground.bed (3430) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_placental_background.bed (932 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_primate_foreground.bed (1500) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_primate_background.bed (447 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_foreground.bed (1639) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_background.bed (957 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_placental_foreground.bed (3177) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_placental_background.bed (1505 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_primate_foreground.bed (1700) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_primate_background.bed (606 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_foreground.bed (1639) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_background.bed (957 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_placental_foreground.bed (3177) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_placental_background.bed (1505 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_primate_foreground.bed (1700) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400_primate_background.bed (606 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_foreground.bed (801) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background.bed (4139 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_placental_foreground.bed (1450) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_placental_background.bed (7266 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_primate_foreground.bed (613) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_primate_background.bed (3064 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_foreground.bed (801) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_background.bed (4139 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_placental_foreground.bed (1653) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_placental_background.bed (6962 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_primate_foreground.bed (741) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_primate_background.bed (3532 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_foreground.bed (801) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_background.bed (4139 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_placental_foreground.bed (1653) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_placental_background.bed (6962 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_primate_foreground.bed (741) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400_primate_background.bed (3532 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_foreground.bed (1092) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background.bed (3848 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_placental_foreground.bed (535) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_placental_background.bed (8181 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_primate_foreground.bed (269) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_primate_background.bed (3408 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_foreground.bed (1092) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_background.bed (3848 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_placental_foreground.bed (1262) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_placental_background.bed (7353 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_primate_foreground.bed (456) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_primate_background.bed (3817 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_foreground.bed (1092) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_background.bed (3848 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_placental_foreground.bed (1262) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_placental_background.bed (7353 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_primate_foreground.bed (456) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400_primate_background.bed (3817 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_foreground.bed (3047) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background.bed (1893 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_placental_foreground.bed (6731) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_placental_background.bed (1985 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_primate_foreground.bed (2795) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_primate_background.bed (882 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_foreground.bed (3047) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_background.bed (1893 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_placental_foreground.bed (5700) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_placental_background.bed (2915 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_primate_foreground.bed (3076) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_primate_background.bed (1197 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_foreground.bed (3047) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_background.bed (1893 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_placental_foreground.bed (5700) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_placental_background.bed (2915 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_primate_foreground.bed (3076) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400_primate_background.bed (1197 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_foreground.bed (1408) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background.bed (6225 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_placental_foreground.bed (2697) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_placental_background.bed (11638 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_primate_foreground.bed (1018) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_primate_background.bed (4647 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_foreground.bed (1408) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_background.bed (6225 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_placental_foreground.bed (2598) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_placental_background.bed (10606 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_primate_foreground.bed (1147) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_primate_background.bed (5229 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_foreground.bed (1408) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_background.bed (6225 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_placental_foreground.bed (2598) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_placental_background.bed (10606 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_primate_foreground.bed (1147) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400_primate_background.bed (5229 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background.bed (7632 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_placental_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_placental_background.bed (14334 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_primate_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_primate_background.bed (5664 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_background.bed (7632 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_placental_foreground.bed (5) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_placental_background.bed (13199 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_primate_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_primate_background.bed (6375 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_background.bed (7632 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_placental_foreground.bed (5) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_placental_background.bed (13199 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_primate_foreground.bed (1) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400_primate_background.bed (6375 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_foreground.bed (8) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background.bed (7625 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_placental_foreground.bed (14) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_placental_background.bed (14321 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_primate_foreground.bed (5) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_primate_background.bed (5660 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_foreground.bed (8) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_background.bed (7625 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_placental_foreground.bed (17) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_placental_background.bed (13187 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_primate_foreground.bed (9) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_primate_background.bed (6367 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_foreground.bed (8) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_background.bed (7625 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_placental_foreground.bed (17) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_placental_background.bed (13187 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_primate_foreground.bed (9) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400_primate_background.bed (6367 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_foreground.bed (1605) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background.bed (6028 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_placental_foreground.bed (896) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_placental_background.bed (13439 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_primate_foreground.bed (464) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_primate_background.bed (5201 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_foreground.bed (1605) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_background.bed (6028 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_placental_foreground.bed (2096) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_placental_background.bed (11108 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_primate_foreground.bed (749) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_primate_background.bed (5627 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_foreground.bed (1605) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_background.bed (6028 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_placental_foreground.bed (2096) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_placental_background.bed (11108 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_primate_foreground.bed (749) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400_primate_background.bed (5627 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_foreground.bed (4611) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background.bed (3022 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_placental_foreground.bed (10727) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_placental_background.bed (3608 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_primate_foreground.bed (4177) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_primate_background.bed (1488 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_foreground.bed (4611) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_background.bed (3022 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_placental_foreground.bed (8488) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_placental_background.bed (4716 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_primate_foreground.bed (4470) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_primate_background.bed (1906 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_foreground.bed (4611) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_background.bed (3022 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_placental_foreground.bed (8488) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_placental_background.bed (4716 entries)\n",
      "ForegroundBackgroundPair:\n",
      "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_primate_foreground.bed (4470) entries\n",
      "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_primate_background.bed (1906 entries)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "homer_dir = '{}/homer'.format(bed_folder)\n",
    "!mkdir -p $homer_dir\n",
    "\n",
    "DIRECTIONS = 'upstream', 'downstream'\n",
    "\n",
    "def construct_homer_command(foreground_filename, background_filename, homer_flags, findMotifsGenome='findMotifsGenome.pl', \n",
    "                            out_dir_prefix=None,\n",
    "                            force=False):\n",
    "    if out_dir_prefix is None:\n",
    "        out_dir = foreground_filename.replace('.bed', '')\n",
    "    else:\n",
    "        out_dir = '{}/{}'.format(out_dir_prefix, os.path.basename(foreground_filename).replace('.bed', ''))\n",
    "    command = '{} {} hg19 {} -bg {} {}'.format(\n",
    "        findMotifsGenome, foreground_filename, out_dir, background_filename, homer_flags)\n",
    "    return command\n",
    "\n",
    "def unique_regions(bed):\n",
    "    return pybedtools.BedTool(list(set(x for x in bed)))\n",
    "\n",
    "def get_flanking_intron(bed, direction, genome, nt):\n",
    "    if direction == 'downstream':\n",
    "        intron = bed.flank(l=0, r=nt, s=True, genome=genome)\n",
    "    elif direction == 'upstream':\n",
    "        intron = bed.flank(l=nt, r=0, s=True, genome=genome)\n",
    "    else:\n",
    "        raise ValueError('Only \"downstream\" and \"upstream\" are accepted '\n",
    "                         'direction values, not \"{}\"'.format(direction))\n",
    "\n",
    "    # Saved every exon that was exactly upstream or downstream of a junction,\n",
    "    # So when taking the flanking sequence, there's a lot of repetition\n",
    "    intron = unique_regions(intron)\n",
    "    return intron\n",
    "\n",
    "\n",
    "bed_folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities'\n",
    "! mkdir $bed_folder\n",
    "import pybedtools\n",
    "\n",
    "DIR = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "exon_bedfile = '{}/exon2.bed'.format(DIR)\n",
    "exon_bed = pybedtools.BedTool(exon_bedfile)\n",
    "\n",
    "commands = []\n",
    "\n",
    "findMotifsGenome = '/home/yeo-lab/software/homer/bin/findMotifsGenome.pl'\n",
    "n_processors = 4\n",
    "homer_flags = '-rna -len 4,5,6 -mset vertebrates -mis 0 -p {} -noweight'.format(n_processors)\n",
    "\n",
    "\n",
    "primate_filename = '/projects/ps-yeolab/genomes/hg19/database/phastConsElements46wayPrimates.bed'\n",
    "primate = pybedtools.BedTool(primate_filename)\n",
    "placental_filename = '/projects/ps-yeolab/genomes/hg19/database/phastConsElements46wayPlacental.bed'\n",
    "placental = pybedtools.BedTool(placental_filename)\n",
    "conserved_regions = {'primate': primate, 'placental': placental}\n",
    "\n",
    "genome = 'hg19'\n",
    "\n",
    "nt = 400\n",
    "\n",
    "class ForegroundBackgroundPair(object):\n",
    "\n",
    "    def __init__(self, foreground, background):\n",
    "        self.foreground = foreground\n",
    "        self.background = background\n",
    "        self.beds = {'foreground': self.foreground,\n",
    "                     'background': self.background}\n",
    "\n",
    "    def __repr__(self):\n",
    "        s = 'ForegroundBackgroundPair:\\nForeground: {} ({}) entries' \\\n",
    "            '\\nBackground: {} ({} entries)'.format(\n",
    "                self.foreground.fn, len(self.foreground),\n",
    "                self.background.fn, len(self.background))\n",
    "        return s\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.__repr__()\n",
    "\n",
    "    @staticmethod\n",
    "    def _prefix(bed, x_ground):\n",
    "        return bed.fn.replace(x_ground, '').replace('.bed', '').rstrip('_')\n",
    "\n",
    "    def intersect(self, other, other_name):\n",
    "        intersections = {}\n",
    "        for x_ground, bed in self.beds.items():\n",
    "            prefix = self._prefix(bed, x_ground)\n",
    "            intersected_filename = '{prefix}_{other_name}_{x_ground}.bed'.format(\n",
    "                prefix=prefix, other_name=other_name, x_ground=x_ground)\n",
    "            bed.intersect(other).saveas(intersected_filename)\n",
    "            intersected = pybedtools.BedTool(intersected_filename)\n",
    "            intersections[x_ground] = intersected\n",
    "        return ForegroundBackgroundPair(**intersections)\n",
    "\n",
    "    def flanking_intron(self, direction, genome, nt):\n",
    "        introns = {}\n",
    "        for x_ground, bed in self.beds.items():\n",
    "            prefix = self._prefix(bed, x_ground)\n",
    "            intron_filename = '{prefix}_{direction}{nt}_{name}.bed'.format(\n",
    "                prefix=prefix, direction=direction, nt=nt, name=x_ground)\n",
    "            intron = get_flanking_intron(bed, direction, genome, nt)\n",
    "            intron.saveas(intron_filename)\n",
    "            intron = pybedtools.BedTool(intron_filename)\n",
    "\n",
    "            introns[x_ground] = intron\n",
    "        return ForegroundBackgroundPair(**introns)\n",
    "\n",
    "\n",
    "for phenotype, phenotype_df in study.supplemental.modalities_tidy.groupby('phenotype'):\n",
    "    background_events = set(phenotype_df.event_name)\n",
    "    for modality, modality_df in phenotype_df.groupby('modality'):\n",
    "        event_names = set(modality_df.event_name)\n",
    "        format_args = bed_folder, phenotype, modality\n",
    "        foreground_bed = exon_bed.filter(lambda x: x.name in event_names)\n",
    "        foreground_filename = '{}/exon2_{}_{}_foreground.bed'.format(*format_args)\n",
    "        foreground = pybedtools.BedTool(foreground_bed.saveas(foreground_filename))\n",
    "\n",
    "        background_bed = exon_bed.filter(lambda x: x.name not in event_names and x.name in background_events)\n",
    "        background_filename = '{}/exon2_{}_{}_background.bed'.format(*format_args)\n",
    "        background = pybedtools.BedTool(background_bed.saveas(background_filename))\n",
    "\n",
    "#         ! wc -l $foreground_filename\n",
    "#         ! wc -l $background_filename\n",
    "        \n",
    "        pair = ForegroundBackgroundPair(foreground, background)\n",
    "        print '\\n', pair \n",
    "        \n",
    "        command = construct_homer_command(foreground.fn, background.fn, homer_flags, findMotifsGenome, out_dir_prefix=homer_dir)\n",
    "        if command is not None:\n",
    "            commands.append(command)\n",
    "        \n",
    "        beds = {'background': background_bed, 'foreground': foreground_bed}\n",
    "        \n",
    "        for clade, conservation_bed in conserved_regions.items():\n",
    "            conserved_pair = pair.intersect(conservation_bed, clade)\n",
    "#             conserved_beds = {}\n",
    "#             conserved_filenames = {}\n",
    "#             for name, bed in beds.items():\n",
    "#                 conserved = bed.intersect(conservation_bed)\n",
    "#                 conserved_filename = '{}/exon2_{}_{}_{clade}_{name}.bed'.format(*format_args, clade=clade, name=name)\n",
    "#                 conserved_beds[name] = conserved\n",
    "#                 conserved_filenames[name] = conserved_filename\n",
    "            print '\\n', conserved_pair\n",
    "            command = construct_homer_command(conserved_pair.foreground.fn, \n",
    "                                              conserved_pair.background.fn, \n",
    "                                              homer_flags, findMotifsGenome, out_dir_prefix=homer_dir)\n",
    "            if command is not None:\n",
    "                commands.append(command)\n",
    "\n",
    "        \n",
    "        for direction in DIRECTIONS:\n",
    "            intron_pair = pair.flanking_intron('downstream', 'hg19', 400)\n",
    "#             direction_beds = {}\n",
    "#             direction_filenames = {}\n",
    "#             for name, bed in beds.items():\n",
    "#                 intron = get_flanking_intron(bed, direction, genome, nt)\n",
    "#                 intron_filename = '{}/exon2_{}_{}_{direction}{nt}_{name}.bed'.format(\n",
    "#                     *format_args, direction=direction, nt=nt, name=name)\n",
    "#                 direction_beds[name] = intron\n",
    "#                 direction_filenames[name] = intron_filename\n",
    "            print '\\n', intron_pair\n",
    "            command = construct_homer_command(intron_pair.foreground.fn, \n",
    "                                              intron_pair.background.fn, \n",
    "                                              homer_flags, findMotifsGenome, out_dir_prefix=homer_dir)\n",
    "            if command is not None:\n",
    "                commands.append(command)\n",
    "\n",
    "            for clade, conservation_bed in conserved_regions.items():\n",
    "                conserved_introns = intron_pair.intersect(conservation_bed, clade)\n",
    "#                 conserved_beds = {}\n",
    "#                 conserved_filenames = {}\n",
    "#                 for name, bed in beds.items():\n",
    "#                     conserved = bed.intersect(conservation_bed)\n",
    "#                     conserved_filename = '{}/exon2_{}_{}_{direction}{nt}_{clade}_{name}.bed'.format(\n",
    "#                         *format_args, direction=direction, nt=nt, clade=clade, name=name)\n",
    "#                     conserved_beds[name] = conserved\n",
    "#                     conserved_filenames[name] = conserved_filename\n",
    "                print '\\n', conserved_introns\n",
    "                command = construct_homer_command(conserved_introns.foreground.fn,\n",
    "                                                  conserved_introns.background.fn, \n",
    "                                                  homer_flags, findMotifsGenome, out_dir_prefix=homer_dir)\n",
    "                if command is not None:\n",
    "                    commands.append(command)\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_placental_foreground.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_bimodal_placental_foreground -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_placental_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "commands[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 108 tasks as an array-job.\n",
      "job ID: 3822542\n"
     ]
    }
   ],
   "source": [
    "from gscripts.qtools import Submitter\n",
    "\n",
    "jobname = 'homer'\n",
    "sub = Submitter(commands, jobname, walltime='0:30:00', queue='home-scrm',\n",
    "               ppn=n_processors, write_and_submit=True, array=True,\n",
    "               out_filename='{}/{}.out'.format(homer_dir, jobname),\n",
    "               err_filename='{}/{}.err'.format(homer_dir, jobname),)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "108"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(commands)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "background = pybedtools.BedTool(background_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "asdf = pybedtools.BedTool(background.saveas('asdf.bed'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'asdf.bed'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asdf.fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ForegroundBackgroundPair:\n",
       "Foreground: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1.bed\n",
       "Background: /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background.bed"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pybedtools\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "foreground = pybedtools.BedTool(foreground_filename)\n",
    "background = pybedtools.BedTool(background_filename)\n",
    "\n",
    "pair = ForegroundBackgroundPair(foreground, background)\n",
    "pair\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'primate'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.ForegroundBackground at 0x2b38bb2ed590>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conserved_pair = pair.intersect(conservation_bed, clade)\n",
    "conserved_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_primate_foreground.bed'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conserved_pair.foreground.fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_primate_background.bed'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conserved_pair.background.fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "intron_pair = pair.flanking_intron('downstream', 'hg19', 400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_background.bed'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_pair.background.fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400_foreground.bed'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_pair.foreground.fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Line magic function `%pdbasd` not found.\n"
     ]
    }
   ],
   "source": [
    "%pdbasd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Genome file \"%s\" does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-7-d4ea9a34aa29>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mget_flanking_intron\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mbed\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdirection\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgenome\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnt\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-4-8f93d24fe5c7>\u001b[0m in \u001b[0;36mget_flanking_intron\u001b[1;34m(bed, direction, genome, nt)\u001b[0m\n\u001b[0;32m     14\u001b[0m         \u001b[0mintron\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mbed\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mflank\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ml\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mnt\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ms\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mg\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mgenome\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     15\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mdirection\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'upstream'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 16\u001b[1;33m         \u001b[0mintron\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mbed\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mflank\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ml\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mnt\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ms\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mg\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mgenome\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     17\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     18\u001b[0m         raise ValueError('Only \"downstream\" and \"upstream\" are accepted '\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pybedtools/bedtool.pyc\u001b[0m in \u001b[0;36mdecorated\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    666\u001b[0m             \u001b[1;31m# this calls the actual method in the first place; *result* is\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    667\u001b[0m             \u001b[1;31m# whatever you get back\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 668\u001b[1;33m             \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    669\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    670\u001b[0m             \u001b[1;31m# add appropriate tags\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pybedtools/bedtool.pyc\u001b[0m in \u001b[0;36mwrapped\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    225\u001b[0m                             \u001b[0mcheck_for_genome\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    226\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mcheck_for_genome\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 227\u001b[1;33m                 \u001b[0mkwargs\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcheck_genome\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    228\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    229\u001b[0m             \u001b[1;31m# For sequence methods, we may need to make a tempfile that will\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pybedtools/bedtool.pyc\u001b[0m in \u001b[0;36mcheck_genome\u001b[1;34m(self, **kwargs)\u001b[0m\n\u001b[0;32m   1276\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1277\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'g'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1278\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Genome file \"%s\" does not exist'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1279\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1280\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Genome file \"%s\" does not exist"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pybedtools/bedtool.py\u001b[0m(1278)\u001b[0;36mcheck_genome\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m   1277 \u001b[1;33m        \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'g'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[1;32m-> 1278 \u001b[1;33m            \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Genome file \"%s\" does not exist'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[1;32m   1279 \u001b[1;33m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> dir()\n",
      "['kwargs', 'self']\n",
      "ipdb> self\n",
      "<generator object <genexpr> at 0x2b38d784c4b0>\n",
      "ipdb> kwargs\n",
      "{'i': <generator object <genexpr> at 0x2b38d784c4b0>, 's': True, 'r': 0, 'l': 400, 'g': 'hg19'}\n",
      "ipdb> q\n"
     ]
    }
   ],
   "source": [
    "get_flanking_intron(bed, direction, genome, nt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  - What are properties that are unique to bimodal events?\n",
    "    - Foreground: any event that has ever been bimodal\n",
    "    - Background: All events that have never been bimodal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities': File exists\n",
      "2155 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal.bed\n",
      "5970 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle.bed\n",
      "8124 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background.bed\n",
      "9 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal.bed\n",
      "8116 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background.bed\n",
      "1770 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0.bed\n",
      "6355 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background.bed\n",
      "4945 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1.bed\n",
      "3180 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background.bed\n",
      "CPU times: user 2.67 s, sys: 170 ms, total: 2.84 s\n",
      "Wall time: 4.38 s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "bed_folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities'\n",
    "! mkdir $bed_folder\n",
    "import pybedtools\n",
    "\n",
    "DIR = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "exon_bedfile = '{}/exon2.bed'.format(DIR)\n",
    "exon_bed = pybedtools.BedTool(exon_bedfile)\n",
    "\n",
    "all_events = set(study.supplemental.modalities_tidy.event_name)\n",
    "\n",
    "for modality, modality_df in study.supplemental.modalities_tidy.groupby('modality'):\n",
    "    event_names = set(modality_df.event_name)\n",
    "    background_events = all_events - event_names\n",
    "    \n",
    "    format_args = bed_folder, modality\n",
    "    filtered_bed = exon_bed.filter(lambda x: x.name in event_names)\n",
    "    filtered_filename = '{}/exon2_{}.bed'.format(*format_args)\n",
    "    filtered_bed.saveas(filtered_filename)\n",
    "\n",
    "    background_bed = exon_bed.filter(lambda x: x.name not in event_names and x.name in background_events)\n",
    "    background_filename = '{}/exon2_{}_background.bed'.format(*format_args)\n",
    "    background_bed.saveas(background_filename)\n",
    "\n",
    "    ! wc -l $filtered_filename\n",
    "    ! wc -l $background_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/home/obotvinnik/projects/singlecell_pnms/analysis/homer/': File exists\r\n"
     ]
    }
   ],
   "source": [
    "mkdir /home/obotvinnik/projects/singlecell_pnms/analysis/homer/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HOMER enrichment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "homer_dir = '{}/homer'.format(bed_folder)\n",
    "!mkdir -p $homer_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### First extract upstream/downstream intron sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon2_~0.bed\n",
      "\t1770 1770\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_upstream400.bed\n",
      "1770 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_upstream400.bed\n",
      "\t1770 1770\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_downstream400.bed\n",
      "1770 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_downstream400.bed\n",
      "exon2_middle_background.bed\n",
      "\t8124 8124\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_upstream400.bed\n",
      "8124 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_upstream400.bed\n",
      "\t8124 8124\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_downstream400.bed\n",
      "8124 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_downstream400.bed\n",
      "exon2_MN_bimodal_background.bed\n",
      "\t2206 2206\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_upstream400.bed\n",
      "2206 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_upstream400.bed\n",
      "\t2206 2206\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_downstream400.bed\n",
      "2206 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_downstream400.bed\n",
      "exon2_iPSC_~1_background.bed\n",
      "\t3022 3022\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_upstream400.bed\n",
      "3022 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_upstream400.bed\n",
      "\t3022 3022\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_downstream400.bed\n",
      "3022 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_downstream400.bed\n",
      "exon2_NPC_~1_background.bed\n",
      "\t1893 1893\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_upstream400.bed\n",
      "1893 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_upstream400.bed\n",
      "\t1893 1893\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_downstream400.bed\n",
      "1893 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_downstream400.bed\n",
      "exon2_middle.bed\n",
      "\t1 1\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_upstream400.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_upstream400.bed\n",
      "\t1 1\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_downstream400.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_downstream400.bed\n",
      "exon2_NPC_~0.bed\n",
      "\t1092 1092\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_upstream400.bed\n",
      "1092 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_upstream400.bed\n",
      "\t1092 1092\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400.bed\n",
      "1092 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400.bed\n",
      "exon2_MN_~0.bed\n",
      "\t566 566\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_upstream400.bed\n",
      "566 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_upstream400.bed\n",
      "\t566 566\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400.bed\n",
      "566 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400.bed\n",
      "exon2_MN_~0_background.bed\n",
      "\t2030 2030\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_upstream400.bed\n",
      "2030 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_upstream400.bed\n",
      "\t2030 2030\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_downstream400.bed\n",
      "2030 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_downstream400.bed\n",
      "exon2_iPSC_bimodal_background.bed\n",
      "\t6225 6225\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_upstream400.bed\n",
      "6225 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_upstream400.bed\n",
      "\t6225 6225\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_downstream400.bed\n",
      "6225 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_downstream400.bed\n",
      "exon2_~0_background.bed\n",
      "\t6355 6355\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_upstream400.bed\n",
      "6355 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_upstream400.bed\n",
      "\t6355 6355\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_downstream400.bed\n",
      "6355 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_downstream400.bed\n",
      "exon2_~1.bed\n",
      "\t4945 4945\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_upstream400.bed\n",
      "4945 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_upstream400.bed\n",
      "\t4945 4945\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_downstream400.bed\n",
      "4945 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_downstream400.bed\n",
      "exon2_NPC_bimodal_background.bed\n",
      "\t4139 4139\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_upstream400.bed\n",
      "4139 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_upstream400.bed\n",
      "\t4139 4139\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_downstream400.bed\n",
      "4139 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_downstream400.bed\n",
      "exon2_multimodal_background.bed\n",
      "\t8116 8116\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_upstream400.bed\n",
      "8116 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_upstream400.bed\n",
      "\t8116 8116\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_downstream400.bed\n",
      "8116 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_downstream400.bed\n",
      "exon2_MN_multimodal.bed\n",
      "\t1 1\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_upstream400.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_upstream400.bed\n",
      "\t1 1\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400.bed\n",
      "exon2_MN_~1.bed\n",
      "\t1639 1639\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_upstream400.bed\n",
      "1639 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_upstream400.bed\n",
      "\t1639 1639\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400.bed\n",
      "1639 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400.bed\n",
      "exon2_NPC_~1.bed\n",
      "\t3047 3047\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_upstream400.bed\n",
      "3047 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_upstream400.bed\n",
      "\t3047 3047\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400.bed\n",
      "3047 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400.bed\n",
      "exon2_multimodal.bed\n",
      "\t9 9\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_upstream400.bed\n",
      "9 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_upstream400.bed\n",
      "\t9 9\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_downstream400.bed\n",
      "9 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_downstream400.bed\n",
      "exon2_iPSC_middle_background.bed\n",
      "\t7632 7632\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_upstream400.bed\n",
      "7632 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_upstream400.bed\n",
      "\t7632 7632\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_downstream400.bed\n",
      "7632 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_downstream400.bed\n",
      "exon2_iPSC_~0.bed\n",
      "\t1605 1605\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_upstream400.bed\n",
      "1605 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_upstream400.bed\n",
      "\t1605 1605\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400.bed\n",
      "1605 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400.bed\n",
      "exon2_~1_background.bed\n",
      "\t3180 3180\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_upstream400.bed\n",
      "3180 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_upstream400.bed\n",
      "\t3180 3180\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_downstream400.bed\n",
      "3180 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_downstream400.bed\n",
      "exon2_iPSC_multimodal.bed\n",
      "\t8 8\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_upstream400.bed\n",
      "8 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_upstream400.bed\n",
      "\t8 8\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400.bed\n",
      "8 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400.bed\n",
      "exon2_MN_multimodal_background.bed\n",
      "\t2595 2595\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_upstream400.bed\n",
      "2595 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_upstream400.bed\n",
      "\t2595 2595\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_downstream400.bed\n",
      "2595 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_downstream400.bed\n",
      "exon2_MN_bimodal.bed\n",
      "\t390 390\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_upstream400.bed\n",
      "390 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_upstream400.bed\n",
      "\t390 390\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400.bed\n",
      "390 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400.bed\n",
      "exon2_bimodal.bed\n",
      "\t2155 2155\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_upstream400.bed\n",
      "2155 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_upstream400.bed\n",
      "\t2155 2155\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_downstream400.bed\n",
      "2155 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_downstream400.bed\n",
      "exon2_NPC_~0_background.bed\n",
      "\t3848 3848\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_upstream400.bed\n",
      "3848 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_upstream400.bed\n",
      "\t3848 3848\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_downstream400.bed\n",
      "3848 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_downstream400.bed\n",
      "exon2_bimodal_background.bed\n",
      "\t5970 5970\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_upstream400.bed\n",
      "5970 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_upstream400.bed\n",
      "\t5970 5970\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_downstream400.bed\n",
      "5970 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_downstream400.bed\n",
      "exon2_iPSC_middle.bed\n",
      "\t1 1\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_upstream400.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_upstream400.bed\n",
      "\t1 1\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400.bed\n",
      "1 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400.bed\n",
      "exon2_MN_~1_background.bed\n",
      "\t957 957\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_upstream400.bed\n",
      "957 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_upstream400.bed\n",
      "\t957 957\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_downstream400.bed\n",
      "957 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_downstream400.bed\n",
      "exon2_iPSC_~1.bed\n",
      "\t4611 4611\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_upstream400.bed\n",
      "4611 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_upstream400.bed\n",
      "\t4611 4611\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400.bed\n",
      "4611 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400.bed\n",
      "exon2_iPSC_bimodal.bed\n",
      "\t1408 1408\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_upstream400.bed\n",
      "1408 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_upstream400.bed\n",
      "\t1408 1408\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400.bed\n",
      "1408 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400.bed\n",
      "exon2_NPC_bimodal.bed\n",
      "\t801 801\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_upstream400.bed\n",
      "801 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_upstream400.bed\n",
      "\t801 801\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400.bed\n",
      "801 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400.bed\n",
      "exon2_iPSC_~0_background.bed\n",
      "\t6028 6028\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_upstream400.bed\n",
      "6028 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_upstream400.bed\n",
      "\t6028 6028\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_downstream400.bed\n",
      "6028 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_downstream400.bed\n",
      "exon2_iPSC_multimodal_background.bed\n",
      "\t7625 7625\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_upstream400.bed\n",
      "7625 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_upstream400.bed\n",
      "\t7625 7625\n",
      "\t/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_downstream400.bed\n",
      "7625 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_downstream400.bed\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import pybedtools\n",
    "\n",
    "nt = 400\n",
    "\n",
    "directions = 'upstream', 'downstream'\n",
    "\n",
    "for bedfile in glob.iglob('{}/exon2*.bed'.format(bed_folder)):\n",
    "    if 'stream' in bedfile:\n",
    "        continue\n",
    "    \n",
    "    basename = os.path.basename(bedfile)\n",
    "    print basename\n",
    "    prefix = basename.split('.bed')[0]\n",
    "    bed = pybedtools.BedTool(bedfile)\n",
    "    \n",
    "    for direction in directions:\n",
    "        # Get downstream intron\n",
    "        if direction == 'downstream':\n",
    "            intron = bed.flank(l=0, r=nt, s=True, g=pybedtools.chromsizes('hg19'))\n",
    "        elif direction == 'upstream':\n",
    "            intron = bed.flank(l=nt, r=0, s=True, g=pybedtools.chromsizes('hg19'))\n",
    "            \n",
    "        # Saved every exon that was exactly upstream or downstream of a junction,\n",
    "        # So when taking the flanking sequence, there's a lot of repetition\n",
    "        unique = pybedtools.BedTool(list(set(x for x in intron)))\n",
    "\n",
    "\n",
    "        print '\\t', len(intron), len(unique)\n",
    "        saveas = '{}/{}_{}{}.bed'.format(bed_folder, prefix, direction, nt)\n",
    "        print '\\t', saveas\n",
    "        unique.saveas(saveas)\n",
    "        ! wc -l $saveas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment of within-phenotype modality\n",
    "\n",
    "#### Background: All other modalities, within that phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from modish import MODALITY_ORDER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 85 tasks as an array-job.\n",
      "job ID: 3798111\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~0 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~0_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~0_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~0_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_middle_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~0_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~1_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~1_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~1_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~1_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_multimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_middle -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~0 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_multimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_middle_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~0 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_bimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_multimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_middle_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_bimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_middle_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~1_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~1_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_bimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_multimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_multimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~0_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~0_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~0_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_bimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~1 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_multimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~0_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~1_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_bimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_bimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_multimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~1 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_bimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~1 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_bimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~1_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_multimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_multimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_multimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~1_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_multimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~0_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_bimodal_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~0_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~0_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~0_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_bimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~0 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~0_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_bimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_multimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_bimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~1_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_bimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_bimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_bimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_bimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~0_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~0_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_middle -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~1_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~1_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~1_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~0_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_upstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_middle_upstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_upstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~0_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~0_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~0_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~0_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_middle_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_multimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_multimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_~1_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_~1_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_multimodal_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_multimodal_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_~1 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_~1_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_bimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_~1_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_~1_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_bimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_bimodal -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_bimodal_background.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_downstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_middle_downstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_middle_background_downstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_NPC_~1_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_NPC_~1_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_MN_bimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_MN_bimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_middle_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_middle_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_upstream200.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_multimodal_upstream200 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_multimodal_background_upstream200.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n",
      "\n",
      "/home/yeo-lab/software/homer/bin/findMotifsGenome.pl /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_downstream400.bed hg19 /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/homer/exon2_iPSC_bimodal_downstream400 -bg /projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/exon2_iPSC_bimodal_background_downstream400.bed -rna -len 4,5,6 -mset vertebrates -mis 0 -p 4 -noweight\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import fileinput\n",
    "\n",
    "from gscripts.qtools import Submitter\n",
    "import glob\n",
    "import itertools\n",
    "\n",
    "findMotifsGenome = '/home/yeo-lab/software/homer/bin/findMotifsGenome.pl'\n",
    "n_processors = 4\n",
    "homer_flags = '-rna -len 4,5,6 -mset vertebrates -mis 0 -p {} -noweight'.format(n_processors)\n",
    "\n",
    "commands = []\n",
    "\n",
    "\n",
    "for bedfile in glob.iglob('{}/exon2*.bed'.format(bed_folder)):\n",
    "    \n",
    "    basename = os.path.basename(bedfile)\n",
    "    if 'background' in basename:\n",
    "        continue\n",
    "    \n",
    "    prefix = basename.split('.')[0]\n",
    "    modality = prefix.split('_')[1]\n",
    "    if modality not in MODALITY_ORDER:\n",
    "        modality = prefix.split('_')[2]\n",
    "#     print modality\n",
    "    \n",
    "    background = bedfile.replace(modality, modality + '_background')\n",
    "    \n",
    "    out_dir = '{}/{}'.format(homer_dir, basename.replace('.bed', ''))\n",
    "#     if os.path.exists('{}/homerResults.html'):\n",
    "#         print 'Already finished {}, moving on'.format(basename)\n",
    "#         continue\n",
    "    command = '{} {} hg19 {} -bg {} {}'.format(\n",
    "        findMotifsGenome, bedfile, out_dir, background, homer_flags)\n",
    "    print '\\n', command\n",
    "    commands.append(command)\n",
    "#     ! $command\n",
    "jobname = 'homer'\n",
    "sub = Submitter(commands, jobname, walltime='0:30:00', queue='home-scrm',\n",
    "               ppn=n_processors, write_and_submit=True, array=True,\n",
    "               out_filename='{}/{}.out'.format(homer_dir, jobname),\n",
    "               err_filename='{}/{}.err'.format(homer_dir, jobname),)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
