{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py:810: UserWarning: Found matplotlib configuration in ~/.matplotlib/. To conform with the XDG base directory standard, this configuration location has been deprecated on Linux, and the new location is now /home/obotvinnik/.config/matplotlib/. Please move your configuration there to ensure that matplotlib will continue to find it in the future.\n",
      "  _get_xdg_config_dir())\n",
      "/home/obotvinnik/workspace-git/flotilla/flotilla/__init__.py:55: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  load_species_data=load_species_data)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-02-10 09:53:22\tReading datapackage from /projects/ps-yeolab/obotvinnik/flotilla_projects/singlecell_pnm_figure2_modalities_bayesian/datapackage.json\n",
      "2016-02-10 09:53:22\tParsing datapackage to create a Study object\n",
      "2016-02-10 09:54:06\tInitializing Study\n",
      "2016-02-10 09:54:06\tInitializing Predictor configuration manager for Study\n",
      "2016-02-10 09:54:06\tPredictor ExtraTreesClassifier is of type <class 'sklearn.ensemble.forest.ExtraTreesClassifier'>\n",
      "2016-02-10 09:54:06\tAdded ExtraTreesClassifier to default predictors\n",
      "2016-02-10 09:54:06\tPredictor ExtraTreesRegressor is of type <class 'sklearn.ensemble.forest.ExtraTreesRegressor'>\n",
      "2016-02-10 09:54:06\tAdded ExtraTreesRegressor to default predictors\n",
      "2016-02-10 09:54:06\tPredictor GradientBoostingClassifier is of type <class 'sklearn.ensemble.gradient_boosting.GradientBoostingClassifier'>\n",
      "2016-02-10 09:54:06\tAdded GradientBoostingClassifier to default predictors\n",
      "2016-02-10 09:54:06\tPredictor GradientBoostingRegressor is of type <class 'sklearn.ensemble.gradient_boosting.GradientBoostingRegressor'>\n",
      "2016-02-10 09:54:06\tAdded GradientBoostingRegressor to default predictors\n",
      "2016-02-10 09:54:11\tBuilding Gene Ontology database...\n",
      "2016-02-10 09:54:20\t\tDone.\n",
      "2016-02-10 09:54:21\tLoading metadata\n",
      "2016-02-10 09:54:21\tLoading expression data\n",
      "2016-02-10 09:54:21\tInitializing expression\n",
      "2016-02-10 09:54:21\tDone initializing expression\n",
      "2016-02-10 09:54:23\tLoading splicing data\n",
      "2016-02-10 09:54:23\tInitializing splicing\n",
      "2016-02-10 09:54:23\tDone initializing splicing\n",
      "2016-02-10 09:54:23\tSuccessfully initialized a Study object!\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "sns.set(style='ticks', context='talk', rc={'font.sans-serif':'Arial', 'pdf.fonttype': 42})\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "import flotilla\n",
    "flotilla_dir = '/projects/ps-yeolab/obotvinnik/flotilla_projects'\n",
    "\n",
    "study = flotilla.embark('singlecell_pnm_figure2_modalities_bayesian', flotilla_dir=flotilla_dir)\n",
    "not_outliers = study.splicing.singles.index.difference(study.splicing.outliers.index)\n",
    "\n",
    "psi = study.splicing.singles.ix[not_outliers]\n",
    "grouped = psi.groupby(study.sample_id_to_phenotype)\n",
    "psi_filtered = grouped.apply(lambda x: x.dropna(axis=1, thresh=20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/home/obotvinnik/Dropbox/figures2/singlecell_pnm/figure2_modalities/bayesian': File exists\r\n"
     ]
    }
   ],
   "source": [
    "folder = '/home/obotvinnik/Dropbox/figures2/singlecell_pnm/figure2_modalities/bayesian'\n",
    "!mkdir $folder\n",
    "\n",
    "figure_folder = '{}/kmer_counting'.format(folder)\n",
    "! mkdir -p $figure_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phenotype  modality  \n",
       "MN         ambivalent       1\n",
       "           bimodal        390\n",
       "           excluded       566\n",
       "           included      1639\n",
       "NPC        ambivalent       2\n",
       "           bimodal        799\n",
       "           excluded      1092\n",
       "           included      3047\n",
       "iPSC       ambivalent      33\n",
       "           bimodal       1383\n",
       "           concurrent       1\n",
       "           excluded      1605\n",
       "           included      4611\n",
       "dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study.supplemental.modalities_tidy.groupby(['phenotype', 'modality']).size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make BED files of each modality within each celltype\n",
    "\n",
    "Use the other modalities, within that celltype, as background.\n",
    "\n",
    "  - What properties are unique to events that are bimodal in iPSC, rather than ~0 or ~1?\n",
    "    - Foreground: Bimodal events in iPSC\n",
    "    - Background: All other events in iPSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory `/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/bayesian': File exists\r\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "bed_folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms/figure2_modalities/bayesian'\n",
    "! mkdir $bed_folder\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pyhomer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_events = study.supplemental.modalities_tidy.event_id.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>102116289</td>\n",
       "      <td>102116521</td>\n",
       "      <td>exon:chr10:102114184-102114389:+@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>102286156</td>\n",
       "      <td>102286311</td>\n",
       "      <td>exon:chr10:102286732-102286831:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>103354415</td>\n",
       "      <td>103354495</td>\n",
       "      <td>exon:chr10:103348089-103348157:+@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>103364897</td>\n",
       "      <td>103364969</td>\n",
       "      <td>exon:chr10:103360960-103361093:+@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>103368592</td>\n",
       "      <td>103368694</td>\n",
       "      <td>exon:chr10:103360960-103361093:+@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0          1          2  \\\n",
       "0  chr10  102116289  102116521   \n",
       "1  chr10  102286156  102286311   \n",
       "2  chr10  103354415  103354495   \n",
       "3  chr10  103364897  103364969   \n",
       "4  chr10  103368592  103368694   \n",
       "\n",
       "                                                   3     4  5  \n",
       "0  exon:chr10:102114184-102114389:+@exon:chr10:10...  1000  +  \n",
       "1  exon:chr10:102286732-102286831:-@exon:chr10:10...  1000  -  \n",
       "2  exon:chr10:103348089-103348157:+@exon:chr10:10...  1000  +  \n",
       "3  exon:chr10:103360960-103361093:+@exon:chr10:10...  1000  +  \n",
       "4  exon:chr10:103360960-103361093:+@exon:chr10:10...  1000  +  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from outrigger.region import Region\n",
    "\n",
    "exon2s = map(lambda x: x.split('@')[1], all_events)\n",
    "exon2_regions = map(Region, exon2s)\n",
    "exon2_bed_table = pd.DataFrame(map(lambda x: [x.chrom, x._start, x._stop, x.name, 1000, x.strand], exon2_regions))\n",
    "exon2_bed_table[3] = all_events\n",
    "exon2_bed_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5163</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101163481</td>\n",
       "      <td>101163631</td>\n",
       "      <td>exon:chr10:101165513-101165617:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5164</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101478103</td>\n",
       "      <td>101478257</td>\n",
       "      <td>exon:chr10:101480744-101480825:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5165</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101510126</td>\n",
       "      <td>101510153</td>\n",
       "      <td>exon:chr10:101507014-101507147:+@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5166</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101996607</td>\n",
       "      <td>101996726</td>\n",
       "      <td>exon:chr10:101997779-101997988:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5167</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101997779</td>\n",
       "      <td>101997988</td>\n",
       "      <td>exon:chr10:102003455-102003529:-@exon:chr10:10...</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2  \\\n",
       "5163  chr10  101163481  101163631   \n",
       "5164  chr10  101478103  101478257   \n",
       "5165  chr10  101510126  101510153   \n",
       "5166  chr10  101996607  101996726   \n",
       "5167  chr10  101997779  101997988   \n",
       "\n",
       "                                                      3     4  5  \n",
       "5163  exon:chr10:101165513-101165617:-@exon:chr10:10...  1000  -  \n",
       "5164  exon:chr10:101480744-101480825:-@exon:chr10:10...  1000  -  \n",
       "5165  exon:chr10:101507014-101507147:+@exon:chr10:10...  1000  +  \n",
       "5166  exon:chr10:101997779-101997988:-@exon:chr10:10...  1000  -  \n",
       "5167  exon:chr10:102003455-102003529:-@exon:chr10:10...  1000  -  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon2_bed_table = exon2_bed_table.sort_values(by=3)\n",
    "exon2_bed_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exon2_bed_table.to_csv('{}/background_events.bed'.format(bed_folder), index=False, header=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---\n",
      "MN ambivalent\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 32.2 ms\n",
      "CPU times: user 399 ms, sys: 7 ms, total: 406 ms\n",
      "Wall time: 801 ms\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 3.41 s\n",
      "CPU times: user 3min 6s, sys: 116 ms, total: 3min 6s\n",
      "Wall time: 3min 6s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 1e+03 µs, sys: 8 ms, total: 9 ms\n",
      "Wall time: 23.5 ms\n",
      "CPU times: user 439 ms, sys: 1e+03 µs, total: 440 ms\n",
      "Wall time: 440 ms\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 181 ms\n",
      "CPU times: user 3min 3s, sys: 117 ms, total: 3min 3s\n",
      "Wall time: 3min 4s\n",
      "---\n",
      "MN bimodal\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 33.9 ms\n",
      "CPU times: user 42.9 s, sys: 26 ms, total: 43 s\n",
      "Wall time: 43 s\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 85.5 ms\n",
      "CPU times: user 3min 7s, sys: 578 ms, total: 3min 7s\n",
      "Wall time: 3min 8s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 33.7 ms\n",
      "CPU times: user 37.4 s, sys: 17 ms, total: 37.4 s\n",
      "Wall time: 37.4 s\n",
      "CPU times: user 0 ns, sys: 10 ms, total: 10 ms\n",
      "Wall time: 85.3 ms\n",
      "CPU times: user 3min 4s, sys: 208 ms, total: 3min 4s\n",
      "Wall time: 3min 4s\n",
      "---\n",
      "MN excluded\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 35.5 ms\n",
      "CPU times: user 43.8 s, sys: 183 ms, total: 44 s\n",
      "Wall time: 44 s\n",
      "CPU times: user 1e+03 µs, sys: 9 ms, total: 10 ms\n",
      "Wall time: 95.7 ms\n",
      "CPU times: user 3min 6s, sys: 455 ms, total: 3min 7s\n",
      "Wall time: 3min 7s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 33.3 ms\n",
      "CPU times: user 40 s, sys: 16 ms, total: 40 s\n",
      "Wall time: 40 s\n",
      "CPU times: user 1e+03 µs, sys: 9 ms, total: 10 ms\n",
      "Wall time: 89 ms\n",
      "CPU times: user 3min 4s, sys: 800 ms, total: 3min 5s\n",
      "Wall time: 3min 6s\n",
      "---\n",
      "MN included\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 0 ns, sys: 9 ms, total: 9 ms\n",
      "Wall time: 67.8 ms\n",
      "CPU times: user 1min 40s, sys: 627 ms, total: 1min 40s\n",
      "Wall time: 1min 41s\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 110 ms\n",
      "CPU times: user 3min 7s, sys: 883 ms, total: 3min 8s\n",
      "Wall time: 3min 8s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 66.7 ms\n",
      "CPU times: user 1min 47s, sys: 477 ms, total: 1min 48s\n",
      "Wall time: 1min 48s\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 4.89 s\n",
      "CPU times: user 3min 5s, sys: 389 ms, total: 3min 5s\n",
      "Wall time: 3min 6s\n",
      "---\n",
      "NPC ambivalent\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 23.3 ms\n",
      "CPU times: user 73 ms, sys: 1 ms, total: 74 ms\n",
      "Wall time: 73.5 ms\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 1.45 s\n",
      "CPU times: user 5min 27s, sys: 1.79 s, total: 5min 29s\n",
      "Wall time: 5min 30s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 3 ms, sys: 7 ms, total: 10 ms\n",
      "Wall time: 21.9 ms\n",
      "CPU times: user 48 ms, sys: 1e+03 µs, total: 49 ms\n",
      "Wall time: 49.3 ms\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 215 ms\n",
      "CPU times: user 5min 24s, sys: 1.41 s, total: 5min 26s\n",
      "Wall time: 5min 26s\n",
      "---\n",
      "NPC bimodal\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 46.4 ms\n",
      "CPU times: user 1min 13s, sys: 597 ms, total: 1min 14s\n",
      "Wall time: 1min 14s\n",
      "CPU times: user 1 ms, sys: 8 ms, total: 9 ms\n",
      "Wall time: 168 ms\n",
      "CPU times: user 5min 26s, sys: 3.01 s, total: 5min 29s\n",
      "Wall time: 5min 30s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 46.9 ms\n",
      "CPU times: user 1min 17s, sys: 656 ms, total: 1min 17s\n",
      "Wall time: 1min 17s\n",
      "CPU times: user 3 ms, sys: 7 ms, total: 10 ms\n",
      "Wall time: 162 ms\n",
      "CPU times: user 5min 26s, sys: 1.19 s, total: 5min 28s\n",
      "Wall time: 5min 28s\n",
      "---\n",
      "NPC excluded\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 3 ms, sys: 7 ms, total: 10 ms\n",
      "Wall time: 42.8 ms\n",
      "CPU times: user 1min 16s, sys: 578 ms, total: 1min 16s\n",
      "Wall time: 1min 16s\n",
      "CPU times: user 1e+03 µs, sys: 9 ms, total: 10 ms\n",
      "Wall time: 166 ms\n",
      "CPU times: user 5min 26s, sys: 1.18 s, total: 5min 28s\n",
      "Wall time: 5min 28s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 46.7 ms\n",
      "CPU times: user 1min 12s, sys: 64 ms, total: 1min 12s\n",
      "Wall time: 1min 12s\n",
      "CPU times: user 1 ms, sys: 9 ms, total: 10 ms\n",
      "Wall time: 151 ms\n",
      "CPU times: user 5min 24s, sys: 2.71 s, total: 5min 27s\n",
      "Wall time: 5min 28s\n",
      "---\n",
      "NPC included\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 3 ms, sys: 7 ms, total: 10 ms\n",
      "Wall time: 114 ms\n",
      "CPU times: user 2min 58s, sys: 1.87 s, total: 3min\n",
      "Wall time: 3min 1s\n",
      "CPU times: user 4 ms, sys: 7 ms, total: 11 ms\n",
      "Wall time: 173 ms\n",
      "CPU times: user 5min 26s, sys: 2.1 s, total: 5min 28s\n",
      "Wall time: 5min 29s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 107 ms\n",
      "CPU times: user 2min 57s, sys: 1.43 s, total: 2min 58s\n",
      "Wall time: 2min 59s\n",
      "CPU times: user 1 ms, sys: 9 ms, total: 10 ms\n",
      "Wall time: 153 ms\n",
      "CPU times: user 5min 26s, sys: 2.62 s, total: 5min 28s\n",
      "Wall time: 5min 29s\n",
      "---\n",
      "iPSC ambivalent\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 32.3 ms\n",
      "CPU times: user 6.03 s, sys: 7 ms, total: 6.03 s\n",
      "Wall time: 6.04 s\n",
      "CPU times: user 1 ms, sys: 10 ms, total: 11 ms\n",
      "Wall time: 866 ms\n",
      "CPU times: user 7min 51s, sys: 1.58 s, total: 7min 53s\n",
      "Wall time: 7min 54s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 1 ms, sys: 9 ms, total: 10 ms\n",
      "Wall time: 27.4 ms\n",
      "CPU times: user 6.39 s, sys: 112 ms, total: 6.5 s\n",
      "Wall time: 6.51 s\n",
      "CPU times: user 3 ms, sys: 8 ms, total: 11 ms\n",
      "Wall time: 310 ms\n",
      "CPU times: user 7min 55s, sys: 2.45 s, total: 7min 58s\n",
      "Wall time: 7min 59s\n",
      "---\n",
      "iPSC bimodal\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 3 ms, sys: 12 ms, total: 15 ms\n",
      "Wall time: 38.4 s\n",
      "CPU times: user 1min 48s, sys: 514 ms, total: 1min 48s\n",
      "Wall time: 1min 48s\n",
      "CPU times: user 4 ms, sys: 9 ms, total: 13 ms\n",
      "Wall time: 40.9 s\n",
      "CPU times: user 7min 52s, sys: 1.96 s, total: 7min 54s\n",
      "Wall time: 7min 54s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 1 ms, sys: 9 ms, total: 10 ms\n",
      "Wall time: 85.5 ms\n",
      "CPU times: user 1min 39s, sys: 1.66 s, total: 1min 40s\n",
      "Wall time: 1min 41s\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 260 ms\n",
      "CPU times: user 7min 52s, sys: 3.68 s, total: 7min 56s\n",
      "Wall time: 7min 57s\n",
      "---\n",
      "iPSC concurrent\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 3 ms, sys: 8 ms, total: 11 ms\n",
      "Wall time: 30.3 ms\n",
      "CPU times: user 93 ms, sys: 19 ms, total: 112 ms\n",
      "Wall time: 120 ms\n",
      "CPU times: user 999 µs, sys: 11 ms, total: 12 ms\n",
      "Wall time: 8.36 s\n",
      "CPU times: user 7min 50s, sys: 2.79 s, total: 7min 52s\n",
      "Wall time: 7min 53s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 0 ns, sys: 12 ms, total: 12 ms\n",
      "Wall time: 97.5 ms\n",
      "CPU times: user 130 ms, sys: 0 ns, total: 130 ms\n",
      "Wall time: 130 ms\n",
      "CPU times: user 1e+03 µs, sys: 11 ms, total: 12 ms\n",
      "Wall time: 39 s\n",
      "CPU times: user 7min 54s, sys: 2.99 s, total: 7min 57s\n",
      "Wall time: 7min 58s\n",
      "---\n",
      "iPSC excluded\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 9 ms, total: 11 ms\n",
      "Wall time: 296 ms\n",
      "CPU times: user 1min 58s, sys: 748 ms, total: 1min 58s\n",
      "Wall time: 1min 59s\n",
      "CPU times: user 1 ms, sys: 10 ms, total: 11 ms\n",
      "Wall time: 337 ms\n",
      "CPU times: user 7min 49s, sys: 2.62 s, total: 7min 52s\n",
      "Wall time: 7min 53s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 1 ms, sys: 9 ms, total: 10 ms\n",
      "Wall time: 56.2 ms\n",
      "CPU times: user 1min 55s, sys: 315 ms, total: 1min 56s\n",
      "Wall time: 1min 56s\n",
      "CPU times: user 1 ms, sys: 9 ms, total: 10 ms\n",
      "Wall time: 479 ms\n",
      "CPU times: user 7min 54s, sys: 975 ms, total: 7min 55s\n",
      "Wall time: 7min 56s\n",
      "---\n",
      "iPSC included\n",
      "\n",
      "\tupstream\n",
      "CPU times: user 2 ms, sys: 7 ms, total: 9 ms\n",
      "Wall time: 172 ms\n",
      "CPU times: user 4min 1s, sys: 2.48 s, total: 4min 3s\n",
      "Wall time: 4min 4s\n",
      "CPU times: user 999 µs, sys: 9 ms, total: 10 ms\n",
      "Wall time: 239 ms\n",
      "CPU times: user 7min 51s, sys: 2.67 s, total: 7min 54s\n",
      "Wall time: 7min 55s\n",
      "\n",
      "\tdownstream\n",
      "CPU times: user 2 ms, sys: 8 ms, total: 10 ms\n",
      "Wall time: 201 ms\n",
      "CPU times: user 4min 14s, sys: 1.89 s, total: 4min 16s\n",
      "Wall time: 4min 17s\n",
      "CPU times: user 999 µs, sys: 9 ms, total: 10 ms\n",
      "Wall time: 250 ms\n",
      "CPU times: user 7min 54s, sys: 3.37 s, total: 7min 58s\n",
      "Wall time: 7min 58s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>upstream400nt_MN_ambivalent_placental</th>\n",
       "      <th>downstream400nt_MN_ambivalent_placental</th>\n",
       "      <th>upstream400nt_MN_bimodal_placental</th>\n",
       "      <th>downstream400nt_MN_bimodal_placental</th>\n",
       "      <th>upstream400nt_MN_excluded_placental</th>\n",
       "      <th>downstream400nt_MN_excluded_placental</th>\n",
       "      <th>upstream400nt_MN_included_placental</th>\n",
       "      <th>downstream400nt_MN_included_placental</th>\n",
       "      <th>upstream400nt_NPC_ambivalent_placental</th>\n",
       "      <th>downstream400nt_NPC_ambivalent_placental</th>\n",
       "      <th>...</th>\n",
       "      <th>upstream400nt_iPSC_ambivalent_placental</th>\n",
       "      <th>downstream400nt_iPSC_ambivalent_placental</th>\n",
       "      <th>upstream400nt_iPSC_bimodal_placental</th>\n",
       "      <th>downstream400nt_iPSC_bimodal_placental</th>\n",
       "      <th>upstream400nt_iPSC_concurrent_placental</th>\n",
       "      <th>downstream400nt_iPSC_concurrent_placental</th>\n",
       "      <th>upstream400nt_iPSC_excluded_placental</th>\n",
       "      <th>downstream400nt_iPSC_excluded_placental</th>\n",
       "      <th>upstream400nt_iPSC_included_placental</th>\n",
       "      <th>downstream400nt_iPSC_included_placental</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAA</th>\n",
       "      <td>0.474524</td>\n",
       "      <td>-0.241359</td>\n",
       "      <td>0.185327</td>\n",
       "      <td>0.115022</td>\n",
       "      <td>0.168236</td>\n",
       "      <td>0.097491</td>\n",
       "      <td>-0.081924</td>\n",
       "      <td>-0.049461</td>\n",
       "      <td>0.475049</td>\n",
       "      <td>-0.247089</td>\n",
       "      <td>...</td>\n",
       "      <td>0.257730</td>\n",
       "      <td>0.473247</td>\n",
       "      <td>0.084969</td>\n",
       "      <td>0.016316</td>\n",
       "      <td>0.264650</td>\n",
       "      <td>-0.244265</td>\n",
       "      <td>0.182092</td>\n",
       "      <td>0.159253</td>\n",
       "      <td>-0.073529</td>\n",
       "      <td>-0.049269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAC</th>\n",
       "      <td>-0.247947</td>\n",
       "      <td>-0.244669</td>\n",
       "      <td>0.147088</td>\n",
       "      <td>0.040678</td>\n",
       "      <td>0.186828</td>\n",
       "      <td>0.152138</td>\n",
       "      <td>-0.076310</td>\n",
       "      <td>-0.044002</td>\n",
       "      <td>-0.244190</td>\n",
       "      <td>-0.242973</td>\n",
       "      <td>...</td>\n",
       "      <td>0.301763</td>\n",
       "      <td>0.098104</td>\n",
       "      <td>0.046665</td>\n",
       "      <td>0.003262</td>\n",
       "      <td>-0.237077</td>\n",
       "      <td>-0.234942</td>\n",
       "      <td>0.155938</td>\n",
       "      <td>0.168336</td>\n",
       "      <td>-0.055657</td>\n",
       "      <td>-0.043485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAG</th>\n",
       "      <td>0.265535</td>\n",
       "      <td>-0.067158</td>\n",
       "      <td>0.100481</td>\n",
       "      <td>0.041271</td>\n",
       "      <td>0.192364</td>\n",
       "      <td>0.123487</td>\n",
       "      <td>-0.067273</td>\n",
       "      <td>-0.038083</td>\n",
       "      <td>0.807236</td>\n",
       "      <td>-0.296720</td>\n",
       "      <td>...</td>\n",
       "      <td>0.459921</td>\n",
       "      <td>0.225355</td>\n",
       "      <td>0.028947</td>\n",
       "      <td>-0.001096</td>\n",
       "      <td>0.502446</td>\n",
       "      <td>-0.288491</td>\n",
       "      <td>0.180525</td>\n",
       "      <td>0.155642</td>\n",
       "      <td>-0.058116</td>\n",
       "      <td>-0.040434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAT</th>\n",
       "      <td>-0.301619</td>\n",
       "      <td>-0.304202</td>\n",
       "      <td>0.201983</td>\n",
       "      <td>0.130975</td>\n",
       "      <td>0.271184</td>\n",
       "      <td>0.145468</td>\n",
       "      <td>-0.108141</td>\n",
       "      <td>-0.064133</td>\n",
       "      <td>0.636355</td>\n",
       "      <td>-0.299703</td>\n",
       "      <td>...</td>\n",
       "      <td>0.331626</td>\n",
       "      <td>0.550518</td>\n",
       "      <td>0.119045</td>\n",
       "      <td>0.031195</td>\n",
       "      <td>0.340351</td>\n",
       "      <td>-0.294706</td>\n",
       "      <td>0.218485</td>\n",
       "      <td>0.194627</td>\n",
       "      <td>-0.093558</td>\n",
       "      <td>-0.063249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACA</th>\n",
       "      <td>-0.285755</td>\n",
       "      <td>-0.257514</td>\n",
       "      <td>0.109675</td>\n",
       "      <td>0.006757</td>\n",
       "      <td>0.134356</td>\n",
       "      <td>0.180587</td>\n",
       "      <td>-0.055661</td>\n",
       "      <td>-0.042285</td>\n",
       "      <td>-0.272352</td>\n",
       "      <td>-0.256995</td>\n",
       "      <td>...</td>\n",
       "      <td>0.405794</td>\n",
       "      <td>0.293585</td>\n",
       "      <td>0.041910</td>\n",
       "      <td>-0.026416</td>\n",
       "      <td>-0.257180</td>\n",
       "      <td>-0.245345</td>\n",
       "      <td>0.108622</td>\n",
       "      <td>0.187163</td>\n",
       "      <td>-0.043230</td>\n",
       "      <td>-0.041479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      upstream400nt_MN_ambivalent_placental  \\\n",
       "AAAA                               0.474524   \n",
       "AAAC                              -0.247947   \n",
       "AAAG                               0.265535   \n",
       "AAAT                              -0.301619   \n",
       "AACA                              -0.285755   \n",
       "\n",
       "      downstream400nt_MN_ambivalent_placental  \\\n",
       "AAAA                                -0.241359   \n",
       "AAAC                                -0.244669   \n",
       "AAAG                                -0.067158   \n",
       "AAAT                                -0.304202   \n",
       "AACA                                -0.257514   \n",
       "\n",
       "      upstream400nt_MN_bimodal_placental  \\\n",
       "AAAA                            0.185327   \n",
       "AAAC                            0.147088   \n",
       "AAAG                            0.100481   \n",
       "AAAT                            0.201983   \n",
       "AACA                            0.109675   \n",
       "\n",
       "      downstream400nt_MN_bimodal_placental  \\\n",
       "AAAA                              0.115022   \n",
       "AAAC                              0.040678   \n",
       "AAAG                              0.041271   \n",
       "AAAT                              0.130975   \n",
       "AACA                              0.006757   \n",
       "\n",
       "      upstream400nt_MN_excluded_placental  \\\n",
       "AAAA                             0.168236   \n",
       "AAAC                             0.186828   \n",
       "AAAG                             0.192364   \n",
       "AAAT                             0.271184   \n",
       "AACA                             0.134356   \n",
       "\n",
       "      downstream400nt_MN_excluded_placental  \\\n",
       "AAAA                               0.097491   \n",
       "AAAC                               0.152138   \n",
       "AAAG                               0.123487   \n",
       "AAAT                               0.145468   \n",
       "AACA                               0.180587   \n",
       "\n",
       "      upstream400nt_MN_included_placental  \\\n",
       "AAAA                            -0.081924   \n",
       "AAAC                            -0.076310   \n",
       "AAAG                            -0.067273   \n",
       "AAAT                            -0.108141   \n",
       "AACA                            -0.055661   \n",
       "\n",
       "      downstream400nt_MN_included_placental  \\\n",
       "AAAA                              -0.049461   \n",
       "AAAC                              -0.044002   \n",
       "AAAG                              -0.038083   \n",
       "AAAT                              -0.064133   \n",
       "AACA                              -0.042285   \n",
       "\n",
       "      upstream400nt_NPC_ambivalent_placental  \\\n",
       "AAAA                                0.475049   \n",
       "AAAC                               -0.244190   \n",
       "AAAG                                0.807236   \n",
       "AAAT                                0.636355   \n",
       "AACA                               -0.272352   \n",
       "\n",
       "      downstream400nt_NPC_ambivalent_placental  \\\n",
       "AAAA                                 -0.247089   \n",
       "AAAC                                 -0.242973   \n",
       "AAAG                                 -0.296720   \n",
       "AAAT                                 -0.299703   \n",
       "AACA                                 -0.256995   \n",
       "\n",
       "                       ...                     \\\n",
       "AAAA                   ...                      \n",
       "AAAC                   ...                      \n",
       "AAAG                   ...                      \n",
       "AAAT                   ...                      \n",
       "AACA                   ...                      \n",
       "\n",
       "      upstream400nt_iPSC_ambivalent_placental  \\\n",
       "AAAA                                 0.257730   \n",
       "AAAC                                 0.301763   \n",
       "AAAG                                 0.459921   \n",
       "AAAT                                 0.331626   \n",
       "AACA                                 0.405794   \n",
       "\n",
       "      downstream400nt_iPSC_ambivalent_placental  \\\n",
       "AAAA                                   0.473247   \n",
       "AAAC                                   0.098104   \n",
       "AAAG                                   0.225355   \n",
       "AAAT                                   0.550518   \n",
       "AACA                                   0.293585   \n",
       "\n",
       "      upstream400nt_iPSC_bimodal_placental  \\\n",
       "AAAA                              0.084969   \n",
       "AAAC                              0.046665   \n",
       "AAAG                              0.028947   \n",
       "AAAT                              0.119045   \n",
       "AACA                              0.041910   \n",
       "\n",
       "      downstream400nt_iPSC_bimodal_placental  \\\n",
       "AAAA                                0.016316   \n",
       "AAAC                                0.003262   \n",
       "AAAG                               -0.001096   \n",
       "AAAT                                0.031195   \n",
       "AACA                               -0.026416   \n",
       "\n",
       "      upstream400nt_iPSC_concurrent_placental  \\\n",
       "AAAA                                 0.264650   \n",
       "AAAC                                -0.237077   \n",
       "AAAG                                 0.502446   \n",
       "AAAT                                 0.340351   \n",
       "AACA                                -0.257180   \n",
       "\n",
       "      downstream400nt_iPSC_concurrent_placental  \\\n",
       "AAAA                                  -0.244265   \n",
       "AAAC                                  -0.234942   \n",
       "AAAG                                  -0.288491   \n",
       "AAAT                                  -0.294706   \n",
       "AACA                                  -0.245345   \n",
       "\n",
       "      upstream400nt_iPSC_excluded_placental  \\\n",
       "AAAA                               0.182092   \n",
       "AAAC                               0.155938   \n",
       "AAAG                               0.180525   \n",
       "AAAT                               0.218485   \n",
       "AACA                               0.108622   \n",
       "\n",
       "      downstream400nt_iPSC_excluded_placental  \\\n",
       "AAAA                                 0.159253   \n",
       "AAAC                                 0.168336   \n",
       "AAAG                                 0.155642   \n",
       "AAAT                                 0.194627   \n",
       "AACA                                 0.187163   \n",
       "\n",
       "      upstream400nt_iPSC_included_placental  \\\n",
       "AAAA                              -0.073529   \n",
       "AAAC                              -0.055657   \n",
       "AAAG                              -0.058116   \n",
       "AAAT                              -0.093558   \n",
       "AACA                              -0.043230   \n",
       "\n",
       "      downstream400nt_iPSC_included_placental  \n",
       "AAAA                                -0.049269  \n",
       "AAAC                                -0.043485  \n",
       "AAAG                                -0.040434  \n",
       "AAAT                                -0.063249  \n",
       "AACA                                -0.041479  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import kvector\n",
    "import pybedtools\n",
    "\n",
    "DIRECTIONS = 'upstream', 'downstream'\n",
    "\n",
    "\n",
    "\n",
    "DIR = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "exon_bedfile = '{}/background_events.bed'.format(bed_folder)\n",
    "exon_bed = pybedtools.BedTool(exon_bedfile)\n",
    "\n",
    "commands = []\n",
    "\n",
    "findMotifsGenome = '/home/yeo-lab/software/homer/bin/findMotifsGenome.pl'\n",
    "n_processors = 4\n",
    "homer_flags = '-rna -len 4,5,6 -mset vertebrates -mis 0 -p {} -noweight'.format(n_processors)\n",
    "\n",
    "\n",
    "primate_filename = '/projects/ps-yeolab/genomes/hg19/database/phastConsElements46wayPrimates.bed'\n",
    "primate = pybedtools.BedTool(primate_filename)\n",
    "placental_filename = '/projects/ps-yeolab/genomes/hg19/database/phastConsElements46wayPlacental.bed'\n",
    "placental = pybedtools.BedTool(placental_filename)\n",
    "conserved_regions = {'primate': primate, 'placental': placental}\n",
    "\n",
    "conservation_bed = placental\n",
    "\n",
    "genome = 'hg19'\n",
    "\n",
    "nt = 400\n",
    "\n",
    "\n",
    "genome_fasta = '/projects/ps-yeolab/genomes/hg19/chromosomes/all.fa'\n",
    "\n",
    "kmer_zscores = []\n",
    "\n",
    "for phenotype, phenotype_df in study.supplemental.modalities_tidy.groupby('phenotype'):\n",
    "    background_events = set(phenotype_df.event_id)\n",
    "    for modality, modality_df in phenotype_df.groupby('modality'):\n",
    "        print '---\\n', phenotype, modality\n",
    "        event_ids = set(modality_df.event_id)\n",
    "        format_args = bed_folder, phenotype, modality\n",
    "        foreground_table = exon2_bed_table.loc[exon2_bed_table[3].isin(event_ids)]\n",
    "        foreground_filename = '{}/exon2_{}_{}_foreground.bed'.format(*format_args)\n",
    "        foreground_table.to_csv(foreground_filename, index=False, header=False, sep='\\t')\n",
    "        foreground = pybedtools.BedTool(foreground_filename)\n",
    "        \n",
    "        \n",
    "        background_table = exon2_bed_table.loc[exon2_bed_table[3].isin(background_events)]\n",
    "        background_filename = '{}/exon2_{}_{}_background.bed'.format(*format_args)\n",
    "        background_table.to_csv(background_filename, index=False, header=False, sep='\\t')\n",
    "        background = pybedtools.BedTool(background_filename)\n",
    "        \n",
    "        \n",
    "        pair = pyhomer.ForegroundBackgroundPair(foreground, background)\n",
    "#         print '\\n', pair \n",
    "        \n",
    "        for direction in DIRECTIONS:\n",
    "            print '\\n\\t', direction\n",
    "            intron_pair = pair.flanking_intron(direction, 'hg19', 400)\n",
    "\n",
    "            conserved_introns = intron_pair.intersect(conservation_bed, 'placental')\n",
    "#             print '\\n', conserved_introns\n",
    "            %time seqs = conserved_introns.foreground.sequence(fi=genome_fasta, s=True)\n",
    "            %time foreground_kmers = kvector.count_kmers(seqs.seqfn)\n",
    "\n",
    "            %time seqs = conserved_introns.background.sequence(fi=genome_fasta, s=True)\n",
    "            %time background_kmers = kvector.count_kmers(seqs.seqfn)\n",
    "            kmer_zscore = (foreground_kmers.mean() - background_kmers.mean())/background_kmers.std()\n",
    "            kmer_zscore.name = '{}{}nt_{}_{}_placental'.format(direction, nt, phenotype, modality)\n",
    "            kmer_zscores.append(kmer_zscore)\n",
    "kmer_zscores_all = pd.concat(kmer_zscores, axis=1)\n",
    "kmer_zscores_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5376, 26)\n",
      "(5376, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>upstream400nt_MN_ambivalent_placental</th>\n",
       "      <th>downstream400nt_MN_ambivalent_placental</th>\n",
       "      <th>upstream400nt_MN_bimodal_placental</th>\n",
       "      <th>downstream400nt_MN_bimodal_placental</th>\n",
       "      <th>upstream400nt_MN_excluded_placental</th>\n",
       "      <th>downstream400nt_MN_excluded_placental</th>\n",
       "      <th>upstream400nt_MN_included_placental</th>\n",
       "      <th>downstream400nt_MN_included_placental</th>\n",
       "      <th>upstream400nt_NPC_ambivalent_placental</th>\n",
       "      <th>downstream400nt_NPC_ambivalent_placental</th>\n",
       "      <th>...</th>\n",
       "      <th>upstream400nt_iPSC_ambivalent_placental</th>\n",
       "      <th>downstream400nt_iPSC_ambivalent_placental</th>\n",
       "      <th>upstream400nt_iPSC_bimodal_placental</th>\n",
       "      <th>downstream400nt_iPSC_bimodal_placental</th>\n",
       "      <th>upstream400nt_iPSC_concurrent_placental</th>\n",
       "      <th>downstream400nt_iPSC_concurrent_placental</th>\n",
       "      <th>upstream400nt_iPSC_excluded_placental</th>\n",
       "      <th>downstream400nt_iPSC_excluded_placental</th>\n",
       "      <th>upstream400nt_iPSC_included_placental</th>\n",
       "      <th>downstream400nt_iPSC_included_placental</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAA</th>\n",
       "      <td>0.474524</td>\n",
       "      <td>-0.241359</td>\n",
       "      <td>0.185327</td>\n",
       "      <td>0.115022</td>\n",
       "      <td>0.168236</td>\n",
       "      <td>0.097491</td>\n",
       "      <td>-0.081924</td>\n",
       "      <td>-0.049461</td>\n",
       "      <td>0.475049</td>\n",
       "      <td>-0.247089</td>\n",
       "      <td>...</td>\n",
       "      <td>0.257730</td>\n",
       "      <td>0.473247</td>\n",
       "      <td>0.084969</td>\n",
       "      <td>0.016316</td>\n",
       "      <td>0.264650</td>\n",
       "      <td>-0.244265</td>\n",
       "      <td>0.182092</td>\n",
       "      <td>0.159253</td>\n",
       "      <td>-0.073529</td>\n",
       "      <td>-0.049269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAC</th>\n",
       "      <td>-0.247947</td>\n",
       "      <td>-0.244669</td>\n",
       "      <td>0.147088</td>\n",
       "      <td>0.040678</td>\n",
       "      <td>0.186828</td>\n",
       "      <td>0.152138</td>\n",
       "      <td>-0.076310</td>\n",
       "      <td>-0.044002</td>\n",
       "      <td>-0.244190</td>\n",
       "      <td>-0.242973</td>\n",
       "      <td>...</td>\n",
       "      <td>0.301763</td>\n",
       "      <td>0.098104</td>\n",
       "      <td>0.046665</td>\n",
       "      <td>0.003262</td>\n",
       "      <td>-0.237077</td>\n",
       "      <td>-0.234942</td>\n",
       "      <td>0.155938</td>\n",
       "      <td>0.168336</td>\n",
       "      <td>-0.055657</td>\n",
       "      <td>-0.043485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAG</th>\n",
       "      <td>0.265535</td>\n",
       "      <td>-0.067158</td>\n",
       "      <td>0.100481</td>\n",
       "      <td>0.041271</td>\n",
       "      <td>0.192364</td>\n",
       "      <td>0.123487</td>\n",
       "      <td>-0.067273</td>\n",
       "      <td>-0.038083</td>\n",
       "      <td>0.807236</td>\n",
       "      <td>-0.296720</td>\n",
       "      <td>...</td>\n",
       "      <td>0.459921</td>\n",
       "      <td>0.225355</td>\n",
       "      <td>0.028947</td>\n",
       "      <td>-0.001096</td>\n",
       "      <td>0.502446</td>\n",
       "      <td>-0.288491</td>\n",
       "      <td>0.180525</td>\n",
       "      <td>0.155642</td>\n",
       "      <td>-0.058116</td>\n",
       "      <td>-0.040434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAT</th>\n",
       "      <td>-0.301619</td>\n",
       "      <td>-0.304202</td>\n",
       "      <td>0.201983</td>\n",
       "      <td>0.130975</td>\n",
       "      <td>0.271184</td>\n",
       "      <td>0.145468</td>\n",
       "      <td>-0.108141</td>\n",
       "      <td>-0.064133</td>\n",
       "      <td>0.636355</td>\n",
       "      <td>-0.299703</td>\n",
       "      <td>...</td>\n",
       "      <td>0.331626</td>\n",
       "      <td>0.550518</td>\n",
       "      <td>0.119045</td>\n",
       "      <td>0.031195</td>\n",
       "      <td>0.340351</td>\n",
       "      <td>-0.294706</td>\n",
       "      <td>0.218485</td>\n",
       "      <td>0.194627</td>\n",
       "      <td>-0.093558</td>\n",
       "      <td>-0.063249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACA</th>\n",
       "      <td>-0.285755</td>\n",
       "      <td>-0.257514</td>\n",
       "      <td>0.109675</td>\n",
       "      <td>0.006757</td>\n",
       "      <td>0.134356</td>\n",
       "      <td>0.180587</td>\n",
       "      <td>-0.055661</td>\n",
       "      <td>-0.042285</td>\n",
       "      <td>-0.272352</td>\n",
       "      <td>-0.256995</td>\n",
       "      <td>...</td>\n",
       "      <td>0.405794</td>\n",
       "      <td>0.293585</td>\n",
       "      <td>0.041910</td>\n",
       "      <td>-0.026416</td>\n",
       "      <td>-0.257180</td>\n",
       "      <td>-0.245345</td>\n",
       "      <td>0.108622</td>\n",
       "      <td>0.187163</td>\n",
       "      <td>-0.043230</td>\n",
       "      <td>-0.041479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      upstream400nt_MN_ambivalent_placental  \\\n",
       "AAAA                               0.474524   \n",
       "AAAC                              -0.247947   \n",
       "AAAG                               0.265535   \n",
       "AAAT                              -0.301619   \n",
       "AACA                              -0.285755   \n",
       "\n",
       "      downstream400nt_MN_ambivalent_placental  \\\n",
       "AAAA                                -0.241359   \n",
       "AAAC                                -0.244669   \n",
       "AAAG                                -0.067158   \n",
       "AAAT                                -0.304202   \n",
       "AACA                                -0.257514   \n",
       "\n",
       "      upstream400nt_MN_bimodal_placental  \\\n",
       "AAAA                            0.185327   \n",
       "AAAC                            0.147088   \n",
       "AAAG                            0.100481   \n",
       "AAAT                            0.201983   \n",
       "AACA                            0.109675   \n",
       "\n",
       "      downstream400nt_MN_bimodal_placental  \\\n",
       "AAAA                              0.115022   \n",
       "AAAC                              0.040678   \n",
       "AAAG                              0.041271   \n",
       "AAAT                              0.130975   \n",
       "AACA                              0.006757   \n",
       "\n",
       "      upstream400nt_MN_excluded_placental  \\\n",
       "AAAA                             0.168236   \n",
       "AAAC                             0.186828   \n",
       "AAAG                             0.192364   \n",
       "AAAT                             0.271184   \n",
       "AACA                             0.134356   \n",
       "\n",
       "      downstream400nt_MN_excluded_placental  \\\n",
       "AAAA                               0.097491   \n",
       "AAAC                               0.152138   \n",
       "AAAG                               0.123487   \n",
       "AAAT                               0.145468   \n",
       "AACA                               0.180587   \n",
       "\n",
       "      upstream400nt_MN_included_placental  \\\n",
       "AAAA                            -0.081924   \n",
       "AAAC                            -0.076310   \n",
       "AAAG                            -0.067273   \n",
       "AAAT                            -0.108141   \n",
       "AACA                            -0.055661   \n",
       "\n",
       "      downstream400nt_MN_included_placental  \\\n",
       "AAAA                              -0.049461   \n",
       "AAAC                              -0.044002   \n",
       "AAAG                              -0.038083   \n",
       "AAAT                              -0.064133   \n",
       "AACA                              -0.042285   \n",
       "\n",
       "      upstream400nt_NPC_ambivalent_placental  \\\n",
       "AAAA                                0.475049   \n",
       "AAAC                               -0.244190   \n",
       "AAAG                                0.807236   \n",
       "AAAT                                0.636355   \n",
       "AACA                               -0.272352   \n",
       "\n",
       "      downstream400nt_NPC_ambivalent_placental  \\\n",
       "AAAA                                 -0.247089   \n",
       "AAAC                                 -0.242973   \n",
       "AAAG                                 -0.296720   \n",
       "AAAT                                 -0.299703   \n",
       "AACA                                 -0.256995   \n",
       "\n",
       "                       ...                     \\\n",
       "AAAA                   ...                      \n",
       "AAAC                   ...                      \n",
       "AAAG                   ...                      \n",
       "AAAT                   ...                      \n",
       "AACA                   ...                      \n",
       "\n",
       "      upstream400nt_iPSC_ambivalent_placental  \\\n",
       "AAAA                                 0.257730   \n",
       "AAAC                                 0.301763   \n",
       "AAAG                                 0.459921   \n",
       "AAAT                                 0.331626   \n",
       "AACA                                 0.405794   \n",
       "\n",
       "      downstream400nt_iPSC_ambivalent_placental  \\\n",
       "AAAA                                   0.473247   \n",
       "AAAC                                   0.098104   \n",
       "AAAG                                   0.225355   \n",
       "AAAT                                   0.550518   \n",
       "AACA                                   0.293585   \n",
       "\n",
       "      upstream400nt_iPSC_bimodal_placental  \\\n",
       "AAAA                              0.084969   \n",
       "AAAC                              0.046665   \n",
       "AAAG                              0.028947   \n",
       "AAAT                              0.119045   \n",
       "AACA                              0.041910   \n",
       "\n",
       "      downstream400nt_iPSC_bimodal_placental  \\\n",
       "AAAA                                0.016316   \n",
       "AAAC                                0.003262   \n",
       "AAAG                               -0.001096   \n",
       "AAAT                                0.031195   \n",
       "AACA                               -0.026416   \n",
       "\n",
       "      upstream400nt_iPSC_concurrent_placental  \\\n",
       "AAAA                                 0.264650   \n",
       "AAAC                                -0.237077   \n",
       "AAAG                                 0.502446   \n",
       "AAAT                                 0.340351   \n",
       "AACA                                -0.257180   \n",
       "\n",
       "      downstream400nt_iPSC_concurrent_placental  \\\n",
       "AAAA                                  -0.244265   \n",
       "AAAC                                  -0.234942   \n",
       "AAAG                                  -0.288491   \n",
       "AAAT                                  -0.294706   \n",
       "AACA                                  -0.245345   \n",
       "\n",
       "      upstream400nt_iPSC_excluded_placental  \\\n",
       "AAAA                               0.182092   \n",
       "AAAC                               0.155938   \n",
       "AAAG                               0.180525   \n",
       "AAAT                               0.218485   \n",
       "AACA                               0.108622   \n",
       "\n",
       "      downstream400nt_iPSC_excluded_placental  \\\n",
       "AAAA                                 0.159253   \n",
       "AAAC                                 0.168336   \n",
       "AAAG                                 0.155642   \n",
       "AAAT                                 0.194627   \n",
       "AACA                                 0.187163   \n",
       "\n",
       "      upstream400nt_iPSC_included_placental  \\\n",
       "AAAA                              -0.073529   \n",
       "AAAC                              -0.055657   \n",
       "AAAG                              -0.058116   \n",
       "AAAT                              -0.093558   \n",
       "AACA                              -0.043230   \n",
       "\n",
       "      downstream400nt_iPSC_included_placental  \n",
       "AAAA                                -0.049269  \n",
       "AAAC                                -0.043485  \n",
       "AAAG                                -0.040434  \n",
       "AAAT                                -0.063249  \n",
       "AACA                                -0.041479  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_zscores_all = pd.concat(kmer_zscores, axis=1)\n",
    "print kmer_zscores_all.shape\n",
    "kmer_zscores_all = kmer_zscores_all.dropna(axis=1, how='all').dropna(how='all', axis=0)\n",
    "print kmer_zscores_all.shape\n",
    "\n",
    "# Replace remaining NAs with zero since they aren't enriched\n",
    "kmer_zscores_all = kmer_zscores_all.fillna(0)\n",
    "kmer_zscores_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "study.supplemental.kmer_zscores = kmer_zscores_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>direction</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>modality</th>\n",
       "      <th>clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>upstream400nt_MN_ambivalent_placental</th>\n",
       "      <td>upstream400nt</td>\n",
       "      <td>MN</td>\n",
       "      <td>ambivalent</td>\n",
       "      <td>placental</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream400nt_MN_ambivalent_placental</th>\n",
       "      <td>downstream400nt</td>\n",
       "      <td>MN</td>\n",
       "      <td>ambivalent</td>\n",
       "      <td>placental</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>upstream400nt_MN_bimodal_placental</th>\n",
       "      <td>upstream400nt</td>\n",
       "      <td>MN</td>\n",
       "      <td>bimodal</td>\n",
       "      <td>placental</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream400nt_MN_bimodal_placental</th>\n",
       "      <td>downstream400nt</td>\n",
       "      <td>MN</td>\n",
       "      <td>bimodal</td>\n",
       "      <td>placental</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>upstream400nt_MN_excluded_placental</th>\n",
       "      <td>upstream400nt</td>\n",
       "      <td>MN</td>\n",
       "      <td>excluded</td>\n",
       "      <td>placental</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               direction phenotype  \\\n",
       "upstream400nt_MN_ambivalent_placental      upstream400nt        MN   \n",
       "downstream400nt_MN_ambivalent_placental  downstream400nt        MN   \n",
       "upstream400nt_MN_bimodal_placental         upstream400nt        MN   \n",
       "downstream400nt_MN_bimodal_placental     downstream400nt        MN   \n",
       "upstream400nt_MN_excluded_placental        upstream400nt        MN   \n",
       "\n",
       "                                           modality      clade  \n",
       "upstream400nt_MN_ambivalent_placental    ambivalent  placental  \n",
       "downstream400nt_MN_ambivalent_placental  ambivalent  placental  \n",
       "upstream400nt_MN_bimodal_placental          bimodal  placental  \n",
       "downstream400nt_MN_bimodal_placental        bimodal  placental  \n",
       "upstream400nt_MN_excluded_placental        excluded  placental  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_zscores_metadata = pd.DataFrame.from_records(\n",
    "    study.supplemental.kmer_zscores.columns.map(lambda x: pd.Series(x.split('_'))), \n",
    "    index=study.supplemental.kmer_zscores.columns)\n",
    "kmer_zscores_metadata.columns = ['direction', 'phenotype', 'modality', 'clade']\n",
    "study.supplemental.kmer_zscores_metadata = kmer_zscores_metadata\n",
    "\n",
    "kmer_zscores_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote datapackage to /projects/ps-yeolab/obotvinnik/flotilla_projects/singlecell_pnm_figure2_modalities_bayesian_kmers/datapackage.json\n"
     ]
    }
   ],
   "source": [
    "study.save('singlecell_pnm_figure2_modalities_bayesian_kmers', flotilla_dir=flotilla_dir)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
