{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import print_function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata.shape (214, 9)\n",
      "expression.shape (214, 34608)\n",
      "mapping_stats.shape (214, 35)\n",
      "splicing.shape (299, 26987)\n",
      "CPU times: user 14.2 s, sys: 376 ms, total: 14.5 s\n",
      "Wall time: 48.3 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "  \"This module will be removed in 0.20.\", DeprecationWarning)\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import matplotlib as mpl\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import six\n",
    "%matplotlib inline\n",
    "\n",
    "import flotilla\n",
    "\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "\n",
    "csv_folder = '{}/csvs_for_paper'.format(folder)\n",
    "outrigger_rmdup_folder = '{}/outrigger_v2_rmdup'.format(csv_folder)\n",
    "\n",
    "metadata = pd.read_csv('{}/metadata.csv'.format(csv_folder), index_col=0)\n",
    "expression = pd.read_csv('{}/expression.csv'.format(csv_folder), index_col=0)\n",
    "mapping_stats = pd.read_csv('{}/mapping_stats.csv'.format(csv_folder), index_col=0)\n",
    "\n",
    "splicing = pd.read_csv('{}/psi.csv'.format(csv_folder), index_col=0)\n",
    "\n",
    "\n",
    "print('metadata.shape', metadata.shape)\n",
    "print('expression.shape', expression.shape)\n",
    "print('mapping_stats.shape', mapping_stats.shape)\n",
    "print('splicing.shape', splicing.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 obotvinnik yeo-group 17M Jan 12 17:31 /projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/psi.csv\r\n"
     ]
    }
   ],
   "source": [
    "! ls -lha $csv_folder/psi.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# psi_rmdup = pd.read_csv('{}/psi/outrigger_psi.csv'.format(outrigger_rmdup_folder))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "outrigger_folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms/outrigger_v2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14621575, 22)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (15,16) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon4</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>incompatible_junctions</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>level</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>event_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:101996727-102003454:-@exon:chr10:101996607-101996726:-@junction:chr10:101995522-101996606:-|isoform2=junction:chr10:101997989-102003454:-@exon:chr10:101997779-101997988:-@junction:chr10:101995522-101997778:-</th>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>junction:chr10:101997989-102003454:-</td>\n",
       "      <td>junction:chr10:101996727-102003454:-</td>\n",
       "      <td>junction:chr10:101995522-101997778:-</td>\n",
       "      <td>junction:chr10:101995522-101996606:-</td>\n",
       "      <td>junction:chr10:101995522-102003454:-|junction:...</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>...</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>nan,OTTHUMG00000018901.5</td>\n",
       "      <td>3,2</td>\n",
       "      <td>nan,ncRNA_host</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>CWF19L1,SNORA12</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>ENSG00000212464</td>\n",
       "      <td>chr10:101995423-102003534:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:101996727-102003454:-@exon:chr10:101996607-101996726:-@junction:chr10:101995522-101996606:-|isoform2=junction:chr10:101997989-102003454:-@exon:chr10:101997779-101997988:-@junction:chr10:101995522-101997778:-</th>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>junction:chr10:101997989-102003454:-</td>\n",
       "      <td>junction:chr10:101996727-102003454:-</td>\n",
       "      <td>junction:chr10:101995522-101997778:-</td>\n",
       "      <td>junction:chr10:101995522-101996606:-</td>\n",
       "      <td>junction:chr10:101995522-102003454:-|junction:...</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>...</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>nan,OTTHUMG00000018901.5</td>\n",
       "      <td>3,2</td>\n",
       "      <td>nan,ncRNA_host</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>CWF19L1,SNORA12</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>ENSG00000212464</td>\n",
       "      <td>chr10:101995423-102003534:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:101996727-102003454:-@exon:chr10:101996607-101996726:-@junction:chr10:101995522-101996606:-|isoform2=junction:chr10:101997989-102003454:-@exon:chr10:101997779-101997988:-@junction:chr10:101995522-101997778:-</th>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>junction:chr10:101997989-102003454:-</td>\n",
       "      <td>junction:chr10:101996727-102003454:-</td>\n",
       "      <td>junction:chr10:101995522-101997778:-</td>\n",
       "      <td>junction:chr10:101995522-101996606:-</td>\n",
       "      <td>junction:chr10:101995522-102003454:-|junction:...</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>...</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>nan,OTTHUMG00000018901.5</td>\n",
       "      <td>3,2</td>\n",
       "      <td>nan,ncRNA_host</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>CWF19L1,SNORA12</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>ENSG00000212464</td>\n",
       "      <td>chr10:101995423-102003529:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:101996727-102003454:-@exon:chr10:101996607-101996726:-@junction:chr10:101995522-101996606:-|isoform2=junction:chr10:101997989-102003454:-@exon:chr10:101997779-101997988:-@junction:chr10:101995522-101997778:-</th>\n",
       "      <td>exon:chr10:102003455-102003534:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>junction:chr10:101997989-102003454:-</td>\n",
       "      <td>junction:chr10:101996727-102003454:-</td>\n",
       "      <td>junction:chr10:101995522-101997778:-</td>\n",
       "      <td>junction:chr10:101995522-101996606:-</td>\n",
       "      <td>junction:chr10:101995522-102003454:-|junction:...</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>...</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>nan,OTTHUMG00000018901.5</td>\n",
       "      <td>3,2</td>\n",
       "      <td>nan,ncRNA_host</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>CWF19L1,SNORA12</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>ENSG00000212464</td>\n",
       "      <td>chr10:101995423-102003529:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:101996727-102003454:-@exon:chr10:101996607-101996726:-@junction:chr10:101995522-101996606:-|isoform2=junction:chr10:101997989-102003454:-@exon:chr10:101997779-101997988:-@junction:chr10:101995522-101997778:-</th>\n",
       "      <td>exon:chr10:102003455-102003529:-</td>\n",
       "      <td>exon:chr10:101997779-101997988:-</td>\n",
       "      <td>exon:chr10:101996607-101996726:-</td>\n",
       "      <td>exon:chr10:101995424-101995521:-</td>\n",
       "      <td>junction:chr10:101997989-102003454:-</td>\n",
       "      <td>junction:chr10:101996727-102003454:-</td>\n",
       "      <td>junction:chr10:101995522-101997778:-</td>\n",
       "      <td>junction:chr10:101995522-101996606:-</td>\n",
       "      <td>junction:chr10:101995522-102003454:-|junction:...</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>...</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>nan,OTTHUMG00000018901.5</td>\n",
       "      <td>3,2</td>\n",
       "      <td>nan,ncRNA_host</td>\n",
       "      <td>ENSG00000212464.1,ENSG00000095485.12</td>\n",
       "      <td>CWF19L1,SNORA12</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>snoRNA,protein_coding</td>\n",
       "      <td>ENSG00000212464</td>\n",
       "      <td>chr10:101995423-102003534:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                               exon1  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:102003455-102003534:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:102003455-102003534:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:102003455-102003534:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:102003455-102003534:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:102003455-102003529:-   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101997779-101997988:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101997779-101997988:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101997779-101997988:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101997779-101997988:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101997779-101997988:-   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101996607-101996726:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101996607-101996726:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101996607-101996726:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101996607-101996726:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101996607-101996726:-   \n",
       "\n",
       "                                                                               exon4  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101995424-101995521:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101995424-101995521:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101995424-101995521:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101995424-101995521:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  exon:chr10:101995424-101995521:-   \n",
       "\n",
       "                                                                              junction12  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101997989-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101997989-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101997989-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101997989-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101997989-102003454:-   \n",
       "\n",
       "                                                                              junction13  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101996727-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101996727-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101996727-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101996727-102003454:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101996727-102003454:-   \n",
       "\n",
       "                                                                              junction24  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101997778:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101997778:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101997778:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101997778:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101997778:-   \n",
       "\n",
       "                                                                              junction34  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101996606:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101996606:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101996606:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101996606:-   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-101996606:-   \n",
       "\n",
       "                                                                               incompatible_junctions  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-102003454:-|junction:...   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-102003454:-|junction:...   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-102003454:-|junction:...   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-102003454:-|junction:...   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  junction:chr10:101995522-102003454:-|junction:...   \n",
       "\n",
       "                                                                                 gene_id  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "\n",
       "                                                               ...               \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             ...                \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             ...                \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             ...                \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             ...                \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             ...                \n",
       "\n",
       "                                                                gene_type  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "\n",
       "                                                                 havana_gene  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  nan,OTTHUMG00000018901.5   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  nan,OTTHUMG00000018901.5   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  nan,OTTHUMG00000018901.5   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  nan,OTTHUMG00000018901.5   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  nan,OTTHUMG00000018901.5   \n",
       "\n",
       "                                                   level             tag  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...   3,2  nan,ncRNA_host   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...   3,2  nan,ncRNA_host   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...   3,2  nan,ncRNA_host   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...   3,2  nan,ncRNA_host   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...   3,2  nan,ncRNA_host   \n",
       "\n",
       "                                                                           transcript_id  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464.1,ENSG00000095485.12   \n",
       "\n",
       "                                                    transcript_name  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  CWF19L1,SNORA12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  CWF19L1,SNORA12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  CWF19L1,SNORA12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  CWF19L1,SNORA12   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  CWF19L1,SNORA12   \n",
       "\n",
       "                                                   transcript_status  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             KNOWN   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             KNOWN   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             KNOWN   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             KNOWN   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...             KNOWN   \n",
       "\n",
       "                                                          transcript_type  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  snoRNA,protein_coding   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464   \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  ENSG00000212464   \n",
       "\n",
       "                                                                 event_location  \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  chr10:101995423-102003534:-  \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  chr10:101995423-102003534:-  \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  chr10:101995423-102003529:-  \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  chr10:101995423-102003529:-  \n",
       "isoform1=junction:chr10:101996727-102003454:-@e...  chr10:101995423-102003534:-  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutually_exclusive_exon_feature_data = pd.read_csv('{}/index/mxe/manual_metadata.csv'.format(outrigger_folder), index_col=0)\n",
    "print(mutually_exclusive_exon_feature_data.shape)\n",
    "mutually_exclusive_exon_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5496430, 28)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (21,23) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon3</th>\n",
       "      <th>event_length</th>\n",
       "      <th>event_location</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>illegal_junctions</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>level</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>exon:chr10:100188907-100189399:-</td>\n",
       "      <td>1520.0</td>\n",
       "      <td>event:chr10:100188908-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>492.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>930.0</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OTTHUMG00000018876.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>exon:chr10:100188907-100189399:-</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>event:chr10:100188913-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>930.0</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OTTHUMG00000018876.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>exon:chr10:100188907-100189399:-</td>\n",
       "      <td>1098.0</td>\n",
       "      <td>event:chr10:100189330-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>930.0</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OTTHUMG00000018876.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>exon:chr10:100188912-100189399:-</td>\n",
       "      <td>1520.0</td>\n",
       "      <td>event:chr10:100188908-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>492.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>930.0</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OTTHUMG00000018876.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>exon:chr10:100188912-100189399:-</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>event:chr10:100188913-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>930.0</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OTTHUMG00000018876.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                               exon1  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188907-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188907-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188907-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188912-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188912-100189399:-   \n",
       "\n",
       "                                                    event_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1520.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1515.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1098.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1520.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1515.0   \n",
       "\n",
       "                                                                       event_location  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188908-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188913-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100189330-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188908-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188913-100190427:-   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "\n",
       "                                                    exon3_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         492.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         487.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          70.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         492.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         487.0   \n",
       "\n",
       "                                                    illegal_junctions  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...                NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...                NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...                NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...                NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...                NaN   \n",
       "\n",
       "                                                    intron_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          930.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          930.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          930.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          930.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          930.0   \n",
       "\n",
       "                                                         ...         \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...          \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...          \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...          \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...          \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...          \n",
       "\n",
       "                                                   gene_status  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       KNOWN   \n",
       "\n",
       "                                                         gene_type  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "\n",
       "                                                             havana_gene  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  OTTHUMG00000018876.1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  OTTHUMG00000018876.1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  OTTHUMG00000018876.1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  OTTHUMG00000018876.1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  OTTHUMG00000018876.1   \n",
       "\n",
       "                                                   level         tag  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...   2.0  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...   2.0  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...   2.0  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...   2.0  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...   2.0  ncRNA_host   \n",
       "\n",
       "                                                         transcript_id  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "\n",
       "                                                   transcript_name  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...            HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...            HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...            HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...            HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...            HPS1   \n",
       "\n",
       "                                                   transcript_status  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "\n",
       "                                                   transcript_type  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding   \n",
       "\n",
       "                                                         ensembl_id  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skipped_exon_feature_data = pd.read_csv('{}/index/se/manual_metadata.csv'.format(outrigger_folder), index_col=0)\n",
    "print(skipped_exon_feature_data.shape)\n",
    "skipped_exon_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20118005, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>event_length</th>\n",
       "      <th>event_location</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>exon3_length</th>\n",
       "      <th>exon4</th>\n",
       "      <th>...</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>junctions</th>\n",
       "      <th>level</th>\n",
       "      <th>strand</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>1520.0</td>\n",
       "      <td>event:chr10:100188908-100190427:-</td>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>99.0</td>\n",
       "      <td>exon:chr10:100188907-100189399:-</td>\n",
       "      <td>492.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>junction:chr10:100189647-100190327:-@junction:...</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>event:chr10:100188913-100190427:-</td>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>99.0</td>\n",
       "      <td>exon:chr10:100188907-100189399:-</td>\n",
       "      <td>487.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>junction:chr10:100189647-100190327:-@junction:...</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>1098.0</td>\n",
       "      <td>event:chr10:100189330-100190427:-</td>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>99.0</td>\n",
       "      <td>exon:chr10:100188907-100189399:-</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>junction:chr10:100189647-100190327:-@junction:...</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>1520.0</td>\n",
       "      <td>event:chr10:100188908-100190427:-</td>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>99.0</td>\n",
       "      <td>exon:chr10:100188912-100189399:-</td>\n",
       "      <td>492.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>junction:chr10:100189647-100190327:-@junction:...</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoform1=junction:chr10:100189400-100190327:-|isoform2=junction:chr10:100189647-100190327:-@exon:chr10:100189548-100189646:-@junction:chr10:100189400-100189547:-</th>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>event:chr10:100188913-100190427:-</td>\n",
       "      <td>exon:chr10:100190327-100190427:-</td>\n",
       "      <td>100.0</td>\n",
       "      <td>exon:chr10:100189547-100189646:-</td>\n",
       "      <td>99.0</td>\n",
       "      <td>exon:chr10:100188912-100189399:-</td>\n",
       "      <td>487.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>junction:chr10:100189647-100190327:-@junction:...</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000107521.14</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         ensembl_id  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521   \n",
       "\n",
       "                                                    event_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1520.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1515.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1098.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1520.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        1515.0   \n",
       "\n",
       "                                                                       event_location  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188908-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188913-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100189330-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188908-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  event:chr10:100188913-100190427:-   \n",
       "\n",
       "                                                                               exon1  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100190327-100190427:-   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         100.0   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100189547-100189646:-   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          99.0   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188907-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188907-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188907-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188912-100189399:-   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  exon:chr10:100188912-100189399:-   \n",
       "\n",
       "                                                    exon3_length exon4  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         492.0   NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         487.0   NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...          70.0   NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         492.0   NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...         487.0   NaN   \n",
       "\n",
       "                                                         ...       junction24  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...              NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...              NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...              NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...              NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...       ...              NaN   \n",
       "\n",
       "                                                   junction34  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        NaN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...        NaN   \n",
       "\n",
       "                                                                                            junctions  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  junction:chr10:100189647-100190327:-@junction:...   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  junction:chr10:100189647-100190327:-@junction:...   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  junction:chr10:100189647-100190327:-@junction:...   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  junction:chr10:100189647-100190327:-@junction:...   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  junction:chr10:100189647-100190327:-@junction:...   \n",
       "\n",
       "                                                   level strand         tag  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...     2      -  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...     2      -  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...     2      -  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...     2      -  ncRNA_host   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...     2      -  ncRNA_host   \n",
       "\n",
       "                                                         transcript_id  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  ENSG00000107521.14   \n",
       "\n",
       "                                                    transcript_name  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             HPS1   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             HPS1   \n",
       "\n",
       "                                                   transcript_status  \\\n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...             KNOWN   \n",
       "\n",
       "                                                   transcript_type  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding  \n",
       "isoform1=junction:chr10:100189400-100190327:-|i...  protein_coding  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splicing_feature_data = pd.concat([skipped_exon_feature_data, mutually_exclusive_exon_feature_data])\n",
    "print(splicing_feature_data.shape)\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing_feature_data_subset = splicing_feature_data.loc[splicing.columns]\n",
    "print(splicing_feature_data_subset.shape)\n",
    "splicing_feature_data_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing_feature_data_subset = splicing_feature_data_subset.drop_duplicates()\n",
    "print(splicing_feature_data_subset.shape)\n",
    "splicing_feature_data_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "skipped_exon_feature_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing.index = splicing.index.map(lambda x: '_'.join(x.split('_')[:2]))\n",
    "splicing.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add shortened ids to the splicing feature data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "pattern = 'exon:chr[0-9A-Z]+:\\d+-\\d+:[-=]'\n",
    "\n",
    "splicing_feature_data['alternative_exons'] = splicing_feature_data.index.map(lambda x: '\\n'.join(re.findall(pattern, x)))\n",
    "splicing_feature_data['shortened_id'] = splicing_feature_data['alternative_exons'] + '\\n' + splicing_feature_data['event_location']\n",
    "splicing_feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pkm_event = 'isoform1=junction:chr15:72494962-72499068:-@exon:chr15:72494795-72494961:-@junction:chr15:72492997-72494794:-|isoform2=junction:chr15:72495530-72499068:-@exon:chr15:72495363-72495529:-@junction:chr15:72492997-72495362:-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing_feature_data.query('transcript_name == \"SNAP25\"').index.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing_feature_data.loc[pkm_event, 'alternative_exons']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "splicing_feature_data.loc[pkm_event, 'event_location']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pkm_shortened = splicing_feature_data.loc[pkm_event, 'shortened_id'].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a flotilla study!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-52a88d6bea9f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mgene_ontology\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'/projects/ps-yeolab/obotvinnik/flotilla_projects/hg19/gene_ontology.csv'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex_col\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[1;32mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgene_ontology\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mgene_ontology\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "gene_ontology = pd.read_csv('/projects/ps-yeolab/obotvinnik/flotilla_projects/hg19/gene_ontology.csv', index_col=0)\n",
    "print(gene_ontology.shape)\n",
    "gene_ontology.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "greens = map(mpl.colors.rgb2hex, sns.color_palette('Greens', n_colors=3))\n",
    "sns.palplot(greens)\n",
    "\n",
    "lightgreen, mediumgreen, darkgreen = greens\n",
    "\n",
    "phenotype_to_color = {'MN': greens[2],\n",
    "                      'NPC': greens[1],\n",
    "                      'iPSC': greens[0]}\n",
    "phenotype_to_marker = {'MN': 's',    # square\n",
    "                       'NPC': '^',   # Triangle\n",
    "                       'iPSC': 'o'}  # circle\n",
    "metadata_phenotype_order = ('iPSC', 'NPC', 'MN')\n",
    "\n",
    "min_samples = 10\n",
    "\n",
    "study = flotilla.Study(metadata, expression_data=expression, expression_log_base=2, \n",
    "                       expression_plus_one=True,\n",
    "                       expression_thresh=1,\n",
    "                       splicing_data=splicing,\n",
    "#                        splicing_feature_data=splicing_feature_data,\n",
    "#                        splicing_feature_rename_col='gene_name',\n",
    "#                        splicing_feature_expression_id_col='ensembl_id',\n",
    "                       # At least 10 samples per feature (either gene or splicing event)\n",
    "                       splicing_feature_shortener_col='shortened_id',\n",
    "                       metadata_minimum_samples=min_samples, \n",
    "                       mapping_stats_data=mapping_stats,\n",
    "                       mapping_stats_min_reads=1e6,\n",
    "                       mapping_stats_number_mapped_col='Uniquely mapped reads number',\n",
    "                       metadata_phenotype_order=('iPSC', 'NPC', 'MN'), \n",
    "                       metadata_phenotype_to_marker=phenotype_to_marker,\n",
    "                       metadata_phenotype_to_color=phenotype_to_color,\n",
    "                       species='hg19')\n",
    "\n",
    "# Set the curated splicing feature data separately because setting species as 'hg19' overrides\n",
    "study.splicing.feature_data = splicing_feature_data\n",
    "\n",
    "study.plot_gene('RBFOX2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "study.splicing.maybe_renamed_to_feature_id('PKM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "study.plot_event(\"PKM\")\n",
    "# fig = plt.gcf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# ax = plt.gca()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "study.plot_event(pkm_event)\n",
    "ax = plt.gca()\n",
    "ax.set_title('PKM')\n",
    "fig = plt.gcf()\n",
    "fig.tight_layout()\n",
    "\n",
    "ax.text(0.5, -0.4, u'\\n'.join(pkm_event.split('@')), transform=ax.transAxes, \n",
    "        horizontalalignment='center', verticalalignment='top')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "study.plot_event(pkm_event)\n",
    "ax = plt.gca()\n",
    "ax.set_title('PKM')\n",
    "fig = plt.gcf()\n",
    "fig.tight_layout()\n",
    "\n",
    "ax.text(0.5, -0.4, u'\\n'.join(pkm_event.split('@')), transform=ax.transAxes, \n",
    "        horizontalalignment='center', verticalalignment='top')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the study!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "flotilla_dir = '/projects/ps-yeolab/obotvinnik/flotilla_projects'\n",
    "study.save('singlecell_pnm', flotilla_dir=flotilla_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
