{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/alternative\n"
     ]
    }
   ],
   "source": [
    "cd /projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/alternative/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downstream400_placental_mammal_conservation.csv\r\n",
      "exon2_merkin2012_ancient.csv\r\n",
      "exon2_phastcons_placental_mammal.bed\r\n",
      "exon2_phastcons_placental_mammal.txt\r\n",
      "exons.bed\r\n",
      "exons.fasta\r\n",
      "exons_downstream200nt.bed\r\n",
      "exons_downstream200nt_kmers.csv\r\n",
      "exons_kmers.csv\r\n",
      "exons_merkin2012_ancient_alt_exons.bed\r\n",
      "exons_phastcons_placental_mammal.bed\r\n",
      "exons_phastcons_placental_mammal.txt\r\n",
      "exons_stop_plus1.bed\r\n",
      "exons_upstream200_foreground.bed\r\n",
      "exons_upstream200_placental_foreground.bed\r\n",
      "exons_upstream200nt.bed\r\n",
      "exons_upstream200nt_kmers.csv\r\n",
      "gc_content.csv\r\n",
      "genes_expanded.csv\r\n",
      "isoform_domain_effects.csv\r\n",
      "isoform_domain_pfam_accession.csv\r\n",
      "isoform_domain_pfam_go_id.csv\r\n",
      "isoform_domain_pfam_go_name.csv\r\n",
      "isoform_domain_pfam_name.csv\r\n",
      "\u001b[0m\u001b[01;34misoform_protein_properties\u001b[0m/\r\n",
      "\u001b[01;36misoform_translations.csv\u001b[0m@\r\n",
      "isoform_translations.fa\r\n",
      "isoform_translations.fa_iupred_out\r\n",
      "isoform_translations.fa_iupred_tmp\r\n",
      "isoform_translations_hmmscan_pfamA.txt\r\n",
      "isoform_translations_hmmscan_pfamA.txt.err\r\n",
      "isoform_translations_hmmscan_pfamA.txt.out\r\n",
      "maxentscan.csv\r\n",
      "\u001b[01;34mnucleotide_content\u001b[0m/\r\n",
      "\u001b[01;36mphastcons_placental_mammal.txt\u001b[0m@\r\n",
      "protein_translations.csv\r\n",
      "repetitive_element_overlap.csv\r\n",
      "test.bed\r\n",
      "upstream400_placental_mammal_conservation.csv\r\n",
      "\u001b[m"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/obotvinnik/anaconda/lib/python2.7/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "  \"This module will be removed in 0.20.\", DeprecationWarning)\n",
      "/home/obotvinnik/workspace-git/flotilla/flotilla/__init__.py:65: DtypeWarning: Columns (19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  load_species_data=load_species_data)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-10-31 15:48:38\tReading datapackage from /projects/ps-yeolab/obotvinnik/flotilla_projects/singlecell_pnm_figure2_modalities_bayesian/datapackage.json\n",
      "2016-10-31 15:48:38\tParsing datapackage to create a Study object\n",
      "https://s3-us-west-2.amazonaws.com/flotilla-projects/ercc/ERCC_Controls.txt has not been downloaded before.\n",
      "\tDownloading now to /home/obotvinnik/flotilla_projects/hg19/ERCC_Controls.txt\n",
      "2016-10-31 15:49:10 \tInitializing Study\n",
      "2016-10-31 15:49:10 \tInitializing Predictor configuration manager for Study\n",
      "2016-10-31 15:49:10\tPredictor ExtraTreesClassifier is of type <class 'sklearn.ensemble.forest.ExtraTreesClassifier'>\n",
      "2016-10-31 15:49:10\tAdded ExtraTreesClassifier to default predictors\n",
      "2016-10-31 15:49:10\tPredictor ExtraTreesRegressor is of type <class 'sklearn.ensemble.forest.ExtraTreesRegressor'>\n",
      "2016-10-31 15:49:10\tAdded ExtraTreesRegressor to default predictors\n",
      "2016-10-31 15:49:10\tPredictor GradientBoostingClassifier is of type <class 'sklearn.ensemble.gradient_boosting.GradientBoostingClassifier'>\n",
      "2016-10-31 15:49:10\tAdded GradientBoostingClassifier to default predictors\n",
      "2016-10-31 15:49:10\tPredictor GradientBoostingRegressor is of type <class 'sklearn.ensemble.gradient_boosting.GradientBoostingRegressor'>\n",
      "2016-10-31 15:49:10\tAdded GradientBoostingRegressor to default predictors\n",
      "2016-10-31 15:49:15 \tBuilding Gene Ontology database...\n",
      "2016-10-31 15:49:24 \t\tDone\n",
      "2016-10-31 15:49:25 \tLoading metadata\n",
      "2016-10-31 15:49:25\tLoading expression data\n",
      "2016-10-31 15:49:25\tInitializing expression\n",
      "2016-10-31 15:49:25\tDone initializing expression\n",
      "2016-10-31 15:49:28\tLoading splicing data\n",
      "2016-10-31 15:49:28\tInitializing splicing\n",
      "2016-10-31 15:49:28\tDone initializing splicing\n",
      "2016-10-31 15:49:28\tSuccessfully initialized a Study object!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error loading species hg19 data: HTTP Error 404: Not Found"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.set(style='ticks', context='paper', rc={'font.sans-serif':'Arial', 'pdf.fonttype': 42})\n",
    "\n",
    "\n",
    "import six\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "import anchor\n",
    "modality_to_color = anchor.MODALITY_TO_COLOR.copy()\n",
    "modality_to_color['constitutive'] = '#262626'\n",
    "modality_order = ['excluded', 'bimodal', 'included', 'constitutive']\n",
    "modality_palette = [modality_to_color[m] for m in modality_order]\n",
    "modality_palette\n",
    "\n",
    "\n",
    "\n",
    "import flotilla\n",
    "flotilla_dir = '/projects/ps-yeolab/obotvinnik/flotilla_projects'\n",
    "\n",
    "study = flotilla.embark('singlecell_pnm_figure2_modalities_bayesian', flotilla_dir=flotilla_dir)\n",
    "\n",
    "not_outliers = study.splicing.singles.index.difference(study.splicing.outliers.index)\n",
    "\n",
    "import pybedtools\n",
    "\n",
    "folder = '/projects/ps-yeolab/obotvinnik/singlecell_pnms'\n",
    "csv_folder = '{}/csvs_for_paper/'.format(folder)\n",
    "bed_folder = '{}/bed'.format(folder)\n",
    "\n",
    "\n",
    "\n",
    "splicing_feature_folder = '{}/splicing_feature_data'.format(csv_folder)\n",
    "alternative_feature_folder = '{}/alternative'.format(splicing_feature_folder)\n",
    "constitutive_feature_folder = '{}/constitutive'.format(splicing_feature_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "File /projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/alternative/isoform_translations.csv  does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-4-21817e662d82>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0misoform_translations\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'/projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/alternative/isoform_translations.csv '\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[1;32mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0misoform_translations\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36mparser_f\u001b[1;34m(filepath_or_buffer, sep, dialect, compression, doublequote, escapechar, quotechar, quoting, skipinitialspace, lineterminator, header, index_col, names, prefix, skiprows, skipfooter, skip_footer, na_values, true_values, false_values, delimiter, converters, dtype, usecols, engine, delim_whitespace, as_recarray, na_filter, compact_ints, use_unsigned, low_memory, buffer_lines, warn_bad_lines, error_bad_lines, keep_default_na, thousands, comment, decimal, parse_dates, keep_date_col, dayfirst, date_parser, memory_map, float_precision, nrows, iterator, chunksize, verbose, encoding, squeeze, mangle_dupe_cols, tupleize_cols, infer_datetime_format, skip_blank_lines)\u001b[0m\n\u001b[0;32m    496\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[0;32m    497\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 498\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    499\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    500\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    273\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    274\u001b[0m     \u001b[1;31m# Create the parser.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 275\u001b[1;33m     \u001b[0mparser\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    276\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    277\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mnrows\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mchunksize\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[0;32m    588\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'has_index_names'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'has_index_names'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    589\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 590\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    591\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    592\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_get_options_with_defaults\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_make_engine\u001b[1;34m(self, engine)\u001b[0m\n\u001b[0;32m    729\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'c'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    730\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'c'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 731\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    732\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    733\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'python'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/obotvinnik/anaconda/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, src, **kwds)\u001b[0m\n\u001b[0;32m   1101\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'allow_leading_cols'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mFalse\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1102\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1103\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_parser\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1104\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1105\u001b[0m         \u001b[1;31m# XXX\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas/parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader.__cinit__ (pandas/parser.c:3246)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas/parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader._setup_parser_source (pandas/parser.c:6111)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mIOError\u001b[0m: File /projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/alternative/isoform_translations.csv  does not exist"
     ]
    }
   ],
   "source": [
    "isoform_translations = pd.read_csv('/projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/alternative/isoform_translations.csv ')\n",
    "print(isoform_translations.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# IUPred \r\n",
      "# Copyright (c) Zsuzsanna Dosztanyi, 2005\r\n",
      "#\r\n",
      "# Z. Dosztanyi, V. Csizmok, P. Tompa and I. Simon\r\n",
      "# J. Mol. Biol. (2005) 347, 827-839. \r\n",
      "#\r\n",
      "#\r\n",
      "# Prediction output \r\n",
      "# exon:chrY:2713687-2713784:+@exon:chrY:2722641-2722812:+@exon:chrY:2733129-2733286:+@exon:chrY:2734834-2734903:+|CDS:chrY:2713687-2713784:+:2@CDS:chrY:2722641-2722812:+:0@CDS:chrY:2734834-2734903:+:0|isoform2\r\n",
      "    1 V     0.8650\r\n",
      "    2 I     0.7418\r\n",
      "    3 S     0.6984\r\n",
      "    4 I     0.5802\r\n",
      "    5 E     0.4749\r\n",
      "    6 K     0.3491\r\n",
      "    7 T     0.2963\r\n",
      "    8 G     0.2483\r\n",
      "    9 E     0.1998\r\n",
      "   10 H     0.1635\r\n",
      "   11 F     0.1266\r\n"
     ]
    }
   ],
   "source": [
    "! head -n 20 isoform_translations.fa_iupred_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>criteria</th>\n",
       "      <th>criteria_additional</th>\n",
       "      <th>criteria_full</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>exon1</th>\n",
       "      <th>exon1_length</th>\n",
       "      <th>exon2</th>\n",
       "      <th>exon2_length</th>\n",
       "      <th>exon3</th>\n",
       "      <th>...</th>\n",
       "      <th>junction12</th>\n",
       "      <th>junction13</th>\n",
       "      <th>junction23</th>\n",
       "      <th>junction24</th>\n",
       "      <th>junction34</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon2_divisible_by_3</th>\n",
       "      <th>one_ensembl_id</th>\n",
       "      <th>splice_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>event_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>100</td>\n",
       "      <td>exon:chr10:100189548-100189646:-</td>\n",
       "      <td>99</td>\n",
       "      <td>exon:chr10:100189330-100189399:-</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10:100189647-100190327:-</td>\n",
       "      <td>chr10:100189400-100190327:-</td>\n",
       "      <td>chr10:100189400-100189547:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100193697-100193848:-@exon:chr10:100190888-100191048:-@exon:chr10:100190328-100190427:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>152</td>\n",
       "      <td>exon:chr10:100190888-100191048:-</td>\n",
       "      <td>161</td>\n",
       "      <td>exon:chr10:100190328-100190427:-</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10:100191049-100193696:-</td>\n",
       "      <td>chr10:100190428-100193696:-</td>\n",
       "      <td>chr10:100190428-100190887:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000414009.1,ENST000003...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100195392-100195529:-@exon:chr10:100195029-100195171:-@exon:chr10:100193697-100193848:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>best</td>\n",
       "      <td>appris_principal</td>\n",
       "      <td>best,appris_principal</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>exon:chr10:100195392-100195529:-</td>\n",
       "      <td>138</td>\n",
       "      <td>exon:chr10:100195029-100195171:-</td>\n",
       "      <td>143</td>\n",
       "      <td>exon:chr10:100193697-100193848:-</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10:100195172-100195391:-</td>\n",
       "      <td>chr10:100193849-100195391:-</td>\n",
       "      <td>chr10:100193849-100195028:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000361490.4,ENST00000325103.6,ENST000004...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000107521</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>exon:chr10:101165513-101165617:-</td>\n",
       "      <td>105</td>\n",
       "      <td>exon:chr10:101163481-101163631:-</td>\n",
       "      <td>151</td>\n",
       "      <td>exon:chr10:101163226-101163391:-</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10:101163632-101165512:-</td>\n",
       "      <td>chr10:101163392-101165512:-</td>\n",
       "      <td>chr10:101163392-101163480:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000543866.1,ENST00000370508.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000120053</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101419263-101419345:+@exon:chr10:101419619-101419721:+@exon:chr10:101421203-101421385:+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>NaN</td>\n",
       "      <td>only one</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>exon:chr10:101419263-101419345:+</td>\n",
       "      <td>83</td>\n",
       "      <td>exon:chr10:101419619-101419721:+</td>\n",
       "      <td>103</td>\n",
       "      <td>exon:chr10:101421203-101421385:+</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10:101419346-101419618:+</td>\n",
       "      <td>chr10:101419346-101421202:+</td>\n",
       "      <td>chr10:101419722-101421202:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000370489.4</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000198018</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Unnamed: 0  criteria  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...         NaN  only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...         NaN      best   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...         NaN  only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...         NaN  only one   \n",
       "\n",
       "                                                   criteria_additional  \\\n",
       "event_id                                                                 \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                 NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...    appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                 NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                 NaN   \n",
       "\n",
       "                                                            criteria_full  \\\n",
       "event_id                                                                    \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...               only one   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  best,appris_principal   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...               only one   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...               only one   \n",
       "\n",
       "                                                         ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                                               exon1  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195392-100195529:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101165513-101165617:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419263-101419345:+   \n",
       "\n",
       "                                                    exon1_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...           100   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           152   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           138   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           105   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...            83   \n",
       "\n",
       "                                                                               exon2  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189548-100189646:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190888-100191048:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100195029-100195171:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163481-101163631:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101419619-101419721:+   \n",
       "\n",
       "                                                    exon2_length  \\\n",
       "event_id                                                           \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...            99   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...           161   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...           143   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...           151   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...           103   \n",
       "\n",
       "                                                                               exon3  \\\n",
       "event_id                                                                               \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  exon:chr10:100189330-100189399:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  exon:chr10:100190328-100190427:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  exon:chr10:100193697-100193848:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  exon:chr10:101163226-101163391:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  exon:chr10:101421203-101421385:+   \n",
       "\n",
       "                                                       ...      \\\n",
       "event_id                                               ...       \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...     ...       \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...     ...       \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...     ...       \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...     ...       \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...     ...       \n",
       "\n",
       "                                                                     junction12  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189647-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100191049-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100195172-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163632-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101419618:+   \n",
       "\n",
       "                                                                     junction13  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100190327:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100193696:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195391:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101165512:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419346-101421202:+   \n",
       "\n",
       "                                                                     junction23  \\\n",
       "event_id                                                                          \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  chr10:100189400-100189547:-   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  chr10:100190428-100190887:-   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  chr10:100193849-100195028:-   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  chr10:101163392-101163480:-   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  chr10:101419722-101421202:+   \n",
       "\n",
       "                                                   junction24 junction34  \\\n",
       "event_id                                                                   \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...        NaN        NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...        NaN        NaN   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...        NaN        NaN   \n",
       "\n",
       "                                                    strand  \\\n",
       "event_id                                                     \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...       -   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...       -   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...       -   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...       -   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...       +   \n",
       "\n",
       "                                                                                        transcript_id  \\\n",
       "event_id                                                                                                \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENST00000361490.4,ENST00000414009.1,ENST000003...   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENST00000361490.4,ENST00000325103.6,ENST000004...   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                ENST00000543866.1,ENST00000370508.5   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                                  ENST00000370489.4   \n",
       "\n",
       "                                                   exon2_divisible_by_3  \\\n",
       "event_id                                                                  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                 True   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...                False   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...                False   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                False   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...                False   \n",
       "\n",
       "                                                     one_ensembl_id  \\\n",
       "event_id                                                              \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...  ENSG00000107521   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...  ENSG00000120053   \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...  ENSG00000198018   \n",
       "\n",
       "                                                   splice_type  \n",
       "event_id                                                        \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...          SE  \n",
       "exon:chr10:100193697-100193848:-@exon:chr10:100...          SE  \n",
       "exon:chr10:100195392-100195529:-@exon:chr10:100...          SE  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...          SE  \n",
       "exon:chr10:101419263-101419345:+@exon:chr10:101...          SE  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study.splicing.feature_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18929, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform1_gravy</th>\n",
       "      <th>isoform2_gravy</th>\n",
       "      <th>isoform1_translation</th>\n",
       "      <th>isoform2_translation</th>\n",
       "      <th>isoform1_translation.1</th>\n",
       "      <th>isoform2_translation.1</th>\n",
       "      <th>isoform1_translation.2</th>\n",
       "      <th>isoform2_translation.2</th>\n",
       "      <th>isoform1_protein_secondary_structure_helix</th>\n",
       "      <th>isoform1_protein_secondary_structure_sheet</th>\n",
       "      <th>isoform1_protein_secondary_structure_turn</th>\n",
       "      <th>isoform2_protein_secondary_structure_helix</th>\n",
       "      <th>isoform2_protein_secondary_structure_sheet</th>\n",
       "      <th>isoform2_protein_secondary_structure_turn</th>\n",
       "      <th>isoform1_translation.3</th>\n",
       "      <th>isoform2_translation.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.664045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.268616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102.239326</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9467.0555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.202247</td>\n",
       "      <td>0.247191</td>\n",
       "      <td>0.370787</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.056180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.138571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.374329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15790.6691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.328571</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.264286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.135714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101421203-101421385:+@exon:chr10:101439018-101439223:+@exon:chr10:101439482-101439632:+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.053073</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.703552</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34.983799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20486.5337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.357542</td>\n",
       "      <td>0.256983</td>\n",
       "      <td>0.229050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.117318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101480744-101480825:-@exon:chr10:101478103-101478257:-@exon:chr10:101476105-101476218:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.393162</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.807312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.441026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13050.1474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.196581</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101507014-101507147:+@exon:chr10:101510126-101510153:+@exon:chr10:101514286-101514391:+</th>\n",
       "      <td>0.442553</td>\n",
       "      <td>-0.017045</td>\n",
       "      <td>4.329041</td>\n",
       "      <td>6.099915</td>\n",
       "      <td>52.5</td>\n",
       "      <td>50.257955</td>\n",
       "      <td>5158.9614</td>\n",
       "      <td>9521.8938</td>\n",
       "      <td>0.361702</td>\n",
       "      <td>0.404255</td>\n",
       "      <td>0.170213</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.306818</td>\n",
       "      <td>0.238636</td>\n",
       "      <td>0.042553</td>\n",
       "      <td>0.045455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    isoform1_gravy  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...             NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...             NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...             NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...        0.442553   \n",
       "\n",
       "                                                    isoform2_gravy  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...       -0.664045   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...       -0.138571   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...       -0.053073   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...        0.393162   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...       -0.017045   \n",
       "\n",
       "                                                    isoform1_translation  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                   NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                   NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                   NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                   NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...              4.329041   \n",
       "\n",
       "                                                    isoform2_translation  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              4.268616   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              5.374329   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...              8.703552   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...              9.807312   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...              6.099915   \n",
       "\n",
       "                                                    isoform1_translation.1  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                    52.5   \n",
       "\n",
       "                                                    isoform2_translation.1  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...              102.239326   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...               33.800000   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...               34.983799   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...               50.441026   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...               50.257955   \n",
       "\n",
       "                                                    isoform1_translation.2  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...               5158.9614   \n",
       "\n",
       "                                                    isoform2_translation.2  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...               9467.0555   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...              15790.6691   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...              20486.5337   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...              13050.1474   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...               9521.8938   \n",
       "\n",
       "                                                    isoform1_protein_secondary_structure_helix  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                         NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                                         NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                                         NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                                         NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                                    0.361702   \n",
       "\n",
       "                                                    isoform1_protein_secondary_structure_sheet  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                         NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                                         NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                                         NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                                         NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                                    0.404255   \n",
       "\n",
       "                                                    isoform1_protein_secondary_structure_turn  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                        NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                                        NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                                        NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                                        NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                                   0.170213   \n",
       "\n",
       "                                                    isoform2_protein_secondary_structure_helix  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    0.202247   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                                    0.328571   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                                    0.357542   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                                    0.384615   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                                    0.272727   \n",
       "\n",
       "                                                    isoform2_protein_secondary_structure_sheet  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                    0.247191   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                                    0.250000   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                                    0.256983   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                                    0.307692   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                                    0.306818   \n",
       "\n",
       "                                                    isoform2_protein_secondary_structure_turn  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                                   0.370787   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                                   0.264286   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                                   0.229050   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                                   0.196581   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                                   0.238636   \n",
       "\n",
       "                                                    isoform1_translation.3  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                     NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                     NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                0.042553   \n",
       "\n",
       "                                                    isoform2_translation.3  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...                0.056180  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...                0.135714  \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...                0.117318  \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...                0.111111  \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...                0.045455  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_properties = pd.read_csv('/projects/ps-yeolab/obotvinnik/singlecell_pnms/csvs_for_paper/splicing_feature_data/isoform_protein_properties.csv', index_col=0)\n",
    "print(protein_properties.shape)\n",
    "protein_properties.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18929, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform1_gravy</th>\n",
       "      <th>isoform2_gravy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.664045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.138571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101421203-101421385:+@exon:chr10:101439018-101439223:+@exon:chr10:101439482-101439632:+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.053073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101480744-101480825:-@exon:chr10:101478103-101478257:-@exon:chr10:101476105-101476218:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.393162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101507014-101507147:+@exon:chr10:101510126-101510153:+@exon:chr10:101514286-101514391:+</th>\n",
       "      <td>0.442553</td>\n",
       "      <td>-0.017045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    isoform1_gravy  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...             NaN   \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...             NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...             NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...        0.442553   \n",
       "\n",
       "                                                    isoform2_gravy  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...       -0.664045  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...       -0.138571  \n",
       "exon:chr10:101421203-101421385:+@exon:chr10:101...       -0.053073  \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...        0.393162  \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...       -0.017045  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_property = 'gravy'\n",
    "\n",
    "subset = protein_properties[[x for x in protein_properties if protein_property in x]]\n",
    "print(subset.shape)\n",
    "subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11205, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform1_gravy</th>\n",
       "      <th>isoform2_gravy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exon:chr10:100190328-100190427:-@exon:chr10:100189548-100189646:-@exon:chr10:100189330-100189399:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.664045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101165513-101165617:-@exon:chr10:101163481-101163631:-@exon:chr10:101163226-101163391:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.138571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101480744-101480825:-@exon:chr10:101478103-101478257:-@exon:chr10:101476105-101476218:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.393162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101507014-101507147:+@exon:chr10:101510126-101510153:+@exon:chr10:101514286-101514391:+</th>\n",
       "      <td>0.442553</td>\n",
       "      <td>-0.017045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon:chr10:101964263-101964414:-@exon:chr10:101961845-101961906:-@exon:chr10:101960428-101960537:-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.485047</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    isoform1_gravy  \\\n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...             NaN   \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...             NaN   \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...             NaN   \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...        0.442553   \n",
       "exon:chr10:101964263-101964414:-@exon:chr10:101...             NaN   \n",
       "\n",
       "                                                    isoform2_gravy  \n",
       "exon:chr10:100190328-100190427:-@exon:chr10:100...       -0.664045  \n",
       "exon:chr10:101165513-101165617:-@exon:chr10:101...       -0.138571  \n",
       "exon:chr10:101480744-101480825:-@exon:chr10:101...        0.393162  \n",
       "exon:chr10:101507014-101507147:+@exon:chr10:101...       -0.017045  \n",
       "exon:chr10:101964263-101964414:-@exon:chr10:101...       -0.485047  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset = subset.loc[study.splicing.data.columns].dropna(how='all')\n",
    "print(subset.shape)\n",
    "subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>weighted_gravy</th>\n",
       "      <th>event_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample_id, weighted_gravy, event_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = []\n",
    "\n",
    "isoforms = 'isoform1', 'isoform2'\n",
    "\n",
    "for i, (event_id, row) in enumerate(subset.iterrows()):\n",
    "    psi = study.splicing.data[event_id].dropna()\n",
    "    if i > 100:\n",
    "        break\n",
    "    for isoform in isoforms:\n",
    "#     if event_id in study.splicing.data:\n",
    "        if isoform == 'isoform2':\n",
    "            df = psi * value\n",
    "        if isoform == 'isoform1':\n",
    "            df = (1 - psi) * value\n",
    "#         df = df.dropna()\n",
    "        df.name = 'weighted_' + protein_property\n",
    "        df = df.reset_index()\n",
    "        df['event_id'] = event_id\n",
    "        dfs.append(df)\n",
    "property_values = pd.concat(dfs, ignore_index=True)\n",
    "print(property_values.shape)\n",
    "property_values.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
